// tsjs-xpx-chain-sdk/dist/src/model/liquidity/LiquidityProvider.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LiquidityProvider = void 0;
/**
 * The liquidity provider structure
 */
class LiquidityProvider {
    mosaicId;
    providerKey;
    owner;
    additionallyMinted;
    slashingAccount;
    slashingPeriod;
    windowSize;
    creationHeight;
    alpha;
    beta;
    turnoverHistory;
    recentTurnover;
    constructor(mosaicId, providerKey, owner, additionallyMinted, slashingAccount, slashingPeriod, windowSize, creationHeight, alpha, beta, turnoverHistory, recentTurnover) {
        this.mosaicId = mosaicId;
        this.providerKey = providerKey;
        this.owner = owner;
        this.additionallyMinted = additionallyMinted;
        this.slashingAccount = slashingAccount;
        this.slashingPeriod = slashingPeriod;
        this.windowSize = windowSize;
        this.creationHeight = creationHeight;
        this.alpha = alpha;
        this.beta = beta;
        this.turnoverHistory = turnoverHistory;
        this.recentTurnover = recentTurnover;
    }
}
exports.LiquidityProvider = LiquidityProvider;
