// tsjs-xpx-chain-sdk/dist/src/model/harvester/HarvesterInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HarvesterInfo = void 0;
const PublicAccount_1 = require("../account/PublicAccount");
const UInt64_1 = require("../UInt64");
class HarvesterInfo {
    harvesterKey;
    owner;
    disabledHeight;
    lastSigningBlockHeight;
    effectiveBalance;
    canHarvest;
    activity;
    greed;
    constructor(harvesterKey, owner, disabledHeight, lastSigningBlockHeight, effectiveBalance, canHarvest, activity, greed) {
        this.harvesterKey = harvesterKey;
        this.owner = owner;
        this.disabledHeight = disabledHeight;
        this.lastSigningBlockHeight = lastSigningBlockHeight;
        this.effectiveBalance = effectiveBalance;
        this.canHarvest = canHarvest;
        this.activity = activity;
        this.greed = greed;
    }
    static createFromDTO(dto, networkType) {
        return new HarvesterInfo(PublicAccount_1.PublicAccount.createFromPublicKey(dto.key, networkType), PublicAccount_1.PublicAccount.createFromPublicKey(dto.owner, networkType), new UInt64_1.UInt64(dto.disabledHeight), new UInt64_1.UInt64(dto.lastSigningBlockHeight), new UInt64_1.UInt64(dto.effectiveBalance), dto.canHarvest, dto.activity, dto.greed);
    }
}
exports.HarvesterInfo = HarvesterInfo;
