// tsjs-xpx-chain-sdk/dist/src/infrastructure/api/apis.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.APIS = exports.HttpError = void 0;
/*
 * Copyright 2023 ProximaX
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
__exportStar(require("./accountRoutesApi"), exports);
const accountRoutesApi_1 = require("./accountRoutesApi");
__exportStar(require("./blockRoutesApi"), exports);
const blockRoutesApi_1 = require("./blockRoutesApi");
__exportStar(require("./chainRoutesApi"), exports);
const chainRoutesApi_1 = require("./chainRoutesApi");
__exportStar(require("./configRoutesApi"), exports);
const configRoutesApi_1 = require("./configRoutesApi");
__exportStar(require("./diagnosticRoutesApi"), exports);
const diagnosticRoutesApi_1 = require("./diagnosticRoutesApi");
__exportStar(require("./exchangeRoutesApi"), exports);
const exchangeRoutesApi_1 = require("./exchangeRoutesApi");
__exportStar(require("./exchangeSdaRoutesApi"), exports);
const exchangeSdaRoutesApi_1 = require("./exchangeSdaRoutesApi");
__exportStar(require("./lockRoutesApi"), exports);
const lockRoutesApi_1 = require("./lockRoutesApi");
__exportStar(require("./metadataRoutesApi"), exports);
const metadataRoutesApi_1 = require("./metadataRoutesApi");
__exportStar(require("./mosaicRoutesApi"), exports);
const mosaicRoutesApi_1 = require("./mosaicRoutesApi");
__exportStar(require("./namespaceRoutesApi"), exports);
const namespaceRoutesApi_1 = require("./namespaceRoutesApi");
__exportStar(require("./networkRoutesApi"), exports);
const networkRoutesApi_1 = require("./networkRoutesApi");
__exportStar(require("./nodeRoutesApi"), exports);
const nodeRoutesApi_1 = require("./nodeRoutesApi");
__exportStar(require("./storageRoutesApi"), exports);
const storageRoutesApi_1 = require("./storageRoutesApi");
__exportStar(require("./transactionRoutesApi"), exports);
const transactionRoutesApi_1 = require("./transactionRoutesApi");
__exportStar(require("./upgradeRoutesApi"), exports);
const upgradeRoutesApi_1 = require("./upgradeRoutesApi");
__exportStar(require("./harvesterRoutesApi"), exports);
const harvesterRoutesApi_1 = require("./harvesterRoutesApi");
__exportStar(require("./liquidityProviderRoutesApi"), exports);
const liquidityProviderRoutesApi_1 = require("./liquidityProviderRoutesApi");
class HttpError extends Error {
    response;
    body;
    statusCode;
    constructor(response, body, statusCode) {
        super('HTTP request failed');
        this.response = response;
        this.body = body;
        this.statusCode = statusCode;
        this.name = 'HttpError';
    }
}
exports.HttpError = HttpError;
exports.APIS = [
    accountRoutesApi_1.AccountRoutesApi,
    blockRoutesApi_1.BlockRoutesApi,
    chainRoutesApi_1.ChainRoutesApi,
    configRoutesApi_1.ConfigRoutesApi,
    diagnosticRoutesApi_1.DiagnosticRoutesApi,
    exchangeRoutesApi_1.ExchangeRoutesApi,
    exchangeSdaRoutesApi_1.ExchangeSdaRoutesApi,
    lockRoutesApi_1.LockRoutesApi,
    metadataRoutesApi_1.MetadataRoutesApi,
    mosaicRoutesApi_1.MosaicRoutesApi,
    namespaceRoutesApi_1.NamespaceRoutesApi,
    networkRoutesApi_1.NetworkRoutesApi,
    nodeRoutesApi_1.NodeRoutesApi,
    storageRoutesApi_1.StorageRoutesApi,
    transactionRoutesApi_1.TransactionRoutesApi,
    upgradeRoutesApi_1.UpgradeRoutesApi,
    harvesterRoutesApi_1.HarvesterRoutesApi,
    liquidityProviderRoutesApi_1.LiquidityProviderRoutesApi
];
