// tsjs-xpx-chain-sdk/dist/src/infrastructure/api/blockRoutesApi.js
"use strict";
/**
 * Sirius REST API Reference
 * No description provided (https://bcdocs.xpxsirius.io/endpoints/v0.9.0/#tag/Block-routes) *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlockRoutesApi = exports.BlockRoutesApiApiKeys = void 0;
const axios_1 = require("axios");
const models_1 = require("../model/models");
let defaultBasePath = 'http://localhost:3000';
var BlockRoutesApiApiKeys;
(function (BlockRoutesApiApiKeys) {
})(BlockRoutesApiApiKeys || (exports.BlockRoutesApiApiKeys = BlockRoutesApiApiKeys = {}));
class BlockRoutesApi {
    _basePath = defaultBasePath;
    _defaultHeaders = {
        'Content-Type': 'application/json',
        'Accept': 'application/json'
    };
    _useQuerystring = false;
    constructor(basePathOrUsername, password, basePath) {
        if (password) {
            if (basePath) {
                this.basePath = basePath;
            }
        }
        else {
            if (basePathOrUsername) {
                this.basePath = basePathOrUsername;
            }
        }
    }
    set useQuerystring(value) {
        this._useQuerystring = value;
    }
    set basePath(basePath) {
        this._basePath = basePath;
    }
    set defaultHeaders(defaultHeaders) {
        this._defaultHeaders = defaultHeaders;
    }
    get defaultHeaders() {
        return this._defaultHeaders;
    }
    get basePath() {
        return this._basePath;
    }
    combineHeaders(reqOptions) {
        return reqOptions ? { ...this._defaultHeaders, ...reqOptions.headers } : this._defaultHeaders;
    }
    /**
     * Gets a block from the chain that has the given height.
     * @summary Get block information
     * @param height The height of the block.
     */
    async getBlockByHeight(height, reqOptions) {
        const localVarPath = '/block/{height}'
            .replace('{' + 'height' + '}', encodeURIComponent(String(height)));
        let localVarQueryParameters = {};
        // verify required parameter 'height' is not null or undefined
        if (height === null || height === undefined) {
            throw new Error('Required parameter height was null or undefined when calling getBlockByHeight.');
        }
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'GET',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json'
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "BlockInfoDTO");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
    /**
     * Returns the [receipts](https://nemtech.github.io/concepts/receipt.html) linked to a block.
     * @summary Get receipts from a block
     * @param height The height of the block.
     */
    async getBlockReceipts(height, reqOptions) {
        const localVarPath = '/block/{height}/receipts'
            .replace('{' + 'height' + '}', encodeURIComponent(String(height)));
        // verify required parameter 'height' is not null or undefined
        if (height === null || height === undefined) {
            throw new Error('Required parameter height was null or undefined when calling getBlockReceipts.');
        }
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'GET',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json'
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "StatementsDTO");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
    /**
     * Returns an array of [transactions](https://nemtech.github.io/concepts/transaction.html) included in a block for a given block height.
     * @summary Get transactions from a block
     * @param transaction Query Params
     * @param id The transaction id up to which transactions are returned.
     */
    async getBlockTransactions(height, transactionQueryParams, reqOptions) {
        const localVarPath = '/transactions/confirmed';
        // verify required parameter 'height' is not null or undefined
        if (height === null || height === undefined) {
            throw new Error('Required parameter height was null or undefined when calling getBlockTransactions.');
        }
        let localVarQueryParameters = {};
        if (transactionQueryParams) {
            localVarQueryParameters = transactionQueryParams.buildQueryParams();
        }
        localVarQueryParameters['height'] = height;
        localVarQueryParameters['embedded'] = "true";
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'GET',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json',
            params: localVarQueryParameters
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "TransactionSearchDTO");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
    /**
     * Gets up to limit number of blocks after given block height.
     * @summary Get blocks information
     * @param height The height of the block. If height -1 is not a multiple of the limit provided, the inferior closest multiple + 1 is used instead.
     * @param limit The number of blocks to be returned.
     */
    async getBlocksByHeightWithLimit(height, limit, reqOptions) {
        const localVarPath = '/blocks/{height}/limit/{limit}'
            .replace('{' + 'height' + '}', encodeURIComponent(String(height)))
            .replace('{' + 'limit' + '}', encodeURIComponent(String(limit)));
        let localVarQueryParameters = {};
        // verify required parameter 'height' is not null or undefined
        if (height === null || height === undefined) {
            throw new Error('Required parameter height was null or undefined when calling getBlocksByHeightWithLimit.');
        }
        // verify required parameter 'limit' is not null or undefined
        if (limit === null || limit === undefined) {
            throw new Error('Required parameter limit was null or undefined when calling getBlocksByHeightWithLimit.');
        }
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'GET',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json'
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "Array<BlockInfoDTO>");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
    /**
     * Returns the merkle path for a [receipt statement or resolution](https://nemtech.github.io/concepts/receipt.html) linked to a block. The path is the complementary data needed to calculate the merkle root. A client can compare if the calculated root equals the one recorded in the block header, verifying that the receipt was linked with the block.
     * @summary Get the merkle path for a given a receipt statement hash and block
     * @param height The height of the block.
     * @param hash The hash of the receipt statement or resolution.
     */
    async getMerkleReceipts(height, hash, reqOptions) {
        const localVarPath = '/block/{height}/receipt/{hash}/merkle'
            .replace('{' + 'height' + '}', encodeURIComponent(String(height)))
            .replace('{' + 'hash' + '}', encodeURIComponent(String(hash)));
        let localVarQueryParameters = {};
        // verify required parameter 'height' is not null or undefined
        if (height === null || height === undefined) {
            throw new Error('Required parameter height was null or undefined when calling getMerkleReceipts.');
        }
        // verify required parameter 'hash' is not null or undefined
        if (hash === null || hash === undefined) {
            throw new Error('Required parameter hash was null or undefined when calling getMerkleReceipts.');
        }
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'GET',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json'
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "MerkleProofInfoDTO");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
    /**
     * Returns the merkle path for a [transaction](https://nemtech.github.io/concepts/transaction.html) included in a block. The path is the complementary data needed to calculate the merkle root. A client can compare if the calculated root equals the one recorded in the block header, verifying that the transaction was included in the block.
     * @summary Get the merkle path for a given a transaction and block
     * @param height The height of the block.
     * @param hash The hash of the transaction.
     */
    async getMerkleTransaction(height, hash, reqOptions) {
        const localVarPath = '/block/{height}/transaction/{hash}/merkle'
            .replace('{' + 'height' + '}', encodeURIComponent(String(height)))
            .replace('{' + 'hash' + '}', encodeURIComponent(String(hash)));
        // verify required parameter 'height' is not null or undefined
        if (height === null || height === undefined) {
            throw new Error('Required parameter height was null or undefined when calling getMerkleTransaction.');
        }
        // verify required parameter 'hash' is not null or undefined
        if (hash === null || hash === undefined) {
            throw new Error('Required parameter hash was null or undefined when calling getMerkleTransaction.');
        }
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'GET',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json'
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "MerkleProofInfoDTO");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
}
exports.BlockRoutesApi = BlockRoutesApi;
