// tsjs-xpx-chain-sdk/dist/src/infrastructure/api/diagnosticRoutesApi.js
"use strict";
/**
 * Sirius REST API Reference
 * No description provided (https://bcdocs.xpxsirius.io/endpoints/v0.9.0/#tag/Diagnostic-routes) *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DiagnosticRoutesApi = exports.DiagnosticRoutesApiApiKeys = void 0;
const axios_1 = require("axios");
const models_1 = require("../model/models");
let defaultBasePath = 'http://localhost:3000';
var DiagnosticRoutesApiApiKeys;
(function (DiagnosticRoutesApiApiKeys) {
})(DiagnosticRoutesApiApiKeys || (exports.DiagnosticRoutesApiApiKeys = DiagnosticRoutesApiApiKeys = {}));
class DiagnosticRoutesApi {
    _basePath = defaultBasePath;
    _defaultHeaders = {
        'Content-Type': 'application/json',
        'Accept': 'application/json'
    };
    _useQuerystring = false;
    constructor(basePathOrUsername, password, basePath) {
        if (password) {
            if (basePath) {
                this.basePath = basePath;
            }
        }
        else {
            if (basePathOrUsername) {
                this.basePath = basePathOrUsername;
            }
        }
    }
    set useQuerystring(value) {
        this._useQuerystring = value;
    }
    set basePath(basePath) {
        this._basePath = basePath;
    }
    set defaultHeaders(defaultHeaders) {
        this._defaultHeaders = defaultHeaders;
    }
    get defaultHeaders() {
        return this._defaultHeaders;
    }
    get basePath() {
        return this._basePath;
    }
    combineHeaders(reqOptions) {
        return reqOptions ? { ...this._defaultHeaders, ...reqOptions.headers } : this._defaultHeaders;
    }
    /**
     * Returns diagnostic information about the node storage.
     * @summary Get the storage information of the node
     */
    async getDiagnosticStorage(reqOptions) {
        const localVarPath = '/diagnostic/storage';
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'GET',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json'
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "StorageInfoDTO");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
    /**
     * Returns the version of the running rest component.
     * @summary Get the version of the running rest component
     */
    async getServerInfo(reqOptions) {
        const localVarPath = '/diagnostic/server';
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'GET',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json'
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "ServerDTO");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
}
exports.DiagnosticRoutesApi = DiagnosticRoutesApi;
