// tsjs-xpx-chain-sdk/dist/src/infrastructure/api/exchangeSdaRoutesApi.js
"use strict";
/**
 * Sirius REST API Reference
 * No description provided (https://bcdocs.xpxsirius.io/endpoints/v0.9.0/#tag/SDA-SDA-Exchange-routes) *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExchangeSdaRoutesApi = exports.ExchangeSdaRoutesApiApiKeys = void 0;
const axios_1 = require("axios");
const models_1 = require("../model/models");
let defaultBasePath = 'http://localhost:3000';
var ExchangeSdaRoutesApiApiKeys;
(function (ExchangeSdaRoutesApiApiKeys) {
})(ExchangeSdaRoutesApiApiKeys || (exports.ExchangeSdaRoutesApiApiKeys = ExchangeSdaRoutesApiApiKeys = {}));
class ExchangeSdaRoutesApi {
    _basePath = defaultBasePath;
    _defaultHeaders = {
        'Content-Type': 'application/json',
        'Accept': 'application/json'
    };
    _useQuerystring = false;
    constructor(basePathOrUsername, password, basePath) {
        if (password) {
            if (basePath) {
                this.basePath = basePath;
            }
        }
        else {
            if (basePathOrUsername) {
                this.basePath = basePathOrUsername;
            }
        }
    }
    set useQuerystring(value) {
        this._useQuerystring = value;
    }
    set basePath(basePath) {
        this._basePath = basePath;
    }
    set defaultHeaders(defaultHeaders) {
        this._defaultHeaders = defaultHeaders;
    }
    get defaultHeaders() {
        return this._defaultHeaders;
    }
    get basePath() {
        return this._basePath;
    }
    combineHeaders(reqOptions) {
        return reqOptions ? { ...this._defaultHeaders, ...reqOptions.headers } : this._defaultHeaders;
    }
    /**
     * Return exchange offer by account id.
     * @summary Get exchange offers by account id
     * @param accountId The public key or address of an account.
     */
    async getAccountSdaExchangeOffers(accountId, reqOptions) {
        const localVarPath = '/account/{accountId}/exchangesda'
            .replace('{' + 'accountId' + '}', encodeURIComponent(String(accountId)));
        // verify required parameter 'accountId' is not null or undefined
        if (accountId === null || accountId === undefined) {
            throw new Error('Required parameter accountId was null or undefined when calling getAccountSdaExchangeOffers.');
        }
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'GET',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json'
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "ExchangeSdaInlineResponse200");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
    /**
     * Get exchange offer by type and mosaic id.
     * @summary Get exchange offer by type and mosaic id
     * @param offerType The type of exchange offer.
     * @param mosaicId The mosaic identifier.
     */
    async getExchangeSdaOffers(offerType, mosaicId, reqOptions) {
        const localVarPath = '/exchangesda/{offerType}/{mosaicId}'
            .replace('{' + 'offerType' + '}', encodeURIComponent(String(offerType)))
            .replace('{' + 'mosaicId' + '}', encodeURIComponent(String(mosaicId)));
        // verify required parameter 'offerType' is not null or undefined
        if (offerType === null || offerType === undefined) {
            throw new Error('Required parameter offerType was null or undefined when calling getExchangeSdaOffers.');
        }
        // verify required parameter 'mosaicId' is not null or undefined
        if (mosaicId === null || mosaicId === undefined) {
            throw new Error('Required parameter mosaicId was null or undefined when calling getExchangeSdaOffers.');
        }
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'GET',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json'
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "Array<SdaOfferInfoWithOwnerDTO>");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
}
exports.ExchangeSdaRoutesApi = ExchangeSdaRoutesApi;
