// tsjs-xpx-chain-sdk/dist/src/infrastructure/api/lockRoutesApi.js
"use strict";
/**
 * Sirius REST API Reference
 * No description provided (https://bcdocs.xpxsirius.io/endpoints/v0.9.0/#tag/Lock-routes) *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LockRoutesApi = exports.LockRoutesApiApiKeys = void 0;
const axios_1 = require("axios");
const models_1 = require("../model/models");
let defaultBasePath = 'http://localhost:3000';
var LockRoutesApiApiKeys;
(function (LockRoutesApiApiKeys) {
})(LockRoutesApiApiKeys || (exports.LockRoutesApiApiKeys = LockRoutesApiApiKeys = {}));
class LockRoutesApi {
    _basePath = defaultBasePath;
    _defaultHeaders = {
        'Content-Type': 'application/json',
        'Accept': 'application/json'
    };
    _useQuerystring = false;
    constructor(basePathOrUsername, password, basePath) {
        if (password) {
            if (basePath) {
                this.basePath = basePath;
            }
        }
        else {
            if (basePathOrUsername) {
                this.basePath = basePathOrUsername;
            }
        }
    }
    set useQuerystring(value) {
        this._useQuerystring = value;
    }
    set basePath(basePath) {
        this._basePath = basePath;
    }
    set defaultHeaders(defaultHeaders) {
        this._defaultHeaders = defaultHeaders;
    }
    get defaultHeaders() {
        return this._defaultHeaders;
    }
    get basePath() {
        return this._basePath;
    }
    combineHeaders(reqOptions) {
        return reqOptions ? { ...this._defaultHeaders, ...reqOptions.headers } : this._defaultHeaders;
    }
    /**
     * Get account lock hash.
     * @summary Get account lock hash
     * @param accountId The public key or address of the account.
     */
    async getAccountLockHash(accountId, reqOptions) {
        const localVarPath = '/account/{accountId}/lock/hash'
            .replace('{' + 'accountId' + '}', encodeURIComponent(String(accountId)));
        // verify required parameter 'accountId' is not null or undefined
        if (accountId === null || accountId === undefined) {
            throw new Error('Required parameter accountId was null or undefined when calling getAccountLockHash.');
        }
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'GET',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json'
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "Array<HashLockWithMeta>");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
    /**
     * Get account lock secret.
     * @summary Get account lock secret
     * @param accountId The public key or address of the account.
     */
    async getAccountLocksecret(accountId, reqOptions) {
        const localVarPath = '/account/{accountId}/lock/secret'
            .replace('{' + 'accountId' + '}', encodeURIComponent(String(accountId)));
        // verify required parameter 'accountId' is not null or undefined
        if (accountId === null || accountId === undefined) {
            throw new Error('Required parameter accountId was null or undefined when calling getAccountLocksecret.');
        }
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'GET',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json'
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "Array<SecretLockWithMeta>");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
    /**
     * Get composite hash.
     * @summary Get composite hash
     * @param compositeHash The composite hash of account addres and secret.
     */
    async getCompositeHash(compositeHash, reqOptions) {
        const localVarPath = '/lock/compositeHash/{compositeHash}'
            .replace('{' + 'compositeHash' + '}', encodeURIComponent(String(compositeHash)));
        // verify required parameter 'compositeHash' is not null or undefined
        if (compositeHash === null || compositeHash === undefined) {
            throw new Error('Required parameter compositeHash was null or undefined when calling getCompositeHash.');
        }
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'GET',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json'
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "SecretLockWithMeta");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
    /**
     * Get lock hash.
     * @summary Get lock hash
     * @param hash The hash.
     */
    async getLockHash(hash, reqOptions) {
        const localVarPath = '/lock/hash/{hash}'
            .replace('{' + 'hash' + '}', encodeURIComponent(String(hash)));
        // verify required parameter 'hash' is not null or undefined
        if (hash === null || hash === undefined) {
            throw new Error('Required parameter hash was null or undefined when calling getLockHash.');
        }
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'GET',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json'
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "HashLockWithMeta");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
    /**
     * Get secret hash.
     * @summary Get secret hash
     * @param secret The proof hashed.
     */
    async getSecretHash(secret, reqOptions) {
        const localVarPath = '/lock/secret/{secret}'
            .replace('{' + 'secret' + '}', encodeURIComponent(String(secret)));
        // verify required parameter 'secret' is not null or undefined
        if (secret === null || secret === undefined) {
            throw new Error('Required parameter secret was null or undefined when calling getSecretHash.');
        }
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'GET',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json'
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "SecretLockWithMeta");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
}
exports.LockRoutesApi = LockRoutesApi;
