// tsjs-xpx-chain-sdk/dist/src/infrastructure/api/metadataRoutesApi.js
"use strict";
/**
 * Sirius REST API Reference
 * No description provided (https://bcdocs.xpxsirius.io/endpoints/v0.9.0/#tag/Metadata-V2-routes) *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetadataRoutesApi = void 0;
const axios_1 = require("axios");
const models_1 = require("../model/models");
let defaultBasePath = 'http://localhost:3000';
class MetadataRoutesApi {
    _basePath = defaultBasePath;
    _defaultHeaders = {
        'Content-Type': 'application/json',
        'Accept': 'application/json'
    };
    _useQuerystring = false;
    constructor(basePathOrUsername, password, basePath) {
        if (password) {
            if (basePath) {
                this.basePath = basePath;
            }
        }
        else {
            if (basePathOrUsername) {
                this.basePath = basePathOrUsername;
            }
        }
    }
    set useQuerystring(value) {
        this._useQuerystring = value;
    }
    set basePath(basePath) {
        this._basePath = basePath;
    }
    set defaultHeaders(defaultHeaders) {
        this._defaultHeaders = defaultHeaders;
    }
    get defaultHeaders() {
        return this._defaultHeaders;
    }
    get basePath() {
        return this._basePath;
    }
    combineHeaders(reqOptions) {
        return reqOptions ? { ...this._defaultHeaders, ...reqOptions.headers } : this._defaultHeaders;
    }
    /**
     * Get the metadata from a compositeHash.
     * @summary Get metadata of namespace/mosaic/account
     * @param compositeHash The metadata compositeHash identifier.
     */
    async getMetadata(compositeHash, reqOptions) {
        const localVarPath = '/metadata_v2/{compositeHash}'
            .replace('{' + 'compositeHash' + '}', encodeURIComponent(String(compositeHash)));
        // verify required parameter 'compositeHash' is not null or undefined
        if (compositeHash === null || compositeHash === undefined) {
            throw new Error('Required parameter compositeHash was null or undefined when calling getMetadata.');
        }
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'GET',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json'
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "MetadataEntryGetInlineResponse");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
    /**
     * Gets an array of metadata.
     * @summary Get metadatas(namespace/mosaic/account) for an array of compositeHashes
     * @param compositeHashes of metadata
     */
    async getMetadatas(compositeHashes, reqOptions) {
        const localVarPath = '/metadata_v2';
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'POST',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json',
            data: models_1.ObjectSerializer.serialize(compositeHashes, "compositeHashes")
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "Array<MetadataEntryGetInlineResponse>");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
    /**
     * Search metadatas based on the metadata query params
     * @summary Search metadata
     * @param metadataQueryParams search filter
     */
    async searchMetadata(metadataQueryParams, reqOptions) {
        const localVarPath = '/metadata_v2';
        let localVarQueryParameters = {};
        if (metadataQueryParams) {
            localVarQueryParameters = metadataQueryParams.buildQueryParams();
        }
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'GET',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json',
            params: localVarQueryParameters
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "MetadataSearchDTO");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
}
exports.MetadataRoutesApi = MetadataRoutesApi;
