// tsjs-xpx-chain-sdk/dist/src/infrastructure/api/namespaceRoutesApi.js
"use strict";
/**
 * Sirius REST API Reference
 * No description provided (https://bcdocs.xpxsirius.io/endpoints/v0.9.0/#tag/Namespace-routes) *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NamespaceRoutesApi = exports.NamespaceRoutesApiApiKeys = void 0;
const axios_1 = require("axios");
const models_1 = require("../model/models");
let defaultBasePath = 'http://localhost:3000';
var NamespaceRoutesApiApiKeys;
(function (NamespaceRoutesApiApiKeys) {
})(NamespaceRoutesApiApiKeys || (exports.NamespaceRoutesApiApiKeys = NamespaceRoutesApiApiKeys = {}));
class NamespaceRoutesApi {
    _basePath = defaultBasePath;
    _defaultHeaders = {
        'Content-Type': 'application/json',
        'Accept': 'application/json'
    };
    _useQuerystring = false;
    constructor(basePathOrUsername, password, basePath) {
        if (password) {
            if (basePath) {
                this.basePath = basePath;
            }
        }
        else {
            if (basePathOrUsername) {
                this.basePath = basePathOrUsername;
            }
        }
    }
    set useQuerystring(value) {
        this._useQuerystring = value;
    }
    set basePath(basePath) {
        this._basePath = basePath;
    }
    set defaultHeaders(defaultHeaders) {
        this._defaultHeaders = defaultHeaders;
    }
    get defaultHeaders() {
        return this._defaultHeaders;
    }
    get basePath() {
        return this._basePath;
    }
    combineHeaders(reqOptions) {
        return reqOptions ? { ...this._defaultHeaders, ...reqOptions.headers } : this._defaultHeaders;
    }
    /**
     * Gets the namespace for a given namespaceId.
     * @summary Get namespace information
     * @param namespaceId The namespace identifier.
     */
    async getNamespace(namespaceId, reqOptions) {
        const localVarPath = '/namespace/{namespaceId}'
            .replace('{' + 'namespaceId' + '}', encodeURIComponent(String(namespaceId)));
        let localVarQueryParameters = {};
        // verify required parameter 'namespaceId' is not null or undefined
        if (namespaceId === null || namespaceId === undefined) {
            throw new Error('Required parameter namespaceId was null or undefined when calling getNamespace.');
        }
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'GET',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json'
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "NamespaceInfoDTO");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
    /**
     * Gets an array of namespaces for a given account address.
     * @summary Get namespaces owned by an account
     * @param accountId The address or public key of the account.
     * @param pageSize The number of namespaces to return.
     * @param id The namespace id up to which namespace objects are returned.
     */
    async getNamespacesFromAccount(accountId, pageSize, id, reqOptions) {
        const localVarPath = '/account/{accountId}/namespaces'
            .replace('{' + 'accountId' + '}', encodeURIComponent(String(accountId)));
        let localVarQueryParameters = {};
        // verify required parameter 'accountId' is not null or undefined
        if (accountId === null || accountId === undefined) {
            throw new Error('Required parameter accountId was null or undefined when calling getNamespacesFromAccount.');
        }
        if (pageSize !== undefined) {
            localVarQueryParameters['pageSize'] = models_1.ObjectSerializer.serialize(pageSize, "number");
        }
        if (id !== undefined) {
            localVarQueryParameters['id'] = models_1.ObjectSerializer.serialize(id, "string");
        }
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'GET',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json',
            params: localVarQueryParameters,
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "Array<NamespaceInfoDTO>");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
    /**
     * Gets namespaces for a given array of addresses.
     * @summary Get namespaces for given array of addresses
     * @param accountIds
     * @param pageSize The number of namespaces to return.
     * @param id The namespace id up to which namespace objects are returned.
     */
    async getNamespacesFromAccounts(accountIds, pageSize, id, reqOptions) {
        const localVarPath = '/account/namespaces';
        let localVarQueryParameters = {};
        // verify required parameter 'accountIds' is not null or undefined
        if (accountIds === null || accountIds === undefined) {
            throw new Error('Required parameter accountIds was null or undefined when calling getNamespacesFromAccounts.');
        }
        if (pageSize !== undefined) {
            localVarQueryParameters['pageSize'] = models_1.ObjectSerializer.serialize(pageSize, "number");
        }
        if (id !== undefined) {
            localVarQueryParameters['id'] = models_1.ObjectSerializer.serialize(id, "string");
        }
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'POST',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json',
            params: localVarQueryParameters,
            data: models_1.ObjectSerializer.serialize(accountIds, "AccountIds")
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "Array<NamespaceInfoDTO>");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
    /**
     * Returns friendly names for namespaces.
     * @summary Get readable names for a set of namespaces
     * @param namespaceIds
     */
    async getNamespacesNames(namespaceIds, reqOptions) {
        const localVarPath = '/namespace/names';
        let localVarQueryParameters = {};
        // verify required parameter 'namespaceIds' is not null or undefined
        if (namespaceIds === null || namespaceIds === undefined) {
            throw new Error('Required parameter namespaceIds was null or undefined when calling getNamespacesNames.');
        }
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'POST',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json',
            data: models_1.ObjectSerializer.serialize(namespaceIds, "NamespaceIds")
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "Array<NamespaceNameDTO>");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
}
exports.NamespaceRoutesApi = NamespaceRoutesApi;
