// tsjs-xpx-chain-sdk/dist/src/infrastructure/api/storageRoutesApi.js
"use strict";
/**
 * Sirius REST API Reference
 * No description provided (https://bcdocs.xpxsirius.io/endpoints/v0.9.0/#tag/Storage-routes) *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StorageRoutesApi = void 0;
const axios_1 = require("axios");
const models_1 = require("../model/models");
let defaultBasePath = 'http://localhost:3000';
class StorageRoutesApi {
    _basePath = defaultBasePath;
    _defaultHeaders = {
        'Content-Type': 'application/json',
        'Accept': 'application/json'
    };
    _useQuerystring = false;
    constructor(basePathOrUsername, password, basePath) {
        if (password) {
            if (basePath) {
                this.basePath = basePath;
            }
        }
        else {
            if (basePathOrUsername) {
                this.basePath = basePathOrUsername;
            }
        }
    }
    set useQuerystring(value) {
        this._useQuerystring = value;
    }
    set basePath(basePath) {
        this._basePath = basePath;
    }
    set defaultHeaders(defaultHeaders) {
        this._defaultHeaders = defaultHeaders;
    }
    get defaultHeaders() {
        return this._defaultHeaders;
    }
    get basePath() {
        return this._basePath;
    }
    combineHeaders(reqOptions) {
        return reqOptions ? { ...this._defaultHeaders, ...reqOptions.headers } : this._defaultHeaders;
    }
    /**
     * Return bc drive info by account id/ drive key.
     * @summary Return bc drive info by account id/ drive key.
     * @param driveKey The public key of the drive.
     */
    async getBcDrive(accountId, reqOptions) {
        const localVarPath = '/bcdrives/{accountId}'
            .replace('{' + 'accountId' + '}', encodeURIComponent(String(accountId)));
        // verify required parameter 'accountId' is not null or undefined
        if (accountId === null || accountId === undefined || accountId === "") {
            throw new Error('Required parameter accountId was null or undefined when calling getBcDrive.');
        }
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'GET',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json'
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "DriveInlineDTO");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
    /**
     * Search bc drives info
     * @summary Search bc drives info with query params
     */
    async searchBcDrives(queryParams, reqOptions) {
        const localVarPath = '/bcdrives';
        let localVarQueryParameters = {};
        if (queryParams) {
            localVarQueryParameters = queryParams.buildQueryParams();
        }
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'GET',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json',
            params: localVarQueryParameters
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "BcDriveSearchDTO");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
    /**
     * Return replicator info by account id/ address key.
     * @summary Return replicator info by account id/ address key.
     * @param accountId The public key or address of an account.
     */
    async getReplicator(accountId, reqOptions) {
        const localVarPath = '/replicators/{accountId}'
            .replace('{' + 'accountId' + '}', encodeURIComponent(String(accountId)));
        // verify required parameter 'accountId' is not null or undefined
        if (accountId === null || accountId === undefined || accountId === "") {
            throw new Error('Required parameter accountId was null or undefined when calling getReplicator.');
        }
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'GET',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json'
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "ReplicatorInlineDTO");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
    /**
     * Search replicators info
     * @summary Search replicators info with query params
     */
    async searchReplicators(queryParams, reqOptions) {
        const localVarPath = '/replicators';
        let localVarQueryParameters = {};
        if (queryParams) {
            localVarQueryParameters = queryParams.buildQueryParams();
        }
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'GET',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json',
            params: localVarQueryParameters
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "ReplicatorSearchDTO");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
    /**
     * Return download channel info by public key.
     * @summary Return download chanel info by public key.
     * @param accountId The public key download channel.
     */
    async getDownloadChannel(channelId, reqOptions) {
        const localVarPath = '/download_channels/{channelId}'
            .replace('{' + 'channelId' + '}', encodeURIComponent(String(channelId)));
        // verify required parameter 'channelId' is not null or undefined
        if (channelId === null || channelId === undefined || channelId === "") {
            throw new Error('Required parameter channelId was null or undefined when calling getDownloadChannel.');
        }
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'GET',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json'
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "DownloadChannelInlineInfoDTO");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
    /**
     * Search download channel info
     * @summary Search download channel info with query params
     */
    async searchDownloadChannel(queryParams, reqOptions) {
        const localVarPath = '/download_channels';
        let localVarQueryParameters = {};
        if (queryParams) {
            localVarQueryParameters = queryParams.buildQueryParams();
        }
        let requestHeaders = this.combineHeaders(reqOptions);
        let localVarRequestOptions = {
            method: 'GET',
            headers: requestHeaders,
            url: localVarPath,
            baseURL: this.basePath,
            responseType: 'json',
            params: localVarQueryParameters
        };
        return new Promise((resolve, reject) => {
            (0, axios_1.default)(localVarRequestOptions).then((response) => {
                let body = models_1.ObjectSerializer.deserialize(response.data, "DownloadChannelSearchDTO");
                if (response.status && response.status >= 200 && response.status <= 299) {
                    resolve({ response: response, body: body });
                }
                else {
                    reject(response);
                }
            }, (error) => {
                reject(error);
            });
        });
    }
}
exports.StorageRoutesApi = StorageRoutesApi;
