// tsjs-xpx-chain-sdk/dist/src/model/account/Account.js
"use strict";
/*
 * Copyright 2023 ProximaX
 * Copyright 2018 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Account = void 0;
const crypto_1 = require("../../core/crypto");
const format_1 = require("../../core/format");
const EncryptedMessage_1 = require("../transaction/EncryptedMessage");
const Transaction_1 = require("../transaction/Transaction");
const Address_1 = require("./Address");
const PublicAccount_1 = require("./PublicAccount");
/**
 * The account structure describes an account private key, public key, address and allows signing transactions.
 */
class Account {
    address;
    keyPair;
    version;
    /**
     * @internal
     * @param address
     * @param keyPair
     * @param version the account version, default to version 2
     */
    constructor(
    /**
     * The account address.
     */
    address, 
    /**
     * The account keyPair, public and private key.
     */
    keyPair, 
    /**
     * The account version.
     */
    version = 1) {
        this.address = address;
        this.keyPair = keyPair;
        this.version = version;
    }
    /**
     * Create an Account from a given private key
     * @param privateKey - Private key from an account
     * @param networkType - Network type
     * @return {Account}
     */
    static createFromPrivateKeyV1(privateKey, networkType) {
        const keyPair = crypto_1.KeyPair.createKeyPairFromPrivateKeyString(privateKey, crypto_1.DerivationScheme.Ed25519Sha3);
        const address = format_1.RawAddress.addressToString(format_1.RawAddress.publicKeyToAddress(keyPair.publicKey, networkType));
        return new Account(Address_1.Address.createFromRawAddress(address), keyPair, 1);
    }
    /**
     * Create an Account from a given private key
     * @param privateKey - Private key from an account
     * @param networkType - Network type
     * @return {Account}
     */
    static createFromPrivateKeyV2(privateKey, networkType) {
        const keyPair = crypto_1.KeyPair.createKeyPairFromPrivateKeyString(privateKey, crypto_1.DerivationScheme.Ed25519Sha2);
        const address = format_1.RawAddress.addressToString(format_1.RawAddress.publicKeyToAddress(keyPair.publicKey, networkType));
        return new Account(Address_1.Address.createFromRawAddress(address), keyPair, 2);
    }
    /**
     * Create an Account from a given private key
     * @param privateKey - Private key from an account
     * @param networkType - Network type
     * @param version - Account version (1 or 2)
     * @return {Account}
     */
    static createFromPrivateKey(privateKey, networkType, version = 1) {
        switch (version) {
            case 1:
                return this.createFromPrivateKeyV1(privateKey, networkType);
            case 2:
                return this.createFromPrivateKeyV2(privateKey, networkType);
            default:
                throw new Error("Invalid version number");
        }
    }
    /**
    * Create an Account from a given mnemonic string
    * mnenonic language supported,
    * need to setDefaultWordlist to specific langauge in order to use other langauge
    * default is english
    * langauge supported:
    * chinese_simplified
    * chinese_traditional
    * english
    * japanese
    * spanish
    * italian
    * french
    * korean
    * czech
    * portuguese
    * @param mnemonic - The mnemonic string
    * @param networkType - Network type
    * @param version - Account version (1 or 2)
    * @return {Account}
    */
    static createFromMnemonic(mnemonic, networkType, version = 1) {
        if (!crypto_1.Crypto.isValidMnemonic(mnemonic)) {
            throw Error(`Invalid mnemonic: ${mnemonic.length}`);
        }
        const privateKey = crypto_1.Crypto.mnemonicToHex(mnemonic);
        switch (version) {
            case 1:
                return this.createFromPrivateKeyV1(privateKey, networkType);
            case 2:
                return this.createFromPrivateKeyV2(privateKey, networkType);
            default:
                throw new Error("Invalid version number");
        }
    }
    /**
     * Generate a new account
     * @param networkType - Network type
     * @param version - Account version
     */
    static generateNewAccount(networkType, version = 1) {
        // Create random bytes
        const randomBytesArray = crypto_1.Crypto.randomBytes(32);
        // Hash random bytes with entropy seed
        // Finalize and keep only 32 bytes
        const hashKey = format_1.Convert.uint8ArrayToHex(randomBytesArray);
        const dScheme = PublicAccount_1.PublicAccount.getDerivationSchemeFromAccVersion(version);
        // Create KeyPair from hash key
        const keyPair = crypto_1.KeyPair.createKeyPairFromPrivateKeyString(hashKey, dScheme);
        const address = Address_1.Address.createFromPublicKey(format_1.Convert.uint8ArrayToHex(keyPair.publicKey), networkType);
        return new Account(address, keyPair, version);
    }
    /**
     * Create a new encrypted Message
     * @param message - Plain message to be encrypted
     * @param recipientPublicAccount - Recipient public account
     * @returns {EncryptedMessage}
     */
    encryptMessage(message, recipientPublicAccount) {
        const dScheme = PublicAccount_1.PublicAccount.getDerivationSchemeFromAccVersion(this.version);
        return EncryptedMessage_1.EncryptedMessage.create(message, recipientPublicAccount, this.privateKey, dScheme);
    }
    /**
     * Decrypts an encrypted message
     * @param encryptedMessage - Encrypted message
     * @param publicAccount - The public account originally encrypted the message
     * @returns {PlainMessage}
     */
    decryptMessage(encryptedMessage, publicAccount) {
        const dScheme = PublicAccount_1.PublicAccount.getDerivationSchemeFromAccVersion(this.version);
        return EncryptedMessage_1.EncryptedMessage.decrypt(encryptedMessage, this.privateKey, publicAccount, dScheme);
    }
    /**
     * Account public key.
     * @return {string}
     */
    get publicKey() {
        return format_1.Convert.uint8ArrayToHex(this.keyPair.publicKey);
    }
    /**
     * Public account.
     * @return {PublicAccount}
     */
    get publicAccount() {
        return PublicAccount_1.PublicAccount.createFromPublicKey(this.publicKey, this.address.networkType, this.version);
    }
    /**
     * Account private key.
     * @return {string}
     */
    get privateKey() {
        return format_1.Convert.uint8ArrayToHex(this.keyPair.privateKey);
    }
    /**
     * Account private key represent by mnemonic
     * @return {string}
     */
    get mnemonic() {
        return crypto_1.Crypto.entropyToMnemonic(this.privateKey);
    }
    /**
     * Sign a transaction
     * @param transaction - The transaction to be signed.
     * @param generationHash - Network generation hash hex
     * @return {SignedTransaction}
     */
    sign(transaction, generationHash) {
        return transaction.signWith(this, generationHash);
    }
    /**
     * Sign a transaction
     * @param transaction - The transaction to be signed.
     * @param generationHash - Network generation hash hex
     * @return {SignedTransaction}
     */
    upgradeV2Sign(transaction, generationHash) {
        return transaction.newSignWith(this, generationHash);
    }
    /**
     * Sign a transaction payload, please make sure the transaction version scheme is correct
     * @param transaction - The transaction payload to be signed.
     * @param generationHash - Network generation hash hex
     * @return {SignedTransaction}
     */
    signTransactionPayload(payload, generationHash) {
        return Transaction_1.Transaction.signPayload(payload, this, generationHash);
    }
    /**
     * Sign transaction with cosignatories creating a new SignedTransaction
     * @param transaction - The aggregate transaction to be signed.
     * @param cosignatories - The array of accounts that will cosign the transaction
     * @param generationHash - Network generation hash hex
     * @return {SignedTransaction}
     */
    signTransactionWithCosignatories(transaction, cosignatories, generationHash) {
        return transaction.signTransactionWithCosignatories(this, cosignatories, generationHash);
    }
    /**
     * Sign transaction with cosignatories creating a new SignedTransaction
     * @param transaction - The aggregate transaction to be signed.
     * @param cosignatories - The array of accounts that will cosign the transaction
     * @param generationHash - Network generation hash hex
     * @return {SignedTransaction}
     */
    signTransactionWithCosignatoriesV1(transaction, cosignatories, generationHash) {
        return transaction.signTransactionWithCosignatoriesV1(this, cosignatories, generationHash);
    }
    /**
     * Sign aggregate signature transaction
     * @param cosignatureTransaction - The aggregate signature transaction.
     * @return {CosignatureSignedTransaction}
     */
    newSignCosignatureTransaction(cosignatureTransaction) {
        const dScheme = PublicAccount_1.PublicAccount.getDerivationSchemeFromAccVersion(this.version);
        return cosignatureTransaction.newSignWith(this);
    }
    /**
     * Sign aggregate signature transaction
     * @param cosignatureTransaction - The aggregate signature transaction.
     * @return {CosignatureSignedTransaction}
     */
    signCosignatureTransaction(cosignatureTransaction) {
        return cosignatureTransaction.signWith(this);
    }
    /**
     * Sign raw data
     * @param data - Data to be signed
     * @return {string} - Signed data result
     */
    signData(data) {
        const dScheme = PublicAccount_1.PublicAccount.getDerivationSchemeFromAccVersion(this.version);
        return format_1.Convert.uint8ArrayToHex(crypto_1.KeyPair.sign(this.keyPair, format_1.Convert.hexToUint8(format_1.Convert.utf8ToHex(data)), dScheme));
    }
    /**
     * Sign hexadecimal string (bytes representation)
     * @param hexString - Hex string to be signed
     * @return {string} - Signed data result
     */
    signHexString(hexString) {
        const dScheme = PublicAccount_1.PublicAccount.getDerivationSchemeFromAccVersion(this.version);
        if (!format_1.Convert.isHexString(hexString))
            throw new Error('Invalid hex string');
        return format_1.Convert.uint8ArrayToHex(crypto_1.KeyPair.sign(this.keyPair, format_1.Convert.hexToUint8(hexString), dScheme));
    }
}
exports.Account = Account;
