// tsjs-xpx-chain-sdk/dist/src/model/account/AccountInfo.js
"use strict";
/*
 * Copyright 2023 ProximaX
 * Copyright 2018 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountInfo = void 0;
const PublicAccount_1 = require("./PublicAccount");
/**
 * The account info structure describes basic information for an account.
 */
class AccountInfo {
    meta;
    address;
    addressHeight;
    publicKey;
    publicKeyHeight;
    accountType;
    linkedAccountKey;
    mosaics;
    version;
    /**
     * Account info get from blockchain
     */
    constructor(
    /**
     * Account metadata
     */
    meta, 
    /**
     * Address of the account.
     */
    address, 
    /**
     * Height when the address was published.
     */
    addressHeight, 
    /**
     * Public key of the account.
     */
    publicKey, 
    /**
     * Height when the public key was published.
     */
    publicKeyHeight, 
    /**
     * Account type
     */
    accountType, 
    /**
     * Linked account key
     */
    linkedAccountKey, 
    /**
     * Mosaics hold by the account.
     */
    mosaics, 
    /**
     * account version
     */
    version = 1) {
        this.meta = meta;
        this.address = address;
        this.addressHeight = addressHeight;
        this.publicKey = publicKey;
        this.publicKeyHeight = publicKeyHeight;
        this.accountType = accountType;
        this.linkedAccountKey = linkedAccountKey;
        this.mosaics = mosaics;
        this.version = version;
    }
    /**
     * Returns account public account.
     * @returns {PublicAccount}
     */
    get publicAccount() {
        return PublicAccount_1.PublicAccount.createFromPublicKey(this.publicKey, this.address.networkType, this.version);
    }
}
exports.AccountInfo = AccountInfo;
