// tsjs-xpx-chain-sdk/dist/src/model/account/PublicAccount.js
"use strict";
/*
 * Copyright 2023 ProximaX
 * Copyright 2018 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PublicAccount = void 0;
const crypto_1 = require("../../core/crypto");
const format_1 = require("../../core/format");
const Address_1 = require("./Address");
const Hash512 = 64;
/**
 * The public account structure contains account's address and public key.
 */
class PublicAccount {
    publicKey;
    address;
    version;
    /**
     * @internal
     * @param publicKey
     * @param address
     * @param version
     */
    constructor(
    /**
     * The account public private.
     */
    publicKey, 
    /**
     * The account address.
     */
    address, 
    /**
     * The account version.
     */
    version = 1) {
        this.publicKey = publicKey;
        this.address = address;
        this.version = version;
    }
    /**
     * Create a PublicAccount from a public key and network type.
     * @param publicKey Public key
     * @param networkType Network type
     * @param version Account version
     * @returns {PublicAccount}
     */
    static createFromPublicKey(publicKey, networkType, version = 1) {
        if (publicKey == null || (publicKey.length !== 64 && publicKey.length !== 66)) {
            throw new Error('Not a valid public key');
        }
        if (publicKey.length === 66 && !publicKey.startsWith("0x")) {
            throw new Error('Not a valid public key');
        }
        else if (publicKey.length === 66) {
            publicKey = publicKey.substring(2);
        }
        const address = Address_1.Address.createFromPublicKey(publicKey, networkType);
        return new PublicAccount(publicKey, address, version);
    }
    /**
     * Verify a signature.
     *
     * @param {string} data - The data to verify.
     * @param {string} signature - The signature to verify.
     * @param {DerivationScheme} dSchema The derivation scheme
     * @return {boolean}  - True if the signature is valid, false otherwise.
     */
    verifySignature(data, signature) {
        if (!signature) {
            throw new Error('Missing argument');
        }
        if (signature.length / 2 !== Hash512) {
            throw new Error('Signature length is incorrect');
        }
        if (!format_1.Convert.isHexString(signature)) {
            throw new Error('Signature must be hexadecimal only');
        }
        if (!this.version) {
            throw new Error('Account version missing, please specify a version');
        }
        // Convert signature key to Uint8Array
        const convertedSignature = format_1.Convert.hexToUint8(signature);
        // Convert to Uint8Array
        const convertedData = format_1.Convert.hexToUint8(format_1.Convert.utf8ToHex(data));
        const dScheme = PublicAccount.getDerivationSchemeFromAccVersion(this.version);
        return crypto_1.KeyPair.verify(format_1.Convert.hexToUint8(this.publicKey), convertedData, convertedSignature, dScheme);
    }
    /**
     * Verify a signature from hexadecimal string (bytes representation).
     *
     * @param {string} hexString - The data to verify.
     * @param {string} signature - The signature to verify.
     * @param {DerivationScheme} dSchema The derivation scheme
     * @return {boolean}  - True if the signature is valid, false otherwise.
     */
    verifySignatureWithHexString(hexString, signature) {
        if (!format_1.Convert.isHexString(hexString)) {
            throw new Error('HexString must be hexadecimal only');
        }
        if (!signature) {
            throw new Error('Missing argument');
        }
        if (signature.length / 2 !== Hash512) {
            throw new Error('Signature length is incorrect');
        }
        if (!format_1.Convert.isHexString(signature)) {
            throw new Error('Signature must be hexadecimal only');
        }
        if (!this.version) {
            throw new Error('Account version missing, please specify a version');
        }
        // Convert signature key to Uint8Array
        const convertedSignature = format_1.Convert.hexToUint8(signature);
        // Convert to Uint8Array
        const convertedData = format_1.Convert.hexToUint8(hexString);
        const dScheme = PublicAccount.getDerivationSchemeFromAccVersion(this.version);
        return crypto_1.KeyPair.verify(format_1.Convert.hexToUint8(this.publicKey), convertedData, convertedSignature, dScheme);
    }
    /**
     * Compares public accounts for equality.
     * @param publicAccount
     * @returns {boolean}
     */
    equals(publicAccount) {
        return this.publicKey === publicAccount.publicKey && this.address.plain() === publicAccount.address.plain();
    }
    /**
     * Create DTO object
     */
    toDTO() {
        return {
            publicKey: this.publicKey,
            address: this.address.toDTO(),
        };
    }
    static getDerivationSchemeFromAccVersion(accVersion) {
        if (accVersion === 2) {
            return crypto_1.DerivationScheme.Ed25519Sha2;
        }
        else {
            return crypto_1.DerivationScheme.Ed25519Sha3;
        }
    }
    static getAccVersionFromDerivationScheme(dScheme) {
        if (dScheme === crypto_1.DerivationScheme.Ed25519Sha2) {
            return 2;
        }
        else {
            return 1;
        }
    }
}
exports.PublicAccount = PublicAccount;
