// tsjs-xpx-chain-sdk/dist/src/infrastructure/builders/AccountMetadataTransaction.js
"use strict";
// Copyright 2019 ProximaX Limited. All rights reserved.
// Use of this source code is governed by the Apache 2.0
// license that can be found in the LICENSE file
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = void 0;
/**
 * @module transactions/AccountMetadataTransaction
 */
const format_1 = require("../../core/format");
const VerifiableTransaction_1 = require("./VerifiableTransaction");
const AccountMetadataTransactionSchema_1 = require("../schemas/AccountMetadataTransactionSchema");
const AccountMetadataTransactionBuffer_1 = require("../buffers/AccountMetadataTransactionBuffer");
const flatbuffers = require("flatbuffers");
class AccountMetadataTransaction extends VerifiableTransaction_1.VerifiableTransaction {
    constructor(bytes) {
        super(bytes, AccountMetadataTransactionSchema_1.default);
    }
}
exports.default = AccountMetadataTransaction;
class Builder {
    size;
    fee;
    version;
    type;
    deadline;
    targetPublicKey;
    scopedMetadataKey;
    valueSizeDelta;
    value;
    oldValue;
    valueSize;
    valueDifferences;
    constructor() {
        this.fee = [0, 0];
        this.version = 1;
    }
    addSize(size) {
        this.size = size;
        return this;
    }
    addMaxFee(fee) {
        this.fee = fee;
        return this;
    }
    addVersion(version) {
        this.version = version;
        return this;
    }
    addType(type) {
        this.type = type;
        return this;
    }
    addDeadline(deadline) {
        this.deadline = deadline;
        return this;
    }
    addTargetPublicKey(targetPublicKey) {
        this.targetPublicKey = targetPublicKey;
        return this;
    }
    addScopedMetadataKey(scopedMetadataKey) {
        this.scopedMetadataKey = scopedMetadataKey;
        return this;
    }
    addValueSizeDelta(valueSizeDelta) {
        this.valueSizeDelta = valueSizeDelta;
        return this;
    }
    addValue(value) {
        this.value = value;
        return this;
    }
    addOldValue(oldValue) {
        this.oldValue = oldValue;
        return this;
    }
    addValueSize(valueSize) {
        this.valueSize = valueSize;
        return this;
    }
    addValueDifferences(valueDifferences) {
        this.valueDifferences = valueDifferences;
        return this;
    }
    build() {
        const builder = new flatbuffers.Builder(1);
        const valueSizeDeltaUint8Array = new Uint8Array([this.valueSizeDelta, this.valueSizeDelta >> 8]);
        const valueSizeUint8Array = new Uint8Array([this.valueSize, this.valueSize >> 8]);
        // Create vectors
        const signatureVector = AccountMetadataTransactionBuffer_1.AccountMetadataTransactionBuffer
            .createSignatureVector(builder, Array(...Array(64)).map(Number.prototype.valueOf, 0));
        const signerVector = AccountMetadataTransactionBuffer_1.AccountMetadataTransactionBuffer.createSignerVector(builder, Array(...Array(32)).map(Number.prototype.valueOf, 0));
        const deadlineVector = AccountMetadataTransactionBuffer_1.AccountMetadataTransactionBuffer.createDeadlineVector(builder, this.deadline);
        const feeVector = AccountMetadataTransactionBuffer_1.AccountMetadataTransactionBuffer.createMaxFeeVector(builder, this.fee);
        const targetKeyVector = AccountMetadataTransactionBuffer_1.AccountMetadataTransactionBuffer.createTargetKeyVector(builder, format_1.Convert.hexToUint8(this.targetPublicKey));
        const scopedMetadataKeyVector = AccountMetadataTransactionBuffer_1.AccountMetadataTransactionBuffer.createScopedMetadataKeyVector(builder, this.scopedMetadataKey);
        const valueSizeDeltaVector = AccountMetadataTransactionBuffer_1.AccountMetadataTransactionBuffer.createValueSizeDeltaVector(builder, valueSizeDeltaUint8Array);
        const valueSizeVector = AccountMetadataTransactionBuffer_1.AccountMetadataTransactionBuffer.createValueSizeDeltaVector(builder, valueSizeUint8Array);
        const valueVector = AccountMetadataTransactionBuffer_1.AccountMetadataTransactionBuffer.createValueVector(builder, this.valueDifferences);
        AccountMetadataTransactionBuffer_1.AccountMetadataTransactionBuffer.startAccountMetadataTransactionBuffer(builder);
        AccountMetadataTransactionBuffer_1.AccountMetadataTransactionBuffer.addSize(builder, this.size);
        AccountMetadataTransactionBuffer_1.AccountMetadataTransactionBuffer.addSignature(builder, signatureVector);
        AccountMetadataTransactionBuffer_1.AccountMetadataTransactionBuffer.addSigner(builder, signerVector);
        AccountMetadataTransactionBuffer_1.AccountMetadataTransactionBuffer.addVersion(builder, this.version);
        AccountMetadataTransactionBuffer_1.AccountMetadataTransactionBuffer.addType(builder, this.type);
        AccountMetadataTransactionBuffer_1.AccountMetadataTransactionBuffer.addMaxFee(builder, feeVector);
        AccountMetadataTransactionBuffer_1.AccountMetadataTransactionBuffer.addDeadline(builder, deadlineVector);
        AccountMetadataTransactionBuffer_1.AccountMetadataTransactionBuffer.addTargetKey(builder, targetKeyVector);
        AccountMetadataTransactionBuffer_1.AccountMetadataTransactionBuffer.addScopedMetadataKey(builder, scopedMetadataKeyVector);
        AccountMetadataTransactionBuffer_1.AccountMetadataTransactionBuffer.addValueSizeDelta(builder, valueSizeDeltaVector);
        AccountMetadataTransactionBuffer_1.AccountMetadataTransactionBuffer.addValueSize(builder, valueSizeVector);
        AccountMetadataTransactionBuffer_1.AccountMetadataTransactionBuffer.addValue(builder, valueVector);
        const codedTransfer = AccountMetadataTransactionBuffer_1.AccountMetadataTransactionBuffer.endAccountMetadataTransactionBuffer(builder);
        builder.finish(codedTransfer);
        const bytes = builder.asUint8Array();
        return new AccountMetadataTransaction(bytes);
    }
    static stringToUint8(stringToConvert) {
        return format_1.Convert.hexToUint8(format_1.Convert.utf8ToHex(stringToConvert));
    }
}
exports.Builder = Builder;
