// tsjs-xpx-chain-sdk/dist/src/infrastructure/builders/AggregateTransaction.js
"use strict";
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = exports.AggregateTransaction = void 0;
/**
 * @module transactions/AggregateTransaction
 */
const crypto_1 = require("../../core/crypto");
const TransactionType_1 = require("../../model/transaction/TransactionType");
const AggregateTransactionBuffer_1 = require("../buffers/AggregateTransactionBuffer");
const AggregateTransactionSchema_1 = require("../schemas/AggregateTransactionSchema");
const CosignatureTransaction_1 = require("./CosignatureTransaction");
const VerifiableTransaction_1 = require("./VerifiableTransaction");
const flatbuffers = require("flatbuffers");
const model_1 = require("../../model/model");
const { AggregateTransactionBuffer, } = AggregateTransactionBuffer_1.default.Buffers;
class AggregateTransaction extends VerifiableTransaction_1.VerifiableTransaction {
    constructor(bytes) {
        super(bytes, AggregateTransactionSchema_1.default);
    }
    static recalculateSize(payload) {
        // Calculate new size
        const size = `00000000${(payload.length / 2).toString(16)}`;
        const formatedSize = size.substring(size.length - 8, size.length);
        const littleEndianSize = formatedSize.substring(6, 8) + formatedSize.substring(4, 6) +
            formatedSize.substring(2, 4) + formatedSize.substring(0, 2);
        return littleEndianSize +
            payload.substring(8, payload.length);
    }
    newSignTransactionWithCosigners(initializer, cosigners, generationHash) {
        const dScheme = model_1.PublicAccount.getDerivationSchemeFromAccVersion(initializer.version);
        const signedTransaction = this.signTransaction(initializer, generationHash, dScheme);
        cosigners.forEach((cosigner) => {
            const signatureTransaction = new CosignatureTransaction_1.CosignatureTransaction(signedTransaction.hash);
            const cosignerDScheme = model_1.PublicAccount.getDerivationSchemeFromAccVersion(cosigner.version);
            const signatureCosignTransaction = signatureTransaction.newSignCosignatoriesTransaction(cosigner, cosignerDScheme);
            signedTransaction.payload = signedTransaction.payload + signatureCosignTransaction.scheme +
                signatureCosignTransaction.signer + signatureCosignTransaction.signature;
        });
        // Calculate new size
        signedTransaction.payload = AggregateTransaction.recalculateSize(signedTransaction.payload);
        return signedTransaction;
    }
    signTransactionWithCosigners(initializer, cosigners, generationHash) {
        return this.signTransactionWithCosignersV1(initializer, cosigners, generationHash);
    }
    signTransactionWithCosignersV1(initializer, cosigners, generationHash) {
        const signedTransaction = this.signTransaction(initializer, generationHash, crypto_1.DerivationScheme.Ed25519Sha3);
        cosigners.forEach((cosigner) => {
            const signatureTransaction = new CosignatureTransaction_1.CosignatureTransaction(signedTransaction.hash);
            const signatureCosignTransaction = signatureTransaction.signCosignatoriesTransaction(cosigner);
            signedTransaction.payload = signedTransaction.payload +
                signatureCosignTransaction.signer + signatureCosignTransaction.signature;
        });
        // Calculate new size
        signedTransaction.payload = AggregateTransaction.recalculateSize(signedTransaction.payload);
        return signedTransaction;
    }
    newSignTransactionGivenSignatures(initializer, cosignedSignedTransactions, generationHash) {
        const dScheme = model_1.PublicAccount.getDerivationSchemeFromAccVersion(initializer.version);
        const signedTransaction = this.signTransaction(initializer, generationHash, dScheme);
        cosignedSignedTransactions.forEach((cosignedTransaction) => {
            signedTransaction.payload = signedTransaction.payload + cosignedTransaction.scheme + cosignedTransaction.signer + cosignedTransaction.signature;
        });
        // Calculate new size
        signedTransaction.payload = AggregateTransaction.recalculateSize(signedTransaction.payload);
        return signedTransaction;
    }
    signTransactionGivenSignatures(initializer, cosignedSignedTransactions, generationHash) {
        return this.signTransactionGivenSignaturesV1(initializer, cosignedSignedTransactions, generationHash);
    }
    signTransactionGivenSignaturesV1(initializer, cosignedSignedTransactions, generationHash) {
        const signedTransaction = this.signTransaction(initializer, generationHash, crypto_1.DerivationScheme.Ed25519Sha3);
        cosignedSignedTransactions.forEach((cosignedTransaction) => {
            signedTransaction.payload = signedTransaction.payload + cosignedTransaction.signer + cosignedTransaction.signature;
        });
        // Calculate new size
        signedTransaction.payload = AggregateTransaction.recalculateSize(signedTransaction.payload);
        return signedTransaction;
    }
    static appendSignatures(signedTransaction, cosignedSignedTransactions) {
        let newPayload = signedTransaction.payload;
        cosignedSignedTransactions.forEach(cosignature => {
            newPayload = newPayload + cosignature.scheme + cosignature.signer + cosignature.signature;
        });
        newPayload = this.recalculateSize(newPayload);
        return {
            payload: newPayload,
            hash: signedTransaction.hash
        };
    }
    static appendSignaturesV1(signedTransaction, cosignedSignedTransactions) {
        let newPayload = signedTransaction.payload;
        cosignedSignedTransactions.forEach(cosignature => {
            newPayload = newPayload + cosignature.signer + cosignature.signature;
        });
        newPayload = this.recalculateSize(newPayload);
        return {
            payload: newPayload,
            hash: signedTransaction.hash
        };
    }
}
exports.AggregateTransaction = AggregateTransaction;
// tslint:disable-next-line:max-classes-per-file
class Builder {
    size;
    maxFee;
    version;
    type;
    deadline;
    transactions;
    constructor() {
        this.maxFee = [0, 0];
        this.type = TransactionType_1.TransactionType.AGGREGATE_COMPLETE_V1;
    }
    addSize(size) {
        this.size = size;
        return this;
    }
    addMaxFee(maxFee) {
        this.maxFee = maxFee;
        return this;
    }
    addVersion(version) {
        this.version = version;
        return this;
    }
    addType(type) {
        this.type = type;
        return this;
    }
    addDeadline(deadline) {
        this.deadline = deadline;
        return this;
    }
    addTransactions(transactions) {
        let tmp = [];
        for (let i = 0; i < transactions.length; ++i) {
            tmp = tmp.concat(transactions[i]);
        }
        this.transactions = tmp;
        return this;
    }
    build() {
        const builder = new flatbuffers.Builder(1);
        // Create vectors
        const signatureVector = AggregateTransactionBuffer
            .createSignatureVector(builder, Array(...Array(64))
            .map(Number.prototype.valueOf, 0));
        const signerVector = AggregateTransactionBuffer
            .createSignerVector(builder, Array(...Array(32))
            .map(Number.prototype.valueOf, 0));
        const deadlineVector = AggregateTransactionBuffer.createDeadlineVector(builder, this.deadline);
        const feeVector = AggregateTransactionBuffer.createMaxFeeVector(builder, this.maxFee);
        const modificationsVector = AggregateTransactionBuffer.createTransactionsVector(builder, this.transactions);
        AggregateTransactionBuffer.startAggregateTransactionBuffer(builder);
        AggregateTransactionBuffer.addSize(builder, this.size);
        AggregateTransactionBuffer.addSignature(builder, signatureVector);
        AggregateTransactionBuffer.addSigner(builder, signerVector);
        AggregateTransactionBuffer.addVersion(builder, this.version);
        AggregateTransactionBuffer.addType(builder, this.type);
        AggregateTransactionBuffer.addMaxFee(builder, feeVector);
        AggregateTransactionBuffer.addDeadline(builder, deadlineVector);
        AggregateTransactionBuffer.addTransactionsSize(builder, 0 !== this.transactions.length ? this.transactions.length : 4294967296);
        AggregateTransactionBuffer.addTransactions(builder, modificationsVector);
        // Calculate size
        const codedAggregate = AggregateTransactionBuffer.endAggregateTransactionBuffer(builder);
        builder.finish(codedAggregate);
        const bytes = builder.asUint8Array();
        return new AggregateTransaction(bytes);
    }
}
exports.Builder = Builder;
