// tsjs-xpx-chain-sdk/dist/src/infrastructure/builders/HarvesterTransaction.js
"use strict";
// Copyright 2019 ProximaX Limited. All rights reserved.
// Use of this source code is governed by the Apache 2.0
// license that can be found in the LICENSE file
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = void 0;
/**
 * @module transactions/HarvesterTransaction
 */
const format_1 = require("../../core/format");
const VerifiableTransaction_1 = require("./VerifiableTransaction");
const HarvesterTransactionSchema_1 = require("../schemas/HarvesterTransactionSchema");
const HarvesterTransactionBuffer_1 = require("../buffers/HarvesterTransactionBuffer");
const flatbuffers = require("flatbuffers");
class HarvesterTransaction extends VerifiableTransaction_1.VerifiableTransaction {
    constructor(bytes) {
        super(bytes, HarvesterTransactionSchema_1.default);
    }
}
exports.default = HarvesterTransaction;
class Builder {
    size;
    fee;
    version;
    type;
    deadline;
    harvesterKey;
    constructor() {
        this.fee = [0, 0];
        this.version = 1;
    }
    addSize(size) {
        this.size = size;
        return this;
    }
    addMaxFee(fee) {
        this.fee = fee;
        return this;
    }
    addVersion(version) {
        this.version = version;
        return this;
    }
    addType(type) {
        this.type = type;
        return this;
    }
    addDeadline(deadline) {
        this.deadline = deadline;
        return this;
    }
    addHarvesterKey(harvesterKey) {
        this.harvesterKey = harvesterKey;
        return this;
    }
    build() {
        const builder = new flatbuffers.Builder(1);
        // Create vectors
        const signatureVector = HarvesterTransactionBuffer_1.HarvesterTransactionBuffer
            .createSignatureVector(builder, Array(...Array(64)).map(Number.prototype.valueOf, 0));
        const signerVector = HarvesterTransactionBuffer_1.HarvesterTransactionBuffer.createSignerVector(builder, Array(...Array(32)).map(Number.prototype.valueOf, 0));
        const deadlineVector = HarvesterTransactionBuffer_1.HarvesterTransactionBuffer.createDeadlineVector(builder, this.deadline);
        const feeVector = HarvesterTransactionBuffer_1.HarvesterTransactionBuffer.createMaxFeeVector(builder, this.fee);
        const harvesterVector = HarvesterTransactionBuffer_1.HarvesterTransactionBuffer.createHarvesterKeyVector(builder, format_1.Convert.hexToUint8(this.harvesterKey));
        HarvesterTransactionBuffer_1.HarvesterTransactionBuffer.startHarvesterTransactionBuffer(builder);
        HarvesterTransactionBuffer_1.HarvesterTransactionBuffer.addSize(builder, this.size);
        HarvesterTransactionBuffer_1.HarvesterTransactionBuffer.addSignature(builder, signatureVector);
        HarvesterTransactionBuffer_1.HarvesterTransactionBuffer.addSigner(builder, signerVector);
        HarvesterTransactionBuffer_1.HarvesterTransactionBuffer.addVersion(builder, this.version);
        HarvesterTransactionBuffer_1.HarvesterTransactionBuffer.addType(builder, this.type);
        HarvesterTransactionBuffer_1.HarvesterTransactionBuffer.addMaxFee(builder, feeVector);
        HarvesterTransactionBuffer_1.HarvesterTransactionBuffer.addDeadline(builder, deadlineVector);
        HarvesterTransactionBuffer_1.HarvesterTransactionBuffer.addHarvesterKey(builder, harvesterVector);
        const codedTransfer = HarvesterTransactionBuffer_1.HarvesterTransactionBuffer.endHarvesterTransactionBuffer(builder);
        builder.finish(codedTransfer);
        const bytes = builder.asUint8Array();
        return new HarvesterTransaction(bytes);
    }
}
exports.Builder = Builder;
