// tsjs-xpx-chain-sdk/dist/src/infrastructure/builders/MosaicRemoveLevyTransaction.js
"use strict";
// Copyright 2019 ProximaX Limited. All rights reserved.
// Use of this source code is governed by the Apache 2.0
// license that can be found in the LICENSE file
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = void 0;
const VerifiableTransaction_1 = require("./VerifiableTransaction");
const MosaicRemoveLevyTransactionSchema_1 = require("../schemas/MosaicRemoveLevyTransactionSchema");
const RemoveMosaicLevyTransactionBuffer_1 = require("../buffers/RemoveMosaicLevyTransactionBuffer");
const flatbuffers = require("flatbuffers");
class MosaicRemoveLevyTransaction extends VerifiableTransaction_1.VerifiableTransaction {
    constructor(bytes) {
        super(bytes, MosaicRemoveLevyTransactionSchema_1.default);
    }
}
exports.default = MosaicRemoveLevyTransaction;
class Builder {
    size;
    fee;
    version;
    type;
    deadline;
    mosaicId;
    constructor() {
        this.fee = [0, 0];
        this.version = 1;
    }
    addSize(size) {
        this.size = size;
        return this;
    }
    addMaxFee(fee) {
        this.fee = fee;
        return this;
    }
    addVersion(version) {
        this.version = version;
        return this;
    }
    addType(type) {
        this.type = type;
        return this;
    }
    addDeadline(deadline) {
        this.deadline = deadline;
        return this;
    }
    addMosaicId(mosaicId) {
        this.mosaicId = mosaicId;
        return this;
    }
    build() {
        const builder = new flatbuffers.Builder(1);
        // Create vectors
        const signatureVector = RemoveMosaicLevyTransactionBuffer_1.RemoveMosaicLevyTransactionBuffer
            .createSignatureVector(builder, Array(...Array(64)).map(Number.prototype.valueOf, 0));
        const signerVector = RemoveMosaicLevyTransactionBuffer_1.RemoveMosaicLevyTransactionBuffer.createSignerVector(builder, Array(...Array(32)).map(Number.prototype.valueOf, 0));
        const deadlineVector = RemoveMosaicLevyTransactionBuffer_1.RemoveMosaicLevyTransactionBuffer.createDeadlineVector(builder, this.deadline);
        const feeVector = RemoveMosaicLevyTransactionBuffer_1.RemoveMosaicLevyTransactionBuffer.createMaxFeeVector(builder, this.fee);
        const mosaicIdVector = RemoveMosaicLevyTransactionBuffer_1.RemoveMosaicLevyTransactionBuffer.createMosaicIdVector(builder, this.mosaicId);
        RemoveMosaicLevyTransactionBuffer_1.RemoveMosaicLevyTransactionBuffer.startRemoveMosaicLevyTransactionBuffer(builder);
        RemoveMosaicLevyTransactionBuffer_1.RemoveMosaicLevyTransactionBuffer.addSize(builder, this.size);
        RemoveMosaicLevyTransactionBuffer_1.RemoveMosaicLevyTransactionBuffer.addSignature(builder, signatureVector);
        RemoveMosaicLevyTransactionBuffer_1.RemoveMosaicLevyTransactionBuffer.addSigner(builder, signerVector);
        RemoveMosaicLevyTransactionBuffer_1.RemoveMosaicLevyTransactionBuffer.addVersion(builder, this.version);
        RemoveMosaicLevyTransactionBuffer_1.RemoveMosaicLevyTransactionBuffer.addType(builder, this.type);
        RemoveMosaicLevyTransactionBuffer_1.RemoveMosaicLevyTransactionBuffer.addMaxFee(builder, feeVector);
        RemoveMosaicLevyTransactionBuffer_1.RemoveMosaicLevyTransactionBuffer.addDeadline(builder, deadlineVector);
        RemoveMosaicLevyTransactionBuffer_1.RemoveMosaicLevyTransactionBuffer.addMosaicId(builder, mosaicIdVector);
        const codedTransfer = RemoveMosaicLevyTransactionBuffer_1.RemoveMosaicLevyTransactionBuffer.endRemoveMosaicLevyTransactionBuffer(builder);
        builder.finish(codedTransfer);
        const bytes = builder.asUint8Array();
        return new MosaicRemoveLevyTransaction(bytes);
    }
}
exports.Builder = Builder;
