// tsjs-xpx-chain-sdk/dist/src/infrastructure/builders/MosaicMetadataTransaction.js
"use strict";
// Copyright 2019 ProximaX Limited. All rights reserved.
// Use of this source code is governed by the Apache 2.0
// license that can be found in the LICENSE file
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = void 0;
/**
 * @module transactions/MosaicMetadataTransaction
 */
const format_1 = require("../../core/format");
const VerifiableTransaction_1 = require("./VerifiableTransaction");
const MosaicMetadataTransactionSchema_1 = require("../schemas/MosaicMetadataTransactionSchema");
const MetadataTransactionBuffer_1 = require("../buffers/MetadataTransactionBuffer");
const flatbuffers = require("flatbuffers");
class MosaicMetadataTransaction extends VerifiableTransaction_1.VerifiableTransaction {
    constructor(bytes) {
        super(bytes, MosaicMetadataTransactionSchema_1.default);
    }
}
exports.default = MosaicMetadataTransaction;
class Builder {
    size;
    fee;
    version;
    type;
    deadline;
    targetPublicKey;
    scopedMetadataKey;
    targetMosaicId;
    valueSizeDelta;
    value;
    oldValue;
    valueSize;
    valueDifferences;
    constructor() {
        this.fee = [0, 0];
        this.version = 1;
    }
    addSize(size) {
        this.size = size;
        return this;
    }
    addMaxFee(fee) {
        this.fee = fee;
        return this;
    }
    addVersion(version) {
        this.version = version;
        return this;
    }
    addType(type) {
        this.type = type;
        return this;
    }
    addDeadline(deadline) {
        this.deadline = deadline;
        return this;
    }
    addTargetPublicKey(targetPublicKey) {
        this.targetPublicKey = targetPublicKey;
        return this;
    }
    addScopedMetadataKey(scopedMetadataKey) {
        this.scopedMetadataKey = scopedMetadataKey;
        return this;
    }
    addTargetMosaicId(targetMosaicId) {
        this.targetMosaicId = targetMosaicId;
        return this;
    }
    addValueSizeDelta(valueSizeDelta) {
        this.valueSizeDelta = valueSizeDelta;
        return this;
    }
    addValue(value) {
        this.value = value;
        return this;
    }
    addOldValue(oldValue) {
        this.oldValue = oldValue;
        return this;
    }
    addValueSize(valueSize) {
        this.valueSize = valueSize;
        return this;
    }
    addValueDifferences(valueDifferences) {
        this.valueDifferences = valueDifferences;
        return this;
    }
    build() {
        const builder = new flatbuffers.Builder(1);
        const targetIdUint32 = new Uint32Array(this.targetMosaicId);
        const valueSizeDeltaUint8Array = new Uint8Array([this.valueSizeDelta, this.valueSizeDelta >> 8]);
        const valueSizeUint8Array = new Uint8Array([this.valueSize, this.valueSize >> 8]);
        // Create vectors
        const signatureVector = MetadataTransactionBuffer_1.MetadataTransactionBuffer
            .createSignatureVector(builder, Array(...Array(64)).map(Number.prototype.valueOf, 0));
        const signerVector = MetadataTransactionBuffer_1.MetadataTransactionBuffer.createSignerVector(builder, Array(...Array(32)).map(Number.prototype.valueOf, 0));
        const deadlineVector = MetadataTransactionBuffer_1.MetadataTransactionBuffer.createDeadlineVector(builder, this.deadline);
        const feeVector = MetadataTransactionBuffer_1.MetadataTransactionBuffer.createMaxFeeVector(builder, this.fee);
        const targetKeyVector = MetadataTransactionBuffer_1.MetadataTransactionBuffer.createTargetKeyVector(builder, format_1.Convert.hexToUint8(this.targetPublicKey));
        const scopedMetadataKeyVector = MetadataTransactionBuffer_1.MetadataTransactionBuffer.createScopedMetadataKeyVector(builder, this.scopedMetadataKey);
        const targetIdVector = MetadataTransactionBuffer_1.MetadataTransactionBuffer.createTargetIdVector(builder, targetIdUint32);
        const valueSizeDeltaVector = MetadataTransactionBuffer_1.MetadataTransactionBuffer.createValueSizeDeltaVector(builder, valueSizeDeltaUint8Array);
        const valueSizeVector = MetadataTransactionBuffer_1.MetadataTransactionBuffer.createValueSizeDeltaVector(builder, valueSizeUint8Array);
        const valueVector = MetadataTransactionBuffer_1.MetadataTransactionBuffer.createValueVector(builder, this.valueDifferences);
        MetadataTransactionBuffer_1.MetadataTransactionBuffer.startMetadataTransactionBuffer(builder);
        MetadataTransactionBuffer_1.MetadataTransactionBuffer.addSize(builder, this.size);
        MetadataTransactionBuffer_1.MetadataTransactionBuffer.addSignature(builder, signatureVector);
        MetadataTransactionBuffer_1.MetadataTransactionBuffer.addSigner(builder, signerVector);
        MetadataTransactionBuffer_1.MetadataTransactionBuffer.addVersion(builder, this.version);
        MetadataTransactionBuffer_1.MetadataTransactionBuffer.addType(builder, this.type);
        MetadataTransactionBuffer_1.MetadataTransactionBuffer.addMaxFee(builder, feeVector);
        MetadataTransactionBuffer_1.MetadataTransactionBuffer.addDeadline(builder, deadlineVector);
        MetadataTransactionBuffer_1.MetadataTransactionBuffer.addTargetKey(builder, targetKeyVector);
        MetadataTransactionBuffer_1.MetadataTransactionBuffer.addScopedMetadataKey(builder, scopedMetadataKeyVector);
        MetadataTransactionBuffer_1.MetadataTransactionBuffer.addTargetId(builder, targetIdVector);
        MetadataTransactionBuffer_1.MetadataTransactionBuffer.addValueSizeDelta(builder, valueSizeDeltaVector);
        MetadataTransactionBuffer_1.MetadataTransactionBuffer.addValueSize(builder, valueSizeVector);
        MetadataTransactionBuffer_1.MetadataTransactionBuffer.addValue(builder, valueVector);
        const codedTransfer = MetadataTransactionBuffer_1.MetadataTransactionBuffer.endMetadataTransactionBuffer(builder);
        builder.finish(codedTransfer);
        const bytes = builder.asUint8Array();
        return new MosaicMetadataTransaction(bytes);
    }
    static stringToUint8(stringToConvert) {
        return format_1.Convert.hexToUint8(format_1.Convert.utf8ToHex(stringToConvert));
    }
}
exports.Builder = Builder;
