// tsjs-xpx-chain-sdk/dist/src/infrastructure/builders/MosaicModifyLevyTransaction.js
"use strict";
// Copyright 2019 ProximaX Limited. All rights reserved.
// Use of this source code is governed by the Apache 2.0
// license that can be found in the LICENSE file
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = void 0;
/**
 * @module transactions/MosaicModifyLevyTransaction
 */
const format_1 = require("../../core/format");
const VerifiableTransaction_1 = require("./VerifiableTransaction");
const MosaicModifyLevyTransactionSchema_1 = require("../schemas/MosaicModifyLevyTransactionSchema");
const ModifyMosaicLevyTransactionBuffer_1 = require("../buffers/ModifyMosaicLevyTransactionBuffer");
const MosaicLevyBuffer_1 = require("../buffers/MosaicLevyBuffer");
const flatbuffers = require("flatbuffers");
class MosaicModifyLevyTransaction extends VerifiableTransaction_1.VerifiableTransaction {
    constructor(bytes) {
        super(bytes, MosaicModifyLevyTransactionSchema_1.default);
    }
}
exports.default = MosaicModifyLevyTransaction;
class Builder {
    size;
    fee;
    version;
    type;
    deadline;
    mosaicId;
    levyType;
    levyRecipeint;
    levyMosaicId;
    levyFee;
    constructor() {
        this.fee = [0, 0];
        this.version = 1;
    }
    addSize(size) {
        this.size = size;
        return this;
    }
    addMaxFee(fee) {
        this.fee = fee;
        return this;
    }
    addVersion(version) {
        this.version = version;
        return this;
    }
    addType(type) {
        this.type = type;
        return this;
    }
    addDeadline(deadline) {
        this.deadline = deadline;
        return this;
    }
    addMosaicId(mosaicId) {
        this.mosaicId = mosaicId;
        return this;
    }
    addLevyType(levyType) {
        this.levyType = levyType;
        return this;
    }
    addLevyRecipeint(plainAddress) {
        this.levyRecipeint = format_1.RawAddress.stringToAddress(plainAddress);
        return this;
    }
    addLevyMosaicId(levyMosaicId) {
        this.levyMosaicId = levyMosaicId;
        return this;
    }
    addLevyFee(levyFee) {
        this.levyFee = levyFee;
        return this;
    }
    build() {
        const builder = new flatbuffers.Builder(1);
        // Create vectors
        const signatureVector = ModifyMosaicLevyTransactionBuffer_1.ModifyMosaicLevyTransactionBuffer
            .createSignatureVector(builder, Array(...Array(64)).map(Number.prototype.valueOf, 0));
        const signerVector = ModifyMosaicLevyTransactionBuffer_1.ModifyMosaicLevyTransactionBuffer.createSignerVector(builder, Array(...Array(32)).map(Number.prototype.valueOf, 0));
        const deadlineVector = ModifyMosaicLevyTransactionBuffer_1.ModifyMosaicLevyTransactionBuffer.createDeadlineVector(builder, this.deadline);
        const feeVector = ModifyMosaicLevyTransactionBuffer_1.ModifyMosaicLevyTransactionBuffer.createMaxFeeVector(builder, this.fee);
        const mosaicIdVector = ModifyMosaicLevyTransactionBuffer_1.ModifyMosaicLevyTransactionBuffer.createMosaicIdVector(builder, this.mosaicId);
        const levyRecipientVector = MosaicLevyBuffer_1.MosaicLevy.createRecipientVector(builder, this.levyRecipeint);
        const levyMosaicIdVector = MosaicLevyBuffer_1.MosaicLevy.createMosaicIdVector(builder, this.levyMosaicId);
        const levyFeeVector = MosaicLevyBuffer_1.MosaicLevy.createFeeVector(builder, this.levyFee);
        const levyVector = MosaicLevyBuffer_1.MosaicLevy.createMosaicLevy(builder, this.levyType, levyRecipientVector, levyMosaicIdVector, levyFeeVector);
        ModifyMosaicLevyTransactionBuffer_1.ModifyMosaicLevyTransactionBuffer.startModifyMosaicLevyTransactionBuffer(builder);
        ModifyMosaicLevyTransactionBuffer_1.ModifyMosaicLevyTransactionBuffer.addSize(builder, this.size);
        ModifyMosaicLevyTransactionBuffer_1.ModifyMosaicLevyTransactionBuffer.addSignature(builder, signatureVector);
        ModifyMosaicLevyTransactionBuffer_1.ModifyMosaicLevyTransactionBuffer.addSigner(builder, signerVector);
        ModifyMosaicLevyTransactionBuffer_1.ModifyMosaicLevyTransactionBuffer.addVersion(builder, this.version);
        ModifyMosaicLevyTransactionBuffer_1.ModifyMosaicLevyTransactionBuffer.addType(builder, this.type);
        ModifyMosaicLevyTransactionBuffer_1.ModifyMosaicLevyTransactionBuffer.addMaxFee(builder, feeVector);
        ModifyMosaicLevyTransactionBuffer_1.ModifyMosaicLevyTransactionBuffer.addDeadline(builder, deadlineVector);
        ModifyMosaicLevyTransactionBuffer_1.ModifyMosaicLevyTransactionBuffer.addMosaicId(builder, mosaicIdVector);
        ModifyMosaicLevyTransactionBuffer_1.ModifyMosaicLevyTransactionBuffer.addLevy(builder, levyVector);
        const codedTransfer = ModifyMosaicLevyTransactionBuffer_1.ModifyMosaicLevyTransactionBuffer.endModifyMosaicLevyTransactionBuffer(builder);
        builder.finish(codedTransfer);
        const bytes = builder.asUint8Array();
        return new MosaicModifyLevyTransaction(bytes);
    }
}
exports.Builder = Builder;
