// tsjs-xpx-chain-sdk/dist/src/infrastructure/builders/PlaceSdaExchangeOfferTransaction.js
"use strict";
// Copyright 2022 ProximaX Limited. All rights reserved.
// Use of this source code is governed by the Apache 2.0
// license that can be found in the LICENSE file
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = void 0;
const VerifiableTransaction_1 = require("./VerifiableTransaction");
const TransactionType_1 = require("../../model/transaction/TransactionType");
const PlaceSdaExchangeOfferTransactionSchema_1 = require("../schemas/PlaceSdaExchangeOfferTransactionSchema");
const PlaceSdaExchangeOfferTransactionBuffer_1 = require("../buffers/PlaceSdaExchangeOfferTransactionBuffer");
const PlaceSdaExchangeOfferBuffer_1 = require("../buffers/PlaceSdaExchangeOfferBuffer");
const flatbuffers = require("flatbuffers");
/**
 * @module transactions/PlaceSdaExchangeOfferTransaction
 */
class PlaceSdaExchangeOfferTransaction extends VerifiableTransaction_1.VerifiableTransaction {
    constructor(bytes) {
        super(bytes, PlaceSdaExchangeOfferTransactionSchema_1.default);
    }
}
exports.default = PlaceSdaExchangeOfferTransaction;
// tslint:disable-next-line:max-classes-per-file
class Builder {
    size;
    maxFee;
    version;
    type;
    deadline;
    offers;
    constructor() {
        this.maxFee = [0, 0];
        this.type = TransactionType_1.TransactionType.PLACE_SDA_EXCHANGE_OFFER;
    }
    addSize(size) {
        this.size = size;
        return this;
    }
    addMaxFee(maxFee) {
        this.maxFee = maxFee;
        return this;
    }
    addVersion(version) {
        this.version = version;
        return this;
    }
    addType(type) {
        this.type = type;
        return this;
    }
    addDeadline(deadline) {
        this.deadline = deadline;
        return this;
    }
    addOffers(offers) {
        this.offers = offers;
        return this;
    }
    build() {
        const builder = new flatbuffers.Builder(1);
        // Constants
        // Create offers
        const offersArray = [];
        this.offers.forEach(offer => {
            const mosaicIdGiveVector = PlaceSdaExchangeOfferBuffer_1.PlaceSdaExchangeOfferBuffer
                .createMosaicIdGiveVector(builder, offer.mosaicIdGive.id.toDTO());
            const mosaicAmountGiveVector = PlaceSdaExchangeOfferBuffer_1.PlaceSdaExchangeOfferBuffer
                .createMosaicAmountGiveVector(builder, offer.mosaicAmountGive.toDTO());
            const mosaicIdGetVector = PlaceSdaExchangeOfferBuffer_1.PlaceSdaExchangeOfferBuffer
                .createMosaicIdGetVector(builder, offer.mosaicIdGet.id.toDTO());
            const mosaicAmountGetVector = PlaceSdaExchangeOfferBuffer_1.PlaceSdaExchangeOfferBuffer
                .createMosaicAmountGetVector(builder, offer.mosaicAmountGet.toDTO());
            const durationVector = PlaceSdaExchangeOfferBuffer_1.PlaceSdaExchangeOfferBuffer
                .createDurationVector(builder, offer.duration.toDTO());
            PlaceSdaExchangeOfferBuffer_1.PlaceSdaExchangeOfferBuffer.startPlaceSdaExchangeOfferBuffer(builder);
            PlaceSdaExchangeOfferBuffer_1.PlaceSdaExchangeOfferBuffer.addMosaicIdGive(builder, mosaicIdGiveVector);
            PlaceSdaExchangeOfferBuffer_1.PlaceSdaExchangeOfferBuffer.addMosaicAmountGive(builder, mosaicAmountGiveVector);
            PlaceSdaExchangeOfferBuffer_1.PlaceSdaExchangeOfferBuffer.addMosaicIdGet(builder, mosaicIdGetVector);
            PlaceSdaExchangeOfferBuffer_1.PlaceSdaExchangeOfferBuffer.addMosaicAmountGet(builder, mosaicAmountGetVector);
            PlaceSdaExchangeOfferBuffer_1.PlaceSdaExchangeOfferBuffer.addDuration(builder, durationVector);
            offersArray.push(PlaceSdaExchangeOfferBuffer_1.PlaceSdaExchangeOfferBuffer.endPlaceSdaExchangeOfferBuffer(builder));
        });
        // Create vectors
        const signatureVector = PlaceSdaExchangeOfferTransactionBuffer_1.PlaceSdaExchangeOfferTransactionBuffer
            .createSignatureVector(builder, Array(...Array(64)).map(Number.prototype.valueOf, 0));
        const signerVector = PlaceSdaExchangeOfferTransactionBuffer_1.PlaceSdaExchangeOfferTransactionBuffer
            .createSignerVector(builder, Array(...Array(32)).map(Number.prototype.valueOf, 0));
        const deadlineVector = PlaceSdaExchangeOfferTransactionBuffer_1.PlaceSdaExchangeOfferTransactionBuffer
            .createDeadlineVector(builder, this.deadline);
        const feeVector = PlaceSdaExchangeOfferTransactionBuffer_1.PlaceSdaExchangeOfferTransactionBuffer
            .createMaxFeeVector(builder, this.maxFee);
        const offersVector = PlaceSdaExchangeOfferTransactionBuffer_1.PlaceSdaExchangeOfferTransactionBuffer
            .createSdaOffersVector(builder, offersArray);
        PlaceSdaExchangeOfferTransactionBuffer_1.PlaceSdaExchangeOfferTransactionBuffer.startPlaceSdaExchangeOfferTransactionBuffer(builder);
        PlaceSdaExchangeOfferTransactionBuffer_1.PlaceSdaExchangeOfferTransactionBuffer.addSize(builder, this.size);
        PlaceSdaExchangeOfferTransactionBuffer_1.PlaceSdaExchangeOfferTransactionBuffer.addSignature(builder, signatureVector);
        PlaceSdaExchangeOfferTransactionBuffer_1.PlaceSdaExchangeOfferTransactionBuffer.addSigner(builder, signerVector);
        PlaceSdaExchangeOfferTransactionBuffer_1.PlaceSdaExchangeOfferTransactionBuffer.addVersion(builder, this.version);
        PlaceSdaExchangeOfferTransactionBuffer_1.PlaceSdaExchangeOfferTransactionBuffer.addType(builder, this.type);
        PlaceSdaExchangeOfferTransactionBuffer_1.PlaceSdaExchangeOfferTransactionBuffer.addMaxFee(builder, feeVector);
        PlaceSdaExchangeOfferTransactionBuffer_1.PlaceSdaExchangeOfferTransactionBuffer.addDeadline(builder, deadlineVector);
        PlaceSdaExchangeOfferTransactionBuffer_1.PlaceSdaExchangeOfferTransactionBuffer.addSdaOfferCount(builder, this.offers.length);
        PlaceSdaExchangeOfferTransactionBuffer_1.PlaceSdaExchangeOfferTransactionBuffer.addSdaOffers(builder, offersVector);
        // Calculate size
        const codedOffset = PlaceSdaExchangeOfferTransactionBuffer_1.PlaceSdaExchangeOfferTransactionBuffer.endPlaceSdaExchangeOfferTransactionBuffer(builder);
        builder.finish(codedOffset);
        const bytes = builder.asUint8Array();
        return new PlaceSdaExchangeOfferTransaction(bytes);
    }
}
exports.Builder = Builder;
