// tsjs-xpx-chain-sdk/dist/src/infrastructure/builders/RemoveSdaExchangeOfferTransaction.js
"use strict";
// Copyright 2022 ProximaX Limited. All rights reserved.
// Use of this source code is governed by the Apache 2.0
// license that can be found in the LICENSE file
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = void 0;
const VerifiableTransaction_1 = require("./VerifiableTransaction");
const TransactionType_1 = require("../../model/transaction/TransactionType");
const RemoveSdaExchangeOfferTransactionSchema_1 = require("../schemas/RemoveSdaExchangeOfferTransactionSchema");
const RemoveSdaExchangeOfferTransactionBuffer_1 = require("../buffers/RemoveSdaExchangeOfferTransactionBuffer");
const RemoveSdaExchangeOfferBuffer_1 = require("../buffers/RemoveSdaExchangeOfferBuffer");
const flatbuffers = require("flatbuffers");
/**
 * @module transactions/RemoveSdaExchangeOfferTransaction
 */
class RemoveSdaExchangeOfferTransaction extends VerifiableTransaction_1.VerifiableTransaction {
    constructor(bytes) {
        super(bytes, RemoveSdaExchangeOfferTransactionSchema_1.default);
    }
}
exports.default = RemoveSdaExchangeOfferTransaction;
// tslint:disable-next-line:max-classes-per-file
class Builder {
    size;
    maxFee;
    version;
    type;
    deadline;
    offers;
    constructor() {
        this.maxFee = [0, 0];
        this.type = TransactionType_1.TransactionType.REMOVE_SDA_EXCHANGE_OFFER;
    }
    addSize(size) {
        this.size = size;
        return this;
    }
    addMaxFee(maxFee) {
        this.maxFee = maxFee;
        return this;
    }
    addVersion(version) {
        this.version = version;
        return this;
    }
    addType(type) {
        this.type = type;
        return this;
    }
    addDeadline(deadline) {
        this.deadline = deadline;
        return this;
    }
    addOffers(sdaExchangeOffers) {
        this.offers = sdaExchangeOffers;
        return this;
    }
    build() {
        const builder = new flatbuffers.Builder(1);
        // Constants
        // Create offers
        const offersArray = [];
        this.offers.forEach(offer => {
            const mosaicIdGiveVector = RemoveSdaExchangeOfferBuffer_1.RemoveSdaExchangeOfferBuffer
                .createMosaicIdGiveVector(builder, offer.mosaicIdGive.id.toDTO());
            const mosaicIdGetVector = RemoveSdaExchangeOfferBuffer_1.RemoveSdaExchangeOfferBuffer
                .createMosaicIdGetVector(builder, offer.mosaicIdGet.id.toDTO());
            RemoveSdaExchangeOfferBuffer_1.RemoveSdaExchangeOfferBuffer.startRemoveSdaExchangeOfferBuffer(builder);
            RemoveSdaExchangeOfferBuffer_1.RemoveSdaExchangeOfferBuffer.addMosaicIdGive(builder, mosaicIdGiveVector);
            RemoveSdaExchangeOfferBuffer_1.RemoveSdaExchangeOfferBuffer.addMosaicIdGet(builder, mosaicIdGetVector);
            offersArray.push(RemoveSdaExchangeOfferBuffer_1.RemoveSdaExchangeOfferBuffer.endRemoveSdaExchangeOfferBuffer(builder));
        });
        // Create vectors
        const signatureVector = RemoveSdaExchangeOfferTransactionBuffer_1.RemoveSdaExchangeOfferTransactionBuffer
            .createSignatureVector(builder, Array(...Array(64)).map(Number.prototype.valueOf, 0));
        const signerVector = RemoveSdaExchangeOfferTransactionBuffer_1.RemoveSdaExchangeOfferTransactionBuffer
            .createSignerVector(builder, Array(...Array(32)).map(Number.prototype.valueOf, 0));
        const deadlineVector = RemoveSdaExchangeOfferTransactionBuffer_1.RemoveSdaExchangeOfferTransactionBuffer
            .createDeadlineVector(builder, this.deadline);
        const feeVector = RemoveSdaExchangeOfferTransactionBuffer_1.RemoveSdaExchangeOfferTransactionBuffer
            .createMaxFeeVector(builder, this.maxFee);
        const offersVector = RemoveSdaExchangeOfferTransactionBuffer_1.RemoveSdaExchangeOfferTransactionBuffer
            .createSdaOffersVector(builder, offersArray);
        RemoveSdaExchangeOfferTransactionBuffer_1.RemoveSdaExchangeOfferTransactionBuffer.startRemoveSdaExchangeOfferTransactionBuffer(builder);
        RemoveSdaExchangeOfferTransactionBuffer_1.RemoveSdaExchangeOfferTransactionBuffer.addSize(builder, this.size);
        RemoveSdaExchangeOfferTransactionBuffer_1.RemoveSdaExchangeOfferTransactionBuffer.addSignature(builder, signatureVector);
        RemoveSdaExchangeOfferTransactionBuffer_1.RemoveSdaExchangeOfferTransactionBuffer.addSigner(builder, signerVector);
        RemoveSdaExchangeOfferTransactionBuffer_1.RemoveSdaExchangeOfferTransactionBuffer.addVersion(builder, this.version);
        RemoveSdaExchangeOfferTransactionBuffer_1.RemoveSdaExchangeOfferTransactionBuffer.addType(builder, this.type);
        RemoveSdaExchangeOfferTransactionBuffer_1.RemoveSdaExchangeOfferTransactionBuffer.addMaxFee(builder, feeVector);
        RemoveSdaExchangeOfferTransactionBuffer_1.RemoveSdaExchangeOfferTransactionBuffer.addDeadline(builder, deadlineVector);
        RemoveSdaExchangeOfferTransactionBuffer_1.RemoveSdaExchangeOfferTransactionBuffer.addSdaOfferCount(builder, this.offers.length);
        RemoveSdaExchangeOfferTransactionBuffer_1.RemoveSdaExchangeOfferTransactionBuffer.addSdaOffers(builder, offersVector);
        // Calculate size
        const codedOffset = RemoveSdaExchangeOfferTransactionBuffer_1.RemoveSdaExchangeOfferTransactionBuffer.endRemoveSdaExchangeOfferTransactionBuffer(builder);
        builder.finish(codedOffset);
        const bytes = builder.asUint8Array();
        return new RemoveSdaExchangeOfferTransaction(bytes);
    }
}
exports.Builder = Builder;
