// tsjs-xpx-chain-sdk/dist/src/infrastructure/buffers/storage/DataModificationTransactionBuffer.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataModificationTransactionBuffer = void 0;
const flatbuffers = require("flatbuffers");
class DataModificationTransactionBuffer {
    bb = null;
    bb_pos = 0;
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsDataModificationTransactionBuffer(bb, obj) {
        return (obj || new DataModificationTransactionBuffer()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsDataModificationTransactionBuffer(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new DataModificationTransactionBuffer()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    size() {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    signature(index) {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? this.bb.readUint8(this.bb.__vector(this.bb_pos + offset) + index) : 0;
    }
    signatureLength() {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    signatureArray() {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? new Uint8Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    signer(index) {
        const offset = this.bb.__offset(this.bb_pos, 8);
        return offset ? this.bb.readUint8(this.bb.__vector(this.bb_pos + offset) + index) : 0;
    }
    signerLength() {
        const offset = this.bb.__offset(this.bb_pos, 8);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    signerArray() {
        const offset = this.bb.__offset(this.bb_pos, 8);
        return offset ? new Uint8Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    version() {
        const offset = this.bb.__offset(this.bb_pos, 10);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    type() {
        const offset = this.bb.__offset(this.bb_pos, 12);
        return offset ? this.bb.readUint16(this.bb_pos + offset) : 0;
    }
    maxFee(index) {
        const offset = this.bb.__offset(this.bb_pos, 14);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    maxFeeLength() {
        const offset = this.bb.__offset(this.bb_pos, 14);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    maxFeeArray() {
        const offset = this.bb.__offset(this.bb_pos, 14);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    deadline(index) {
        const offset = this.bb.__offset(this.bb_pos, 16);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    deadlineLength() {
        const offset = this.bb.__offset(this.bb_pos, 16);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    deadlineArray() {
        const offset = this.bb.__offset(this.bb_pos, 16);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    driveKey(index) {
        const offset = this.bb.__offset(this.bb_pos, 18);
        return offset ? this.bb.readUint8(this.bb.__vector(this.bb_pos + offset) + index) : 0;
    }
    driveKeyLength() {
        const offset = this.bb.__offset(this.bb_pos, 18);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    driveKeyArray() {
        const offset = this.bb.__offset(this.bb_pos, 18);
        return offset ? new Uint8Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    downloadDataCdi(index) {
        const offset = this.bb.__offset(this.bb_pos, 20);
        return offset ? this.bb.readUint8(this.bb.__vector(this.bb_pos + offset) + index) : 0;
    }
    downloadDataCdiLength() {
        const offset = this.bb.__offset(this.bb_pos, 20);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    downloadDataCdiArray() {
        const offset = this.bb.__offset(this.bb_pos, 20);
        return offset ? new Uint8Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    uploadSize(index) {
        const offset = this.bb.__offset(this.bb_pos, 22);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    uploadSizeLength() {
        const offset = this.bb.__offset(this.bb_pos, 22);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    uploadSizeArray() {
        const offset = this.bb.__offset(this.bb_pos, 22);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    feedbackFeeAmount(index) {
        const offset = this.bb.__offset(this.bb_pos, 24);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    feedbackFeeAmountLength() {
        const offset = this.bb.__offset(this.bb_pos, 24);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    feedbackFeeAmountArray() {
        const offset = this.bb.__offset(this.bb_pos, 24);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    static startDataModificationTransactionBuffer(builder) {
        builder.startObject(11);
    }
    static addSize(builder, size) {
        builder.addFieldInt32(0, size, 0);
    }
    static addSignature(builder, signatureOffset) {
        builder.addFieldOffset(1, signatureOffset, 0);
    }
    static createSignatureVector(builder, data) {
        builder.startVector(1, data.length, 1);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt8(data[i]);
        }
        return builder.endVector();
    }
    static startSignatureVector(builder, numElems) {
        builder.startVector(1, numElems, 1);
    }
    static addSigner(builder, signerOffset) {
        builder.addFieldOffset(2, signerOffset, 0);
    }
    static createSignerVector(builder, data) {
        builder.startVector(1, data.length, 1);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt8(data[i]);
        }
        return builder.endVector();
    }
    static startSignerVector(builder, numElems) {
        builder.startVector(1, numElems, 1);
    }
    static addVersion(builder, version) {
        builder.addFieldInt32(3, version, 0);
    }
    static addType(builder, type) {
        builder.addFieldInt16(4, type, 0);
    }
    static addMaxFee(builder, maxFeeOffset) {
        builder.addFieldOffset(5, maxFeeOffset, 0);
    }
    static createMaxFeeVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startMaxFeeVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addDeadline(builder, deadlineOffset) {
        builder.addFieldOffset(6, deadlineOffset, 0);
    }
    static createDeadlineVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startDeadlineVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addDriveKey(builder, driveKeyOffset) {
        builder.addFieldOffset(7, driveKeyOffset, 0);
    }
    static createDriveKeyVector(builder, data) {
        builder.startVector(1, data.length, 1);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt8(data[i]);
        }
        return builder.endVector();
    }
    static startDriveKeyVector(builder, numElems) {
        builder.startVector(1, numElems, 1);
    }
    static addDownloadDataCdi(builder, downloadDataCdiOffset) {
        builder.addFieldOffset(8, downloadDataCdiOffset, 0);
    }
    static createDownloadDataCdiVector(builder, data) {
        builder.startVector(1, data.length, 1);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt8(data[i]);
        }
        return builder.endVector();
    }
    static startDownloadDataCdiVector(builder, numElems) {
        builder.startVector(1, numElems, 1);
    }
    static addUploadSize(builder, uploadSizeOffset) {
        builder.addFieldOffset(9, uploadSizeOffset, 0);
    }
    static createUploadSizeVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startUploadSizeVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addFeedbackFeeAmount(builder, feedbackFeeAmountOffset) {
        builder.addFieldOffset(10, feedbackFeeAmountOffset, 0);
    }
    static createFeedbackFeeAmountVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startFeedbackFeeAmountVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static endDataModificationTransactionBuffer(builder) {
        const offset = builder.endObject();
        return offset;
    }
    static createDataModificationTransactionBuffer(builder, size, signatureOffset, signerOffset, version, type, maxFeeOffset, deadlineOffset, driveKeyOffset, downloadDataCdiOffset, uploadSizeOffset, feedbackFeeAmountOffset) {
        DataModificationTransactionBuffer.startDataModificationTransactionBuffer(builder);
        DataModificationTransactionBuffer.addSize(builder, size);
        DataModificationTransactionBuffer.addSignature(builder, signatureOffset);
        DataModificationTransactionBuffer.addSigner(builder, signerOffset);
        DataModificationTransactionBuffer.addVersion(builder, version);
        DataModificationTransactionBuffer.addType(builder, type);
        DataModificationTransactionBuffer.addMaxFee(builder, maxFeeOffset);
        DataModificationTransactionBuffer.addDeadline(builder, deadlineOffset);
        DataModificationTransactionBuffer.addDriveKey(builder, driveKeyOffset);
        DataModificationTransactionBuffer.addDownloadDataCdi(builder, downloadDataCdiOffset);
        DataModificationTransactionBuffer.addUploadSize(builder, uploadSizeOffset);
        DataModificationTransactionBuffer.addFeedbackFeeAmount(builder, feedbackFeeAmountOffset);
        return DataModificationTransactionBuffer.endDataModificationTransactionBuffer(builder);
    }
}
exports.DataModificationTransactionBuffer = DataModificationTransactionBuffer;
