// tsjs-xpx-chain-sdk/dist/src/infrastructure/buffers/storage/DownloadApprovalTransactionBuffer.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
Object.defineProperty(exports, "__esModule", { value: true });
exports.DownloadApprovalTransactionBuffer = void 0;
const flatbuffers = require("flatbuffers");
const KeysBuffer_1 = require("./KeysBuffer");
const OpinionsBuffer_1 = require("./OpinionsBuffer");
const SignaturesBuffer_1 = require("./SignaturesBuffer");
class DownloadApprovalTransactionBuffer {
    bb = null;
    bb_pos = 0;
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsDownloadApprovalTransactionBuffer(bb, obj) {
        return (obj || new DownloadApprovalTransactionBuffer()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsDownloadApprovalTransactionBuffer(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new DownloadApprovalTransactionBuffer()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    size() {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    signature(index) {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? this.bb.readUint8(this.bb.__vector(this.bb_pos + offset) + index) : 0;
    }
    signatureLength() {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    signatureArray() {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? new Uint8Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    signer(index) {
        const offset = this.bb.__offset(this.bb_pos, 8);
        return offset ? this.bb.readUint8(this.bb.__vector(this.bb_pos + offset) + index) : 0;
    }
    signerLength() {
        const offset = this.bb.__offset(this.bb_pos, 8);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    signerArray() {
        const offset = this.bb.__offset(this.bb_pos, 8);
        return offset ? new Uint8Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    version() {
        const offset = this.bb.__offset(this.bb_pos, 10);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    type() {
        const offset = this.bb.__offset(this.bb_pos, 12);
        return offset ? this.bb.readUint16(this.bb_pos + offset) : 0;
    }
    maxFee(index) {
        const offset = this.bb.__offset(this.bb_pos, 14);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    maxFeeLength() {
        const offset = this.bb.__offset(this.bb_pos, 14);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    maxFeeArray() {
        const offset = this.bb.__offset(this.bb_pos, 14);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    deadline(index) {
        const offset = this.bb.__offset(this.bb_pos, 16);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    deadlineLength() {
        const offset = this.bb.__offset(this.bb_pos, 16);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    deadlineArray() {
        const offset = this.bb.__offset(this.bb_pos, 16);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    downloadChannelId(index) {
        const offset = this.bb.__offset(this.bb_pos, 18);
        return offset ? this.bb.readUint8(this.bb.__vector(this.bb_pos + offset) + index) : 0;
    }
    downloadChannelIdLength() {
        const offset = this.bb.__offset(this.bb_pos, 18);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    downloadChannelIdArray() {
        const offset = this.bb.__offset(this.bb_pos, 18);
        return offset ? new Uint8Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    approvalTrigger(index) {
        const offset = this.bb.__offset(this.bb_pos, 20);
        return offset ? this.bb.readUint8(this.bb.__vector(this.bb_pos + offset) + index) : 0;
    }
    approvalTriggerLength() {
        const offset = this.bb.__offset(this.bb_pos, 20);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    approvalTriggerArray() {
        const offset = this.bb.__offset(this.bb_pos, 20);
        return offset ? new Uint8Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    judgingCount() {
        const offset = this.bb.__offset(this.bb_pos, 22);
        return offset ? this.bb.readUint8(this.bb_pos + offset) : 0;
    }
    overlappingCount() {
        const offset = this.bb.__offset(this.bb_pos, 24);
        return offset ? this.bb.readUint8(this.bb_pos + offset) : 0;
    }
    judgedCount() {
        const offset = this.bb.__offset(this.bb_pos, 26);
        return offset ? this.bb.readUint8(this.bb_pos + offset) : 0;
    }
    opinionElementCount() {
        const offset = this.bb.__offset(this.bb_pos, 28);
        return offset ? this.bb.readUint8(this.bb_pos + offset) : 0;
    }
    publicKeys(index, obj) {
        const offset = this.bb.__offset(this.bb_pos, 30);
        return offset ? (obj || new KeysBuffer_1.KeysBuffer()).__init(this.bb.__indirect(this.bb.__vector(this.bb_pos + offset) + index * 4), this.bb) : null;
    }
    publicKeysLength() {
        const offset = this.bb.__offset(this.bb_pos, 30);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    signatures(index, obj) {
        const offset = this.bb.__offset(this.bb_pos, 32);
        return offset ? (obj || new SignaturesBuffer_1.SignaturesBuffer()).__init(this.bb.__indirect(this.bb.__vector(this.bb_pos + offset) + index * 4), this.bb) : null;
    }
    signaturesLength() {
        const offset = this.bb.__offset(this.bb_pos, 32);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    presentOpinions(index) {
        const offset = this.bb.__offset(this.bb_pos, 34);
        return offset ? this.bb.readUint8(this.bb.__vector(this.bb_pos + offset) + index) : 0;
    }
    presentOpinionsLength() {
        const offset = this.bb.__offset(this.bb_pos, 34);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    presentOpinionsArray() {
        const offset = this.bb.__offset(this.bb_pos, 34);
        return offset ? new Uint8Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    opinions(index, obj) {
        const offset = this.bb.__offset(this.bb_pos, 36);
        return offset ? (obj || new OpinionsBuffer_1.OpinionsBuffer()).__init(this.bb.__indirect(this.bb.__vector(this.bb_pos + offset) + index * 4), this.bb) : null;
    }
    opinionsLength() {
        const offset = this.bb.__offset(this.bb_pos, 36);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    static startDownloadApprovalTransactionBuffer(builder) {
        builder.startObject(17);
    }
    static addSize(builder, size) {
        builder.addFieldInt32(0, size, 0);
    }
    static addSignature(builder, signatureOffset) {
        builder.addFieldOffset(1, signatureOffset, 0);
    }
    static createSignatureVector(builder, data) {
        builder.startVector(1, data.length, 1);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt8(data[i]);
        }
        return builder.endVector();
    }
    static startSignatureVector(builder, numElems) {
        builder.startVector(1, numElems, 1);
    }
    static addSigner(builder, signerOffset) {
        builder.addFieldOffset(2, signerOffset, 0);
    }
    static createSignerVector(builder, data) {
        builder.startVector(1, data.length, 1);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt8(data[i]);
        }
        return builder.endVector();
    }
    static startSignerVector(builder, numElems) {
        builder.startVector(1, numElems, 1);
    }
    static addVersion(builder, version) {
        builder.addFieldInt32(3, version, 0);
    }
    static addType(builder, type) {
        builder.addFieldInt16(4, type, 0);
    }
    static addMaxFee(builder, maxFeeOffset) {
        builder.addFieldOffset(5, maxFeeOffset, 0);
    }
    static createMaxFeeVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startMaxFeeVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addDeadline(builder, deadlineOffset) {
        builder.addFieldOffset(6, deadlineOffset, 0);
    }
    static createDeadlineVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startDeadlineVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addDownloadChannelId(builder, downloadChannelIdOffset) {
        builder.addFieldOffset(7, downloadChannelIdOffset, 0);
    }
    static createDownloadChannelIdVector(builder, data) {
        builder.startVector(1, data.length, 1);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt8(data[i]);
        }
        return builder.endVector();
    }
    static startDownloadChannelIdVector(builder, numElems) {
        builder.startVector(1, numElems, 1);
    }
    static addApprovalTrigger(builder, approvalTriggerOffset) {
        builder.addFieldOffset(8, approvalTriggerOffset, 0);
    }
    static createApprovalTriggerVector(builder, data) {
        builder.startVector(1, data.length, 1);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt8(data[i]);
        }
        return builder.endVector();
    }
    static startApprovalTriggerVector(builder, numElems) {
        builder.startVector(1, numElems, 1);
    }
    static addJudgingCount(builder, judgingCount) {
        builder.addFieldInt8(9, judgingCount, 0);
    }
    static addOverlappingCount(builder, overlappingCount) {
        builder.addFieldInt8(10, overlappingCount, 0);
    }
    static addJudgedCount(builder, judgedCount) {
        builder.addFieldInt8(11, judgedCount, 0);
    }
    static addOpinionElementCount(builder, opinionElementCount) {
        builder.addFieldInt8(12, opinionElementCount, 0);
    }
    static addPublicKeys(builder, publicKeysOffset) {
        builder.addFieldOffset(13, publicKeysOffset, 0);
    }
    static createPublicKeysVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addOffset(data[i]);
        }
        return builder.endVector();
    }
    static startPublicKeysVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addSignatures(builder, signaturesOffset) {
        builder.addFieldOffset(14, signaturesOffset, 0);
    }
    static createSignaturesVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addOffset(data[i]);
        }
        return builder.endVector();
    }
    static startSignaturesVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addPresentOpinions(builder, presentOpinionsOffset) {
        builder.addFieldOffset(15, presentOpinionsOffset, 0);
    }
    static createPresentOpinionsVector(builder, data) {
        builder.startVector(1, data.length, 1);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt8(data[i]);
        }
        return builder.endVector();
    }
    static startPresentOpinionsVector(builder, numElems) {
        builder.startVector(1, numElems, 1);
    }
    static addOpinions(builder, opinionsOffset) {
        builder.addFieldOffset(16, opinionsOffset, 0);
    }
    static createOpinionsVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addOffset(data[i]);
        }
        return builder.endVector();
    }
    static startOpinionsVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static endDownloadApprovalTransactionBuffer(builder) {
        const offset = builder.endObject();
        return offset;
    }
    static finishDownloadApprovalTransactionBufferBuffer(builder, offset) {
        builder.finish(offset);
    }
    static finishSizePrefixedDownloadApprovalTransactionBufferBuffer(builder, offset) {
        builder.finish(offset, undefined, true);
    }
    static createDownloadApprovalTransactionBuffer(builder, size, signatureOffset, signerOffset, version, type, maxFeeOffset, deadlineOffset, downloadChannelIdOffset, approvalTriggerOffset, judgingCount, overlappingCount, judgedCount, opinionElementCount, publicKeysOffset, signaturesOffset, presentOpinionsOffset, opinionsOffset) {
        DownloadApprovalTransactionBuffer.startDownloadApprovalTransactionBuffer(builder);
        DownloadApprovalTransactionBuffer.addSize(builder, size);
        DownloadApprovalTransactionBuffer.addSignature(builder, signatureOffset);
        DownloadApprovalTransactionBuffer.addSigner(builder, signerOffset);
        DownloadApprovalTransactionBuffer.addVersion(builder, version);
        DownloadApprovalTransactionBuffer.addType(builder, type);
        DownloadApprovalTransactionBuffer.addMaxFee(builder, maxFeeOffset);
        DownloadApprovalTransactionBuffer.addDeadline(builder, deadlineOffset);
        DownloadApprovalTransactionBuffer.addDownloadChannelId(builder, downloadChannelIdOffset);
        DownloadApprovalTransactionBuffer.addApprovalTrigger(builder, approvalTriggerOffset);
        DownloadApprovalTransactionBuffer.addJudgingCount(builder, judgingCount);
        DownloadApprovalTransactionBuffer.addOverlappingCount(builder, overlappingCount);
        DownloadApprovalTransactionBuffer.addJudgedCount(builder, judgedCount);
        DownloadApprovalTransactionBuffer.addOpinionElementCount(builder, opinionElementCount);
        DownloadApprovalTransactionBuffer.addPublicKeys(builder, publicKeysOffset);
        DownloadApprovalTransactionBuffer.addSignatures(builder, signaturesOffset);
        DownloadApprovalTransactionBuffer.addPresentOpinions(builder, presentOpinionsOffset);
        DownloadApprovalTransactionBuffer.addOpinions(builder, opinionsOffset);
        return DownloadApprovalTransactionBuffer.endDownloadApprovalTransactionBuffer(builder);
    }
}
exports.DownloadApprovalTransactionBuffer = DownloadApprovalTransactionBuffer;
