// tsjs-xpx-chain-sdk/dist/src/infrastructure/buffers/storage/EndDriveVerificationTransactionV2Buffer.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
Object.defineProperty(exports, "__esModule", { value: true });
exports.EndDriveVerificationTransactionV2Buffer = void 0;
const flatbuffers = require("flatbuffers");
const KeysBuffer_1 = require("./KeysBuffer");
const SignaturesBuffer_1 = require("./SignaturesBuffer");
class EndDriveVerificationTransactionV2Buffer {
    bb = null;
    bb_pos = 0;
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsEndDriveVerificationTransactionV2Buffer(bb, obj) {
        return (obj || new EndDriveVerificationTransactionV2Buffer()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsEndDriveVerificationTransactionV2Buffer(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new EndDriveVerificationTransactionV2Buffer()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    size() {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    signature(index) {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? this.bb.readUint8(this.bb.__vector(this.bb_pos + offset) + index) : 0;
    }
    signatureLength() {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    signatureArray() {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? new Uint8Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    signer(index) {
        const offset = this.bb.__offset(this.bb_pos, 8);
        return offset ? this.bb.readUint8(this.bb.__vector(this.bb_pos + offset) + index) : 0;
    }
    signerLength() {
        const offset = this.bb.__offset(this.bb_pos, 8);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    signerArray() {
        const offset = this.bb.__offset(this.bb_pos, 8);
        return offset ? new Uint8Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    version() {
        const offset = this.bb.__offset(this.bb_pos, 10);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    type() {
        const offset = this.bb.__offset(this.bb_pos, 12);
        return offset ? this.bb.readUint16(this.bb_pos + offset) : 0;
    }
    maxFee(index) {
        const offset = this.bb.__offset(this.bb_pos, 14);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    maxFeeLength() {
        const offset = this.bb.__offset(this.bb_pos, 14);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    maxFeeArray() {
        const offset = this.bb.__offset(this.bb_pos, 14);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    deadline(index) {
        const offset = this.bb.__offset(this.bb_pos, 16);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    deadlineLength() {
        const offset = this.bb.__offset(this.bb_pos, 16);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    deadlineArray() {
        const offset = this.bb.__offset(this.bb_pos, 16);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    driveKey(index) {
        const offset = this.bb.__offset(this.bb_pos, 18);
        return offset ? this.bb.readUint8(this.bb.__vector(this.bb_pos + offset) + index) : 0;
    }
    driveKeyLength() {
        const offset = this.bb.__offset(this.bb_pos, 18);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    driveKeyArray() {
        const offset = this.bb.__offset(this.bb_pos, 18);
        return offset ? new Uint8Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    verificationTrigger(index) {
        const offset = this.bb.__offset(this.bb_pos, 20);
        return offset ? this.bb.readUint8(this.bb.__vector(this.bb_pos + offset) + index) : 0;
    }
    verificationTriggerLength() {
        const offset = this.bb.__offset(this.bb_pos, 20);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    verificationTriggerArray() {
        const offset = this.bb.__offset(this.bb_pos, 20);
        return offset ? new Uint8Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    shardId(index) {
        const offset = this.bb.__offset(this.bb_pos, 22);
        return offset ? this.bb.readUint8(this.bb.__vector(this.bb_pos + offset) + index) : 0;
    }
    shardIdLength() {
        const offset = this.bb.__offset(this.bb_pos, 22);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    shardIdArray() {
        const offset = this.bb.__offset(this.bb_pos, 22);
        return offset ? new Uint8Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    keyCount() {
        const offset = this.bb.__offset(this.bb_pos, 24);
        return offset ? this.bb.readUint8(this.bb_pos + offset) : 0;
    }
    judgingKeyCount() {
        const offset = this.bb.__offset(this.bb_pos, 26);
        return offset ? this.bb.readUint8(this.bb_pos + offset) : 0;
    }
    keys(index, obj) {
        const offset = this.bb.__offset(this.bb_pos, 28);
        return offset ? (obj || new KeysBuffer_1.KeysBuffer()).__init(this.bb.__indirect(this.bb.__vector(this.bb_pos + offset) + index * 4), this.bb) : null;
    }
    keysLength() {
        const offset = this.bb.__offset(this.bb_pos, 28);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    signatures(index, obj) {
        const offset = this.bb.__offset(this.bb_pos, 30);
        return offset ? (obj || new SignaturesBuffer_1.SignaturesBuffer()).__init(this.bb.__indirect(this.bb.__vector(this.bb_pos + offset) + index * 4), this.bb) : null;
    }
    signaturesLength() {
        const offset = this.bb.__offset(this.bb_pos, 30);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    opinions() {
        const offset = this.bb.__offset(this.bb_pos, 32);
        return offset ? this.bb.readUint8(this.bb_pos + offset) : 0;
    }
    static startEndDriveVerificationTransactionV2Buffer(builder) {
        builder.startObject(15);
    }
    static addSize(builder, size) {
        builder.addFieldInt32(0, size, 0);
    }
    static addSignature(builder, signatureOffset) {
        builder.addFieldOffset(1, signatureOffset, 0);
    }
    static createSignatureVector(builder, data) {
        builder.startVector(1, data.length, 1);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt8(data[i]);
        }
        return builder.endVector();
    }
    static startSignatureVector(builder, numElems) {
        builder.startVector(1, numElems, 1);
    }
    static addSigner(builder, signerOffset) {
        builder.addFieldOffset(2, signerOffset, 0);
    }
    static createSignerVector(builder, data) {
        builder.startVector(1, data.length, 1);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt8(data[i]);
        }
        return builder.endVector();
    }
    static startSignerVector(builder, numElems) {
        builder.startVector(1, numElems, 1);
    }
    static addVersion(builder, version) {
        builder.addFieldInt32(3, version, 0);
    }
    static addType(builder, type) {
        builder.addFieldInt16(4, type, 0);
    }
    static addMaxFee(builder, maxFeeOffset) {
        builder.addFieldOffset(5, maxFeeOffset, 0);
    }
    static createMaxFeeVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startMaxFeeVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addDeadline(builder, deadlineOffset) {
        builder.addFieldOffset(6, deadlineOffset, 0);
    }
    static createDeadlineVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startDeadlineVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addDriveKey(builder, driveKeyOffset) {
        builder.addFieldOffset(7, driveKeyOffset, 0);
    }
    static createDriveKeyVector(builder, data) {
        builder.startVector(1, data.length, 1);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt8(data[i]);
        }
        return builder.endVector();
    }
    static startDriveKeyVector(builder, numElems) {
        builder.startVector(1, numElems, 1);
    }
    static addVerificationTrigger(builder, verificationTriggerOffset) {
        builder.addFieldOffset(8, verificationTriggerOffset, 0);
    }
    static createVerificationTriggerVector(builder, data) {
        builder.startVector(1, data.length, 1);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt8(data[i]);
        }
        return builder.endVector();
    }
    static startVerificationTriggerVector(builder, numElems) {
        builder.startVector(1, numElems, 1);
    }
    static addShardId(builder, shardIdOffset) {
        builder.addFieldOffset(9, shardIdOffset, 0);
    }
    static createShardIdVector(builder, data) {
        builder.startVector(1, data.length, 1);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt8(data[i]);
        }
        return builder.endVector();
    }
    static startShardIdVector(builder, numElems) {
        builder.startVector(1, numElems, 1);
    }
    static addKeyCount(builder, keyCount) {
        builder.addFieldInt8(10, keyCount, 0);
    }
    static addJudgingKeyCount(builder, judgingKeyCount) {
        builder.addFieldInt8(11, judgingKeyCount, 0);
    }
    static addKeys(builder, keysOffset) {
        builder.addFieldOffset(12, keysOffset, 0);
    }
    static createKeysVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addOffset(data[i]);
        }
        return builder.endVector();
    }
    static startKeysVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addSignatures(builder, signaturesOffset) {
        builder.addFieldOffset(13, signaturesOffset, 0);
    }
    static createSignaturesVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addOffset(data[i]);
        }
        return builder.endVector();
    }
    static startSignaturesVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addOpinions(builder, opinions) {
        builder.addFieldInt8(14, opinions, 0);
    }
    static endEndDriveVerificationTransactionV2Buffer(builder) {
        const offset = builder.endObject();
        return offset;
    }
    static createEndDriveVerificationTransactionV2Buffer(builder, size, signatureOffset, signerOffset, version, type, maxFeeOffset, deadlineOffset, driveKeyOffset, verificationTriggerOffset, shardIdOffset, keyCount, judgingKeyCount, keysOffset, signaturesOffset, opinions) {
        EndDriveVerificationTransactionV2Buffer.startEndDriveVerificationTransactionV2Buffer(builder);
        EndDriveVerificationTransactionV2Buffer.addSize(builder, size);
        EndDriveVerificationTransactionV2Buffer.addSignature(builder, signatureOffset);
        EndDriveVerificationTransactionV2Buffer.addSigner(builder, signerOffset);
        EndDriveVerificationTransactionV2Buffer.addVersion(builder, version);
        EndDriveVerificationTransactionV2Buffer.addType(builder, type);
        EndDriveVerificationTransactionV2Buffer.addMaxFee(builder, maxFeeOffset);
        EndDriveVerificationTransactionV2Buffer.addDeadline(builder, deadlineOffset);
        EndDriveVerificationTransactionV2Buffer.addDriveKey(builder, driveKeyOffset);
        EndDriveVerificationTransactionV2Buffer.addVerificationTrigger(builder, verificationTriggerOffset);
        EndDriveVerificationTransactionV2Buffer.addShardId(builder, shardIdOffset);
        EndDriveVerificationTransactionV2Buffer.addKeyCount(builder, keyCount);
        EndDriveVerificationTransactionV2Buffer.addJudgingKeyCount(builder, judgingKeyCount);
        EndDriveVerificationTransactionV2Buffer.addKeys(builder, keysOffset);
        EndDriveVerificationTransactionV2Buffer.addSignatures(builder, signaturesOffset);
        EndDriveVerificationTransactionV2Buffer.addOpinions(builder, opinions);
        return EndDriveVerificationTransactionV2Buffer.endEndDriveVerificationTransactionV2Buffer(builder);
    }
}
exports.EndDriveVerificationTransactionV2Buffer = EndDriveVerificationTransactionV2Buffer;
