// tsjs-xpx-chain-sdk/dist/src/infrastructure/buffers/storage/KeysBuffer.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeysBuffer = void 0;
const flatbuffers = require("flatbuffers");
class KeysBuffer {
    bb = null;
    bb_pos = 0;
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsKeysBuffer(bb, obj) {
        return (obj || new KeysBuffer()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsKeysBuffer(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new KeysBuffer()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    key(index) {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.readUint8(this.bb.__vector(this.bb_pos + offset) + index) : 0;
    }
    keyLength() {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    keyArray() {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? new Uint8Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    static startKeysBuffer(builder) {
        builder.startObject(1);
    }
    static addKey(builder, keyOffset) {
        builder.addFieldOffset(0, keyOffset, 0);
    }
    static createKeyVector(builder, data) {
        builder.startVector(1, data.length, 1);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt8(data[i]);
        }
        return builder.endVector();
    }
    static startKeyVector(builder, numElems) {
        builder.startVector(1, numElems, 1);
    }
    static endKeysBuffer(builder) {
        const offset = builder.endObject();
        return offset;
    }
    static createKeysBuffer(builder, keyOffset) {
        KeysBuffer.startKeysBuffer(builder);
        KeysBuffer.addKey(builder, keyOffset);
        return KeysBuffer.endKeysBuffer(builder);
    }
}
exports.KeysBuffer = KeysBuffer;
