// tsjs-xpx-chain-sdk/dist/src/infrastructure/buffers/storage/OpinionsBuffer.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpinionsBuffer = void 0;
const flatbuffers = require("flatbuffers");
class OpinionsBuffer {
    bb = null;
    bb_pos = 0;
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsOpinionsBuffer(bb, obj) {
        return (obj || new OpinionsBuffer()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsOpinionsBuffer(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new OpinionsBuffer()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    opinion(index) {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    opinionLength() {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    opinionArray() {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    static startOpinionsBuffer(builder) {
        builder.startObject(1);
    }
    static addOpinion(builder, opinionOffset) {
        builder.addFieldOffset(0, opinionOffset, 0);
    }
    static createOpinionVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startOpinionVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static endOpinionsBuffer(builder) {
        const offset = builder.endObject();
        return offset;
    }
    static createOpinionsBuffer(builder, opinionOffset) {
        OpinionsBuffer.startOpinionsBuffer(builder);
        OpinionsBuffer.addOpinion(builder, opinionOffset);
        return OpinionsBuffer.endOpinionsBuffer(builder);
    }
}
exports.OpinionsBuffer = OpinionsBuffer;
