// tsjs-xpx-chain-sdk/dist/src/infrastructure/buffers/storage/SignaturesBuffer.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignaturesBuffer = void 0;
const flatbuffers = require("flatbuffers");
class SignaturesBuffer {
    bb = null;
    bb_pos = 0;
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsSignaturesBuffer(bb, obj) {
        return (obj || new SignaturesBuffer()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsSignaturesBuffer(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new SignaturesBuffer()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    signature(index) {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.readUint8(this.bb.__vector(this.bb_pos + offset) + index) : 0;
    }
    signatureLength() {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    signatureArray() {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? new Uint8Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    static startSignaturesBuffer(builder) {
        builder.startObject(1);
    }
    static addSignature(builder, signatureOffset) {
        builder.addFieldOffset(0, signatureOffset, 0);
    }
    static createSignatureVector(builder, data) {
        builder.startVector(1, data.length, 1);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt8(data[i]);
        }
        return builder.endVector();
    }
    static startSignatureVector(builder, numElems) {
        builder.startVector(1, numElems, 1);
    }
    static endSignaturesBuffer(builder) {
        const offset = builder.endObject();
        return offset;
    }
    static createSignaturesBuffer(builder, signatureOffset) {
        SignaturesBuffer.startSignaturesBuffer(builder);
        SignaturesBuffer.addSignature(builder, signatureOffset);
        return SignaturesBuffer.endSignaturesBuffer(builder);
    }
}
exports.SignaturesBuffer = SignaturesBuffer;
