// tsjs-xpx-chain-sdk/dist/src/model/mosaic/MosaicProperties.js
"use strict";
/*
 * Copyright 2023 ProximaX
 * Copyright 2018 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MosaicProperties = void 0;
const UInt64_1 = require("../UInt64");
const MosaicPropertyType_1 = require("./MosaicPropertyType");
/**
 * Mosaic properties model
 */
class MosaicProperties {
    divisibility;
    duration;
    /**
     * The creator can choose between a definition that allows a mosaic supply change at a later point or an immutable supply.
     * Allowed values for the property are "true" and "false". The default value is "false".
     */
    supplyMutable;
    /**
     * The creator can choose if the mosaic definition should allow for transfers of the mosaic among accounts other than the creator.
     * If the property 'transferable' is set to "false", only transfer transactions
     * having the creator as sender or as recipient can transfer mosaics of that type.
     * If set to "true" the mosaics can be transferred to and from arbitrary accounts.
     * Allowed values for the property are thus "true" and "false". The default value is "true".
     */
    transferable;
    /**
     * @param flags
     * @param divisibility
     * @param duration
     */
    constructor(flags, 
    /**
     * The divisibility determines up to what decimal place the mosaic can be divided into.
     * Thus a divisibility of 3 means that a mosaic can be divided into smallest parts of 0.001 mosaics
     * i.e. milli mosaics is the smallest sub-unit.
     * When transferring mosaics via a transfer transaction the quantity transferred
     * is given in multiples of those smallest parts.
     * The divisibility must be in the range of 0 and 6. The default value is "0".
     */
    divisibility, 
    /**
     * The duration in blocks a mosaic will be available.
     * After the duration finishes mosaic is inactive and can be renewed.
     * Duration is optional when defining the mosaic
     */
    duration) {
        this.divisibility = divisibility;
        this.duration = duration;
        let binaryFlags = '00' + (flags.lower >>> 0).toString(2);
        binaryFlags = binaryFlags.slice(-2);
        this.supplyMutable = binaryFlags[1] === '1';
        this.transferable = binaryFlags[0] === '1';
    }
    /**
     * Static constructor function with default parameters
     * @returns {MosaicProperties}
     * @param params
     */
    static create(params) {
        const flags = (params.supplyMutable ? 1 : 0) + (params.transferable ? 2 : 0);
        return new MosaicProperties(UInt64_1.UInt64.fromUint(flags), params.divisibility, params.duration);
    }
    /**
     * Create DTO object
     */
    toDTO() {
        const dto = [
            { id: MosaicPropertyType_1.MosaicPropertyType.MosaicFlags, value: UInt64_1.UInt64.fromUint((this.supplyMutable ? 1 : 0) +
                    (this.transferable ? 2 : 0)).toDTO() },
            { id: MosaicPropertyType_1.MosaicPropertyType.Divisibility, value: UInt64_1.UInt64.fromUint(this.divisibility).toDTO() },
        ];
        if (this.duration !== undefined) {
            dto.push({ id: MosaicPropertyType_1.MosaicPropertyType.Duration, value: this.duration.toDTO() });
        }
        return dto;
    }
}
exports.MosaicProperties = MosaicProperties;
