// tsjs-xpx-chain-sdk/dist/src/service/MosaicService.js
"use strict";
/*
 * Copyright 2023 ProximaX
 * Copyright 2018 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MosaicService = void 0;
const rxjs_1 = require("rxjs");
const operators_1 = require("rxjs/operators");
const MosaicAmountView_1 = require("./MosaicAmountView");
const MosaicView_1 = require("./MosaicView");
/**
 * Mosaic service
 */
class MosaicService {
    accountHttp;
    mosaicHttp;
    chainHttp;
    /**
     * Constructor
     * @param accountHttp
     * @param mosaicHttp
     */
    constructor(accountHttp, mosaicHttp, chainHttp) {
        this.accountHttp = accountHttp;
        this.mosaicHttp = mosaicHttp;
        this.chainHttp = chainHttp;
    }
    /**
     * Get mosaic view given mosaicIds
     * @param mosaicIds - The ids of the mosaics
     * @returns {Observable<MosaicView[]>}
     */
    mosaicsView(mosaicIds) {
        return (0, rxjs_1.of)(mosaicIds).pipe((0, operators_1.mergeMap)((_) => this.mosaicHttp.getMosaics(mosaicIds).pipe((0, operators_1.mergeMap)((_) => _), (0, operators_1.map)((mosaicInfo) => {
            return new MosaicView_1.MosaicView(mosaicInfo);
        }), (0, operators_1.toArray)())));
    }
    /**
     * Get mosaic amount view given mosaic array
     * @param mosaics
     * @returns {Observable<MosaicAmountView[]>}
     */
    mosaicsAmountView(mosaics) {
        return (0, rxjs_1.of)(mosaics).pipe((0, operators_1.mergeMap)((_) => _), (0, operators_1.mergeMap)((mosaic) => this.mosaicsView([mosaic.id]).pipe((0, operators_1.filter)((_) => _.length !== 0), (0, operators_1.map)((mosaicViews) => {
            return new MosaicAmountView_1.MosaicAmountView(mosaicViews[0].mosaicInfo, mosaic.amount);
        }), (0, operators_1.toArray)())));
    }
    /**
     * Get balance mosaics in form of MosaicAmountViews for a given account address
     * @param address - Account address
     * @returns {Observable<MosaicAmountView[]>}
     */
    mosaicsAmountViewFromAddress(address) {
        return (0, rxjs_1.of)(address).pipe((0, operators_1.mergeMap)((_) => this.accountHttp.getAccountInfo(_)), (0, operators_1.mergeMap)((_) => this.mosaicsAmountView(_.mosaics)));
    }
    /**
     * Get mosaic remaining blocks for a given mosaicId
     * @param mosaicId - MosaicId
     * @returns {Observable<number>}
     */
    mosaicRemainingBlock(mosaicId) {
        return (0, rxjs_1.of)(mosaicId).pipe((0, operators_1.mergeMap)((_) => this.mosaicHttp.getMosaic(_)), (0, operators_1.mergeMap)((_) => _.duration && _.duration.compact() !== 0 ?
            this.chainHttp.getBlockchainHeight()
                .pipe((0, operators_1.map)((chainHeight) => (_.duration.compact() + _.height.compact()) - chainHeight.compact())) : (0, rxjs_1.of)(Number.POSITIVE_INFINITY)));
    }
}
exports.MosaicService = MosaicService;
