// tsjs-xpx-chain-sdk/dist/src/service/NamespaceService.js
"use strict";
/*
 * Copyright 2023 ProximaX
 * Copyright 2018 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NamespaceService = void 0;
const rxjs_1 = require("rxjs");
const operators_1 = require("rxjs/operators");
/**
 * Namespace service
 */
class NamespaceService {
    namespaceHttp;
    chainHttp;
    /**
     * Constructor
     * @param namespaceHttp
     */
    constructor(namespaceHttp, chainHttp) {
        this.namespaceHttp = namespaceHttp;
        this.chainHttp = chainHttp;
    }
    /**
     * Get namespace info and name from namespace Id
     * @param id
     * @returns {Observable<Namespace>}
     */
    namespace(id) {
        return this.namespaceHttp.getNamespace(id).pipe((0, operators_1.mergeMap)((namespaceInfo) => this.namespaceHttp
            .getNamespacesName(namespaceInfo.levels).pipe((0, operators_1.map)((names) => Object.assign({ __proto__: Object.getPrototypeOf(namespaceInfo) }, namespaceInfo, { name: this.extractFullNamespace(namespaceInfo, names) })))));
    }
    extractFullNamespace(namespace, namespaceNames) {
        return namespace.levels.map((level) => {
            const namespaceName = namespaceNames.find((name) => name.namespaceId.equals(level));
            if (namespace === undefined) {
                throw new Error('Not found');
            }
            return namespaceName;
        })
            .map((namespaceName) => namespaceName.name)
            .join('.');
    }
    /**
     * Get namespace remaining blocks for a given namespaceId
     * @param namespaceId - NamespaceId
     * @returns {Observable<number>}
     */
    namespaceRemainingBlock(namespaceId) {
        return (0, rxjs_1.of)(namespaceId).pipe((0, operators_1.mergeMap)((_) => this.namespaceHttp.getNamespace(_)), (0, operators_1.mergeMap)((nsInfo) => {
            if (nsInfo.endHeight.lower === 0xFFFFFFFF && nsInfo.endHeight.higher === 0xFFFFFFFF) {
                return (0, rxjs_1.of)(Number.POSITIVE_INFINITY);
            }
            else {
                return this.chainHttp.getBlockchainHeight()
                    .pipe((0, operators_1.map)((chainHeight) => nsInfo.endHeight.compact() - chainHeight.compact()));
            }
        }));
    }
}
exports.NamespaceService = NamespaceService;
