// tsjs-xpx-chain-sdk/dist/src/service/NftService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NftService = void 0;
const model_1 = require("../model/model");
const format_1 = require("../core/format");
const infrastructure_1 = require("../infrastructure/infrastructure");
const rxjs_1 = require("rxjs");
const Utilities_1 = require("../model/transaction/Utilities");
class NftService {
    txnBuilder;
    bcApiUrl;
    networkType = 0;
    maxDurationBlock = 0;
    maxBondedTransactionLifetimeInSec = 0;
    nftScopeKey = model_1.UInt64.fromHex(format_1.Convert.utf8ToHex('nft.json'));
    networkHttp;
    constructor(txnBuilder, bcApiUrl) {
        this.txnBuilder = txnBuilder;
        this.bcApiUrl = bcApiUrl;
        this.networkHttp = new infrastructure_1.NetworkHttp(bcApiUrl);
    }
    async getNftByCreatorPublicKey(creatorPublicKey, assetId) {
        let compositeHash = (0, Utilities_1.genMetadataCompositeHash)(model_1.PublicAccount.createFromPublicKey(creatorPublicKey, this.networkType), this.nftScopeKey, model_1.MetadataType.MOSAIC, assetId);
        let assetIdObj;
        if (assetId instanceof model_1.UInt64) {
            assetIdObj = new model_1.MosaicId(assetId.toHex());
        }
        else {
            assetIdObj = assetId;
        }
        let metadata = "";
        let currentOwner = "";
        let currentOwnerAddress = "";
        try {
            const metadataHttp = new infrastructure_1.MetadataHttp(this.bcApiUrl, this.networkHttp);
            const assetHttp = new infrastructure_1.MosaicHttp(this.bcApiUrl, this.networkHttp);
            let metadataEntry = await (0, rxjs_1.firstValueFrom)(metadataHttp.getMetadata(compositeHash));
            let assetRichlist = await (0, rxjs_1.firstValueFrom)(assetHttp.getMosaicRichlist(assetIdObj));
            metadata = metadataEntry.value;
            currentOwner = assetRichlist[0].publicKey;
            currentOwnerAddress = assetRichlist[0].address.pretty();
        }
        catch (error) {
            throw new Error("NFT data not found");
        }
        try {
            return {
                metadata: JSON.parse(metadata),
                currentOwner: currentOwner,
                currentOwnerAddress: currentOwnerAddress
            };
        }
        catch (error) {
            return {
                metadata: JSON.parse(metadata),
                currentOwner: currentOwner,
                currentOwnerAddress: currentOwnerAddress
            };
        }
    }
    async getNftList(filters) {
        const metadataHttp = new infrastructure_1.MetadataHttp(this.bcApiUrl, this.networkHttp);
        let metadataQP = new infrastructure_1.MetadataQueryParams();
        metadataQP.metadataType = model_1.MetadataType.MOSAIC;
        metadataQP.scopedMetadataKey = this.nftScopeKey;
        if (filters.creatorPublicKey) {
            metadataQP.targetKey = filters.creatorPublicKey;
        }
        if (filters.pageSize) {
            metadataQP.pageSize = filters.pageSize;
        }
        if (filters.pageNumber) {
            metadataQP.pageNumber = filters.pageNumber;
        }
        try {
            const metadataSearch = await (0, rxjs_1.firstValueFrom)(metadataHttp.searchMetadata(metadataQP));
            return metadataSearch;
        }
        catch (error) {
            throw new Error("Error retrieving NFT entry");
        }
    }
    async getNftByOwner(ownerPublicKey, part = 1, chunkSize = 50) {
        const accountHttp = new infrastructure_1.AccountHttp(this.bcApiUrl, this.networkHttp);
        const publicAcc = model_1.PublicAccount.createFromPublicKey(ownerPublicKey, await this.getNetwork());
        try {
            const accInfo = await (0, rxjs_1.firstValueFrom)(accountHttp.getAccountInfo(publicAcc.address));
            let nftCandidates = [];
            for (const asset of accInfo.mosaics) {
                if (asset.amount.toBigInt() > BigInt(1)) {
                    continue;
                }
                nftCandidates.push(asset.id);
            }
            let chunkedNftCandidates = nftCandidates.slice((part - 1) * chunkSize, part * chunkSize);
            let nftInfo = await this.findNft(chunkedNftCandidates);
            return nftInfo;
        }
        catch (error) {
            throw new Error("Error retrieving NFT entry");
        }
    }
    async findNft(nftCandidates) {
        let nfts = [];
        let assetHttp = new infrastructure_1.MosaicHttp(this.bcApiUrl, this.networkHttp);
        if (nftCandidates.length > 80) {
            throw new Error("SDAs to check are greater than recommended size, maximum are 80");
        }
        let assetsInfo = await (0, rxjs_1.firstValueFrom)(assetHttp.getMosaics(nftCandidates));
        await this.getNetwork();
        let nftAssets = assetsInfo.filter(x => !x.isSupplyMutable() && x.divisibility === 0)
            .map(x => {
            return {
                assetId: x.mosaicId,
                compositeHash: (0, Utilities_1.genMetadataCompositeHash)(model_1.PublicAccount.createFromPublicKey(x.owner.publicKey, this.networkType), this.nftScopeKey, model_1.MetadataType.MOSAIC, x.mosaicId)
            };
        });
        let metadataHttp = new infrastructure_1.MetadataHttp(this.bcApiUrl, this.networkHttp);
        let metadataEntries = await (0, rxjs_1.firstValueFrom)(metadataHttp.getMetadatas(nftAssets.map(x => x.compositeHash)));
        for (const nftMetadata of metadataEntries) {
            let nftData;
            try {
                nftData = {
                    metadata: JSON.parse(nftMetadata.value),
                    creator: nftMetadata.targetKey,
                    creatorAddress: nftMetadata.sourceAddress.pretty()
                };
            }
            catch (error) {
                nftData = {
                    metadata: nftMetadata.value,
                    creator: nftMetadata.targetKey,
                    creatorAddress: nftMetadata.sourceAddress.pretty()
                };
            }
            nfts.push({
                assetId: nftMetadata.targetId.toHex(),
                data: nftData
            });
        }
        return nfts;
    }
    async getNftData(assetId) {
        let assetIdObj;
        if (typeof assetId === 'string') {
            assetIdObj = new model_1.MosaicId(assetId);
        }
        else if (assetId instanceof model_1.UInt64) {
            assetIdObj = new model_1.MosaicId(assetId.toHex());
        }
        else {
            assetIdObj = assetId;
        }
        const metadataHttp = new infrastructure_1.MetadataHttp(this.bcApiUrl, this.networkHttp);
        const assetHttp = new infrastructure_1.MosaicHttp(this.bcApiUrl, this.networkHttp);
        let metadata = "";
        let creator = "";
        let creatorAddress = "";
        let currentOwner = "";
        let currentOwnerAddress = "";
        try {
            let metadataQP = new infrastructure_1.MetadataQueryParams();
            metadataQP.targetId = assetIdObj;
            metadataQP.metadataType = model_1.MetadataType.MOSAIC;
            metadataQP.scopedMetadataKey = this.nftScopeKey;
            let metadataSearch = await (0, rxjs_1.firstValueFrom)(metadataHttp.searchMetadata(metadataQP));
            let assetRichlist = await (0, rxjs_1.firstValueFrom)(assetHttp.getMosaicRichlist(assetIdObj));
            if (metadataSearch.metadataEntries.length === 0) {
                // throw new Error("No metadata found");
                return null;
            }
            metadata = metadataSearch.metadataEntries[0].value;
            creator = metadataSearch.metadataEntries[0].targetKey;
            creatorAddress = metadataSearch.metadataEntries[0].sourceAddress.pretty();
            currentOwner = assetRichlist[0].publicKey;
            currentOwnerAddress = assetRichlist[0].address.pretty();
        }
        catch (error) {
            throw new Error("Failed to get NFT metadata");
        }
        try {
            return {
                metadata: JSON.parse(metadata),
                creator: creator,
                creatorAddress: creatorAddress,
                currentOwner: currentOwner,
                currentOwnerAddress: currentOwnerAddress
            };
        }
        catch (error) {
            return {
                metadata: metadata,
                creator: creator,
                creatorAddress: creatorAddress,
                currentOwner: currentOwner,
                currentOwnerAddress: currentOwnerAddress
            };
        }
    }
    async createAndAnnounce(owner, metadata) {
        let signedTxn = await this.createSigned(owner, metadata);
        let txnHttp = new infrastructure_1.TransactionHttp(this.bcApiUrl);
        try {
            await (0, rxjs_1.firstValueFrom)(txnHttp.announce(signedTxn.txn));
            return {
                assetId: signedTxn.assetId,
                txnHash: signedTxn.txn.hash
            };
        }
        catch (error) {
            throw new Error("Failed to announce transaction");
        }
    }
    async createSigned(owner, metadata) {
        let nftTxn = await this.createUnsigned(owner.publicAccount, metadata);
        let signedTxn = owner.sign(nftTxn.txn, this.txnBuilder.generationHash);
        return {
            assetId: nftTxn.assetId,
            txn: signedTxn
        };
    }
    async createUnsigned(owner, metadata, deadline, useBonded = false) {
        let aggregateTxn = useBonded ? this.txnBuilder.aggregateBondedV1() :
            this.txnBuilder.aggregateCompleteV1();
        if (deadline) {
            aggregateTxn = aggregateTxn.deadline(deadline);
        }
        let mosaicNonce = model_1.MosaicNonce.createRandom();
        let mosaicId = model_1.MosaicId.createFromNonce(mosaicNonce, owner);
        let assetId = mosaicId.toHex();
        let assetCheckPassed = false;
        const assetHttp = new infrastructure_1.MosaicHttp(this.bcApiUrl, this.networkHttp);
        while (!assetCheckPassed) {
            try {
                let mosaicInfo = await (0, rxjs_1.firstValueFrom)(assetHttp.getMosaic(mosaicId));
                mosaicId = model_1.MosaicId.createFromNonce(mosaicNonce, owner);
                assetId = mosaicId.toHex();
            }
            catch (error) {
                assetCheckPassed = true;
            }
        }
        // const ipfsCid = await this.uploadIPFS(uploadedFile);
        // create mosaic
        const assetDefinitionTxn = this.txnBuilder.mosaicDefinition()
            .mosaicNonce(mosaicNonce)
            .mosaicId(mosaicId)
            .mosaicProperties(model_1.MosaicProperties.create({
            divisibility: 0,
            supplyMutable: false,
            transferable: true,
        }))
            .build();
        // set supply
        const assetSupplyChangeTxn = this.txnBuilder.mosaicSupplyChange()
            .direction(model_1.MosaicSupplyType.Increase)
            .delta(model_1.UInt64.fromUint(1))
            .mosaicId(mosaicId)
            .build();
        if (format_1.Convert.hexToUint8(format_1.Convert.utf8ToHex(JSON.stringify(metadata))).length > 1024) {
            throw new Error("Invalid size of metadata, maximum size of 1024");
        }
        // set meta data
        const assetMetadataTxn = this.txnBuilder.mosaicMetadata()
            .oldValue("")
            .value(JSON.stringify(metadata))
            .scopedMetadataKey(this.nftScopeKey)
            .targetMosaicId(mosaicId)
            .targetPublicKey(owner)
            .calculateDifferences()
            .build();
        const innerTxn = [
            assetDefinitionTxn.toAggregateV1(owner),
            assetSupplyChangeTxn.toAggregateV1(owner),
            assetMetadataTxn.toAggregateV1(owner),
        ];
        return {
            assetId: assetId,
            txn: aggregateTxn.innerTransactions(innerTxn).build()
        };
    }
    async creteMultisigNftAnnounce(myAcc, multisigAcc, metadata, nativeCurrency) {
        let resultTxn = await this.createMultisigNFTPayload(myAcc, multisigAcc, metadata, nativeCurrency);
        let txnHttp = new infrastructure_1.TransactionHttp(this.bcApiUrl);
        let chainHttp = new infrastructure_1.ChainHttp(this.bcApiUrl);
        if (resultTxn.hashLockTxn) {
            try {
                await (0, rxjs_1.firstValueFrom)(txnHttp.announce(resultTxn.hashLockTxn));
            }
            catch (error) {
                throw new Error("Failed to announce transaction");
            }
            let hashLockConfirmed = false;
            let confirmationBlock = BigInt(0);
            let attemptCount = 0;
            while (!hashLockConfirmed) {
                // wait 5 seconds
                await new Promise(resolve => setTimeout(resolve, 5000));
                attemptCount++;
                let hashLockTxnStatus = null;
                try {
                    hashLockTxnStatus = await (0, rxjs_1.firstValueFrom)(txnHttp.getTransactionStatus(resultTxn.hashLockTxn.hash));
                }
                catch (error) {
                }
                if (hashLockTxnStatus) {
                    if (hashLockTxnStatus.group === model_1.TransactionGroupType.CONFIRMED) {
                        hashLockConfirmed = true;
                        confirmationBlock = hashLockTxnStatus.height.toBigInt();
                        break;
                    }
                    else if (hashLockTxnStatus.group === model_1.TransactionGroupType.FAILED) {
                        throw new Error("HashLock transaction failed");
                    }
                }
                if (attemptCount > 5) {
                    throw new Error("HashLock transaction not confirmed");
                }
            }
            attemptCount = 0;
            let abtAnnounced = false;
            while (!abtAnnounced) {
                await new Promise(resolve => setTimeout(resolve, 5000));
                attemptCount++;
                let currentBlockHeight = null;
                try {
                    currentBlockHeight = await (0, rxjs_1.firstValueFrom)(chainHttp.getBlockchainHeight());
                }
                catch (error) {
                    throw new Error("Unable to check current block height");
                }
                if (currentBlockHeight.toBigInt() > (confirmationBlock + BigInt(1))) {
                    await (0, rxjs_1.firstValueFrom)(txnHttp.announceAggregateBonded(resultTxn.txn));
                    abtAnnounced = true;
                    break;
                }
                if (attemptCount > 8) {
                    throw new Error("Sirius Chain not generating new block");
                }
            }
        }
        else {
            try {
                await (0, rxjs_1.firstValueFrom)(txnHttp.announce(resultTxn.txn));
            }
            catch (error) {
                throw new Error("Failed to announce transaction");
            }
        }
        return {
            assetId: resultTxn.assetId,
            txnHash: resultTxn.txn.hash,
            hashLockTxnHash: resultTxn.hashLockTxn?.hash
        };
    }
    async createMultisigNFTPayload(myAcc, multisigAcc, metadata, nativeCurrency) {
        let usingMultisig;
        if (typeof multisigAcc === 'string') {
            usingMultisig = model_1.PublicAccount.createFromPublicKey(multisigAcc, await this.getNetwork(), 1);
        }
        else {
            usingMultisig = multisigAcc;
        }
        let networkConfig = "";
        let multisigAccGraph;
        try {
            const chainConfigHttp = new infrastructure_1.ChainConfigHttp(this.bcApiUrl);
            const chainHttp = new infrastructure_1.ChainHttp(this.bcApiUrl);
            const chainHeight = await (0, rxjs_1.firstValueFrom)(chainHttp.getBlockchainHeight());
            const chainConfig = await (0, rxjs_1.firstValueFrom)(chainConfigHttp.getChainConfig(Number(chainHeight.toBigInt())));
            networkConfig = chainConfig.networkConfig;
        }
        catch (error) {
            throw new Error("Unable to get network config");
        }
        try {
            const accountHttp = new infrastructure_1.AccountHttp(this.bcApiUrl, this.networkHttp);
            const multisigGraphInfo = await (0, rxjs_1.firstValueFrom)(accountHttp.getMultisigAccountGraphInfo(usingMultisig.address));
            multisigAccGraph = multisigGraphInfo;
        }
        catch (error) {
            throw new Error("Unable to get multisig account graph info");
        }
        const cosignerCheckResult = (0, Utilities_1.checkCosigner)(myAcc.publicKey, usingMultisig.publicKey, multisigAccGraph);
        let unsignedResult = await this.createUnsigned(usingMultisig, metadata);
        let signedTxn = myAcc.sign(unsignedResult.txn, this.txnBuilder.generationHash);
        let signedHashLockTxn = null;
        if (!cosignerCheckResult.aggComplete) {
            try {
                const nodeHttp = new infrastructure_1.NodeHttp(this.bcApiUrl);
                const nodeTime = await (0, rxjs_1.firstValueFrom)(nodeHttp.getNodeTime());
                const nodeTimeStamp = new model_1.UInt64(nodeTime.sendTimeStamp);
                const parsedConfig = (0, Utilities_1.parseConfig)(networkConfig);
                const maxBondedTransactionLifetime = (0, Utilities_1.getConfigSetting)(parsedConfig, ["plugin", "catapult.plugins.aggregate", "maxBondedTransactionLifetime"], Utilities_1.TimeUnit.SECOND);
                const maxHashLockDuration = (0, Utilities_1.getConfigSetting)(parsedConfig, ["plugin", "catapult.plugins.lockhash", "maxHashLockDuration"], Utilities_1.TimeUnit.SECOND);
                const lockedFundsPerAggregate = (0, Utilities_1.getConfigSetting)(parsedConfig, ["plugin", "catapult.plugins.lockhash", "lockedFundsPerAggregate"]);
                const blockGenerationTargetTime = (0, Utilities_1.getConfigSetting)(parsedConfig, ["chain", "blockGenerationTargetTime"], Utilities_1.TimeUnit.SECOND);
                const abtDeadline = (0, Utilities_1.getABTMaxSafeDeadlineWithSeconds)(maxBondedTransactionLifetime, maxHashLockDuration, nodeTimeStamp);
                const abtDuration = (0, Utilities_1.getABTMaxSafeDurationWithSeconds)(blockGenerationTargetTime, maxBondedTransactionLifetime, maxHashLockDuration);
                unsignedResult = await this.createUnsigned(usingMultisig, metadata, abtDeadline, true);
                signedTxn = myAcc.sign(unsignedResult.txn, this.txnBuilder.generationHash);
                let unsignedHashLock = await this.txnBuilder.hashLock()
                    .duration(model_1.UInt64.fromUint(abtDuration))
                    .mosaic(new model_1.Mosaic(nativeCurrency ?? new model_1.NamespaceId("prx.xpx"), model_1.UInt64.fromBigInt(BigInt(lockedFundsPerAggregate))))
                    .transactionHash(signedTxn)
                    .build();
                signedHashLockTxn = myAcc.sign(unsignedHashLock, this.txnBuilder.generationHash);
            }
            catch (error) {
                throw new Error("Unable to set Aggregate Bonded transaction");
            }
        }
        return {
            assetId: unsignedResult.assetId,
            txn: signedTxn,
            hashLockTxn: signedHashLockTxn
        };
    }
    async getNetwork() {
        if (this.networkType !== 0) {
            return this.networkType;
        }
        else {
            try {
                let chainNetworkType = await (0, rxjs_1.firstValueFrom)(this.networkHttp.getNetworkType());
                this.networkType = chainNetworkType;
            }
            catch (error) {
                throw new Error("Error getting network type");
            }
            return this.networkType;
        }
    }
}
exports.NftService = NftService;
