// tsjs-xpx-chain-sdk/dist/src/infrastructure/model/accountLinkTypeEnum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountLinkTypeEnum = void 0;
/**
* The account link types:
* 0 -  Unlinked. Account is not linked to another account.
* 1 -  Main. Account is a balance-holding account that is linked to a remote harvester account.
* 2 -  Remote. Account is a remote harvester account that is linked to a balance-holding account.
* 3 -  Remote_Unlinked. Account is a remote harvester eligible account that is unlinked.
* 4 -  Locked. Account that is locked after v2 upgrade transaction
*/
var AccountLinkTypeEnum;
(function (AccountLinkTypeEnum) {
    AccountLinkTypeEnum[AccountLinkTypeEnum["NUMBER_0"] = 0] = "NUMBER_0";
    AccountLinkTypeEnum[AccountLinkTypeEnum["NUMBER_1"] = 1] = "NUMBER_1";
    AccountLinkTypeEnum[AccountLinkTypeEnum["NUMBER_2"] = 2] = "NUMBER_2";
    AccountLinkTypeEnum[AccountLinkTypeEnum["NUMBER_3"] = 3] = "NUMBER_3";
    AccountLinkTypeEnum[AccountLinkTypeEnum["NUMBER_4"] = 4] = "NUMBER_4";
})(AccountLinkTypeEnum || (exports.AccountLinkTypeEnum = AccountLinkTypeEnum = {}));
