// tsjs-xpx-chain-sdk/dist/src/infrastructure/builders/storage/NewDriveClosureTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = void 0;
/**
 * @module transactions/DriveClosureTransaction
 */
const format_1 = require("../../../core/format");
const TransactionType_1 = require("../../../model/transaction/TransactionType");
const VerifiableTransaction_1 = require("../VerifiableTransaction");
const DriveClosureTransactionSchema_1 = require("../../schemas/storage/DriveClosureTransactionSchema");
const DriveClosureTransactionBuffer_1 = require("../../buffers/storage/DriveClosureTransactionBuffer");
const flatbuffers = require("flatbuffers");
class NewDriveClosureTransaction extends VerifiableTransaction_1.VerifiableTransaction {
    constructor(bytes) {
        super(bytes, DriveClosureTransactionSchema_1.default);
    }
}
exports.default = NewDriveClosureTransaction;
// tslint:disable-next-line:max-classes-per-file
class Builder {
    size;
    fee;
    version;
    type;
    deadline;
    driveKey;
    constructor() {
        this.fee = [0, 0];
        this.type = TransactionType_1.TransactionType.Drive_Closure;
    }
    addSize(size) {
        this.size = size;
        return this;
    }
    addMaxFee(fee) {
        this.fee = fee;
        return this;
    }
    addVersion(version) {
        this.version = version;
        return this;
    }
    addType(type) {
        this.type = type;
        return this;
    }
    addDeadline(deadline) {
        this.deadline = deadline;
        return this;
    }
    addDriveKey(driveKey) {
        this.driveKey = driveKey;
        return this;
    }
    build() {
        const builder = new flatbuffers.Builder(1);
        // Create vectors
        const signatureVector = DriveClosureTransactionBuffer_1.DriveClosureTransactionBuffer
            .createSignatureVector(builder, Array(...Array(64)).map(Number.prototype.valueOf, 0));
        const signerVector = DriveClosureTransactionBuffer_1.DriveClosureTransactionBuffer.createSignerVector(builder, Array(...Array(32)).map(Number.prototype.valueOf, 0));
        const deadlineVector = DriveClosureTransactionBuffer_1.DriveClosureTransactionBuffer.createDeadlineVector(builder, this.deadline);
        const feeVector = DriveClosureTransactionBuffer_1.DriveClosureTransactionBuffer.createMaxFeeVector(builder, this.fee);
        const driveKeyVector = DriveClosureTransactionBuffer_1.DriveClosureTransactionBuffer.createDriveKeyVector(builder, format_1.Convert.hexToUint8(this.driveKey));
        DriveClosureTransactionBuffer_1.DriveClosureTransactionBuffer.startDriveClosureTransactionBuffer(builder);
        DriveClosureTransactionBuffer_1.DriveClosureTransactionBuffer.addSize(builder, this.size);
        DriveClosureTransactionBuffer_1.DriveClosureTransactionBuffer.addSignature(builder, signatureVector);
        DriveClosureTransactionBuffer_1.DriveClosureTransactionBuffer.addSigner(builder, signerVector);
        DriveClosureTransactionBuffer_1.DriveClosureTransactionBuffer.addVersion(builder, this.version);
        DriveClosureTransactionBuffer_1.DriveClosureTransactionBuffer.addType(builder, this.type);
        DriveClosureTransactionBuffer_1.DriveClosureTransactionBuffer.addMaxFee(builder, feeVector);
        DriveClosureTransactionBuffer_1.DriveClosureTransactionBuffer.addDeadline(builder, deadlineVector);
        DriveClosureTransactionBuffer_1.DriveClosureTransactionBuffer.addDriveKey(builder, driveKeyVector);
        const codedTransfer = DriveClosureTransactionBuffer_1.DriveClosureTransactionBuffer.endDriveClosureTransactionBuffer(builder);
        builder.finish(codedTransfer);
        const bytes = builder.asUint8Array();
        return new NewDriveClosureTransaction(bytes);
    }
    static stringToUint8(stringToConvert) {
        return format_1.Convert.hexToUint8(format_1.Convert.utf8ToHex(stringToConvert));
    }
}
exports.Builder = Builder;
