// tsjs-xpx-chain-sdk/dist/src/infrastructure/builders/storage/NewDownloadPaymentTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = void 0;
/**
 * @module transactions/DownloadPaymentTransaction
 */
const format_1 = require("../../../core/format");
const TransactionType_1 = require("../../../model/transaction/TransactionType");
const VerifiableTransaction_1 = require("../VerifiableTransaction");
const DownloadPaymentTransactionSchema_1 = require("../../schemas/storage/DownloadPaymentTransactionSchema");
const DownloadPaymentTransactionBuffer_1 = require("../../buffers/storage/DownloadPaymentTransactionBuffer");
const flatbuffers = require("flatbuffers");
class NewDownloadPaymentTransaction extends VerifiableTransaction_1.VerifiableTransaction {
    constructor(bytes) {
        super(bytes, DownloadPaymentTransactionSchema_1.default);
    }
}
exports.default = NewDownloadPaymentTransaction;
// tslint:disable-next-line:max-classes-per-file
class Builder {
    size;
    fee;
    version;
    type;
    deadline;
    downloadChannelId;
    downloadSize;
    feedbackFeeAmount;
    constructor() {
        this.fee = [0, 0];
        this.type = TransactionType_1.TransactionType.Download_Payment;
    }
    addSize(size) {
        this.size = size;
        return this;
    }
    addMaxFee(fee) {
        this.fee = fee;
        return this;
    }
    addVersion(version) {
        this.version = version;
        return this;
    }
    addType(type) {
        this.type = type;
        return this;
    }
    addDeadline(deadline) {
        this.deadline = deadline;
        return this;
    }
    addDownloadChannelId(downloadChannelId) {
        this.downloadChannelId = downloadChannelId;
        return this;
    }
    addDownloadSize(downloadSize) {
        this.downloadSize = downloadSize;
        return this;
    }
    addFeedbackFeeAmount(feedbackFeeAmount) {
        this.feedbackFeeAmount = feedbackFeeAmount;
        return this;
    }
    build() {
        const builder = new flatbuffers.Builder(1);
        // Create vectors
        const signatureVector = DownloadPaymentTransactionBuffer_1.DownloadPaymentTransactionBuffer
            .createSignatureVector(builder, Array(...Array(64)).map(Number.prototype.valueOf, 0));
        const signerVector = DownloadPaymentTransactionBuffer_1.DownloadPaymentTransactionBuffer.createSignerVector(builder, Array(...Array(32)).map(Number.prototype.valueOf, 0));
        const deadlineVector = DownloadPaymentTransactionBuffer_1.DownloadPaymentTransactionBuffer.createDeadlineVector(builder, this.deadline);
        const feeVector = DownloadPaymentTransactionBuffer_1.DownloadPaymentTransactionBuffer.createMaxFeeVector(builder, this.fee);
        const downloadChannelIdVector = DownloadPaymentTransactionBuffer_1.DownloadPaymentTransactionBuffer.createDownloadChannelIdVector(builder, format_1.Convert.hexToUint8(this.downloadChannelId));
        const downloadSizeVector = DownloadPaymentTransactionBuffer_1.DownloadPaymentTransactionBuffer.createDownloadSizeVector(builder, this.downloadSize);
        const feedbackFeeAmountVector = DownloadPaymentTransactionBuffer_1.DownloadPaymentTransactionBuffer.createFeedbackFeeAmountVector(builder, this.feedbackFeeAmount);
        DownloadPaymentTransactionBuffer_1.DownloadPaymentTransactionBuffer.startDownloadPaymentTransactionBuffer(builder);
        DownloadPaymentTransactionBuffer_1.DownloadPaymentTransactionBuffer.addSize(builder, this.size);
        DownloadPaymentTransactionBuffer_1.DownloadPaymentTransactionBuffer.addSignature(builder, signatureVector);
        DownloadPaymentTransactionBuffer_1.DownloadPaymentTransactionBuffer.addSigner(builder, signerVector);
        DownloadPaymentTransactionBuffer_1.DownloadPaymentTransactionBuffer.addVersion(builder, this.version);
        DownloadPaymentTransactionBuffer_1.DownloadPaymentTransactionBuffer.addType(builder, this.type);
        DownloadPaymentTransactionBuffer_1.DownloadPaymentTransactionBuffer.addMaxFee(builder, feeVector);
        DownloadPaymentTransactionBuffer_1.DownloadPaymentTransactionBuffer.addDeadline(builder, deadlineVector);
        DownloadPaymentTransactionBuffer_1.DownloadPaymentTransactionBuffer.addDownloadChannelId(builder, downloadChannelIdVector);
        DownloadPaymentTransactionBuffer_1.DownloadPaymentTransactionBuffer.addDownloadSize(builder, downloadSizeVector);
        DownloadPaymentTransactionBuffer_1.DownloadPaymentTransactionBuffer.addFeedbackFeeAmount(builder, feedbackFeeAmountVector);
        const codedTransfer = DownloadPaymentTransactionBuffer_1.DownloadPaymentTransactionBuffer.endDownloadPaymentTransactionBuffer(builder);
        builder.finish(codedTransfer);
        const bytes = builder.asUint8Array();
        return new NewDownloadPaymentTransaction(bytes);
    }
    static stringToUint8(stringToConvert) {
        return format_1.Convert.hexToUint8(format_1.Convert.utf8ToHex(stringToConvert));
    }
}
exports.Builder = Builder;
