// tsjs-xpx-chain-sdk/dist/src/infrastructure/builders/storage/NewDataModificationTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = void 0;
/**
 * @module transactions/DataModificationTransaction
 */
const format_1 = require("../../../core/format");
const TransactionType_1 = require("../../../model/transaction/TransactionType");
const VerifiableTransaction_1 = require("../VerifiableTransaction");
const DataModificationTransactionSchema_1 = require("../../schemas/storage/DataModificationTransactionSchema");
const DataModificationTransactionBuffer_1 = require("../../buffers/storage/DataModificationTransactionBuffer");
const flatbuffers = require("flatbuffers");
class NewDataModificationTransaction extends VerifiableTransaction_1.VerifiableTransaction {
    constructor(bytes) {
        super(bytes, DataModificationTransactionSchema_1.default);
    }
}
exports.default = NewDataModificationTransaction;
// tslint:disable-next-line:max-classes-per-file
class Builder {
    size;
    fee;
    version;
    type;
    deadline;
    driveKey;
    downloadDataCdi;
    uploadSize;
    feedbackFeeAmount;
    constructor() {
        this.fee = [0, 0];
        this.type = TransactionType_1.TransactionType.Data_Modification;
    }
    addSize(size) {
        this.size = size;
        return this;
    }
    addMaxFee(fee) {
        this.fee = fee;
        return this;
    }
    addVersion(version) {
        this.version = version;
        return this;
    }
    addType(type) {
        this.type = type;
        return this;
    }
    addDeadline(deadline) {
        this.deadline = deadline;
        return this;
    }
    addDriveKey(driveKey) {
        this.driveKey = driveKey;
        return this;
    }
    addDownloadDataCdi(downloadDataCdi) {
        this.downloadDataCdi = downloadDataCdi;
        return this;
    }
    addUploadSize(uploadSize) {
        this.uploadSize = uploadSize;
        return this;
    }
    addFeedbackFeeAmount(feedbackFeeAmount) {
        this.feedbackFeeAmount = feedbackFeeAmount;
        return this;
    }
    build() {
        const builder = new flatbuffers.Builder(1);
        // Create vectors
        const signatureVector = DataModificationTransactionBuffer_1.DataModificationTransactionBuffer
            .createSignatureVector(builder, Array(...Array(64)).map(Number.prototype.valueOf, 0));
        const signerVector = DataModificationTransactionBuffer_1.DataModificationTransactionBuffer.createSignerVector(builder, Array(...Array(32)).map(Number.prototype.valueOf, 0));
        const deadlineVector = DataModificationTransactionBuffer_1.DataModificationTransactionBuffer.createDeadlineVector(builder, this.deadline);
        const feeVector = DataModificationTransactionBuffer_1.DataModificationTransactionBuffer.createMaxFeeVector(builder, this.fee);
        const driveKeyVector = DataModificationTransactionBuffer_1.DataModificationTransactionBuffer.createDriveKeyVector(builder, format_1.Convert.hexToUint8(this.driveKey));
        const downloadDataCdiVector = DataModificationTransactionBuffer_1.DataModificationTransactionBuffer.createDownloadDataCdiVector(builder, format_1.Convert.hexToUint8(this.downloadDataCdi));
        const uploadSizeVector = DataModificationTransactionBuffer_1.DataModificationTransactionBuffer.createUploadSizeVector(builder, this.uploadSize);
        const feedbackFeeAmountVector = DataModificationTransactionBuffer_1.DataModificationTransactionBuffer.createFeedbackFeeAmountVector(builder, this.feedbackFeeAmount);
        DataModificationTransactionBuffer_1.DataModificationTransactionBuffer.startDataModificationTransactionBuffer(builder);
        DataModificationTransactionBuffer_1.DataModificationTransactionBuffer.addSize(builder, this.size);
        DataModificationTransactionBuffer_1.DataModificationTransactionBuffer.addSignature(builder, signatureVector);
        DataModificationTransactionBuffer_1.DataModificationTransactionBuffer.addSigner(builder, signerVector);
        DataModificationTransactionBuffer_1.DataModificationTransactionBuffer.addVersion(builder, this.version);
        DataModificationTransactionBuffer_1.DataModificationTransactionBuffer.addType(builder, this.type);
        DataModificationTransactionBuffer_1.DataModificationTransactionBuffer.addMaxFee(builder, feeVector);
        DataModificationTransactionBuffer_1.DataModificationTransactionBuffer.addDeadline(builder, deadlineVector);
        DataModificationTransactionBuffer_1.DataModificationTransactionBuffer.addDriveKey(builder, driveKeyVector);
        DataModificationTransactionBuffer_1.DataModificationTransactionBuffer.addDownloadDataCdi(builder, downloadDataCdiVector);
        DataModificationTransactionBuffer_1.DataModificationTransactionBuffer.addUploadSize(builder, uploadSizeVector);
        DataModificationTransactionBuffer_1.DataModificationTransactionBuffer.addFeedbackFeeAmount(builder, feedbackFeeAmountVector);
        const codedTransfer = DataModificationTransactionBuffer_1.DataModificationTransactionBuffer.endDataModificationTransactionBuffer(builder);
        builder.finish(codedTransfer);
        const bytes = builder.asUint8Array();
        return new NewDataModificationTransaction(bytes);
    }
    static stringToUint8(stringToConvert) {
        return format_1.Convert.hexToUint8(format_1.Convert.utf8ToHex(stringToConvert));
    }
}
exports.Builder = Builder;
