// tsjs-xpx-chain-sdk/dist/src/infrastructure/builders/storage/NewDataModificationCancelTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = void 0;
/**
 * @module transactions/DataModificationCancelTransaction
 */
const format_1 = require("../../../core/format");
const TransactionType_1 = require("../../../model/transaction/TransactionType");
const VerifiableTransaction_1 = require("../VerifiableTransaction");
const DataModificationCancelTransactionSchema_1 = require("../../schemas/storage/DataModificationCancelTransactionSchema");
const DataModificationCancelTransactionBuffer_1 = require("../../buffers/storage/DataModificationCancelTransactionBuffer");
const flatbuffers = require("flatbuffers");
class NewDataModificationCancelTransaction extends VerifiableTransaction_1.VerifiableTransaction {
    constructor(bytes) {
        super(bytes, DataModificationCancelTransactionSchema_1.default);
    }
}
exports.default = NewDataModificationCancelTransaction;
// tslint:disable-next-line:max-classes-per-file
class Builder {
    size;
    fee;
    version;
    type;
    deadline;
    driveKey;
    dataModificationId;
    feedbackFeeAmount;
    constructor() {
        this.fee = [0, 0];
        this.type = TransactionType_1.TransactionType.Data_Modification_Cancel;
    }
    addSize(size) {
        this.size = size;
        return this;
    }
    addMaxFee(fee) {
        this.fee = fee;
        return this;
    }
    addVersion(version) {
        this.version = version;
        return this;
    }
    addType(type) {
        this.type = type;
        return this;
    }
    addDeadline(deadline) {
        this.deadline = deadline;
        return this;
    }
    addDriveKey(driveKey) {
        this.driveKey = driveKey;
        return this;
    }
    addDataModificationId(dataModificationId) {
        this.dataModificationId = dataModificationId;
        return this;
    }
    build() {
        const builder = new flatbuffers.Builder(1);
        // Create vectors
        const signatureVector = DataModificationCancelTransactionBuffer_1.DataModificationCancelTransactionBuffer
            .createSignatureVector(builder, Array(...Array(64)).map(Number.prototype.valueOf, 0));
        const signerVector = DataModificationCancelTransactionBuffer_1.DataModificationCancelTransactionBuffer.createSignerVector(builder, Array(...Array(32)).map(Number.prototype.valueOf, 0));
        const deadlineVector = DataModificationCancelTransactionBuffer_1.DataModificationCancelTransactionBuffer.createDeadlineVector(builder, this.deadline);
        const feeVector = DataModificationCancelTransactionBuffer_1.DataModificationCancelTransactionBuffer.createMaxFeeVector(builder, this.fee);
        const driveKeyVector = DataModificationCancelTransactionBuffer_1.DataModificationCancelTransactionBuffer.createDriveKeyVector(builder, format_1.Convert.hexToUint8(this.driveKey));
        const dataModificationIdVector = DataModificationCancelTransactionBuffer_1.DataModificationCancelTransactionBuffer.createDataModificationIdVector(builder, format_1.Convert.hexToUint8(this.dataModificationId));
        DataModificationCancelTransactionBuffer_1.DataModificationCancelTransactionBuffer.startDataModificationCancelTransactionBuffer(builder);
        DataModificationCancelTransactionBuffer_1.DataModificationCancelTransactionBuffer.addSize(builder, this.size);
        DataModificationCancelTransactionBuffer_1.DataModificationCancelTransactionBuffer.addSignature(builder, signatureVector);
        DataModificationCancelTransactionBuffer_1.DataModificationCancelTransactionBuffer.addSigner(builder, signerVector);
        DataModificationCancelTransactionBuffer_1.DataModificationCancelTransactionBuffer.addVersion(builder, this.version);
        DataModificationCancelTransactionBuffer_1.DataModificationCancelTransactionBuffer.addType(builder, this.type);
        DataModificationCancelTransactionBuffer_1.DataModificationCancelTransactionBuffer.addMaxFee(builder, feeVector);
        DataModificationCancelTransactionBuffer_1.DataModificationCancelTransactionBuffer.addDeadline(builder, deadlineVector);
        DataModificationCancelTransactionBuffer_1.DataModificationCancelTransactionBuffer.addDriveKey(builder, driveKeyVector);
        DataModificationCancelTransactionBuffer_1.DataModificationCancelTransactionBuffer.addDataModificationId(builder, dataModificationIdVector);
        const codedTransfer = DataModificationCancelTransactionBuffer_1.DataModificationCancelTransactionBuffer.endDataModificationCancelTransactionBuffer(builder);
        builder.finish(codedTransfer);
        const bytes = builder.asUint8Array();
        return new NewDataModificationCancelTransaction(bytes);
    }
    static stringToUint8(stringToConvert) {
        return format_1.Convert.hexToUint8(format_1.Convert.utf8ToHex(stringToConvert));
    }
}
exports.Builder = Builder;
