// tsjs-xpx-chain-sdk/dist/src/infrastructure/builders/storage/NewDownloadTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = void 0;
/**
 * @module transactions/DownloadTransaction
 */
const format_1 = require("../../../core/format");
const TransactionType_1 = require("../../../model/transaction/TransactionType");
const VerifiableTransaction_1 = require("../VerifiableTransaction");
const DownloadTransactionSchema_1 = require("../../schemas/storage/DownloadTransactionSchema");
const DownloadTransactionBuffer_1 = require("../../buffers/storage/DownloadTransactionBuffer");
const KeysBuffer_1 = require("../../buffers/storage/KeysBuffer");
const flatbuffers = require("flatbuffers");
class NewDownloadTransaction extends VerifiableTransaction_1.VerifiableTransaction {
    constructor(bytes) {
        super(bytes, DownloadTransactionSchema_1.default);
    }
}
exports.default = NewDownloadTransaction;
// tslint:disable-next-line:max-classes-per-file
class Builder {
    size;
    fee;
    version;
    type;
    deadline;
    driveKey;
    downloadSize;
    feedbackFeeAmount;
    listOfPublicKeys;
    constructor() {
        this.fee = [0, 0];
        this.type = TransactionType_1.TransactionType.Download;
    }
    addSize(size) {
        this.size = size;
        return this;
    }
    addMaxFee(fee) {
        this.fee = fee;
        return this;
    }
    addVersion(version) {
        this.version = version;
        return this;
    }
    addType(type) {
        this.type = type;
        return this;
    }
    addDeadline(deadline) {
        this.deadline = deadline;
        return this;
    }
    addDriveKey(driveKey) {
        this.driveKey = driveKey;
        return this;
    }
    addDownloadSize(downloadSize) {
        this.downloadSize = downloadSize;
        return this;
    }
    addFeedbackFeeAmount(feedbackFeeAmount) {
        this.feedbackFeeAmount = feedbackFeeAmount;
        return this;
    }
    addListOfPublicKeys(listOfPublicKeys) {
        this.listOfPublicKeys = listOfPublicKeys;
        return this;
    }
    build() {
        const builder = new flatbuffers.Builder(1);
        const totalPublicKeys = this.listOfPublicKeys.length;
        const totalPublicKeysUint8Array = new Uint8Array([totalPublicKeys, totalPublicKeys >> 8]);
        let publicKeyVectorArray = [];
        this.listOfPublicKeys.forEach((publicKey) => {
            const publicKeyVector = KeysBuffer_1.KeysBuffer
                .createKeyVector(builder, format_1.Convert.hexToUint8(publicKey));
            KeysBuffer_1.KeysBuffer.startKeysBuffer(builder);
            KeysBuffer_1.KeysBuffer.addKey(builder, publicKeyVector);
            publicKeyVectorArray.push(KeysBuffer_1.KeysBuffer.endKeysBuffer(builder));
        });
        // Create vectors
        const signatureVector = DownloadTransactionBuffer_1.DownloadTransactionBuffer
            .createSignatureVector(builder, Array(...Array(64)).map(Number.prototype.valueOf, 0));
        const signerVector = DownloadTransactionBuffer_1.DownloadTransactionBuffer.createSignerVector(builder, Array(...Array(32)).map(Number.prototype.valueOf, 0));
        const deadlineVector = DownloadTransactionBuffer_1.DownloadTransactionBuffer.createDeadlineVector(builder, this.deadline);
        const feeVector = DownloadTransactionBuffer_1.DownloadTransactionBuffer.createMaxFeeVector(builder, this.fee);
        const driveKeyVector = DownloadTransactionBuffer_1.DownloadTransactionBuffer.createDriveKeyVector(builder, format_1.Convert.hexToUint8(this.driveKey));
        const downloadSizeVector = DownloadTransactionBuffer_1.DownloadTransactionBuffer.createDownloadSizeVector(builder, this.downloadSize);
        const feedbackFeeAmountVector = DownloadTransactionBuffer_1.DownloadTransactionBuffer.createFeedbackFeeAmountVector(builder, this.feedbackFeeAmount);
        const listOfPublicKeysSizeVector = DownloadTransactionBuffer_1.DownloadTransactionBuffer.createListOfPublicKeysSizeVector(builder, totalPublicKeysUint8Array);
        const listOfPublicKeysVector = DownloadTransactionBuffer_1.DownloadTransactionBuffer.createListOfPublicKeysVector(builder, publicKeyVectorArray);
        DownloadTransactionBuffer_1.DownloadTransactionBuffer.startDownloadTransactionBuffer(builder);
        DownloadTransactionBuffer_1.DownloadTransactionBuffer.addSize(builder, this.size);
        DownloadTransactionBuffer_1.DownloadTransactionBuffer.addSignature(builder, signatureVector);
        DownloadTransactionBuffer_1.DownloadTransactionBuffer.addSigner(builder, signerVector);
        DownloadTransactionBuffer_1.DownloadTransactionBuffer.addVersion(builder, this.version);
        DownloadTransactionBuffer_1.DownloadTransactionBuffer.addType(builder, this.type);
        DownloadTransactionBuffer_1.DownloadTransactionBuffer.addMaxFee(builder, feeVector);
        DownloadTransactionBuffer_1.DownloadTransactionBuffer.addDeadline(builder, deadlineVector);
        DownloadTransactionBuffer_1.DownloadTransactionBuffer.addDriveKey(builder, driveKeyVector);
        DownloadTransactionBuffer_1.DownloadTransactionBuffer.addDownloadSize(builder, downloadSizeVector);
        DownloadTransactionBuffer_1.DownloadTransactionBuffer.addFeedbackFeeAmount(builder, feedbackFeeAmountVector);
        DownloadTransactionBuffer_1.DownloadTransactionBuffer.addListOfPublicKeysSize(builder, listOfPublicKeysSizeVector);
        DownloadTransactionBuffer_1.DownloadTransactionBuffer.addListOfPublicKeys(builder, listOfPublicKeysVector);
        const codedTransfer = DownloadTransactionBuffer_1.DownloadTransactionBuffer.endDownloadTransactionBuffer(builder);
        builder.finish(codedTransfer);
        const bytes = builder.asUint8Array();
        return new NewDownloadTransaction(bytes);
    }
    static stringToUint8(stringToConvert) {
        return format_1.Convert.hexToUint8(format_1.Convert.utf8ToHex(stringToConvert));
    }
}
exports.Builder = Builder;
