// tsjs-xpx-chain-sdk/dist/src/infrastructure/builders/storage/NewEndDriveVerificationV2Transaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = void 0;
/**
 * @module transactions/EndDriveVerificationV2Transaction
 */
const format_1 = require("../../../core/format");
const TransactionType_1 = require("../../../model/transaction/TransactionType");
const VerifiableTransaction_1 = require("../VerifiableTransaction");
const EndDriveVerificationV2TransactionSchema_1 = require("../../schemas/storage/EndDriveVerificationV2TransactionSchema");
const EndDriveVerificationTransactionV2Buffer_1 = require("../../buffers/storage/EndDriveVerificationTransactionV2Buffer");
const flatbuffers = require("flatbuffers");
const KeysBuffer_1 = require("../../buffers/storage/KeysBuffer");
const SignaturesBuffer_1 = require("../../buffers/storage/SignaturesBuffer");
class NewEndDriveVerificationV2Transaction extends VerifiableTransaction_1.VerifiableTransaction {
    constructor(bytes) {
        super(bytes, EndDriveVerificationV2TransactionSchema_1.default);
    }
}
exports.default = NewEndDriveVerificationV2Transaction;
// tslint:disable-next-line:max-classes-per-file
class Builder {
    size;
    fee;
    version;
    type;
    deadline;
    driveKey;
    verificationTrigger;
    shardId;
    keys;
    signatures;
    opinions;
    constructor() {
        this.fee = [0, 0];
        this.type = TransactionType_1.TransactionType.End_Drive_Verification_V2;
    }
    addSize(size) {
        this.size = size;
        return this;
    }
    addMaxFee(fee) {
        this.fee = fee;
        return this;
    }
    addVersion(version) {
        this.version = version;
        return this;
    }
    addType(type) {
        this.type = type;
        return this;
    }
    addDeadline(deadline) {
        this.deadline = deadline;
        return this;
    }
    addDriveKey(driveKey) {
        this.driveKey = driveKey;
        return this;
    }
    addVerificationTrigger(verificationTrigger) {
        this.verificationTrigger = verificationTrigger;
        return this;
    }
    addShardId(shardId) {
        this.shardId = shardId;
        return this;
    }
    addKeys(keys) {
        this.keys = keys;
        return this;
    }
    addSignatures(signatures) {
        this.signatures = signatures;
        return this;
    }
    addOpinions(opinions) {
        this.opinions = opinions;
        return this;
    }
    build() {
        const builder = new flatbuffers.Builder(1);
        // Create vectors
        const signatureVector = EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer
            .createSignatureVector(builder, Array(...Array(64)).map(Number.prototype.valueOf, 0));
        const signerVector = EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer.createSignerVector(builder, Array(...Array(32)).map(Number.prototype.valueOf, 0));
        const deadlineVector = EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer.createDeadlineVector(builder, this.deadline);
        const feeVector = EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer.createMaxFeeVector(builder, this.fee);
        const driveKeyVector = EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer.createDriveKeyVector(builder, format_1.Convert.hexToUint8(this.driveKey));
        const verificationTriggerVector = EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer.createVerificationTriggerVector(builder, format_1.Convert.hexToUint8(this.verificationTrigger));
        const shardIdUint8Array = new Uint8Array([this.shardId, this.shardId >> 8]);
        const shardIdVector = EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer.createShardIdVector(builder, shardIdUint8Array);
        const keysVector = EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer.createKeysVector(builder, this.keys.map((key) => {
            const keyBufferVector = KeysBuffer_1.KeysBuffer.createKeyVector(builder, format_1.Convert.hexToUint8(key));
            KeysBuffer_1.KeysBuffer.startKeysBuffer(builder);
            KeysBuffer_1.KeysBuffer.addKey(builder, keyBufferVector);
            return KeysBuffer_1.KeysBuffer.endKeysBuffer(builder);
        }));
        const signaturesVector = EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer.createSignaturesVector(builder, this.signatures.map((signature) => {
            const signatureBufferVector = SignaturesBuffer_1.SignaturesBuffer.createSignatureVector(builder, format_1.Convert.hexToUint8(signature));
            SignaturesBuffer_1.SignaturesBuffer.startSignaturesBuffer(builder);
            SignaturesBuffer_1.SignaturesBuffer.addSignature(builder, signatureBufferVector);
            return SignaturesBuffer_1.SignaturesBuffer.endSignaturesBuffer(builder);
        }));
        EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer.startEndDriveVerificationTransactionV2Buffer(builder);
        EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer.addSize(builder, this.size);
        EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer.addSignature(builder, signatureVector);
        EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer.addSigner(builder, signerVector);
        EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer.addVersion(builder, this.version);
        EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer.addType(builder, this.type);
        EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer.addMaxFee(builder, feeVector);
        EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer.addDeadline(builder, deadlineVector);
        EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer.addDriveKey(builder, driveKeyVector);
        EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer.addVerificationTrigger(builder, verificationTriggerVector);
        EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer.addShardId(builder, shardIdVector);
        EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer.addKeyCount(builder, this.keys.length);
        EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer.addJudgingKeyCount(builder, this.signatures.length);
        EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer.addKeys(builder, keysVector);
        EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer.addSignatures(builder, signaturesVector);
        EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer.addOpinions(builder, this.opinions);
        const codedTransfer = EndDriveVerificationTransactionV2Buffer_1.EndDriveVerificationTransactionV2Buffer.endEndDriveVerificationTransactionV2Buffer(builder);
        builder.finish(codedTransfer);
        const bytes = builder.asUint8Array();
        return new NewEndDriveVerificationV2Transaction(bytes);
    }
    static stringToUint8(stringToConvert) {
        return format_1.Convert.hexToUint8(format_1.Convert.utf8ToHex(stringToConvert));
    }
}
exports.Builder = Builder;
