// tsjs-xpx-chain-sdk/dist/src/infrastructure/builders/storage/NewFinishDownloadTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = void 0;
/**
 * @module transactions/FinishDownloadTransaction
 */
const format_1 = require("../../../core/format");
const TransactionType_1 = require("../../../model/transaction/TransactionType");
const VerifiableTransaction_1 = require("../VerifiableTransaction");
const FinishDownloadTransactionSchema_1 = require("../../schemas/storage/FinishDownloadTransactionSchema");
const FinishDownloadTransactionBuffer_1 = require("../../buffers/storage/FinishDownloadTransactionBuffer");
const flatbuffers = require("flatbuffers");
class NewFinishDownloadTransaction extends VerifiableTransaction_1.VerifiableTransaction {
    constructor(bytes) {
        super(bytes, FinishDownloadTransactionSchema_1.default);
    }
}
exports.default = NewFinishDownloadTransaction;
// tslint:disable-next-line:max-classes-per-file
class Builder {
    size;
    fee;
    version;
    type;
    deadline;
    downloadChannelId;
    feedbackFeeAmount;
    constructor() {
        this.fee = [0, 0];
        this.type = TransactionType_1.TransactionType.Finish_Download;
    }
    addSize(size) {
        this.size = size;
        return this;
    }
    addMaxFee(fee) {
        this.fee = fee;
        return this;
    }
    addVersion(version) {
        this.version = version;
        return this;
    }
    addType(type) {
        this.type = type;
        return this;
    }
    addDeadline(deadline) {
        this.deadline = deadline;
        return this;
    }
    addDownloadChannelId(downloadChannelId) {
        this.downloadChannelId = downloadChannelId;
        return this;
    }
    addFeedbackFeeAmount(feedbackFeeAmount) {
        this.feedbackFeeAmount = feedbackFeeAmount;
        return this;
    }
    build() {
        const builder = new flatbuffers.Builder(1);
        // Create vectors
        const signatureVector = FinishDownloadTransactionBuffer_1.FinishDownloadTransactionBuffer
            .createSignatureVector(builder, Array(...Array(64)).map(Number.prototype.valueOf, 0));
        const signerVector = FinishDownloadTransactionBuffer_1.FinishDownloadTransactionBuffer.createSignerVector(builder, Array(...Array(32)).map(Number.prototype.valueOf, 0));
        const deadlineVector = FinishDownloadTransactionBuffer_1.FinishDownloadTransactionBuffer.createDeadlineVector(builder, this.deadline);
        const feeVector = FinishDownloadTransactionBuffer_1.FinishDownloadTransactionBuffer.createMaxFeeVector(builder, this.fee);
        const downloadChannelIdVector = FinishDownloadTransactionBuffer_1.FinishDownloadTransactionBuffer.createDownloadChannelIdVector(builder, format_1.Convert.hexToUint8(this.downloadChannelId));
        const feedbackFeeAmountVector = FinishDownloadTransactionBuffer_1.FinishDownloadTransactionBuffer.createFeedbackFeeAmountVector(builder, this.feedbackFeeAmount);
        FinishDownloadTransactionBuffer_1.FinishDownloadTransactionBuffer.startFinishDownloadTransactionBuffer(builder);
        FinishDownloadTransactionBuffer_1.FinishDownloadTransactionBuffer.addSize(builder, this.size);
        FinishDownloadTransactionBuffer_1.FinishDownloadTransactionBuffer.addSignature(builder, signatureVector);
        FinishDownloadTransactionBuffer_1.FinishDownloadTransactionBuffer.addSigner(builder, signerVector);
        FinishDownloadTransactionBuffer_1.FinishDownloadTransactionBuffer.addVersion(builder, this.version);
        FinishDownloadTransactionBuffer_1.FinishDownloadTransactionBuffer.addType(builder, this.type);
        FinishDownloadTransactionBuffer_1.FinishDownloadTransactionBuffer.addMaxFee(builder, feeVector);
        FinishDownloadTransactionBuffer_1.FinishDownloadTransactionBuffer.addDeadline(builder, deadlineVector);
        FinishDownloadTransactionBuffer_1.FinishDownloadTransactionBuffer.addDownloadChannelId(builder, downloadChannelIdVector);
        FinishDownloadTransactionBuffer_1.FinishDownloadTransactionBuffer.addFeedbackFeeAmount(builder, feedbackFeeAmountVector);
        const codedTransfer = FinishDownloadTransactionBuffer_1.FinishDownloadTransactionBuffer.endFinishDownloadTransactionBuffer(builder);
        builder.finish(codedTransfer);
        const bytes = builder.asUint8Array();
        return new NewFinishDownloadTransaction(bytes);
    }
    static stringToUint8(stringToConvert) {
        return format_1.Convert.hexToUint8(format_1.Convert.utf8ToHex(stringToConvert));
    }
}
exports.Builder = Builder;
