// tsjs-xpx-chain-sdk/dist/src/infrastructure/builders/storage/NewPrepareBcDriveTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = void 0;
/**
 * @module transactions/PrepareBcDriveTransaction
 */
const format_1 = require("../../../core/format");
const TransactionType_1 = require("../../../model/transaction/TransactionType");
const VerifiableTransaction_1 = require("../VerifiableTransaction");
const PrepareBcDriveTransactionSchema_1 = require("../../schemas/storage/PrepareBcDriveTransactionSchema");
const PrepareBcDriveTransactionBuffer_1 = require("../../buffers/storage/PrepareBcDriveTransactionBuffer");
const flatbuffers = require("flatbuffers");
class NewPrepareBcDriveTransaction extends VerifiableTransaction_1.VerifiableTransaction {
    constructor(bytes) {
        super(bytes, PrepareBcDriveTransactionSchema_1.default);
    }
}
exports.default = NewPrepareBcDriveTransaction;
// tslint:disable-next-line:max-classes-per-file
class Builder {
    size;
    fee;
    version;
    type;
    deadline;
    driveSize;
    verificationFeeAmount;
    replicatorCount;
    constructor() {
        this.fee = [0, 0];
        this.type = TransactionType_1.TransactionType.Prepare_Bc_Drive;
    }
    addSize(size) {
        this.size = size;
        return this;
    }
    addMaxFee(fee) {
        this.fee = fee;
        return this;
    }
    addVersion(version) {
        this.version = version;
        return this;
    }
    addType(type) {
        this.type = type;
        return this;
    }
    addDeadline(deadline) {
        this.deadline = deadline;
        return this;
    }
    addDriveSize(driveSize) {
        this.driveSize = driveSize;
        return this;
    }
    addVerificationFeeAmount(verificationFeeAmount) {
        this.verificationFeeAmount = verificationFeeAmount;
        return this;
    }
    addReplicatorCount(replicatorCount) {
        this.replicatorCount = replicatorCount;
        return this;
    }
    build() {
        const builder = new flatbuffers.Builder(1);
        // Create vectors
        const signatureVector = PrepareBcDriveTransactionBuffer_1.PrepareBcDriveTransactionBuffer
            .createSignatureVector(builder, Array(...Array(64)).map(Number.prototype.valueOf, 0));
        const signerVector = PrepareBcDriveTransactionBuffer_1.PrepareBcDriveTransactionBuffer.createSignerVector(builder, Array(...Array(32)).map(Number.prototype.valueOf, 0));
        const deadlineVector = PrepareBcDriveTransactionBuffer_1.PrepareBcDriveTransactionBuffer.createDeadlineVector(builder, this.deadline);
        const feeVector = PrepareBcDriveTransactionBuffer_1.PrepareBcDriveTransactionBuffer.createMaxFeeVector(builder, this.fee);
        const driveSizeVector = PrepareBcDriveTransactionBuffer_1.PrepareBcDriveTransactionBuffer.createDriveSizeVector(builder, this.driveSize);
        const verificationFeeAmountVector = PrepareBcDriveTransactionBuffer_1.PrepareBcDriveTransactionBuffer.createDriveSizeVector(builder, this.verificationFeeAmount);
        PrepareBcDriveTransactionBuffer_1.PrepareBcDriveTransactionBuffer.startPrepareBcDriveTransactionBuffer(builder);
        PrepareBcDriveTransactionBuffer_1.PrepareBcDriveTransactionBuffer.addSize(builder, this.size);
        PrepareBcDriveTransactionBuffer_1.PrepareBcDriveTransactionBuffer.addSignature(builder, signatureVector);
        PrepareBcDriveTransactionBuffer_1.PrepareBcDriveTransactionBuffer.addSigner(builder, signerVector);
        PrepareBcDriveTransactionBuffer_1.PrepareBcDriveTransactionBuffer.addVersion(builder, this.version);
        PrepareBcDriveTransactionBuffer_1.PrepareBcDriveTransactionBuffer.addType(builder, this.type);
        PrepareBcDriveTransactionBuffer_1.PrepareBcDriveTransactionBuffer.addMaxFee(builder, feeVector);
        PrepareBcDriveTransactionBuffer_1.PrepareBcDriveTransactionBuffer.addDeadline(builder, deadlineVector);
        PrepareBcDriveTransactionBuffer_1.PrepareBcDriveTransactionBuffer.addDriveSize(builder, driveSizeVector);
        PrepareBcDriveTransactionBuffer_1.PrepareBcDriveTransactionBuffer.addVerificationFeeAmount(builder, verificationFeeAmountVector);
        PrepareBcDriveTransactionBuffer_1.PrepareBcDriveTransactionBuffer.addReplicatorCount(builder, this.replicatorCount);
        const codedTransfer = PrepareBcDriveTransactionBuffer_1.PrepareBcDriveTransactionBuffer.endPrepareBcDriveTransactionBuffer(builder);
        builder.finish(codedTransfer);
        const bytes = builder.asUint8Array();
        return new NewPrepareBcDriveTransaction(bytes);
    }
    static stringToUint8(stringToConvert) {
        return format_1.Convert.hexToUint8(format_1.Convert.utf8ToHex(stringToConvert));
    }
}
exports.Builder = Builder;
