// tsjs-xpx-chain-sdk/dist/src/infrastructure/builders/storage/NewReplicatorOnboardingTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = void 0;
/**
 * @module transactions/NewReplicatorOnboardingTransaction
 */
const format_1 = require("../../../core/format");
const TransactionType_1 = require("../../../model/transaction/TransactionType");
const VerifiableTransaction_1 = require("../VerifiableTransaction");
const ReplicatorOnboardingTransactionSchema_1 = require("../../schemas/storage/ReplicatorOnboardingTransactionSchema");
const ReplicatorOnboardingTransactionBuffer_1 = require("../../buffers/storage/ReplicatorOnboardingTransactionBuffer");
const flatbuffers = require("flatbuffers");
class NewReplicatorOnboardingTransaction extends VerifiableTransaction_1.VerifiableTransaction {
    constructor(bytes) {
        super(bytes, ReplicatorOnboardingTransactionSchema_1.default);
    }
}
exports.default = NewReplicatorOnboardingTransaction;
// tslint:disable-next-line:max-classes-per-file
class Builder {
    size;
    fee;
    version;
    type;
    deadline;
    capacity;
    constructor() {
        this.fee = [0, 0];
        this.type = TransactionType_1.TransactionType.Replicator_Onboarding;
    }
    addSize(size) {
        this.size = size;
        return this;
    }
    addMaxFee(fee) {
        this.fee = fee;
        return this;
    }
    addVersion(version) {
        this.version = version;
        return this;
    }
    addType(type) {
        this.type = type;
        return this;
    }
    addDeadline(deadline) {
        this.deadline = deadline;
        return this;
    }
    addCapacity(capacity) {
        this.capacity = capacity;
        return this;
    }
    build() {
        const builder = new flatbuffers.Builder(1);
        // Create vectors
        const signatureVector = ReplicatorOnboardingTransactionBuffer_1.ReplicatorOnboardingTransactionBuffer
            .createSignatureVector(builder, Array(...Array(64)).map(Number.prototype.valueOf, 0));
        const signerVector = ReplicatorOnboardingTransactionBuffer_1.ReplicatorOnboardingTransactionBuffer.createSignerVector(builder, Array(...Array(32)).map(Number.prototype.valueOf, 0));
        const deadlineVector = ReplicatorOnboardingTransactionBuffer_1.ReplicatorOnboardingTransactionBuffer.createDeadlineVector(builder, this.deadline);
        const feeVector = ReplicatorOnboardingTransactionBuffer_1.ReplicatorOnboardingTransactionBuffer.createMaxFeeVector(builder, this.fee);
        const capacityVector = ReplicatorOnboardingTransactionBuffer_1.ReplicatorOnboardingTransactionBuffer.createCapacityVector(builder, this.capacity);
        ReplicatorOnboardingTransactionBuffer_1.ReplicatorOnboardingTransactionBuffer.startReplicatorOnboardingTransactionBuffer(builder);
        ReplicatorOnboardingTransactionBuffer_1.ReplicatorOnboardingTransactionBuffer.addSize(builder, this.size);
        ReplicatorOnboardingTransactionBuffer_1.ReplicatorOnboardingTransactionBuffer.addSignature(builder, signatureVector);
        ReplicatorOnboardingTransactionBuffer_1.ReplicatorOnboardingTransactionBuffer.addSigner(builder, signerVector);
        ReplicatorOnboardingTransactionBuffer_1.ReplicatorOnboardingTransactionBuffer.addVersion(builder, this.version);
        ReplicatorOnboardingTransactionBuffer_1.ReplicatorOnboardingTransactionBuffer.addType(builder, this.type);
        ReplicatorOnboardingTransactionBuffer_1.ReplicatorOnboardingTransactionBuffer.addMaxFee(builder, feeVector);
        ReplicatorOnboardingTransactionBuffer_1.ReplicatorOnboardingTransactionBuffer.addDeadline(builder, deadlineVector);
        ReplicatorOnboardingTransactionBuffer_1.ReplicatorOnboardingTransactionBuffer.addCapacity(builder, capacityVector);
        const codedTransfer = ReplicatorOnboardingTransactionBuffer_1.ReplicatorOnboardingTransactionBuffer.endReplicatorOnboardingTransactionBuffer(builder);
        builder.finish(codedTransfer);
        const bytes = builder.asUint8Array();
        return new NewReplicatorOnboardingTransaction(bytes);
    }
    static stringToUint8(stringToConvert) {
        return format_1.Convert.hexToUint8(format_1.Convert.utf8ToHex(stringToConvert));
    }
}
exports.Builder = Builder;
