// tsjs-xpx-chain-sdk/dist/src/infrastructure/builders/storage/NewStoragePaymentTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = void 0;
/**
 * @module transactions/StoragePaymentTransaction
 */
const format_1 = require("../../../core/format");
const TransactionType_1 = require("../../../model/transaction/TransactionType");
const VerifiableTransaction_1 = require("../VerifiableTransaction");
const StoragePaymentTransactionSchema_1 = require("../../schemas/storage/StoragePaymentTransactionSchema");
const StoragePaymentTransactionBuffer_1 = require("../../buffers/storage/StoragePaymentTransactionBuffer");
const flatbuffers = require("flatbuffers");
class NewStoragePaymentTransaction extends VerifiableTransaction_1.VerifiableTransaction {
    constructor(bytes) {
        super(bytes, StoragePaymentTransactionSchema_1.default);
    }
}
exports.default = NewStoragePaymentTransaction;
// tslint:disable-next-line:max-classes-per-file
class Builder {
    size;
    fee;
    version;
    type;
    deadline;
    driveKey;
    storageUnits;
    constructor() {
        this.fee = [0, 0];
        this.type = TransactionType_1.TransactionType.Storage_Payment;
    }
    addSize(size) {
        this.size = size;
        return this;
    }
    addMaxFee(fee) {
        this.fee = fee;
        return this;
    }
    addVersion(version) {
        this.version = version;
        return this;
    }
    addType(type) {
        this.type = type;
        return this;
    }
    addDeadline(deadline) {
        this.deadline = deadline;
        return this;
    }
    addDriveKey(driveKey) {
        this.driveKey = driveKey;
        return this;
    }
    addStorageUnits(storageUnits) {
        this.storageUnits = storageUnits;
        return this;
    }
    build() {
        const builder = new flatbuffers.Builder(1);
        // Create vectors
        const signatureVector = StoragePaymentTransactionBuffer_1.StoragePaymentTransactionBuffer
            .createSignatureVector(builder, Array(...Array(64)).map(Number.prototype.valueOf, 0));
        const signerVector = StoragePaymentTransactionBuffer_1.StoragePaymentTransactionBuffer.createSignerVector(builder, Array(...Array(32)).map(Number.prototype.valueOf, 0));
        const deadlineVector = StoragePaymentTransactionBuffer_1.StoragePaymentTransactionBuffer.createDeadlineVector(builder, this.deadline);
        const feeVector = StoragePaymentTransactionBuffer_1.StoragePaymentTransactionBuffer.createMaxFeeVector(builder, this.fee);
        const driveKeyVector = StoragePaymentTransactionBuffer_1.StoragePaymentTransactionBuffer.createDriveKeyVector(builder, format_1.Convert.hexToUint8(this.driveKey));
        const storageUnitsVector = StoragePaymentTransactionBuffer_1.StoragePaymentTransactionBuffer.createStorageUnitsVector(builder, this.storageUnits);
        StoragePaymentTransactionBuffer_1.StoragePaymentTransactionBuffer.startStoragePaymentTransactionBuffer(builder);
        StoragePaymentTransactionBuffer_1.StoragePaymentTransactionBuffer.addSize(builder, this.size);
        StoragePaymentTransactionBuffer_1.StoragePaymentTransactionBuffer.addSignature(builder, signatureVector);
        StoragePaymentTransactionBuffer_1.StoragePaymentTransactionBuffer.addSigner(builder, signerVector);
        StoragePaymentTransactionBuffer_1.StoragePaymentTransactionBuffer.addVersion(builder, this.version);
        StoragePaymentTransactionBuffer_1.StoragePaymentTransactionBuffer.addType(builder, this.type);
        StoragePaymentTransactionBuffer_1.StoragePaymentTransactionBuffer.addMaxFee(builder, feeVector);
        StoragePaymentTransactionBuffer_1.StoragePaymentTransactionBuffer.addDeadline(builder, deadlineVector);
        StoragePaymentTransactionBuffer_1.StoragePaymentTransactionBuffer.addDriveKey(builder, driveKeyVector);
        StoragePaymentTransactionBuffer_1.StoragePaymentTransactionBuffer.addStorageUnits(builder, storageUnitsVector);
        const codedTransfer = StoragePaymentTransactionBuffer_1.StoragePaymentTransactionBuffer.endStoragePaymentTransactionBuffer(builder);
        builder.finish(codedTransfer);
        const bytes = builder.asUint8Array();
        return new NewStoragePaymentTransaction(bytes);
    }
    static stringToUint8(stringToConvert) {
        return format_1.Convert.hexToUint8(format_1.Convert.utf8ToHex(stringToConvert));
    }
}
exports.Builder = Builder;
