// tsjs-xpx-chain-sdk/dist/src/infrastructure/builders/storage/NewVerificationPaymentTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = void 0;
/**
 * @module transactions/VerificationPaymentTransaction
 */
const format_1 = require("../../../core/format");
const TransactionType_1 = require("../../../model/transaction/TransactionType");
const VerifiableTransaction_1 = require("../VerifiableTransaction");
const VerificationPaymentTransactionSchema_1 = require("../../schemas/storage/VerificationPaymentTransactionSchema");
const VerificationPaymentTransactionBuffer_1 = require("../../buffers/storage/VerificationPaymentTransactionBuffer");
const flatbuffers = require("flatbuffers");
class NewVerificationPaymentTransaction extends VerifiableTransaction_1.VerifiableTransaction {
    constructor(bytes) {
        super(bytes, VerificationPaymentTransactionSchema_1.default);
    }
}
exports.default = NewVerificationPaymentTransaction;
// tslint:disable-next-line:max-classes-per-file
class Builder {
    size;
    fee;
    version;
    type;
    deadline;
    driveKey;
    verificationFeeAmount;
    constructor() {
        this.fee = [0, 0];
        this.type = TransactionType_1.TransactionType.Verification_Payment;
    }
    addSize(size) {
        this.size = size;
        return this;
    }
    addMaxFee(fee) {
        this.fee = fee;
        return this;
    }
    addVersion(version) {
        this.version = version;
        return this;
    }
    addType(type) {
        this.type = type;
        return this;
    }
    addDeadline(deadline) {
        this.deadline = deadline;
        return this;
    }
    addDriveKey(driveKey) {
        this.driveKey = driveKey;
        return this;
    }
    addVerificationFeeAmount(verificationFeeAmount) {
        this.verificationFeeAmount = verificationFeeAmount;
        return this;
    }
    build() {
        const builder = new flatbuffers.Builder(1);
        // Create vectors
        const signatureVector = VerificationPaymentTransactionBuffer_1.VerificationPaymentTransactionBuffer
            .createSignatureVector(builder, Array(...Array(64)).map(Number.prototype.valueOf, 0));
        const signerVector = VerificationPaymentTransactionBuffer_1.VerificationPaymentTransactionBuffer.createSignerVector(builder, Array(...Array(32)).map(Number.prototype.valueOf, 0));
        const deadlineVector = VerificationPaymentTransactionBuffer_1.VerificationPaymentTransactionBuffer.createDeadlineVector(builder, this.deadline);
        const feeVector = VerificationPaymentTransactionBuffer_1.VerificationPaymentTransactionBuffer.createMaxFeeVector(builder, this.fee);
        const driveKeyVector = VerificationPaymentTransactionBuffer_1.VerificationPaymentTransactionBuffer.createDriveKeyVector(builder, format_1.Convert.hexToUint8(this.driveKey));
        const verificationFeeAmountVector = VerificationPaymentTransactionBuffer_1.VerificationPaymentTransactionBuffer.createVerificationFeeAmountVector(builder, this.verificationFeeAmount);
        VerificationPaymentTransactionBuffer_1.VerificationPaymentTransactionBuffer.startVerificationPaymentTransactionBuffer(builder);
        VerificationPaymentTransactionBuffer_1.VerificationPaymentTransactionBuffer.addSize(builder, this.size);
        VerificationPaymentTransactionBuffer_1.VerificationPaymentTransactionBuffer.addSignature(builder, signatureVector);
        VerificationPaymentTransactionBuffer_1.VerificationPaymentTransactionBuffer.addSigner(builder, signerVector);
        VerificationPaymentTransactionBuffer_1.VerificationPaymentTransactionBuffer.addVersion(builder, this.version);
        VerificationPaymentTransactionBuffer_1.VerificationPaymentTransactionBuffer.addType(builder, this.type);
        VerificationPaymentTransactionBuffer_1.VerificationPaymentTransactionBuffer.addMaxFee(builder, feeVector);
        VerificationPaymentTransactionBuffer_1.VerificationPaymentTransactionBuffer.addDeadline(builder, deadlineVector);
        VerificationPaymentTransactionBuffer_1.VerificationPaymentTransactionBuffer.addDriveKey(builder, driveKeyVector);
        VerificationPaymentTransactionBuffer_1.VerificationPaymentTransactionBuffer.addVerificationFeeAmount(builder, verificationFeeAmountVector);
        const codedTransfer = VerificationPaymentTransactionBuffer_1.VerificationPaymentTransactionBuffer.endVerificationPaymentTransactionBuffer(builder);
        builder.finish(codedTransfer);
        const bytes = builder.asUint8Array();
        return new NewVerificationPaymentTransaction(bytes);
    }
    static stringToUint8(stringToConvert) {
        return format_1.Convert.hexToUint8(format_1.Convert.utf8ToHex(stringToConvert));
    }
}
exports.Builder = Builder;
