// tsjs-xpx-chain-sdk/dist/src/infrastructure/builders/storage/ReplicatorOffboardingTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = void 0;
/**
 * @module transactions/ReplicatorOffboardingTransaction
 */
const format_1 = require("../../../core/format");
const TransactionType_1 = require("../../../model/transaction/TransactionType");
const VerifiableTransaction_1 = require("../VerifiableTransaction");
const ReplicatorOffboardingTransactionSchema_1 = require("../../schemas/storage/ReplicatorOffboardingTransactionSchema");
const ReplicatorOffboardingTransactionBuffer_1 = require("../../buffers/storage/ReplicatorOffboardingTransactionBuffer");
const flatbuffers = require("flatbuffers");
class ReplicatorOffboardingTransaction extends VerifiableTransaction_1.VerifiableTransaction {
    constructor(bytes) {
        super(bytes, ReplicatorOffboardingTransactionSchema_1.default);
    }
}
exports.default = ReplicatorOffboardingTransaction;
// tslint:disable-next-line:max-classes-per-file
class Builder {
    size;
    fee;
    version;
    type;
    deadline;
    driveKey;
    constructor() {
        this.fee = [0, 0];
        this.type = TransactionType_1.TransactionType.Replicator_Offboarding;
    }
    addSize(size) {
        this.size = size;
        return this;
    }
    addMaxFee(fee) {
        this.fee = fee;
        return this;
    }
    addVersion(version) {
        this.version = version;
        return this;
    }
    addType(type) {
        this.type = type;
        return this;
    }
    addDeadline(deadline) {
        this.deadline = deadline;
        return this;
    }
    addDriveKey(driveKey) {
        this.driveKey = driveKey;
        return this;
    }
    build() {
        const builder = new flatbuffers.Builder(1);
        // Create vectors
        const signatureVector = ReplicatorOffboardingTransactionBuffer_1.ReplicatorOffboardingTransactionBuffer
            .createSignatureVector(builder, Array(...Array(64)).map(Number.prototype.valueOf, 0));
        const signerVector = ReplicatorOffboardingTransactionBuffer_1.ReplicatorOffboardingTransactionBuffer.createSignerVector(builder, Array(...Array(32)).map(Number.prototype.valueOf, 0));
        const deadlineVector = ReplicatorOffboardingTransactionBuffer_1.ReplicatorOffboardingTransactionBuffer.createDeadlineVector(builder, this.deadline);
        const feeVector = ReplicatorOffboardingTransactionBuffer_1.ReplicatorOffboardingTransactionBuffer.createMaxFeeVector(builder, this.fee);
        const driveKeyVector = ReplicatorOffboardingTransactionBuffer_1.ReplicatorOffboardingTransactionBuffer.createDriveKeyVector(builder, format_1.Convert.hexToUint8(this.driveKey));
        ReplicatorOffboardingTransactionBuffer_1.ReplicatorOffboardingTransactionBuffer.startReplicatorOffboardingTransactionBuffer(builder);
        ReplicatorOffboardingTransactionBuffer_1.ReplicatorOffboardingTransactionBuffer.addSize(builder, this.size);
        ReplicatorOffboardingTransactionBuffer_1.ReplicatorOffboardingTransactionBuffer.addSignature(builder, signatureVector);
        ReplicatorOffboardingTransactionBuffer_1.ReplicatorOffboardingTransactionBuffer.addSigner(builder, signerVector);
        ReplicatorOffboardingTransactionBuffer_1.ReplicatorOffboardingTransactionBuffer.addVersion(builder, this.version);
        ReplicatorOffboardingTransactionBuffer_1.ReplicatorOffboardingTransactionBuffer.addType(builder, this.type);
        ReplicatorOffboardingTransactionBuffer_1.ReplicatorOffboardingTransactionBuffer.addMaxFee(builder, feeVector);
        ReplicatorOffboardingTransactionBuffer_1.ReplicatorOffboardingTransactionBuffer.addDeadline(builder, deadlineVector);
        ReplicatorOffboardingTransactionBuffer_1.ReplicatorOffboardingTransactionBuffer.addDriveKey(builder, driveKeyVector);
        const codedTransfer = ReplicatorOffboardingTransactionBuffer_1.ReplicatorOffboardingTransactionBuffer.endReplicatorOffboardingTransactionBuffer(builder);
        builder.finish(codedTransfer);
        const bytes = builder.asUint8Array();
        return new ReplicatorOffboardingTransaction(bytes);
    }
    static stringToUint8(stringToConvert) {
        return format_1.Convert.hexToUint8(format_1.Convert.utf8ToHex(stringToConvert));
    }
}
exports.Builder = Builder;
