// tsjs-xpx-chain-sdk/dist/src/model/receipt/OfferCreationReceipt.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OfferCreationReceipt = void 0;
const Receipt_1 = require("./Receipt");
/**
 * Offer creation: A creation of Sda Exchange offer.
 */
class OfferCreationReceipt extends Receipt_1.Receipt {
    sender;
    mosaicIdGive;
    mosaicIdGet;
    mosaicAmountGive;
    mosaicAmountGet;
    /**
     * Offer creation receipt
     * @param sender - public key.
     * @param mosaicIdGive - The amount of mosaic.
     * @param mosaicIdGet - The receipt version
     * @param mosaicAmountGive - The receipt type
     * @param mosaicAmountGet - the receipt size
     */
    constructor(sender, mosaicIdGive, mosaicIdGet, mosaicAmountGive, mosaicAmountGet, version, type) {
        super(version, type);
        this.sender = sender;
        this.mosaicIdGive = mosaicIdGive;
        this.mosaicIdGet = mosaicIdGet;
        this.mosaicAmountGive = mosaicAmountGive;
        this.mosaicAmountGet = mosaicAmountGet;
    }
}
exports.OfferCreationReceipt = OfferCreationReceipt;
