// tsjs-xpx-chain-sdk/dist/src/model/receipt/ReceiptGroupType.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, type 2.0 (the "License"),
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http =//www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReceiptGroupType = void 0;
/**
 * Receipt group type enums.
 */
var ReceiptGroupType;
(function (ReceiptGroupType) {
    ReceiptGroupType[ReceiptGroupType["Balance_Transfer"] = 1] = "Balance_Transfer";
    ReceiptGroupType[ReceiptGroupType["Balance_Change"] = 2] = "Balance_Change";
    // Balance_Debit = 3,
    ReceiptGroupType[ReceiptGroupType["ArtifactExpiry"] = 4] = "ArtifactExpiry";
    ReceiptGroupType[ReceiptGroupType["Inflation"] = 5] = "Inflation";
    // Operation = 7, 
    // Signer_Balance = 8,
    ReceiptGroupType[ReceiptGroupType["Offer_Creation"] = 10] = "Offer_Creation";
    ReceiptGroupType[ReceiptGroupType["Offer_Exchange"] = 11] = "Offer_Exchange";
    ReceiptGroupType[ReceiptGroupType["Offer_Removal"] = 12] = "Offer_Removal";
})(ReceiptGroupType || (exports.ReceiptGroupType = ReceiptGroupType = {}));
