// tsjs-xpx-chain-sdk/dist/src/infrastructure/AccountHttp.js
"use strict";
/*
 * Copyright 2023 ProximaX
 * Copyright 2018 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountHttp = void 0;
const rxjs_1 = require("rxjs");
const operators_1 = require("rxjs/operators");
const DtoMapping_1 = require("../core/utils/DtoMapping");
const AccountInfo_1 = require("../model/account/AccountInfo");
const AccountNames_1 = require("../model/account/AccountNames");
const Address_1 = require("../model/account/Address");
const MultisigAccountGraphInfo_1 = require("../model/account/MultisigAccountGraphInfo");
const MultisigAccountInfo_1 = require("../model/account/MultisigAccountInfo");
const PublicAccount_1 = require("../model/account/PublicAccount");
const Mosaic_1 = require("../model/mosaic/Mosaic");
const MosaicId_1 = require("../model/mosaic/MosaicId");
const NamespaceId_1 = require("../model/namespace/NamespaceId");
const NamespaceName_1 = require("../model/namespace/NamespaceName");
const TransactionSearch_1 = require("../model/transaction/TransactionSearch");
const UInt64_1 = require("../model/UInt64");
const api_1 = require("./api");
const Http_1 = require("./Http");
const NetworkHttp_1 = require("./NetworkHttp");
const CreateTransactionFromDTO_1 = require("./transaction/CreateTransactionFromDTO");
const Pagination_1 = require("../model/Pagination");
/**
 * Account http repository.
 *
 * @since 1.0
 */
class AccountHttp extends Http_1.Http {
    /**
     * @internal
     * xpx chain Library account routes api
     */
    accountRoutesApi;
    url;
    /**
     * Constructor
     * @param url
     * @param networkHttp
     */
    constructor(url, networkHttp) {
        networkHttp = networkHttp == null ? new NetworkHttp_1.NetworkHttp(url) : networkHttp;
        super(networkHttp);
        this.url = url;
        this.accountRoutesApi = new api_1.AccountRoutesApi(url);
    }
    /**
     * Gets an AccountInfo for an account.
     * @param address Address
     * @returns Observable<AccountInfo>
     */
    getAccountInfo(address, requestOptions) {
        return (0, rxjs_1.from)(this.accountRoutesApi.getAccountInfo(address.plain(), requestOptions)).pipe((0, operators_1.map)((response) => {
            const accountInfoDTO = response.body;
            return new AccountInfo_1.AccountInfo(accountInfoDTO.meta, Address_1.Address.createFromEncoded(accountInfoDTO.account.address), new UInt64_1.UInt64(accountInfoDTO.account.addressHeight), accountInfoDTO.account.publicKey, new UInt64_1.UInt64(accountInfoDTO.account.publicKeyHeight), accountInfoDTO.account.accountType.valueOf(), accountInfoDTO.account.linkedAccountKey, accountInfoDTO.account.mosaics.map((mosaicDTO) => new Mosaic_1.Mosaic(new MosaicId_1.MosaicId(mosaicDTO.id), new UInt64_1.UInt64(mosaicDTO.amount))));
        }));
    }
    /**
     * Get Account restrictions.
     * @param publicAccount public account
     * @returns Observable<AccountRestrictionsInfo>
     */
    getAccountRestrictions(address, requestOptions) {
        return (0, rxjs_1.from)(this.accountRoutesApi.getAccountProperties(address.plain(), requestOptions))
            .pipe((0, operators_1.map)((response) => {
            return DtoMapping_1.DtoMapping.extractAccountRestrictionFromDto(response.body);
        }));
    }
    /**
     * Get Account restrictions.
     * @param address list of addresses
     * @returns Observable<AccountRestrictionsInfo[]>
     */
    getAccountRestrictionsFromAccounts(addresses, requestOptions) {
        const accountIds = {
            addresses: addresses.map((address) => address.plain()),
        };
        return (0, rxjs_1.from)(this.accountRoutesApi.getAccountPropertiesFromAccounts(accountIds, requestOptions))
            .pipe((0, operators_1.map)((response) => {
            return response.body.map((restriction) => {
                return DtoMapping_1.DtoMapping.extractAccountRestrictionFromDto(restriction);
            });
        }));
    }
    /**
     * Gets AccountsInfo for different accounts.
     * @param addresses List of Address
     * @returns Observable<AccountInfo[]>
     */
    getAccountsInfo(addresses, requestOptions) {
        const accountIdsBody = {
            addresses: addresses.map((address) => address.plain()),
        };
        return (0, rxjs_1.from)(this.accountRoutesApi.getAccountsInfo(accountIdsBody, requestOptions))
            .pipe((0, operators_1.map)((response) => {
            return response.body.map((accountInfoDTO) => {
                return new AccountInfo_1.AccountInfo(accountInfoDTO.meta, Address_1.Address.createFromEncoded(accountInfoDTO.account.address), new UInt64_1.UInt64(accountInfoDTO.account.addressHeight), accountInfoDTO.account.publicKey, new UInt64_1.UInt64(accountInfoDTO.account.publicKeyHeight), accountInfoDTO.account.accountType.valueOf(), accountInfoDTO.account.linkedAccountKey, accountInfoDTO.account.mosaics.map((mosaicDTO) => new Mosaic_1.Mosaic(new MosaicId_1.MosaicId(mosaicDTO.id), new UInt64_1.UInt64(mosaicDTO.amount))));
            });
        }));
    }
    getAccountsNames(addresses, requestOptions) {
        const accountIdsBody = {
            addresses: addresses.map((address) => address.plain()),
        };
        return (0, rxjs_1.from)(this.accountRoutesApi.getAccountsNames(accountIdsBody, requestOptions)).pipe((0, operators_1.map)((response) => {
            return response.body.map((accountName) => {
                return new AccountNames_1.AccountNames(Address_1.Address.createFromEncoded(accountName.address), accountName.names.map((name) => {
                    return new NamespaceName_1.NamespaceName(new NamespaceId_1.NamespaceId(name), name);
                }));
            });
        }));
    }
    /**
     * Gets a MultisigAccountInfo for an account.
     * @param address - User address
     * @returns Observable<MultisigAccountInfo>
     */
    getMultisigAccountInfo(address, requestOptions) {
        return this.getNetworkTypeObservable(requestOptions).pipe((0, operators_1.mergeMap)((networkType) => (0, rxjs_1.from)(this.accountRoutesApi.getAccountMultisig(address.plain(), requestOptions))
            .pipe((0, operators_1.map)((response) => {
            const multisigAccountInfoDTO = response.body;
            return new MultisigAccountInfo_1.MultisigAccountInfo(PublicAccount_1.PublicAccount.createFromPublicKey(multisigAccountInfoDTO.multisig.account, networkType), multisigAccountInfoDTO.multisig.minApproval, multisigAccountInfoDTO.multisig.minRemoval, multisigAccountInfoDTO.multisig.cosignatories
                .map((cosigner) => PublicAccount_1.PublicAccount.createFromPublicKey(cosigner, networkType)), multisigAccountInfoDTO.multisig.multisigAccounts
                .map((multisigAccount) => PublicAccount_1.PublicAccount.createFromPublicKey(multisigAccount, networkType)));
        }))));
    }
    /**
     * Gets a MultisigAccountGraphInfo for an account.
     * @param address - User address
     * @returns Observable<MultisigAccountGraphInfo>
     */
    getMultisigAccountGraphInfo(address, requestOptions) {
        return this.getNetworkTypeObservable(requestOptions).pipe((0, operators_1.mergeMap)((networkType) => (0, rxjs_1.from)(this.accountRoutesApi.getAccountMultisigGraph(address.plain(), requestOptions))
            .pipe((0, operators_1.map)((response) => {
            const multisigAccounts = new Map();
            response.body.map((multisigAccountGraphInfoDTO) => {
                multisigAccounts.set(multisigAccountGraphInfoDTO.level, multisigAccountGraphInfoDTO.multisigEntries.map((multisigAccountInfoDTO) => {
                    return new MultisigAccountInfo_1.MultisigAccountInfo(PublicAccount_1.PublicAccount.createFromPublicKey(multisigAccountInfoDTO.multisig.account, networkType), multisigAccountInfoDTO.multisig.minApproval, multisigAccountInfoDTO.multisig.minRemoval, multisigAccountInfoDTO.multisig.cosignatories
                        .map((cosigner) => PublicAccount_1.PublicAccount.createFromPublicKey(cosigner, networkType)), multisigAccountInfoDTO.multisig.multisigAccounts
                        .map((multisigAccountDTO) => PublicAccount_1.PublicAccount.createFromPublicKey(multisigAccountDTO, networkType)));
                }));
            });
            return new MultisigAccountGraphInfo_1.MultisigAccountGraphInfo(multisigAccounts);
        }))));
    }
    /**
     * Gets an array of confirmed transactions for which an account is signer or receiver.
     * @param publicAccount - User public account
     * @param txnQueryParams - (Optional) Transaction Query params
     * @returns Observable<Transaction[]>
     */
    transactions(publicAccount, txnQueryParams, requestOptions) {
        const plainAddress = publicAccount.address.plain();
        return (0, rxjs_1.from)(this.accountRoutesApi.transactions(plainAddress, txnQueryParams, requestOptions)).pipe((0, operators_1.map)((response) => {
            let transactions = [];
            transactions = response.body.data.map((transactionDTO) => {
                return (0, CreateTransactionFromDTO_1.CreateTransactionFromDTO)(transactionDTO);
            });
            return transactions;
        }));
    }
    /**
     * Gets an array of confirmed transactions for which an account is signer or receiver.
     * @param publicAccount - User public account
     * @param txnQueryParams - (Optional) Transaction Query params
     * @returns Observable<TransactionSearch>
     */
    transactionsWithPagination(publicAccount, txnQueryParams, requestOptions) {
        const plainAddress = publicAccount.address.plain();
        return (0, rxjs_1.from)(this.accountRoutesApi.transactions(plainAddress, txnQueryParams, requestOptions)).pipe((0, operators_1.map)((response) => {
            let transactions = [];
            transactions = response.body.data.map((transactionDTO) => {
                return (0, CreateTransactionFromDTO_1.CreateTransactionFromDTO)(transactionDTO);
            });
            let paginationData = new Pagination_1.Pagination(response.body.pagination.totalEntries, response.body.pagination.pageNumber, response.body.pagination.pageSize, response.body.pagination.totalPages);
            return new TransactionSearch_1.TransactionSearch(transactions, paginationData);
        }));
    }
    /**
     * Gets an array of transactions for which an account is the recipient of a transaction.
     * A transaction is said to be incoming with respect to an account if the account is the recipient of a transaction.
     * @param accountId - User public account or address (you can use address if public account is not known to the network just yet)
     * @param txnQueryParams - (Optional) Transaction Query params
     * @returns Observable<Transaction[]>
     */
    incomingTransactions(accountId, txnQueryParams, requestOptions) {
        const plainAddress = accountId instanceof PublicAccount_1.PublicAccount ? accountId.address.plain() : accountId.plain();
        return (0, rxjs_1.from)(this.accountRoutesApi.incomingTransactions(plainAddress, txnQueryParams, requestOptions)).pipe((0, operators_1.map)((response) => {
            if (response.body.data.length) {
                return response.body.data.map((transactionDTO) => {
                    return (0, CreateTransactionFromDTO_1.CreateTransactionFromDTO)(transactionDTO);
                });
            }
            else {
                return [];
            }
        }));
    }
    /**
     * Gets an array of transactions for which an account is the recipient of a transaction.
     * A transaction is said to be incoming with respect to an account if the account is the recipient of a transaction.
     * @param accountId - User public account or address (you can use address if public account is not known to the network just yet)
     * @param txnQueryParams - (Optional) Transaction Query params
     * @returns Observable<TransactionSearch>
     */
    incomingTransactionsWithPagination(accountId, txnQueryParams, requestOptions) {
        const plainAddress = accountId instanceof PublicAccount_1.PublicAccount ? accountId.address.plain() : accountId.plain();
        return (0, rxjs_1.from)(this.accountRoutesApi.incomingTransactions(plainAddress, txnQueryParams, requestOptions)).pipe((0, operators_1.map)((response) => {
            let transactions = [];
            transactions = response.body.data.map((transactionDTO) => {
                return (0, CreateTransactionFromDTO_1.CreateTransactionFromDTO)(transactionDTO);
            });
            let paginationData = new Pagination_1.Pagination(response.body.pagination.totalEntries, response.body.pagination.pageNumber, response.body.pagination.pageSize, response.body.pagination.totalPages);
            return new TransactionSearch_1.TransactionSearch(transactions, paginationData);
        }));
    }
    /**
     * Gets an array of transactions for which an account is the sender a transaction.
     * A transaction is said to be outgoing with respect to an account if the account is the sender of a transaction.
     * @param publicAccount - User public account
     * @param txnQueryParams - (Optional) Transaction Query params
     * @returns Observable<Transaction[]>
     */
    outgoingTransactions(publicAccount, txnQueryParams, requestOptions) {
        return (0, rxjs_1.from)(this.accountRoutesApi.outgoingTransactions(publicAccount.publicKey, txnQueryParams, requestOptions)).pipe((0, operators_1.map)((response) => {
            if (response.body.data.length) {
                return response.body.data.map((transactionDTO) => {
                    return (0, CreateTransactionFromDTO_1.CreateTransactionFromDTO)(transactionDTO);
                });
            }
            else {
                return [];
            }
        }));
    }
    /**
     * Gets an array of transactions for which an account is the sender a transaction.
     * A transaction is said to be outgoing with respect to an account if the account is the sender of a transaction.
     * @param publicAccount - User public account
     * @param txnQueryParams - (Optional) Transaction Query params
     * @returns Observable<TransactionSearch>
     */
    outgoingTransactionsWithPagination(publicAccount, txnQueryParams, requestOptions) {
        return (0, rxjs_1.from)(this.accountRoutesApi.outgoingTransactions(publicAccount.publicKey, txnQueryParams, requestOptions)).pipe((0, operators_1.map)((response) => {
            let transactions = [];
            transactions = response.body.data.map((transactionDTO) => {
                return (0, CreateTransactionFromDTO_1.CreateTransactionFromDTO)(transactionDTO);
            });
            let paginationData = new Pagination_1.Pagination(response.body.pagination.totalEntries, response.body.pagination.pageNumber, response.body.pagination.pageSize, response.body.pagination.totalPages);
            return new TransactionSearch_1.TransactionSearch(transactions, paginationData);
        }));
    }
    /**
     * Gets the array of transactions for which an account is the sender or receiver and which have not yet been included in a block.
     * Unconfirmed transactions are those transactions that have not yet been included in a block.
     * Unconfirmed transactions are not guaranteed to be included in any block.
     * @param publicAccount - User public account
     * @param txnQueryParams - (Optional) Transaction Query params
     * @returns Observable<Transaction[]>
     */
    unconfirmedTransactions(publicAccount, txnQueryParams, requestOptions) {
        const plainAddress = publicAccount.address.plain();
        return (0, rxjs_1.from)(this.accountRoutesApi.unconfirmedTransactions(plainAddress, txnQueryParams, requestOptions)).pipe((0, operators_1.map)((response) => {
            if (response.body.data.length) {
                return response.body.data.map((transactionDTO) => {
                    return (0, CreateTransactionFromDTO_1.CreateTransactionFromDTO)(transactionDTO);
                });
            }
            else {
                return [];
            }
        }));
    }
    /**
     * Gets the array of transactions for which an account is the sender or receiver and which have not yet been included in a block.
     * Unconfirmed transactions are those transactions that have not yet been included in a block.
     * Unconfirmed transactions are not guaranteed to be included in any block.
     * @param publicAccount - User public account
     * @param txnQueryParams - (Optional) Transaction Query params
     * @returns Observable<TransactionSearch>
     */
    unconfirmedTransactionsWithPagination(publicAccount, txnQueryParams, requestOptions) {
        const plainAddress = publicAccount.address.plain();
        return (0, rxjs_1.from)(this.accountRoutesApi.unconfirmedTransactions(plainAddress, txnQueryParams, requestOptions))
            .pipe((0, operators_1.map)((response) => {
            let transactions = [];
            transactions = response.body.data.map((transactionDTO) => {
                return (0, CreateTransactionFromDTO_1.CreateTransactionFromDTO)(transactionDTO);
            });
            let paginationData = new Pagination_1.Pagination(response.body.pagination.totalEntries, response.body.pagination.pageNumber, response.body.pagination.pageSize, response.body.pagination.totalPages);
            return new TransactionSearch_1.TransactionSearch(transactions, paginationData);
        }));
    }
    /**
     * Gets an array of transactions for which an account is the sender or has sign the transaction.
     * A transaction is said to be aggregate bonded with respect to an account if there are missing signatures.
     * @param publicAccount - User public account
     * @param txnQueryParams - (Optional) Transaction Query params
     * @returns Observable<AggregateTransaction[]>
     */
    aggregateBondedTransactions(publicAccount, txnQueryParams, requestOptions) {
        const plainAddress = publicAccount.address.plain();
        // let transactionRoutesApi = getCompleteTransaction ? new TransactionRoutesApi(this.url) : null;
        let firstObservable = (0, rxjs_1.from)(this.accountRoutesApi.partialTransactions(plainAddress, txnQueryParams, requestOptions))
            .pipe((0, operators_1.map)((response) => {
            if (response.body.data.length) {
                return response.body.data.map((transactionDTO) => {
                    return (0, CreateTransactionFromDTO_1.CreateTransactionFromDTO)(transactionDTO);
                });
            }
            else {
                //return new TransactionSearch([], response.body.pagination);
                return [];
            }
        }));
        return firstObservable;
    }
}
exports.AccountHttp = AccountHttp;
