// tsjs-xpx-chain-sdk/dist/src/infrastructure/DownloadChannelQueryParams.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DownloadChannelQueryParams = exports.DownloadChannelFieldOrder = exports.DownloadChannelSortingField = void 0;
const QueryParams_1 = require("./QueryParams");
const model_1 = require("../model/model");
const PaginationQueryParams_1 = require("./PaginationQueryParams");
var DownloadChannelSortingField;
(function (DownloadChannelSortingField) {
    DownloadChannelSortingField["ID"] = "id";
})(DownloadChannelSortingField || (exports.DownloadChannelSortingField = DownloadChannelSortingField = {}));
class DownloadChannelFieldOrder {
    orderBy;
    order;
    constructor(orderBy, order) {
        this.orderBy = orderBy;
        this.order = order;
    }
    static setAscending(orderBy) {
        return new DownloadChannelFieldOrder(orderBy, QueryParams_1.Order_v2.ASC);
    }
    static setDescending(orderBy) {
        return new DownloadChannelFieldOrder(orderBy, QueryParams_1.Order_v2.DESC);
    }
}
exports.DownloadChannelFieldOrder = DownloadChannelFieldOrder;
class DownloadChannelQueryParams extends PaginationQueryParams_1.PaginationQueryParams {
    /**
    * Download size in MB.
    */
    downloadSize;
    ;
    /**
    * From download size in MB.
    */
    fromDownloadSize;
    ;
    /**
    * To download size in MB.
    */
    toDownloadSize;
    ;
    /**
    * Number of completed download approval transactions.
    */
    downloadApprovalCount;
    /**
    * From number of completed download approval transactions.
    */
    fromDownloadApprovalCount;
    /**
    * To number of completed download approval transactions.
    */
    toDownloadApprovalCount;
    consumerKey;
    order;
    orderBy;
    offset;
    /**
     * Constructor
     */
    constructor() {
        super();
    }
    updateFieldOrder(DownloadChannelFieldOrder) {
        this.orderBy = DownloadChannelFieldOrder.orderBy;
        this.order = DownloadChannelFieldOrder.order;
    }
    buildQueryParams() {
        let queryParams = Object.assign({}, this);
        DownloadChannelQueryParams.adjustNonComplianceParams(queryParams);
        DownloadChannelQueryParams.convertToPrimitiveType(queryParams);
        let flattenedQueryParams = super.buildQueryParams(queryParams);
        return flattenedQueryParams;
    }
    buildQueryParamsString() {
        let queryParams = this.buildQueryParams();
        const entries = Object.entries(queryParams);
        let queryParamsArray = entries.map(data => {
            if (data[1] instanceof Array) {
                return data[1].map(arrayData => {
                    return data[0] + "[]" + "=" + arrayData;
                }).join("&");
            }
            return data[0] + "=" + data[1];
        });
        return queryParamsArray.join("&");
    }
    static convertToPrimitiveType(queryParams) {
        if (queryParams.downloadSize) {
            if (queryParams.downloadSize instanceof model_1.UInt64) {
                queryParams.downloadSize = queryParams.downloadSize.toBigInt();
            }
        }
        if (queryParams.fromDownloadSize) {
            if (queryParams.fromDownloadSize instanceof model_1.UInt64) {
                queryParams.fromDownloadSize = queryParams.fromDownloadSize.toBigInt();
            }
        }
        if (queryParams.toDownloadSize) {
            if (queryParams.toDownloadSize instanceof model_1.UInt64) {
                queryParams.toDownloadSize = queryParams.toDownloadSize.toBigInt();
            }
        }
    }
    static adjustNonComplianceParams(queryParams) {
        if (queryParams.downloadApprovalCount) {
            if (queryParams.toDownloadApprovalCount) {
                delete queryParams.toDownloadApprovalCount;
            }
            if (queryParams.fromDownloadApprovalCount) {
                delete queryParams.fromDownloadApprovalCount;
            }
        }
        if (queryParams.pageSize) {
            if (queryParams.pageSize < 0) {
                delete queryParams.pageSize; // default to 20
            }
            else if (queryParams.pageSize < 10) {
                queryParams.pageSize = 10;
            }
            else if (queryParams.pageSize > 100) {
                queryParams.pageSize = 100;
            }
        }
        if (queryParams.pageNumber) {
            if (queryParams.pageNumber <= 0) {
                queryParams.pageNumber = 1;
            }
        }
    }
}
exports.DownloadChannelQueryParams = DownloadChannelQueryParams;
