// tsjs-xpx-chain-sdk/dist/src/infrastructure/DriveQueryParams.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DriveQueryParams = exports.DriveFieldOrder = exports.DriveSortingField = void 0;
const PublicAccount_1 = require("../model/account/PublicAccount");
const QueryParams_1 = require("./QueryParams");
const model_1 = require("../model/model");
const PaginationQueryParams_1 = require("./PaginationQueryParams");
var DriveSortingField;
(function (DriveSortingField) {
    DriveSortingField["ID"] = "id";
})(DriveSortingField || (exports.DriveSortingField = DriveSortingField = {}));
class DriveFieldOrder {
    orderBy;
    order;
    constructor(orderBy, order) {
        this.orderBy = orderBy;
        this.order = order;
    }
    static setAscending(orderBy) {
        return new DriveFieldOrder(orderBy, QueryParams_1.Order_v2.ASC);
    }
    static setDescending(orderBy) {
        return new DriveFieldOrder(orderBy, QueryParams_1.Order_v2.DESC);
    }
}
exports.DriveFieldOrder = DriveFieldOrder;
class DriveQueryParams extends PaginationQueryParams_1.PaginationQueryParams {
    /**
    * Size of drive (MB).
    */
    size;
    /**
    * From size of drive (MB).
    */
    fromSize;
    /**
    * To size of drive (MB).
    */
    toSize;
    /**
    * Used size of drive.
    */
    usedSize;
    /**
    * From used size of drive.
    */
    fromUsedSize;
    /**
    * To used size of drive.
    */
    toUsedSize;
    /**
     * Meta file size of drive.
    */
    metaFilesSize;
    /**
     * From meta file size of drive.
    */
    fromMetaFilesSize;
    /**
     * To meta file size of drive.
    */
    toMetaFilesSize;
    /**
     * Number of replicators.
    */
    replicatorCount;
    /**
     * From number of replicators.
    */
    fromReplicatorCount;
    /**
     * To number of replicators.
    */
    toReplicatorCount;
    /**
     * Public key of an Owner.
     */
    owner;
    order;
    orderBy;
    offset;
    /**
     * Constructor
     */
    constructor() {
        super();
    }
    updateFieldOrder(DriveFieldOrder) {
        this.orderBy = DriveFieldOrder.orderBy;
        this.order = DriveFieldOrder.order;
    }
    buildQueryParams() {
        let queryParams = Object.assign({}, this);
        DriveQueryParams.adjustNonComplianceParams(queryParams);
        DriveQueryParams.convertToPrimitiveType(queryParams);
        let flattenedQueryParams = super.buildQueryParams(queryParams);
        return flattenedQueryParams;
    }
    buildQueryParamsString() {
        let queryParams = this.buildQueryParams();
        const entries = Object.entries(queryParams);
        let queryParamsArray = entries.map(data => {
            if (data[1] instanceof Array) {
                return data[1].map(arrayData => {
                    return data[0] + "[]" + "=" + arrayData;
                }).join("&");
            }
            return data[0] + "=" + data[1];
        });
        return queryParamsArray.join("&");
    }
    static convertToPrimitiveType(queryParams) {
        if (queryParams.size) {
            if (queryParams.size instanceof model_1.UInt64) {
                queryParams.size = queryParams.size.toBigInt();
            }
        }
        if (queryParams.fromSize) {
            if (queryParams.fromSize instanceof model_1.UInt64) {
                queryParams.fromSize = queryParams.fromSize.toBigInt();
            }
        }
        if (queryParams.toSize) {
            if (queryParams.toSize instanceof model_1.UInt64) {
                queryParams.toSize = queryParams.toSize.toBigInt();
            }
        }
        if (queryParams.usedSize) {
            if (queryParams.usedSize instanceof model_1.UInt64) {
                queryParams.usedSize = queryParams.usedSize.toBigInt();
            }
        }
        if (queryParams.fromUsedSize) {
            if (queryParams.fromUsedSize instanceof model_1.UInt64) {
                queryParams.fromUsedSize = queryParams.fromUsedSize.toBigInt();
            }
        }
        if (queryParams.toUsedSize) {
            if (queryParams.toUsedSize instanceof model_1.UInt64) {
                queryParams.toSize = queryParams.toUsedSize.toBigInt();
            }
        }
        if (queryParams.metaFilesSize) {
            if (queryParams.metaFilesSize instanceof model_1.UInt64) {
                queryParams.metaFilesSize = queryParams.metaFilesSize.toBigInt();
            }
        }
        if (queryParams.fromMetaFilesSize) {
            if (queryParams.fromMetaFilesSize instanceof model_1.UInt64) {
                queryParams.fromMetaFilesSize = queryParams.fromMetaFilesSize.toBigInt();
            }
        }
        if (queryParams.toMetaFilesSize) {
            if (queryParams.toMetaFilesSize instanceof model_1.UInt64) {
                queryParams.toMetaFilesSize = queryParams.toMetaFilesSize.toBigInt();
            }
        }
        if (queryParams.owner) {
            if (queryParams.owner instanceof PublicAccount_1.PublicAccount) {
                queryParams.owner = queryParams.owner.publicKey;
            }
        }
    }
    static adjustNonComplianceParams(queryParams) {
        if (queryParams.size) {
            if (queryParams.fromSize) {
                delete queryParams.fromSize;
            }
            if (queryParams.toSize) {
                delete queryParams.toSize;
            }
        }
        if (queryParams.usedSize) {
            if (queryParams.fromUsedSize) {
                delete queryParams.fromUsedSize;
            }
            if (queryParams.toUsedSize) {
                delete queryParams.toUsedSize;
            }
        }
        if (queryParams.metaFilesSize) {
            if (queryParams.fromMetaFilesSize) {
                delete queryParams.fromMetaFilesSize;
            }
            if (queryParams.toMetaFilesSize) {
                delete queryParams.toMetaFilesSize;
            }
        }
        if (queryParams.pageSize) {
            if (queryParams.pageSize < 0) {
                delete queryParams.pageSize; // default to 20
            }
            else if (queryParams.pageSize < 10) {
                queryParams.pageSize = 10;
            }
            else if (queryParams.pageSize > 100) {
                queryParams.pageSize = 100;
            }
        }
        if (queryParams.pageNumber) {
            if (queryParams.pageNumber <= 0) {
                queryParams.pageNumber = 1;
            }
        }
    }
}
exports.DriveQueryParams = DriveQueryParams;
