// tsjs-xpx-chain-sdk/dist/src/infrastructure/MetadataQueryParams.js
"use strict";
/*
 * Copyright 2021 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetadataQueryParams = exports.MetadataFieldOrder = exports.MetadataSortingField = void 0;
const QueryParams_1 = require("./QueryParams");
const Address_1 = require("../model/account/Address");
const NamespaceId_1 = require("../model/namespace/NamespaceId");
const MosaicId_1 = require("../model/mosaic/MosaicId");
const PublicAccount_1 = require("../model/account/PublicAccount");
const UInt64_1 = require("../model/UInt64");
var MetadataSortingField;
(function (MetadataSortingField) {
    MetadataSortingField["VALUE"] = "metadataEntry.value";
    MetadataSortingField["VALUE_SIZE"] = "metadataEntry.valueSize";
})(MetadataSortingField || (exports.MetadataSortingField = MetadataSortingField = {}));
class MetadataFieldOrder {
    order;
    sortingField;
    constructor(order, sortingField) {
        this.order = order;
        this.sortingField = sortingField;
    }
    static setAscending(sortingField) {
        return new MetadataFieldOrder(QueryParams_1.Order_v2.ASC, sortingField);
    }
    static setDescending(sortingField) {
        return new MetadataFieldOrder(QueryParams_1.Order_v2.DESC, sortingField);
    }
}
exports.MetadataFieldOrder = MetadataFieldOrder;
class MetadataQueryParams {
    pageSize;
    pageNumber;
    metadataType;
    order;
    sortField;
    sourceAddress;
    targetId; // mosaic/ namespace hex string
    targetKey; // public key
    scopedMetadataKey; // hex string
    /**
     * Constructor
     */
    constructor() {
    }
    updateFieldOrder(fieldOrder) {
        this.sortField = fieldOrder.sortingField;
        this.order = fieldOrder.order;
    }
    buildQueryParams() {
        let queryParams = Object.assign({}, this);
        MetadataQueryParams.adjustNonComplianceParams(queryParams);
        MetadataQueryParams.convertToPrimitiveType(queryParams);
        return queryParams;
    }
    buildQueryParamsString() {
        let queryParams = this.buildQueryParams();
        const entries = Object.entries(queryParams);
        let queryParamsArray = entries.map(data => {
            if (data[1] instanceof Array) {
                return data[1].map(arrayData => {
                    return data[0] + "[]" + "=" + arrayData;
                }).join("&");
            }
            return data[0] + "=" + data[1];
        });
        return queryParamsArray.join("&");
    }
    static convertToPrimitiveType(queryParams) {
        if (queryParams.sourceAddress) {
            if (queryParams.sourceAddress instanceof Address_1.Address) {
                queryParams.sourceAddress = queryParams.sourceAddress.plain();
            }
        }
        if (queryParams.targetId) {
            if (queryParams.targetId instanceof NamespaceId_1.NamespaceId) {
                queryParams.targetId = queryParams.targetId.toHex();
            }
            else if (queryParams.targetId instanceof MosaicId_1.MosaicId) {
                queryParams.targetId = queryParams.targetId.toHex();
            }
        }
        if (queryParams.targetKey) {
            if (queryParams.targetKey instanceof PublicAccount_1.PublicAccount) {
                queryParams.targetKey = queryParams.targetKey.publicKey;
            }
        }
        if (queryParams.scopedMetadataKey) {
            if (queryParams.scopedMetadataKey instanceof UInt64_1.UInt64) {
                queryParams.scopedMetadataKey = queryParams.scopedMetadataKey.toHex();
            }
        }
    }
    static adjustNonComplianceParams(queryParams) {
        if (queryParams.pageSize) {
            if (queryParams.pageSize < 0) {
                delete queryParams.pageSize; // default to 20
            }
            else if (queryParams.pageSize < 10) {
                queryParams.pageSize = 10;
            }
            else if (queryParams.pageSize > 100) {
                queryParams.pageSize = 100;
            }
        }
        if (queryParams.pageNumber) {
            if (queryParams.pageNumber <= 0) {
                queryParams.pageNumber = 1;
            }
        }
        if (queryParams.order === undefined || queryParams.sortField === undefined) {
            if (queryParams.order) {
                delete queryParams.order;
            }
            if (queryParams.sortField) {
                delete queryParams.sortField;
            }
        }
    }
}
exports.MetadataQueryParams = MetadataQueryParams;
