// tsjs-xpx-chain-sdk/dist/src/infrastructure/MosaicQueryParams.js
"use strict";
/*
 * Copyright 2021 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MosaicQueryParams = exports.MosaicCreatorFilters = exports.MosaicFieldOrder = exports.MosaicSortingField = void 0;
const PublicAccount_1 = require("../model/account/PublicAccount");
const QueryParams_1 = require("./QueryParams");
const PaginationQueryParams_1 = require("./PaginationQueryParams");
var MosaicSortingField;
(function (MosaicSortingField) {
    MosaicSortingField["BLOCK_HEIGHT"] = "mosaic.height";
    MosaicSortingField["SUPPLY"] = "mosaic.supply";
})(MosaicSortingField || (exports.MosaicSortingField = MosaicSortingField = {}));
class MosaicFieldOrder {
    sortingField;
    order;
    constructor(sortingField, order) {
        this.sortingField = sortingField;
        this.order = order;
    }
    static setAscending(sortingField) {
        return new MosaicFieldOrder(sortingField, QueryParams_1.Order_v2.ASC);
    }
    static setDescending(sortingField) {
        return new MosaicFieldOrder(sortingField, QueryParams_1.Order_v2.DESC);
    }
}
exports.MosaicFieldOrder = MosaicFieldOrder;
class MosaicCreatorFilters {
    ownerPubKey;
    /**
     * optional when ownerPubKey defined
     * true = returns only mosaics creator holding
     * false = returns only mosaics creator not holding
    */
    holding;
    constructor(ownerPublicKey, holding) {
        this.ownerPubKey = ownerPublicKey;
        if (holding !== undefined) {
            this.holding = holding;
        }
    }
    buildQueryParamsString() {
        let queryParams = this.buildQueryParams();
        const entries = Object.entries(queryParams);
        let queryParamsArray = entries.map(data => {
            return data[0] + "=" + data[1];
        });
        return queryParamsArray.join("&");
    }
    buildQueryParams() {
        let queryParams = Object.assign({}, this);
        MosaicCreatorFilters.convertToPrimitiveType(queryParams);
        return queryParams;
    }
    static convertToPrimitiveType(queryParams) {
        if (queryParams.ownerPubKey) {
            if (queryParams.ownerPubKey instanceof PublicAccount_1.PublicAccount) {
                queryParams.ownerPubKey = queryParams.ownerPubKey.publicKey;
            }
        }
    }
}
exports.MosaicCreatorFilters = MosaicCreatorFilters;
class MosaicQueryParams extends PaginationQueryParams_1.PaginationQueryParams {
    order;
    sortField;
    ownerFilters;
    mutable;
    transferable;
    supply;
    /**
     * Constructor
     */
    constructor() {
        super();
    }
    updateFieldOrder(fieldOrder) {
        this.sortField = fieldOrder.sortingField;
        this.order = fieldOrder.order;
    }
    buildQueryParams() {
        let queryParams = Object.assign({}, this);
        MosaicQueryParams.adjustNonComplianceParams(queryParams);
        let flattenedQueryParams = super.buildQueryParams(queryParams);
        return flattenedQueryParams;
    }
    buildQueryParamsString() {
        let queryParams = this.buildQueryParams();
        const entries = Object.entries(queryParams);
        let queryParamsArray = entries.map(data => {
            if (data[1] instanceof Array) {
                return data[1].map(arrayData => {
                    return data[0] + "[]" + "=" + arrayData;
                }).join("&");
            }
            return data[0] + "=" + data[1];
        });
        return queryParamsArray.join("&");
    }
    static adjustNonComplianceParams(queryParams) {
        if (queryParams.pageSize) {
            if (queryParams.pageSize < 0) {
                delete queryParams.pageSize; // default to 20
            }
            else if (queryParams.pageSize < 10) {
                queryParams.pageSize = 10;
            }
            else if (queryParams.pageSize > 100) {
                queryParams.pageSize = 100;
            }
        }
        if (queryParams.pageNumber) {
            if (queryParams.pageNumber <= 0) {
                queryParams.pageNumber = 1;
            }
        }
        if (queryParams.order === undefined || queryParams.sortField === undefined) {
            if (queryParams.order) {
                delete queryParams.order;
            }
            if (queryParams.sortField) {
                delete queryParams.sortField;
            }
        }
    }
}
exports.MosaicQueryParams = MosaicQueryParams;
