// tsjs-xpx-chain-sdk/dist/src/infrastructure/MosaicHttp.js
"use strict";
/*
 * Copyright 2023 ProximaX
 * Copyright 2018 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MosaicHttp = void 0;
const rxjs_1 = require("rxjs");
const operators_1 = require("rxjs/operators");
const PublicAccount_1 = require("../model/account/PublicAccount");
const MosaicId_1 = require("../model/mosaic/MosaicId");
const MosaicInfo_1 = require("../model/mosaic/MosaicInfo");
const MosaicNames_1 = require("../model/mosaic/MosaicNames");
const MosaicProperties_1 = require("../model/mosaic/MosaicProperties");
const MosaicPropertyType_1 = require("../model/mosaic/MosaicPropertyType");
const NamespaceId_1 = require("../model/namespace/NamespaceId");
const NamespaceName_1 = require("../model/namespace/NamespaceName");
const UInt64_1 = require("../model/UInt64");
const api_1 = require("./api");
const Http_1 = require("./Http");
const NetworkHttp_1 = require("./NetworkHttp");
const model_1 = require("../model/model");
const MosaicLevy_1 = require("../model/mosaic/MosaicLevy");
const Pagination_1 = require("../model/Pagination");
/**
 * Mosaic http repository.
 *
 * @since 1.0
 */
class MosaicHttp extends Http_1.Http {
    /**
     * @internal
     * xpx chain Library mosaic routes api
     */
    mosaicRoutesApi;
    /**
     * Constructor
     * @param url
     * @param networkHttp
     */
    constructor(url, networkHttp) {
        networkHttp = networkHttp == null ? new NetworkHttp_1.NetworkHttp(url) : networkHttp;
        super(networkHttp);
        this.mosaicRoutesApi = new api_1.MosaicRoutesApi(url);
    }
    /**
     * Gets the MosaicInfo for a given mosaicId
     * @param mosaicId - Mosaic id
     * @returns Observable<MosaicInfo>
     */
    getMosaic(mosaicId, requestOptions) {
        return this.getNetworkTypeObservable(requestOptions).pipe((0, operators_1.mergeMap)((networkType) => (0, rxjs_1.from)(this.mosaicRoutesApi.getMosaic(mosaicId.toHex(), requestOptions))
            .pipe((0, operators_1.map)((response) => {
            const mosaicInfoDTO = response.body;
            let mosaicFlag;
            let divisibility;
            let duration;
            if (mosaicInfoDTO.mosaic.properties[MosaicPropertyType_1.MosaicPropertyType.MosaicFlags].value) {
                mosaicFlag = mosaicInfoDTO.mosaic.properties[MosaicPropertyType_1.MosaicPropertyType.MosaicFlags].value;
            }
            if (mosaicInfoDTO.mosaic.properties[MosaicPropertyType_1.MosaicPropertyType.Divisibility].value) {
                divisibility = mosaicInfoDTO.mosaic.properties[MosaicPropertyType_1.MosaicPropertyType.Divisibility].value;
            }
            if (mosaicInfoDTO.mosaic.properties[MosaicPropertyType_1.MosaicPropertyType.Duration].value) {
                duration = mosaicInfoDTO.mosaic.properties[MosaicPropertyType_1.MosaicPropertyType.Duration].value;
            }
            return new MosaicInfo_1.MosaicInfo(mosaicInfoDTO.meta.id, new MosaicId_1.MosaicId(mosaicInfoDTO.mosaic.mosaicId), new UInt64_1.UInt64(mosaicInfoDTO.mosaic.supply), new UInt64_1.UInt64(mosaicInfoDTO.mosaic.height), PublicAccount_1.PublicAccount.createFromPublicKey(mosaicInfoDTO.mosaic.owner, networkType), mosaicInfoDTO.mosaic.revision, new MosaicProperties_1.MosaicProperties(mosaicFlag ? new UInt64_1.UInt64(mosaicFlag) : UInt64_1.UInt64.fromUint(0), (divisibility ? new UInt64_1.UInt64(divisibility) : UInt64_1.UInt64.fromUint(0)).compact(), duration ? new UInt64_1.UInt64(duration) : undefined));
        }))));
    }
    /**
     * Gets MosaicInfo for different mosaicIds.
     * @param mosaicIds - Array of mosaic ids
     * @returns Observable<MosaicInfo[]>
     */
    getMosaics(mosaicIds, requestOptions) {
        const mosaicIdsBody = {
            mosaicIds: mosaicIds.map((id) => id.toHex()),
        };
        return this.getNetworkTypeObservable(requestOptions).pipe((0, operators_1.mergeMap)((networkType) => (0, rxjs_1.from)(this.mosaicRoutesApi.getMosaics(mosaicIdsBody, requestOptions))
            .pipe((0, operators_1.map)((response) => {
            return response.body.map((mosaicInfoDTO) => {
                let mosaicFlag;
                let divisibility;
                let duration;
                if (mosaicInfoDTO.mosaic.properties[MosaicPropertyType_1.MosaicPropertyType.MosaicFlags].value) {
                    mosaicFlag = mosaicInfoDTO.mosaic.properties[MosaicPropertyType_1.MosaicPropertyType.MosaicFlags].value;
                }
                if (mosaicInfoDTO.mosaic.properties[MosaicPropertyType_1.MosaicPropertyType.Divisibility].value) {
                    divisibility = mosaicInfoDTO.mosaic.properties[MosaicPropertyType_1.MosaicPropertyType.Divisibility].value;
                }
                if (mosaicInfoDTO.mosaic.properties[MosaicPropertyType_1.MosaicPropertyType.Duration].value) {
                    duration = mosaicInfoDTO.mosaic.properties[MosaicPropertyType_1.MosaicPropertyType.Duration].value;
                }
                return new MosaicInfo_1.MosaicInfo(mosaicInfoDTO.meta.id, new MosaicId_1.MosaicId(mosaicInfoDTO.mosaic.mosaicId), new UInt64_1.UInt64(mosaicInfoDTO.mosaic.supply), new UInt64_1.UInt64(mosaicInfoDTO.mosaic.height), PublicAccount_1.PublicAccount.createFromPublicKey(mosaicInfoDTO.mosaic.owner, networkType), mosaicInfoDTO.mosaic.revision, new MosaicProperties_1.MosaicProperties(mosaicFlag ? new UInt64_1.UInt64(mosaicFlag) : UInt64_1.UInt64.fromUint(0), (divisibility ? new UInt64_1.UInt64(divisibility) : UInt64_1.UInt64.fromUint(0)).compact(), duration ? new UInt64_1.UInt64(duration) : undefined));
            });
        }))));
    }
    /**
     * Get readable names for a set of mosaics
     * Returns friendly names for mosaics.
     * @param mosaicIds - Array of mosaic ids
     * @return Observable<MosaicNames[]>
     */
    getMosaicsNames(mosaicIds, requestOptions) {
        const mosaicIdsBody = {
            mosaicIds: mosaicIds.map((id) => id.toHex()),
        };
        return (0, rxjs_1.from)(this.mosaicRoutesApi.getMosaicsNames(mosaicIdsBody, requestOptions))
            .pipe((0, operators_1.map)((response) => {
            return response.body.map((mosaic) => {
                return new MosaicNames_1.MosaicNames(new MosaicId_1.MosaicId(mosaic.mosaicId), mosaic.names.map((name) => {
                    return new NamespaceName_1.NamespaceName(new NamespaceId_1.NamespaceId(name), name);
                }));
            });
        }));
    }
    /**
     * Gets mosaic richlist
     * @param mosaicId - Mosaic id
     * @param queryParams - (Optional) Page query params
     * @returns Observable<RichlistEntry[]>
     */
    getMosaicRichlist(mosaicId, queryParams, requestOptions) {
        return (0, rxjs_1.from)(this.mosaicRoutesApi.getMosaicRichList(mosaicId.toHex(), this.pageQueryParams(queryParams).page, this.pageQueryParams(queryParams).pageSize, requestOptions)).pipe((0, operators_1.map)((response) => {
            return response.body.map((richlistEntryDTO) => {
                return model_1.RichlistEntry.create(model_1.Address.createFromEncoded(richlistEntryDTO.address), richlistEntryDTO.publicKey ? richlistEntryDTO.publicKey : '0'.repeat(64), new UInt64_1.UInt64(richlistEntryDTO.amount));
            });
        }));
    }
    /**
     * Gets mosaic levy
     * @param mosaicId - Mosaic id
     * @returns Observable<MosaicLevy>
     */
    getMosaicLevy(mosaicId, requestOptions) {
        return (0, rxjs_1.from)(this.mosaicRoutesApi.getMosaicLevy(mosaicId.toHex(), requestOptions)).pipe((0, operators_1.map)((mosaicLevyDTO) => {
            return new MosaicLevy_1.MosaicLevy(mosaicLevyDTO.body.type, model_1.Address.createFromEncoded(mosaicLevyDTO.body.recipient), new MosaicId_1.MosaicId(mosaicLevyDTO.body.mosaicId), new UInt64_1.UInt64(mosaicLevyDTO.body.fee));
        }));
    }
    /**
     * Gets MosaicInfo for different mosaicIds.
     * @param mosaicQueryParams - conditions of the mosaic search
     * @returns Observable<MosaicInfo[]>
     */
    searchMosaics(mosaicQueryParams, requestOptions) {
        return this.getNetworkTypeObservable(requestOptions).pipe((0, operators_1.mergeMap)((networkType) => (0, rxjs_1.from)(this.mosaicRoutesApi.searchMosaics(mosaicQueryParams, requestOptions))
            .pipe((0, operators_1.map)((response) => {
            let mosaicsInfo = response.body.data.map((mosaicInfoDTO) => {
                let mosaicFlag;
                let divisibility;
                let duration;
                if (mosaicInfoDTO.mosaic.properties[MosaicPropertyType_1.MosaicPropertyType.MosaicFlags].value) {
                    mosaicFlag = mosaicInfoDTO.mosaic.properties[MosaicPropertyType_1.MosaicPropertyType.MosaicFlags].value;
                }
                if (mosaicInfoDTO.mosaic.properties[MosaicPropertyType_1.MosaicPropertyType.Divisibility].value) {
                    divisibility = mosaicInfoDTO.mosaic.properties[MosaicPropertyType_1.MosaicPropertyType.Divisibility].value;
                }
                if (mosaicInfoDTO.mosaic.properties[MosaicPropertyType_1.MosaicPropertyType.Duration].value) {
                    duration = mosaicInfoDTO.mosaic.properties[MosaicPropertyType_1.MosaicPropertyType.Duration].value;
                }
                return new MosaicInfo_1.MosaicInfo(mosaicInfoDTO.meta.id, new MosaicId_1.MosaicId(mosaicInfoDTO.mosaic.mosaicId), new UInt64_1.UInt64(mosaicInfoDTO.mosaic.supply), new UInt64_1.UInt64(mosaicInfoDTO.mosaic.height), PublicAccount_1.PublicAccount.createFromPublicKey(mosaicInfoDTO.mosaic.owner, networkType), mosaicInfoDTO.mosaic.revision, new MosaicProperties_1.MosaicProperties(mosaicFlag ? new UInt64_1.UInt64(mosaicFlag) : UInt64_1.UInt64.fromUint(0), (divisibility ? new UInt64_1.UInt64(divisibility) : UInt64_1.UInt64.fromUint(0)).compact(), duration ? new UInt64_1.UInt64(duration) : undefined));
            });
            let paginationData = new Pagination_1.Pagination(response.body.pagination.totalEntries, response.body.pagination.pageNumber, response.body.pagination.pageSize, response.body.pagination.totalPages);
            return new model_1.MosaicSearch(mosaicsInfo, paginationData);
        }))));
    }
}
exports.MosaicHttp = MosaicHttp;
