// tsjs-xpx-chain-sdk/dist/src/infrastructure/ReplicatorQueryParams.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReplicatorQueryParams = exports.ReplicatorFieldOrder = exports.ReplicatorSortingField = void 0;
const QueryParams_1 = require("./QueryParams");
const model_1 = require("../model/model");
const PaginationQueryParams_1 = require("./PaginationQueryParams");
var ReplicatorSortingField;
(function (ReplicatorSortingField) {
    ReplicatorSortingField["ID"] = "id";
})(ReplicatorSortingField || (exports.ReplicatorSortingField = ReplicatorSortingField = {}));
class ReplicatorFieldOrder {
    orderBy;
    order;
    constructor(orderBy, order) {
        this.orderBy = orderBy;
        this.order = order;
    }
    static setAscending(orderBy) {
        return new ReplicatorFieldOrder(orderBy, QueryParams_1.Order_v2.ASC);
    }
    static setDescending(orderBy) {
        return new ReplicatorFieldOrder(orderBy, QueryParams_1.Order_v2.DESC);
    }
}
exports.ReplicatorFieldOrder = ReplicatorFieldOrder;
class ReplicatorQueryParams extends PaginationQueryParams_1.PaginationQueryParams {
    /**
    * Version of replicator.
    */
    version;
    /**
    * From version of replicator.
    */
    fromVersion;
    /**
    * To version of replicator.
    */
    toVersion;
    /**
    * Storage size that the replicator provides.
    */
    capacity;
    /**
    * From storage size that the replicator provides.
    */
    fromCapacity;
    /**
    * To storage size that the replicator provides.
    */
    toCapacity;
    offset;
    orderBy;
    order;
    /**
     * Constructor
     */
    constructor() {
        super();
    }
    updateFieldOrder(replicatorFieldOrder) {
        this.orderBy = replicatorFieldOrder.orderBy;
        this.order = replicatorFieldOrder.order;
    }
    buildQueryParams() {
        let queryParams = Object.assign({}, this);
        ReplicatorQueryParams.adjustNonComplianceParams(queryParams);
        ReplicatorQueryParams.convertToPrimitiveType(queryParams);
        let flattenedQueryParams = super.buildQueryParams(queryParams);
        return flattenedQueryParams;
    }
    buildQueryParamsString() {
        let queryParams = this.buildQueryParams();
        const entries = Object.entries(queryParams);
        let queryParamsArray = entries.map(data => {
            if (data[1] instanceof Array) {
                return data[1].map(arrayData => {
                    return data[0] + "[]" + "=" + arrayData;
                }).join("&");
            }
            return data[0] + "=" + data[1];
        });
        return queryParamsArray.join("&");
    }
    static convertToPrimitiveType(queryParams) {
        if (queryParams.capacity) {
            if (queryParams.capacity instanceof model_1.UInt64) {
                queryParams.capacity = queryParams.capacity.toBigInt();
            }
        }
        if (queryParams.toCapacity) {
            if (queryParams.toCapacity instanceof model_1.UInt64) {
                queryParams.toCapacity = queryParams.toCapacity.toBigInt();
            }
        }
        if (queryParams.fromCapacity) {
            if (queryParams.fromCapacity instanceof model_1.UInt64) {
                queryParams.fromCapacity = queryParams.fromCapacity.toBigInt();
            }
        }
    }
    static adjustNonComplianceParams(queryParams) {
        if (queryParams.capacity) {
            if (queryParams.toCapacity) {
                delete queryParams.toCapacity;
            }
            if (queryParams.fromCapacity) {
                delete queryParams.fromCapacity;
            }
        }
        if (queryParams.version) {
            if (queryParams.toVersion) {
                delete queryParams.toVersion;
            }
            if (queryParams.fromVersion) {
                delete queryParams.fromVersion;
            }
        }
        if (queryParams.pageSize) {
            if (queryParams.pageSize < 0) {
                delete queryParams.pageSize; // default to 20
            }
            else if (queryParams.pageSize < 10) {
                queryParams.pageSize = 10;
            }
            else if (queryParams.pageSize > 100) {
                queryParams.pageSize = 100;
            }
        }
        if (queryParams.pageNumber) {
            if (queryParams.pageNumber <= 0) {
                queryParams.pageNumber = 1;
            }
        }
    }
}
exports.ReplicatorQueryParams = ReplicatorQueryParams;
