// tsjs-xpx-chain-sdk/dist/src/infrastructure/StorageHttp.js
"use strict";
// Copyright 2023 ProximaX Limited. All rights reserved.
// Use of this source code is governed by the Apache 2.0
// license that can be found in the LICENSE file
Object.defineProperty(exports, "__esModule", { value: true });
exports.StorageHttp = void 0;
const rxjs_1 = require("rxjs");
const operators_1 = require("rxjs/operators");
const apis_1 = require("./api/apis");
const NetworkHttp_1 = require("./NetworkHttp");
const Http_1 = require("./Http");
const model_1 = require("../model/model");
const Pagination_1 = require("../model/Pagination");
/**
 * Storage http repository.
 *
 * @since 0.9.0
 */
class StorageHttp extends Http_1.Http {
    /**
     * @internal
     * xpx chain Library Storage routes api
     */
    storageRoutesApi;
    /**
     * Constructor
     * @param url
     * @param networkHttp
     */
    constructor(url, networkHttp) {
        networkHttp = networkHttp == null ? new NetworkHttp_1.NetworkHttp(url) : networkHttp;
        super(networkHttp);
        this.storageRoutesApi = new apis_1.StorageRoutesApi(url);
    }
    /**
     * Get the drive for a given accountId
     * @param accountId - public key
     * @returns Observable<DriveInfo | undefined>
     */
    getBcDrive(accountId, requestOptions) {
        let accountIdArg = "";
        if (accountId instanceof model_1.PublicAccount) {
            accountIdArg = accountId.publicKey;
        }
        else if (accountId instanceof model_1.Address) {
            accountIdArg = accountId.plain();
        }
        else {
            accountIdArg = accountId;
        }
        return (0, rxjs_1.from)(this.storageRoutesApi.getBcDrive(accountIdArg, requestOptions))
            .pipe((0, operators_1.map)(response => model_1.DriveInfo.createFromDriveDTO(response.body.drive)));
    }
    /**
     * Get Drives info with query params.
     * @param driveQueryParams - conditions of the drive info search
     * @returns Observable<DriveInfoSearch>
     */
    searchBcDrives(queryParams, requestOptions) {
        return (0, rxjs_1.from)(this.storageRoutesApi.searchBcDrives(queryParams, requestOptions)).pipe((0, operators_1.map)((response) => {
            let driveInfos = response.body.data.map((driveInfoInlineDTO) => {
                return model_1.DriveInfo.createFromDriveDTO(driveInfoInlineDTO.drive, driveInfoInlineDTO.meta.id);
            });
            let paginationData = new Pagination_1.Pagination(response.body.pagination.totalEntries, response.body.pagination.pageNumber, response.body.pagination.pageSize, response.body.pagination.totalPages);
            return new model_1.DriveInfoSearch(driveInfos, paginationData);
        }));
    }
    /**
     * Get the download channel
     * @param channelId - Download channel public key or address
     * @returns Observable<DownloadChannel | undefined>
     */
    getDownloadChannel(channelId, requestOptions) {
        const accountIdArg = (channelId instanceof model_1.PublicAccount) ? channelId.publicKey : channelId;
        return (0, rxjs_1.from)(this.storageRoutesApi.getDownloadChannel(accountIdArg, requestOptions))
            .pipe((0, operators_1.map)(response => model_1.DownloadChannel.createFromDownloadChannelInfoDTO(response.body.downloadChannelInfo)));
    }
    /**
     * Get Download Channels info with query params.
     * @param downloadChannelQueryParams - conditions of the Download Channel info search
     * @returns Observable<DownloadChannelSearch>
     */
    searchDownloadChannels(queryParams, requestOptions) {
        return (0, rxjs_1.from)(this.storageRoutesApi.searchDownloadChannel(queryParams, requestOptions)).pipe((0, operators_1.map)((response) => {
            let downloadChannels = response.body.data.map((downloadChannelInlineDTO) => {
                return model_1.DownloadChannel.createFromDownloadChannelInfoDTO(downloadChannelInlineDTO.downloadChannelInfo, downloadChannelInlineDTO.meta.id);
            });
            let paginationData = new Pagination_1.Pagination(response.body.pagination.totalEntries, response.body.pagination.pageNumber, response.body.pagination.pageSize, response.body.pagination.totalPages);
            return new model_1.DownloadChannelSearch(downloadChannels, paginationData);
        }));
    }
    /**
     * Gets the exchanges for a given accountId
     * @param accountId - Account public key or address
     * @returns Observable<Replicator | undefined>
     */
    getReplicator(accountId, requestOptions) {
        const accountIdArg = accountId instanceof model_1.PublicAccount ? accountId.publicKey : accountId;
        return (0, rxjs_1.from)(this.storageRoutesApi.getReplicator(accountIdArg, requestOptions))
            .pipe((0, operators_1.map)(response => model_1.Replicator.createFromReplicatorDTO(response.body.replicator)));
    }
    /**
     * Get replicator info with query params.
     * @param replicatorQueryParams - conditions of the replicator info search
     * @returns Observable<ReplicatorSearch>
     */
    searchReplicators(queryParams, requestOptions) {
        return (0, rxjs_1.from)(this.storageRoutesApi.searchReplicators(queryParams, requestOptions)).pipe((0, operators_1.map)((response) => {
            let replicators = response.body.data.map((replicatorInlineDTO) => {
                return model_1.Replicator.createFromReplicatorDTO(replicatorInlineDTO.replicator, replicatorInlineDTO.meta.id);
            });
            let paginationData = new Pagination_1.Pagination(response.body.pagination.totalEntries, response.body.pagination.pageNumber, response.body.pagination.pageSize, response.body.pagination.totalPages);
            return new model_1.ReplicatorSearch(replicators, paginationData);
        }));
    }
}
exports.StorageHttp = StorageHttp;
