// tsjs-xpx-chain-sdk/dist/src/infrastructure/TransactionHttp.js
"use strict";
/*
 * Copyright 2023 ProximaX
 * Copyright 2018 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionHttp = void 0;
const rxjs_1 = require("rxjs");
const operators_1 = require("rxjs/operators");
const Deadline_1 = require("../model/transaction/Deadline");
const TransactionAnnounceResponse_1 = require("../model/transaction/TransactionAnnounceResponse");
const TransactionStatus_1 = require("../model/transaction/TransactionStatus");
const TransactionType_1 = require("../model/transaction/TransactionType");
const TransactionCount_1 = require("../model/transaction/TransactionCount");
const UInt64_1 = require("../model/UInt64");
const api_1 = require("./api");
const Http_1 = require("./Http");
const CreateTransactionFromDTO_1 = require("./transaction/CreateTransactionFromDTO");
const TransactionGroupType_1 = require("../model/transaction/TransactionGroupType");
const TransactionSearch_1 = require("../model/transaction/TransactionSearch");
const Pagination_1 = require("../model/Pagination");
/**
 * Transaction http repository.
 *
 * @since 1.0
 */
class TransactionHttp extends Http_1.Http {
    url;
    /**
     * @internal
     * xpx chain Library transaction routes api
     */
    transactionRoutesApi;
    /**
     * @internal
     * xpx chain Library blockchain routes api
     */
    blockRoutesApi;
    /**
     * Constructor
     * @param url
     */
    constructor(url) {
        super();
        this.url = url;
        this.transactionRoutesApi = new api_1.TransactionRoutesApi(url);
        this.blockRoutesApi = new api_1.BlockRoutesApi(url);
    }
    /**
     * Gets a transaction for a transactionId
     * @param transactionId - Transaction id or hash.
     * @returns Observable<Transaction>
     */
    getTransaction(transactionId, requestOptions) {
        return (0, rxjs_1.from)(this.transactionRoutesApi.getTransaction(transactionId, requestOptions)).pipe((0, operators_1.map)((response) => {
            return (0, CreateTransactionFromDTO_1.CreateTransactionFromDTO)(response.body);
        }));
    }
    /**
     * Gets an array of transactions for different transaction ids
     * @param transactionIds - Array of transactions id and/or hash.
     * @returns Observable<Transaction[]>
     */
    getTransactions(transactionIds, transactionGroupType = TransactionGroupType_1.TransactionGroupType.CONFIRMED, requestOptions) {
        const transactionIdsBody = {
            transactionIds,
        };
        return (0, rxjs_1.from)(this.transactionRoutesApi.getTransactions(transactionIdsBody, transactionGroupType, requestOptions)).pipe((0, operators_1.map)((response) => {
            return response.body.map((transactionDTO) => {
                return (0, CreateTransactionFromDTO_1.CreateTransactionFromDTO)(transactionDTO);
            });
        }));
    }
    /**
     * Gets an array of transactions for different transaction ids
     * @param transactionIds - Array of transactions id and/or hash.
     * @returns Observable<Transaction[]>
     */
    getTransactionsCount(transactionTypes, transactionGroupType = TransactionGroupType_1.TransactionGroupType.CONFIRMED, requestOptions) {
        const transactionTypesBody = {
            transactionTypes
        };
        return (0, rxjs_1.from)(this.transactionRoutesApi.getTransactionsCount(transactionTypesBody, transactionGroupType, requestOptions)).pipe((0, operators_1.map)((response) => {
            return response.body.map((transactionCountDTO) => {
                return new TransactionCount_1.TransactionCount(transactionCountDTO.type, transactionCountDTO.count);
            });
        }));
    }
    /**
     * Gets a transaction status for a transaction hash
     * @param hash - Transaction hash.
     * @returns Observable<TransactionStatus>
     */
    getTransactionStatus(transactionHash, requestOptions) {
        return (0, rxjs_1.from)(this.transactionRoutesApi.getTransactionStatus(transactionHash, requestOptions)).pipe((0, operators_1.map)((response) => {
            const transactionStatusDTO = response.body;
            return new TransactionStatus_1.TransactionStatus(transactionStatusDTO.status, transactionStatusDTO.group, transactionStatusDTO.hash, transactionStatusDTO.deadline ? Deadline_1.Deadline.createFromDTO(transactionStatusDTO.deadline) : undefined, transactionStatusDTO.height ? new UInt64_1.UInt64(transactionStatusDTO.height) : undefined);
        }));
    }
    /**
     * Gets an array of transaction status for different transaction hashes
     * @param transactionHashes - Array of transaction hash
     * @returns Observable<TransactionStatus[]>
     */
    getTransactionsStatuses(transactionHashes, requestOptions) {
        const transactionHashesBody = {
            hashes: transactionHashes,
        };
        return (0, rxjs_1.from)(this.transactionRoutesApi.getTransactionsStatuses(transactionHashesBody, requestOptions)).pipe((0, operators_1.map)((response) => {
            return response.body.map((transactionStatusDTO) => {
                return new TransactionStatus_1.TransactionStatus(transactionStatusDTO.status, transactionStatusDTO.group, transactionStatusDTO.hash, transactionStatusDTO.deadline ? Deadline_1.Deadline.createFromDTO(transactionStatusDTO.deadline) : undefined, transactionStatusDTO.height ? new UInt64_1.UInt64(transactionStatusDTO.height) : undefined);
            });
        }));
    }
    /**
     * Send a signed transaction
     * @param signedTransaction - Signed transaction
     * @returns Observable<TransactionAnnounceResponse>
     */
    announce(signedTransaction, requestOptions) {
        return (0, rxjs_1.from)(this.transactionRoutesApi.announceTransaction(signedTransaction, requestOptions)).pipe((0, operators_1.map)((response) => {
            return new TransactionAnnounceResponse_1.TransactionAnnounceResponse(response.body.message);
        }));
    }
    /**
     * Send a signed transaction with missing signatures
     * @param signedTransaction - Signed transaction
     * @returns Observable<TransactionAnnounceResponse>
     */
    announceAggregateBonded(signedTransaction, requestOptions) {
        if (![TransactionType_1.TransactionType.AGGREGATE_BONDED_V1, TransactionType_1.TransactionType.AGGREGATE_BONDED_V2].includes(signedTransaction.type)) {
            throw new Error('Only Transaction Type Bonded v1 or v2 is allowed for announce aggregate bonded');
        }
        return (0, rxjs_1.from)(this.transactionRoutesApi.announcePartialTransaction(signedTransaction, requestOptions)).pipe((0, operators_1.map)((response) => {
            return new TransactionAnnounceResponse_1.TransactionAnnounceResponse(response.body.message);
        }));
    }
    /**
     * Send a cosignature signed transaction of an already announced transaction
     * @param cosignatureSignedTransaction - Cosignature signed transaction
     * @returns Observable<TransactionAnnounceResponse>
     */
    announceAggregateBondedCosignature(cosignatureSignedTransaction, requestOptions) {
        return (0, rxjs_1.from)(this.transactionRoutesApi.announceCosignatureTransaction(cosignatureSignedTransaction, requestOptions)).pipe((0, operators_1.map)((response) => {
            return new TransactionAnnounceResponse_1.TransactionAnnounceResponse(response.body.message);
        }));
    }
    /**
     * Gets a transaction's effective paid fee
     * @param transactionId - Transaction id or hash.
     * @returns Observable<number>
     */
    getTransactionEffectiveFee(transactionId, requestOptions) {
        return (0, rxjs_1.from)(this.transactionRoutesApi.getTransaction(transactionId, requestOptions)).pipe((0, operators_1.mergeMap)((response) => {
            // parse transaction to take advantage of `size` getter overload
            const transaction = (0, CreateTransactionFromDTO_1.CreateTransactionFromDTO)(response.body);
            const uintHeight = transaction.transactionInfo.height;
            // now read block details
            return (0, rxjs_1.from)(this.blockRoutesApi.getBlockByHeight(uintHeight.compact(), requestOptions)).pipe((0, operators_1.map)((response) => {
                // @see https://bcdocs.xpxsirius.io/docs/cheatsheet/#fee
                // effective_fee = feeMultiplier x transaction::size
                return response.body.block.feeMultiplier * transaction.size;
            }));
        }), (0, operators_1.catchError)((err) => {
            return (0, rxjs_1.throwError)(() => new Error(err));
        }));
    }
    /**
     * Search transactions
     * @param searchType - Transaction group type.
     * @param queryParams - Transaction Query Params
     * @returns Observable<TransactionSearch>
     */
    searchTransactions(searchType, queryParams, requestOptions) {
        return (0, rxjs_1.from)(this.transactionRoutesApi.searchTransactions(searchType, queryParams, requestOptions)).pipe((0, operators_1.map)((response) => {
            let transactions = response.body.data.map((transactionDTO) => {
                return (0, CreateTransactionFromDTO_1.CreateTransactionFromDTO)(transactionDTO);
            });
            let paginationData = new Pagination_1.Pagination(response.body.pagination.totalEntries, response.body.pagination.pageNumber, response.body.pagination.pageSize, response.body.pagination.totalPages);
            return new TransactionSearch_1.TransactionSearch(transactions, paginationData);
        }));
    }
    /**
     * Search unconfirmed transaction by transaction hash
     * @param txnHash - Transaction hash.
     * @returns Observable<Transaction>
     */
    getUnconfirmedTransaction(txnHash, requestOptions) {
        return (0, rxjs_1.from)(this.transactionRoutesApi.searchTransaction(TransactionGroupType_1.TransactionGroupType.UNCONFIRMED, txnHash, requestOptions)).pipe((0, operators_1.map)((response) => {
            return (0, CreateTransactionFromDTO_1.CreateTransactionFromDTO)(response.body);
        }));
    }
    /**
     * Search partial transaction by transaction hash
     * @param txnHash - Transaction hash.
     * @returns Observable<Transaction>
     */
    getPartialTransaction(txnHash, requestOptions) {
        return (0, rxjs_1.from)(this.transactionRoutesApi.searchTransaction(TransactionGroupType_1.TransactionGroupType.PARTIAL, txnHash, requestOptions)).pipe((0, operators_1.map)((response) => {
            return (0, CreateTransactionFromDTO_1.CreateTransactionFromDTO)(response.body);
        }));
    }
}
exports.TransactionHttp = TransactionHttp;
