// tsjs-xpx-chain-sdk/dist/src/infrastructure/TransactionQueryParams.js
"use strict";
/*
 * Copyright 2021 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionQueryParams = exports.TransactionFieldOrder = exports.TransactionSortingField = exports.TransactionMosaicSearchFilters = void 0;
const PublicAccount_1 = require("../model/account/PublicAccount");
const Address_1 = require("../model/account/Address");
const QueryParams_1 = require("./QueryParams");
const model_1 = require("../model/model");
const PaginationQueryParams_1 = require("./PaginationQueryParams");
class TransactionMosaicSearchFilters {
    transferMosaicId;
    toTransferAmount;
    fromTransferAmount;
    constructor(transferMosaicId) {
        this.transferMosaicId = transferMosaicId;
    }
    buildQueryParamsString() {
        let queryParams = this.buildQueryParams();
        const entries = Object.entries(queryParams);
        let queryParamsArray = entries.map(data => {
            return data[0] + "=" + data[1];
        });
        return queryParamsArray.join("&");
    }
    buildQueryParams() {
        let queryParams = Object.assign({}, this);
        TransactionMosaicSearchFilters.adjustNonComplianceParams(queryParams);
        TransactionMosaicSearchFilters.convertToPrimitiveType(queryParams);
        return queryParams;
    }
    static adjustNonComplianceParams(queryParams) {
        if (queryParams.fromTransferAmount !== undefined && queryParams.fromTransferAmount < 1) {
            delete queryParams.fromTransferAmount;
        }
        if (queryParams.toTransferAmount !== undefined && queryParams.toTransferAmount < 1) {
            delete queryParams.toTransferAmount;
        }
    }
    static convertToPrimitiveType(queryParams) {
        if (queryParams.transferMosaicId instanceof model_1.MosaicId || queryParams.transferMosaicId instanceof model_1.NamespaceId) {
            queryParams.transferMosaicId = queryParams.transferMosaicId.toHex();
        }
    }
}
exports.TransactionMosaicSearchFilters = TransactionMosaicSearchFilters;
var TransactionSortingField;
(function (TransactionSortingField) {
    TransactionSortingField["BLOCK"] = "meta.height";
})(TransactionSortingField || (exports.TransactionSortingField = TransactionSortingField = {}));
class TransactionFieldOrder {
    sortingField;
    order;
    constructor(sortingField, order) {
        this.sortingField = sortingField;
        this.order = order;
    }
    static setAscending(sortingField) {
        return new TransactionFieldOrder(sortingField, QueryParams_1.Order_v2.ASC);
    }
    static setDescending(sortingField) {
        return new TransactionFieldOrder(sortingField, QueryParams_1.Order_v2.DESC);
    }
}
exports.TransactionFieldOrder = TransactionFieldOrder;
class TransactionQueryParams extends PaginationQueryParams_1.PaginationQueryParams {
    /**
    * filter by transaction types
    */
    type;
    /**
    * include innerTransactions as single transaction into the search result, default to false
    */
    embedded;
    order;
    sortField;
    toHeight;
    fromHeight;
    height;
    signerPublicKey;
    recipientAddress;
    /**
     * address involved in the transaction. It can be sender, recipient, or cosigner that required to cosign the transaction).
     * cannot be combined with recipientAddress and signerPublicKey
     */
    address;
    /**
     * combination of signerPublicKey and recipientAddress
     */
    publicKey;
    /**
    * include only first level of Aggregate transactions (exclude innerTransactions), default to true
    * set to false to get the complete aggregate transactions
    */
    firstLevel;
    /**
     * include only transfer transaction with the mosaicId provided,
     * will return error when provided along with transaction types
     */
    transferMosaicIdFilters;
    /* non-applicable for now
    offset?: string;
    */
    /**
     * Constructor
     */
    constructor() {
        super();
    }
    updateFieldOrder(transactionFieldOrder) {
        this.sortField = transactionFieldOrder.sortingField;
        this.order = transactionFieldOrder.order;
    }
    buildQueryParams() {
        let queryParams = Object.assign({}, this);
        TransactionQueryParams.adjustNonComplianceParams(queryParams);
        TransactionQueryParams.convertToPrimitiveType(queryParams);
        let flattenedQueryParams = super.buildQueryParams(queryParams);
        return flattenedQueryParams;
    }
    buildQueryParamsString() {
        let queryParams = this.buildQueryParams();
        const entries = Object.entries(queryParams);
        let queryParamsArray = entries.map(data => {
            if (data[1] instanceof Array) {
                return data[1].map(arrayData => {
                    return data[0] + "[]" + "=" + arrayData;
                }).join("&");
            }
            return data[0] + "=" + data[1];
        });
        return queryParamsArray.join("&");
    }
    static convertToPrimitiveType(queryParams) {
        if (queryParams.recipientAddress) {
            if (queryParams.recipientAddress instanceof Address_1.Address) {
                queryParams.recipientAddress = queryParams.recipientAddress.plain();
            }
        }
        if (queryParams.signerPublicKey) {
            if (queryParams.signerPublicKey instanceof PublicAccount_1.PublicAccount) {
                queryParams.signerPublicKey = queryParams.signerPublicKey.publicKey;
            }
        }
        if (queryParams.publicKey) {
            if (queryParams.publicKey instanceof PublicAccount_1.PublicAccount) {
                queryParams.publicKey = queryParams.publicKey.publicKey;
            }
        }
        if (queryParams.address) {
            if (queryParams.address instanceof Address_1.Address) {
                queryParams.address = queryParams.address.plain();
            }
        }
    }
    static adjustNonComplianceParams(queryParams) {
        if (queryParams.height) {
            if (queryParams.toHeight) {
                delete queryParams.toHeight;
            }
            if (queryParams.fromHeight) {
                delete queryParams.fromHeight;
            }
        }
        if (queryParams.pageSize) {
            if (queryParams.pageSize < 0) {
                delete queryParams.pageSize; // default to 20
            }
            else if (queryParams.pageSize < 10) {
                queryParams.pageSize = 10;
            }
            else if (queryParams.pageSize > 100) {
                queryParams.pageSize = 100;
            }
        }
        if (queryParams.pageNumber) {
            if (queryParams.pageNumber <= 0) {
                queryParams.pageNumber = 1;
            }
        }
        if (queryParams.order === undefined || queryParams.sortField === undefined) {
            if (queryParams.order) {
                delete queryParams.order;
            }
            if (queryParams.sortField) {
                delete queryParams.sortField;
            }
        }
        if (queryParams.publicKey) {
            if (queryParams.address) {
                delete queryParams.address;
            }
            if (queryParams.signerPublicKey) {
                delete queryParams.signerPublicKey;
            }
            if (queryParams.recipientAddress) {
                delete queryParams.recipientAddress;
            }
        }
        else if (queryParams.address) {
            if (queryParams.signerPublicKey) {
                delete queryParams.signerPublicKey;
            }
            if (queryParams.recipientAddress) {
                delete queryParams.recipientAddress;
            }
        }
        else if (queryParams.signerPublicKey) {
            if (queryParams.recipientAddress) {
                delete queryParams.recipientAddress;
            }
        }
    }
}
exports.TransactionQueryParams = TransactionQueryParams;
