// tsjs-xpx-chain-sdk/dist/src/model/transaction/liquidityProvider/CreateLiquidityProviderTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateLiquidityProviderTransactionBuilder = exports.CreateLiquidityProviderTransaction = void 0;
const CreateLiquidityProviderTransaction_1 = require("../../../infrastructure/builders/liquidityProvider/CreateLiquidityProviderTransaction");
const Transaction_1 = require("../Transaction");
const Transaction_2 = require("../Transaction");
const TransactionType_1 = require("../TransactionType");
const TransactionTypeVersion_1 = require("../TransactionTypeVersion");
const FeeCalculationStrategy_1 = require("../FeeCalculationStrategy");
class CreateLiquidityProviderTransaction extends Transaction_2.Transaction {
    providerMosaicId;
    currencyDeposit;
    initialMosaicsMinting;
    slashingPeriod;
    windowSize;
    slashingAccount;
    alpha;
    beta;
    /**
     * Create a create liquidity provider transaction object
     * @param deadline - The deadline to include the transaction.
     * @param providerMosaicId MosaicId
     * @param currencyDeposit Amount
     * @param initialMosaicsMinting Amount
     * @param slashingPeriod uint32
     * @param windowSize uint16
     * @param slashingAccount PublicAccount
     * @param alpha uint32
     * @param beta uint32
     * @param networkType Network Type
     * @param maxFee - (Optional) Max fee defined by the sender
     * @returns {CreateLiquidityProviderTransaction}
     */
    static create(deadline, providerMosaicId, currencyDeposit, initialMosaicsMinting, slashingPeriod, windowSize, slashingAccount, alpha, beta, networkType, maxFee) {
        return new CreateLiquidityProviderTransactionBuilder()
            .networkType(networkType)
            .deadline(deadline)
            .maxFee(maxFee)
            .providerMosaicId(providerMosaicId)
            .currencyDeposit(currencyDeposit)
            .initialMosaicsMinting(initialMosaicsMinting)
            .slashingPeriod(slashingPeriod)
            .windowSize(windowSize)
            .slashingAccount(slashingAccount)
            .alpha(alpha)
            .beta(beta)
            .build();
    }
    /**
     * @param networkType
     * @param version
     * @param deadline
     * @param maxFee
     * @param providerMosaicId MosaicId
     * @param currencyDeposit Amount
     * @param initialMosaicsMinting Amount
     * @param slashingPeriod uint32
     * @param windowSize uint16
     * @param slashingAccount PublicAccount
     * @param alpha uint32
     * @param beta uint32
     * @param signature
     * @param signer
     * @param transactionInfo
     */
    constructor(networkType, version, deadline, maxFee, providerMosaicId, currencyDeposit, initialMosaicsMinting, slashingPeriod, windowSize, slashingAccount, alpha, beta, signature, signer, transactionInfo) {
        super(TransactionType_1.TransactionType.Create_Liquidity_Provider, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
        this.providerMosaicId = providerMosaicId;
        this.currencyDeposit = currencyDeposit;
        this.initialMosaicsMinting = initialMosaicsMinting;
        this.slashingPeriod = slashingPeriod;
        this.windowSize = windowSize;
        this.slashingAccount = slashingAccount;
        this.alpha = alpha;
        this.beta = beta;
        if (slashingPeriod > 0xFFFFFFFF || slashingPeriod <= 0) {
            throw new Error("slashingPeriod out of range, should be uint32 positive number");
        }
        if (windowSize > 0xFFFF || windowSize <= 0) {
            throw new Error("windowSize out of range, should be uint16 positive number");
        }
        if (alpha > 0xFFFFFFFF || alpha <= 0) {
            throw new Error("slashingPeriod out of range, should be uint32 positive number");
        }
        if (beta > 0xFFFFFFFF || beta <= 0) {
            throw new Error("slashingPeriod out of range, should be uint32 positive number");
        }
    }
    /**
     * @override Transaction.size()
     * @description get the byte size of a CreateLiquidityProviderTransaction
     * @returns {number}
     * @memberof CreateLiquidityProviderTransaction
     */
    get size() {
        return CreateLiquidityProviderTransaction.calculateSize();
    }
    static calculateSize() {
        const byteSize = Transaction_2.Transaction.getHeaderSize();
        // set static byte size fields
        const providerMosaicId = 8;
        const currencyDeposit = 8;
        const initialMosaicsMinting = 8;
        const slashingPeriod = 4;
        const windowSize = 2;
        const slashingAccount = 32;
        const alpha = 4;
        const beta = 4;
        return byteSize + providerMosaicId + currencyDeposit + initialMosaicsMinting + slashingPeriod
            + windowSize + slashingAccount + alpha + beta;
    }
    /**
     * @override Transaction.toJSON()
     * @description Serialize a transaction object - add own fields to the result of Transaction.toJSON()
     * @return {Object}
     * @memberof CreateLiquidityProviderTransaction
     */
    toJSON() {
        const parent = super.toJSON();
        return {
            ...parent,
            transaction: {
                ...parent.transaction,
                providerMosaicId: this.providerMosaicId.toDTO(),
                currencyDeposit: this.currencyDeposit.toDTO(),
                initialMosaicsMinting: this.initialMosaicsMinting.toDTO(),
                slashingPeriod: this.slashingPeriod,
                windowSize: this.windowSize,
                slashingAccount: this.slashingAccount.publicKey,
                alpha: this.alpha,
                beta: this.beta
            }
        };
    }
    /**
     * @internal
     * @returns {VerifiableTransaction}
     */
    buildTransaction() {
        return new CreateLiquidityProviderTransaction_1.Builder()
            .addSize(this.size)
            .addDeadline(this.deadline.toDTO())
            .addMaxFee(this.maxFee.toDTO())
            .addVersion(this.versionToDTO())
            .addType(this.type)
            .addProviderMosaicId(this.providerMosaicId.id.toDTO())
            .addCurrencyDeposit(this.currencyDeposit.toDTO())
            .addInitialMosaicsMinting(this.initialMosaicsMinting.toDTO())
            .addSlashingPeriod(this.slashingPeriod)
            .addWindowSize(this.windowSize)
            .addSlashingAccount(this.slashingAccount.publicKey)
            .addAlpha(this.alpha)
            .addBeta(this.beta)
            .build();
    }
}
exports.CreateLiquidityProviderTransaction = CreateLiquidityProviderTransaction;
class CreateLiquidityProviderTransactionBuilder extends Transaction_1.TransactionBuilder {
    _providerMosaicId;
    _currencyDeposit;
    _initialMosaicsMinting;
    _slashingPeriod;
    _windowSize;
    _slashingAccount;
    _alpha;
    _beta;
    providerMosaicId(providerMosaicId) {
        this._providerMosaicId = providerMosaicId;
        return this;
    }
    currencyDeposit(currencyDeposit) {
        this._currencyDeposit = currencyDeposit;
        return this;
    }
    initialMosaicsMinting(initialMosaicsMinting) {
        this._initialMosaicsMinting = initialMosaicsMinting;
        return this;
    }
    slashingPeriod(slashingPeriod) {
        this._slashingPeriod = slashingPeriod;
        return this;
    }
    windowSize(windowSize) {
        this._windowSize = windowSize;
        return this;
    }
    slashingAccount(slashingAccount) {
        this._slashingAccount = slashingAccount;
        return this;
    }
    alpha(alpha) {
        this._alpha = alpha;
        return this;
    }
    beta(beta) {
        this._beta = beta;
        return this;
    }
    build() {
        return new CreateLiquidityProviderTransaction(this._networkType, this._version || TransactionTypeVersion_1.TransactionTypeVersion.Create_Liquidity_Provider, this._deadline ? this._deadline : this._createNewDeadlineFn(), this._maxFee ? this._maxFee : (0, FeeCalculationStrategy_1.calculateFee)(CreateLiquidityProviderTransaction.calculateSize(), this._feeCalculationStrategy), this._providerMosaicId, this._currencyDeposit, this._initialMosaicsMinting, this._slashingPeriod, this._windowSize, this._slashingAccount, this._alpha, this._beta, this._signature, this._signer, this._transactionInfo);
    }
}
exports.CreateLiquidityProviderTransactionBuilder = CreateLiquidityProviderTransactionBuilder;
