// tsjs-xpx-chain-sdk/dist/src/model/transaction/liquidityProvider/ManualRateChangeTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ManualRateChangeTransactionBuilder = exports.ManualRateChangeTransaction = void 0;
const ManualRateChangeTransaction_1 = require("../../../infrastructure/builders/liquidityProvider/ManualRateChangeTransaction");
const Transaction_1 = require("../Transaction");
const Transaction_2 = require("../Transaction");
const TransactionType_1 = require("../TransactionType");
const TransactionTypeVersion_1 = require("../TransactionTypeVersion");
const FeeCalculationStrategy_1 = require("../FeeCalculationStrategy");
/**
 * ManualRateChangeTransaction for liquidity provider
 */
class ManualRateChangeTransaction extends Transaction_2.Transaction {
    providerMosaicId;
    currencyBalanceIncrease;
    currencyBalanceChange;
    mosaicBalanceIncrease;
    mosaicBalanceChange;
    /**
     * Create a manual rate change transaction object
     * @param deadline - The deadline to include the transaction.
     * @param providerMosaicId MosaicId
     * @param currencyBalanceIncrease boolean
     * @param currencyBalanceChange Amount
     * @param mosaicBalanceIncrease booloan
     * @param mosaicBalanceChange Amount
     * @param maxFee - (Optional) Max fee defined by the sender
     * @returns {ManualRateChangeTransaction}
     */
    static create(deadline, providerMosaicId, currencyBalanceIncrease, currencyBalanceChange, mosaicBalanceIncrease, mosaicBalanceChange, networkType, maxFee) {
        return new ManualRateChangeTransactionBuilder()
            .networkType(networkType)
            .deadline(deadline)
            .maxFee(maxFee)
            .providerMosaicId(providerMosaicId)
            .currencyBalanceIncrease(currencyBalanceIncrease)
            .currencyBalanceChange(currencyBalanceChange)
            .mosaicBalanceIncrease(mosaicBalanceIncrease)
            .mosaicBalanceChange(mosaicBalanceChange)
            .build();
    }
    /**
     * @param networkType
     * @param version
     * @param deadline
     * @param maxFee
     * @param providerMosaicId MosaicId
     * @param currencyBalanceIncrease boolean
     * @param currencyBalanceChange Amount
     * @param mosaicBalanceIncrease booloan
     * @param mosaicBalanceChange Amount
     * @param signature
     * @param signer
     * @param transactionInfo
     */
    constructor(networkType, version, deadline, maxFee, providerMosaicId, currencyBalanceIncrease, currencyBalanceChange, mosaicBalanceIncrease, mosaicBalanceChange, signature, signer, transactionInfo) {
        super(TransactionType_1.TransactionType.Manual_Rate_Change, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
        this.providerMosaicId = providerMosaicId;
        this.currencyBalanceIncrease = currencyBalanceIncrease;
        this.currencyBalanceChange = currencyBalanceChange;
        this.mosaicBalanceIncrease = mosaicBalanceIncrease;
        this.mosaicBalanceChange = mosaicBalanceChange;
    }
    /**
     * @override Transaction.size()
     * @description get the byte size of a ManualRateChangeTransaction
     * @returns {number}
     * @memberof ManualRateChangeTransaction
     */
    get size() {
        return ManualRateChangeTransaction.calculateSize();
    }
    static calculateSize() {
        const byteSize = Transaction_2.Transaction.getHeaderSize();
        // set static byte size fields
        const providerMosaicId = 8;
        const currencyBalanceIncrease = 1;
        const currencyBalanceChange = 8;
        const mosaicBalanceIncrease = 1;
        const mosaicBalanceChange = 8;
        return byteSize + providerMosaicId + currencyBalanceIncrease + currencyBalanceChange + mosaicBalanceIncrease + mosaicBalanceChange;
    }
    /**
     * @override Transaction.toJSON()
     * @description Serialize a transaction object - add own fields to the result of Transaction.toJSON()
     * @return {Object}
     * @memberof ManualRateChangeTransaction
     */
    toJSON() {
        const parent = super.toJSON();
        return {
            ...parent,
            transaction: {
                ...parent.transaction,
                providerMosaicId: this.providerMosaicId.toHex(),
                currencyBalanceIncrease: this.currencyBalanceIncrease,
                currencyBalanceChange: this.currencyBalanceChange.toDTO(),
                mosaicBalanceIncrease: this.mosaicBalanceIncrease,
                mosaicBalanceChange: this.mosaicBalanceChange.toDTO()
            }
        };
    }
    /**
     * @internal
     * @returns {VerifiableTransaction}
     */
    buildTransaction() {
        return new ManualRateChangeTransaction_1.Builder()
            .addSize(this.size)
            .addDeadline(this.deadline.toDTO())
            .addMaxFee(this.maxFee.toDTO())
            .addVersion(this.versionToDTO())
            .addType(this.type)
            .addProviderMosaicId(this.providerMosaicId.id.toDTO())
            .addCurrencyBalanceIncrease(this.currencyBalanceIncrease)
            .addCurrencyBalanceChange(this.currencyBalanceChange.toDTO())
            .addMosaicBalanceIncrease(this.mosaicBalanceIncrease)
            .addMosaicBalanceChange(this.mosaicBalanceChange.toDTO())
            .build();
    }
}
exports.ManualRateChangeTransaction = ManualRateChangeTransaction;
class ManualRateChangeTransactionBuilder extends Transaction_1.TransactionBuilder {
    _providerMosaicId;
    _currencyBalanceIncrease;
    _currencyBalanceChange;
    _mosaicBalanceIncrease;
    _mosaicBalanceChange;
    providerMosaicId(providerMosaicId) {
        this._providerMosaicId = providerMosaicId;
        return this;
    }
    currencyBalanceIncrease(currencyBalanceIncrease) {
        this._currencyBalanceIncrease = currencyBalanceIncrease;
        return this;
    }
    currencyBalanceChange(currencyBalanceChange) {
        this._currencyBalanceChange = currencyBalanceChange;
        return this;
    }
    mosaicBalanceIncrease(mosaicBalanceIncrease) {
        this._mosaicBalanceIncrease = mosaicBalanceIncrease;
        return this;
    }
    mosaicBalanceChange(mosaicBalanceChange) {
        this._mosaicBalanceChange = mosaicBalanceChange;
        return this;
    }
    build() {
        return new ManualRateChangeTransaction(this._networkType, this._version || TransactionTypeVersion_1.TransactionTypeVersion.Manual_Rate_Change, this._deadline ? this._deadline : this._createNewDeadlineFn(), this._maxFee ? this._maxFee : (0, FeeCalculationStrategy_1.calculateFee)(ManualRateChangeTransaction.calculateSize(), this._feeCalculationStrategy), this._providerMosaicId, this._currencyBalanceIncrease, this._currencyBalanceChange, this._mosaicBalanceIncrease, this._mosaicBalanceChange, this._signature, this._signer, this._transactionInfo);
    }
}
exports.ManualRateChangeTransactionBuilder = ManualRateChangeTransactionBuilder;
