// tsjs-xpx-chain-sdk/dist/src/infrastructure/buffers/PlaceSdaExchangeOfferBuffer.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlaceSdaExchangeOfferBuffer = void 0;
const flatbuffers = require("flatbuffers");
class PlaceSdaExchangeOfferBuffer {
    bb = null;
    bb_pos = 0;
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsPlaceSdaExchangeOfferBuffer(bb, obj) {
        return (obj || new PlaceSdaExchangeOfferBuffer()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsPlaceSdaExchangeOfferBuffer(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new PlaceSdaExchangeOfferBuffer()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    mosaicIdGive(index) {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    mosaicIdGiveLength() {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    mosaicIdGiveArray() {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    mosaicAmountGive(index) {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    mosaicAmountGiveLength() {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    mosaicAmountGiveArray() {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    mosaicIdGet(index) {
        const offset = this.bb.__offset(this.bb_pos, 8);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    mosaicIdGetLength() {
        const offset = this.bb.__offset(this.bb_pos, 8);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    mosaicIdGetArray() {
        const offset = this.bb.__offset(this.bb_pos, 8);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    mosaicAmountGet(index) {
        const offset = this.bb.__offset(this.bb_pos, 10);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    mosaicAmountGetLength() {
        const offset = this.bb.__offset(this.bb_pos, 10);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    mosaicAmountGetArray() {
        const offset = this.bb.__offset(this.bb_pos, 10);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    duration(index) {
        const offset = this.bb.__offset(this.bb_pos, 12);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    durationLength() {
        const offset = this.bb.__offset(this.bb_pos, 12);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    durationArray() {
        const offset = this.bb.__offset(this.bb_pos, 12);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    static startPlaceSdaExchangeOfferBuffer(builder) {
        builder.startObject(5);
    }
    static addMosaicIdGive(builder, mosaicIdGiveOffset) {
        builder.addFieldOffset(0, mosaicIdGiveOffset, 0);
    }
    static createMosaicIdGiveVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startMosaicIdGiveVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addMosaicAmountGive(builder, mosaicAmountGiveOffset) {
        builder.addFieldOffset(1, mosaicAmountGiveOffset, 0);
    }
    static createMosaicAmountGiveVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startMosaicAmountGiveVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addMosaicIdGet(builder, mosaicIdGetOffset) {
        builder.addFieldOffset(2, mosaicIdGetOffset, 0);
    }
    static createMosaicIdGetVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startMosaicIdGetVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addMosaicAmountGet(builder, mosaicAmountGetOffset) {
        builder.addFieldOffset(3, mosaicAmountGetOffset, 0);
    }
    static createMosaicAmountGetVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startMosaicAmountGetVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addDuration(builder, durationOffset) {
        builder.addFieldOffset(4, durationOffset, 0);
    }
    static createDurationVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startDurationVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static endPlaceSdaExchangeOfferBuffer(builder) {
        const offset = builder.endObject();
        return offset;
    }
    static createPlaceSdaExchangeOfferBuffer(builder, mosaicIdGiveOffset, mosaicAmountGiveOffset, mosaicIdGetOffset, mosaicAmountGetOffset, durationOffset) {
        PlaceSdaExchangeOfferBuffer.startPlaceSdaExchangeOfferBuffer(builder);
        PlaceSdaExchangeOfferBuffer.addMosaicIdGive(builder, mosaicIdGiveOffset);
        PlaceSdaExchangeOfferBuffer.addMosaicAmountGive(builder, mosaicAmountGiveOffset);
        PlaceSdaExchangeOfferBuffer.addMosaicIdGet(builder, mosaicIdGetOffset);
        PlaceSdaExchangeOfferBuffer.addMosaicAmountGet(builder, mosaicAmountGetOffset);
        PlaceSdaExchangeOfferBuffer.addDuration(builder, durationOffset);
        return PlaceSdaExchangeOfferBuffer.endPlaceSdaExchangeOfferBuffer(builder);
    }
}
exports.PlaceSdaExchangeOfferBuffer = PlaceSdaExchangeOfferBuffer;
