// tsjs-xpx-chain-sdk/dist/src/infrastructure/buffers/RemoveSdaExchangeOfferBuffer.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoveSdaExchangeOfferBuffer = void 0;
const flatbuffers = require("flatbuffers");
class RemoveSdaExchangeOfferBuffer {
    bb = null;
    bb_pos = 0;
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsRemoveSdaExchangeOfferBuffer(bb, obj) {
        return (obj || new RemoveSdaExchangeOfferBuffer()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsRemoveSdaExchangeOfferBuffer(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new RemoveSdaExchangeOfferBuffer()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    mosaicIdGive(index) {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    mosaicIdGiveLength() {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    mosaicIdGiveArray() {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    mosaicIdGet(index) {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    mosaicIdGetLength() {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    mosaicIdGetArray() {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    static startRemoveSdaExchangeOfferBuffer(builder) {
        builder.startObject(2);
    }
    static addMosaicIdGive(builder, mosaicIdGiveOffset) {
        builder.addFieldOffset(0, mosaicIdGiveOffset, 0);
    }
    static createMosaicIdGiveVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startMosaicIdGiveVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addMosaicIdGet(builder, mosaicIdGetOffset) {
        builder.addFieldOffset(1, mosaicIdGetOffset, 0);
    }
    static createMosaicIdGetVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startMosaicIdGetVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static endRemoveSdaExchangeOfferBuffer(builder) {
        const offset = builder.endObject();
        return offset;
    }
    static finishRemoveSdaExchangeOfferBufferBuffer(builder, offset) {
        builder.finish(offset);
    }
    static finishSizePrefixedRemoveSdaExchangeOfferBufferBuffer(builder, offset) {
        builder.finish(offset, undefined, true);
    }
    static createRemoveSdaExchangeOfferBuffer(builder, mosaicIdGiveOffset, mosaicIdGetOffset) {
        RemoveSdaExchangeOfferBuffer.startRemoveSdaExchangeOfferBuffer(builder);
        RemoveSdaExchangeOfferBuffer.addMosaicIdGive(builder, mosaicIdGiveOffset);
        RemoveSdaExchangeOfferBuffer.addMosaicIdGet(builder, mosaicIdGetOffset);
        return RemoveSdaExchangeOfferBuffer.endRemoveSdaExchangeOfferBuffer(builder);
    }
}
exports.RemoveSdaExchangeOfferBuffer = RemoveSdaExchangeOfferBuffer;
