// tsjs-xpx-chain-sdk/dist/src/infrastructure/builders/liquidityProvider/CreateLiquidityProviderTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = void 0;
/**
 * @module transactions/CreateLiquidityProviderTransaction
 */
const format_1 = require("../../../core/format");
const CreateLiquidityProviderBuffer_1 = require("../../buffers/liquidityProvider/CreateLiquidityProviderBuffer");
const CreateLiquidityProviderTransactionSchema_1 = require("../../schemas/liquidityProvider/CreateLiquidityProviderTransactionSchema");
const VerifiableTransaction_1 = require("../VerifiableTransaction");
const flatbuffers = require("flatbuffers");
class CreateLiquidityProviderTransaction extends VerifiableTransaction_1.VerifiableTransaction {
    constructor(bytes) {
        super(bytes, CreateLiquidityProviderTransactionSchema_1.default);
    }
}
exports.default = CreateLiquidityProviderTransaction;
class Builder {
    size;
    fee;
    version;
    type;
    deadline;
    providerMosaicId;
    currencyDeposit;
    initialMosaicsMinting;
    slashingPeriod; // uint32
    windowSize; // uint16
    slashingAccount;
    alpha; // uint32
    beta; // uint32
    constructor() {
        this.fee = [0, 0];
        this.version = 1;
    }
    addSize(size) {
        this.size = size;
        return this;
    }
    addMaxFee(fee) {
        this.fee = fee;
        return this;
    }
    addVersion(version) {
        this.version = version;
        return this;
    }
    addType(type) {
        this.type = type;
        return this;
    }
    addDeadline(deadline) {
        this.deadline = deadline;
        return this;
    }
    addProviderMosaicId(providerMosaicId) {
        this.providerMosaicId = providerMosaicId;
        return this;
    }
    addCurrencyDeposit(currencyDeposit) {
        this.currencyDeposit = currencyDeposit;
        return this;
    }
    addInitialMosaicsMinting(initialMosaicsMinting) {
        this.initialMosaicsMinting = initialMosaicsMinting;
        return this;
    }
    addSlashingPeriod(slashingPeriod) {
        this.slashingPeriod = slashingPeriod;
        return this;
    }
    addWindowSize(windowSize) {
        this.windowSize = windowSize;
        return this;
    }
    addSlashingAccount(slashingAccount) {
        this.slashingAccount = slashingAccount;
        return this;
    }
    addAlpha(alpha) {
        this.alpha = alpha;
        return this;
    }
    addBeta(beta) {
        this.beta = beta;
        return this;
    }
    build() {
        const builder = new flatbuffers.Builder(1);
        const providerMosaicIdUint32 = new Uint32Array(this.providerMosaicId);
        // Create vectors
        const signatureVector = CreateLiquidityProviderBuffer_1.CreateLiquidityProviderBuffer
            .createSignatureVector(builder, Array(...Array(64)).map(Number.prototype.valueOf, 0));
        const signerVector = CreateLiquidityProviderBuffer_1.CreateLiquidityProviderBuffer.createSignerVector(builder, Array(...Array(32)).map(Number.prototype.valueOf, 0));
        const deadlineVector = CreateLiquidityProviderBuffer_1.CreateLiquidityProviderBuffer.createDeadlineVector(builder, this.deadline);
        const feeVector = CreateLiquidityProviderBuffer_1.CreateLiquidityProviderBuffer.createMaxFeeVector(builder, this.fee);
        const slashingAccountVector = CreateLiquidityProviderBuffer_1.CreateLiquidityProviderBuffer.createSlashingAccountVector(builder, format_1.Convert.hexToUint8(this.slashingAccount));
        const providerMosaicIdVector = CreateLiquidityProviderBuffer_1.CreateLiquidityProviderBuffer.createProviderMosaicIdVector(builder, providerMosaicIdUint32);
        const currencyDepositVector = CreateLiquidityProviderBuffer_1.CreateLiquidityProviderBuffer.createCurrencyDepositVector(builder, this.currencyDeposit);
        const initialMosaicsMintingVector = CreateLiquidityProviderBuffer_1.CreateLiquidityProviderBuffer.createInitialMosaicsMintingVector(builder, this.initialMosaicsMinting);
        CreateLiquidityProviderBuffer_1.CreateLiquidityProviderBuffer.startCreateLiquidityProviderBuffer(builder);
        CreateLiquidityProviderBuffer_1.CreateLiquidityProviderBuffer.addSize(builder, this.size);
        CreateLiquidityProviderBuffer_1.CreateLiquidityProviderBuffer.addSignature(builder, signatureVector);
        CreateLiquidityProviderBuffer_1.CreateLiquidityProviderBuffer.addSigner(builder, signerVector);
        CreateLiquidityProviderBuffer_1.CreateLiquidityProviderBuffer.addVersion(builder, this.version);
        CreateLiquidityProviderBuffer_1.CreateLiquidityProviderBuffer.addType(builder, this.type);
        CreateLiquidityProviderBuffer_1.CreateLiquidityProviderBuffer.addMaxFee(builder, feeVector);
        CreateLiquidityProviderBuffer_1.CreateLiquidityProviderBuffer.addDeadline(builder, deadlineVector);
        CreateLiquidityProviderBuffer_1.CreateLiquidityProviderBuffer.addProviderMosaicId(builder, providerMosaicIdVector);
        CreateLiquidityProviderBuffer_1.CreateLiquidityProviderBuffer.addCurrencyDeposit(builder, currencyDepositVector);
        CreateLiquidityProviderBuffer_1.CreateLiquidityProviderBuffer.addInitialMosaicsMinting(builder, initialMosaicsMintingVector);
        CreateLiquidityProviderBuffer_1.CreateLiquidityProviderBuffer.addSlashingPeriod(builder, this.slashingPeriod);
        CreateLiquidityProviderBuffer_1.CreateLiquidityProviderBuffer.addWindowSize(builder, this.windowSize);
        CreateLiquidityProviderBuffer_1.CreateLiquidityProviderBuffer.addSlashingAccount(builder, slashingAccountVector);
        CreateLiquidityProviderBuffer_1.CreateLiquidityProviderBuffer.addAlpha(builder, this.alpha);
        CreateLiquidityProviderBuffer_1.CreateLiquidityProviderBuffer.addBeta(builder, this.beta);
        const codedTransfer = CreateLiquidityProviderBuffer_1.CreateLiquidityProviderBuffer.endCreateLiquidityProviderBuffer(builder);
        builder.finish(codedTransfer);
        const bytes = builder.asUint8Array();
        return new CreateLiquidityProviderTransaction(bytes);
    }
    static stringToUint8(stringToConvert) {
        return format_1.Convert.hexToUint8(format_1.Convert.utf8ToHex(stringToConvert));
    }
}
exports.Builder = Builder;
