// tsjs-xpx-chain-sdk/dist/src/infrastructure/builders/liquidityProvider/ManualRateChangeTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = void 0;
/**
 * @module transactions/ManualRateChangeTransaction
 */
const format_1 = require("../../../core/format");
const ManualRateChangeBuffer_1 = require("../../buffers/liquidityProvider/ManualRateChangeBuffer");
const ManualRateChangeTransactionSchema_1 = require("../../schemas/liquidityProvider/ManualRateChangeTransactionSchema");
const VerifiableTransaction_1 = require("../VerifiableTransaction");
const flatbuffers = require("flatbuffers");
class ManualRateChangeTransaction extends VerifiableTransaction_1.VerifiableTransaction {
    constructor(bytes) {
        super(bytes, ManualRateChangeTransactionSchema_1.default);
    }
}
exports.default = ManualRateChangeTransaction;
class Builder {
    size;
    fee;
    version;
    type;
    deadline;
    providerMosaicId;
    currencyBalanceIncrease;
    currencyBalanceChange;
    mosaicBalanceIncrease;
    mosaicBalanceChange;
    constructor() {
        this.fee = [0, 0];
        this.version = 1;
    }
    addSize(size) {
        this.size = size;
        return this;
    }
    addMaxFee(fee) {
        this.fee = fee;
        return this;
    }
    addVersion(version) {
        this.version = version;
        return this;
    }
    addType(type) {
        this.type = type;
        return this;
    }
    addDeadline(deadline) {
        this.deadline = deadline;
        return this;
    }
    addProviderMosaicId(providerMosaicId) {
        this.providerMosaicId = providerMosaicId;
        return this;
    }
    addCurrencyBalanceIncrease(currencyBalanceIncrease) {
        this.currencyBalanceIncrease = currencyBalanceIncrease;
        return this;
    }
    addCurrencyBalanceChange(currencyBalanceChange) {
        this.currencyBalanceChange = currencyBalanceChange;
        return this;
    }
    addMosaicBalanceIncrease(mosaicBalanceIncrease) {
        this.mosaicBalanceIncrease = mosaicBalanceIncrease;
        return this;
    }
    addMosaicBalanceChange(mosaicBalanceChange) {
        this.mosaicBalanceChange = mosaicBalanceChange;
        return this;
    }
    build() {
        const builder = new flatbuffers.Builder(1);
        const providerMosaicIdUint32 = new Uint32Array(this.providerMosaicId);
        // Create vectors
        const signatureVector = ManualRateChangeBuffer_1.ManualRateChangeBuffer
            .createSignatureVector(builder, Array(...Array(64)).map(Number.prototype.valueOf, 0));
        const signerVector = ManualRateChangeBuffer_1.ManualRateChangeBuffer.createSignerVector(builder, Array(...Array(32)).map(Number.prototype.valueOf, 0));
        const deadlineVector = ManualRateChangeBuffer_1.ManualRateChangeBuffer.createDeadlineVector(builder, this.deadline);
        const feeVector = ManualRateChangeBuffer_1.ManualRateChangeBuffer.createMaxFeeVector(builder, this.fee);
        const providerMosaicIdVector = ManualRateChangeBuffer_1.ManualRateChangeBuffer.createProviderMosaicIdVector(builder, providerMosaicIdUint32);
        const currencyBalanceChangeVector = ManualRateChangeBuffer_1.ManualRateChangeBuffer.createCurrencyBalanceChangeVector(builder, this.currencyBalanceChange);
        const mosaicBalanceChangeVector = ManualRateChangeBuffer_1.ManualRateChangeBuffer.createMosaicBalanceChangeVector(builder, this.mosaicBalanceChange);
        ManualRateChangeBuffer_1.ManualRateChangeBuffer.startManualRateChangeBuffer(builder);
        ManualRateChangeBuffer_1.ManualRateChangeBuffer.addSize(builder, this.size);
        ManualRateChangeBuffer_1.ManualRateChangeBuffer.addSignature(builder, signatureVector);
        ManualRateChangeBuffer_1.ManualRateChangeBuffer.addSigner(builder, signerVector);
        ManualRateChangeBuffer_1.ManualRateChangeBuffer.addVersion(builder, this.version);
        ManualRateChangeBuffer_1.ManualRateChangeBuffer.addType(builder, this.type);
        ManualRateChangeBuffer_1.ManualRateChangeBuffer.addMaxFee(builder, feeVector);
        ManualRateChangeBuffer_1.ManualRateChangeBuffer.addDeadline(builder, deadlineVector);
        ManualRateChangeBuffer_1.ManualRateChangeBuffer.addProviderMosaicId(builder, providerMosaicIdVector);
        ManualRateChangeBuffer_1.ManualRateChangeBuffer.addCurrencyBalanceIncrease(builder, this.currencyBalanceIncrease);
        ManualRateChangeBuffer_1.ManualRateChangeBuffer.addCurrencyBalanceChange(builder, currencyBalanceChangeVector);
        ManualRateChangeBuffer_1.ManualRateChangeBuffer.addMosaicBalanceIncrease(builder, this.mosaicBalanceIncrease);
        ManualRateChangeBuffer_1.ManualRateChangeBuffer.addMosaicBalanceChange(builder, mosaicBalanceChangeVector);
        const codedTransfer = ManualRateChangeBuffer_1.ManualRateChangeBuffer.endManualRateChangeBuffer(builder);
        builder.finish(codedTransfer);
        const bytes = builder.asUint8Array();
        return new ManualRateChangeTransaction(bytes);
    }
    static stringToUint8(stringToConvert) {
        return format_1.Convert.hexToUint8(format_1.Convert.utf8ToHex(stringToConvert));
    }
}
exports.Builder = Builder;
