// tsjs-xpx-chain-sdk/dist/src/model/wallet/SimpleWallet.js
"use strict";
/*
 * Copyright 2023 ProximaX
 * Copyright 2018 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleWallet = void 0;
const core_1 = require("@js-joda/core");
const crypto_1 = require("../../core/crypto");
const format_1 = require("../../core/format");
const Account_1 = require("../account/Account");
const PublicAccount_1 = require("../account/PublicAccount");
const EncryptedPrivateKey_1 = require("./EncryptedPrivateKey");
const Wallet_1 = require("./Wallet");
/**
 * Simple wallet model generates a private key from a PRNG
 */
class SimpleWallet extends Wallet_1.Wallet {
    encryptedPrivateKey;
    version;
    /**
     * @internal
     * @param name
     * @param network
     * @param address
     * @param creationDate
     * @param encryptedPrivateKey
     * @param version - Account version
     */
    constructor(name, network, publicAccount, creationDate, 
    /**
     * The encrypted private key and information to decrypt it
     */
    encryptedPrivateKey, 
    /**
     * The account version
     */
    version = 1) {
        super(name, network, publicAccount, creationDate, 'sirius_dual');
        this.encryptedPrivateKey = encryptedPrivateKey;
        this.version = version;
    }
    /**
     * Create a Simple wallet
     * @param name - Wallet name
     * @param password - Password to encrypt wallet
     * @param network - Network id
     * @param version - The account version
     * @returns {SimpleWallet}
     */
    static create(name, password, network, version = 1) {
        // Create random bytes
        const randomBytesArray = crypto_1.Crypto.randomBytes(32);
        // Hash random bytes with entropy seed
        // Finalize and keep only 32 bytes
        const hashKey = format_1.Convert.uint8ArrayToHex(randomBytesArray); // TODO: derive private key correctly
        const dScheme = PublicAccount_1.PublicAccount.getDerivationSchemeFromAccVersion(version);
        // Create KeyPair from hash key
        const keyPair = crypto_1.KeyPair.createKeyPairFromPrivateKeyString(hashKey, dScheme);
        // Create publicAccount from public key
        const publicAccount = PublicAccount_1.PublicAccount.createFromPublicKey(format_1.Convert.uint8ArrayToHex(keyPair.publicKey), network);
        // Encrypt private key using password
        const encrypted = crypto_1.Crypto.encodePrivateKey(hashKey, password.value);
        const encryptedPrivateKey = new EncryptedPrivateKey_1.EncryptedPrivateKey(encrypted.ciphertext, encrypted.iv);
        return new SimpleWallet(name, network, publicAccount, core_1.LocalDateTime.now(), encryptedPrivateKey, version);
    }
    /**
     * Create a SimpleWallet from private key
     * @param name - Wallet name
     * @param password - Password to encrypt wallet
     * @param privateKey - Wallet private key
     * @param network - Network id
     * @param version - Account version
     * @returns {SimpleWallet}
     */
    static createFromPrivateKey(name, password, privateKey, network, version = 1) {
        // Create KeyPair from hash key
        const keyPair = crypto_1.KeyPair.createKeyPairFromPrivateKeyString(privateKey, PublicAccount_1.PublicAccount.getDerivationSchemeFromAccVersion(version));
        // Create publicAccount from public key
        const publicAccount = PublicAccount_1.PublicAccount.createFromPublicKey(format_1.Convert.uint8ArrayToHex(keyPair.publicKey), network);
        // Encrypt private key using password
        const encrypted = crypto_1.Crypto.encodePrivateKey(privateKey, password.value);
        const encryptedPrivateKey = new EncryptedPrivateKey_1.EncryptedPrivateKey(encrypted.ciphertext, encrypted.iv);
        return new SimpleWallet(name, network, publicAccount, core_1.LocalDateTime.now(), encryptedPrivateKey, version);
    }
    /**
     * Create a SimpleWallet from private key
     * @param name - Wallet name
     * @param encryptedPrivateKey - EncryptedPrivateKey
     * @param publicKey - Public Key
     * @param network - Network id
     * @param version - Account version
     * @returns {SimpleWallet}
     */
    static createFromEncryptedKey(name, encryptedKey, iv, publicKey, network, version = 1) {
        // Create publicAccount from public key
        const publicAccount = PublicAccount_1.PublicAccount.createFromPublicKey(publicKey, network);
        const encryptedPrivateKey = new EncryptedPrivateKey_1.EncryptedPrivateKey(encryptedKey, iv);
        return new SimpleWallet(name, network, publicAccount, core_1.LocalDateTime.now(), encryptedPrivateKey, version);
    }
    /**
     * Open a wallet and generate an Account
     * @param password - Password to decrypt private key
     * @returns {Account}
     */
    open(password) {
        return Account_1.Account.createFromPrivateKey(this.encryptedPrivateKey.decrypt(password), this.network, this.version);
    }
}
exports.SimpleWallet = SimpleWallet;
