// tsjs-xpx-chain-sdk/dist/src/model/transaction/AggregateV2TransactionCosignature.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AggregateV2TransactionCosignature = void 0;
/**
 * Model representing cosignature of an aggregate transaction.
 */
class AggregateV2TransactionCosignature {
    signatureDScheme;
    signature;
    signer;
    /**
     * @param signatureDScheme
     * @param signature
     * @param signer
     */
    constructor(
    /**
     * The derivation scheme of cosigner.
     */
    signatureDScheme, 
    /**
     * The signature of aggregate transaction done by the cosigner.
     */
    signature, 
    /**
     * The cosigner public account.
     */
    signer) {
        this.signatureDScheme = signatureDScheme;
        this.signature = signature;
        this.signer = signer;
    }
    /**
     * Create DTO object
     */
    toDTO() {
        return {
            derivationScheme: this.signatureDScheme,
            signature: this.signature,
            signer: this.signer.toDTO(),
        };
    }
}
exports.AggregateV2TransactionCosignature = AggregateV2TransactionCosignature;
