// tsjs-xpx-chain-sdk/dist/src/model/transaction/AggregateTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 * Copyright 2018 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AggregateBondedV1TransactionBuilder = exports.AggregateBondedTransactionBuilder = exports.AggregateCompleteV1TransactionBuilder = exports.AggregateCompleteTransactionBuilder = exports.AggregateTransaction = void 0;
const crypto_1 = require("../../core/crypto");
const AggregateTransaction_1 = require("../../infrastructure/builders/AggregateTransaction");
const PublicAccount_1 = require("../account/PublicAccount");
const SignedTransaction_1 = require("./SignedTransaction");
const Transaction_1 = require("./Transaction");
const TransactionType_1 = require("./TransactionType");
const TransactionTypeVersion_1 = require("./TransactionTypeVersion");
const FeeCalculationStrategy_1 = require("./FeeCalculationStrategy");
/**
 * Aggregate innerTransactions contain multiple innerTransactions that can be initiated by different accounts.
 */
class AggregateTransaction extends Transaction_1.Transaction {
    innerTransactions;
    cosignatures;
    /**
     * @param networkType
     * @param type
     * @param version
     * @param deadline
     * @param maxFee
     * @param innerTransactions
     * @param cosignatures
     * @param signature
     * @param signer
     * @param transactionInfo
     */
    constructor(networkType, type, version, deadline, maxFee, 
    /**
     * The array of innerTransactions included in the aggregate transaction.
     */
    innerTransactions, 
    /**
     * The array of transaction cosigners signatures.
     */
    cosignatures, signature, signer, transactionInfo) {
        super(type, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
        this.innerTransactions = innerTransactions;
        this.cosignatures = cosignatures;
    }
    /**
     * Create an aggregate complete v2 transaction object
     * @param deadline - The deadline to include the transaction.
     * @param innerTransactions - The array of inner innerTransactions.
     * @param networkType - The network type.
     * @param cosignatures
     * @param maxFee - (Optional) Max fee defined by the sender
     * @returns {AggregateTransaction}
     */
    static createComplete(deadline, innerTransactions, networkType, cosignatures = [], maxFee) {
        return new AggregateCompleteV1TransactionBuilder()
            .networkType(networkType)
            .deadline(deadline)
            .maxFee(maxFee)
            .innerTransactions(innerTransactions)
            .cosignatures(cosignatures)
            .build();
    }
    /**
     * Create an aggregate complete transaction object
     * @param deadline - The deadline to include the transaction.
     * @param innerTransactions - The array of inner innerTransactions.
     * @param networkType - The network type.
     * @param cosignatures
     * @param maxFee - (Optional) Max fee defined by the sender
     * @returns {AggregateTransaction}
     */
    static createCompleteV1(deadline, innerTransactions, networkType, cosignatures = [], maxFee) {
        return new AggregateCompleteV1TransactionBuilder()
            .networkType(networkType)
            .deadline(deadline)
            .maxFee(maxFee)
            .innerTransactions(innerTransactions)
            .cosignatures(cosignatures)
            .build();
    }
    /**
     * Create an aggregate bonded v2 transaction object
     * @param {Deadline} deadline
     * @param {InnerTransaction[]} innerTransactions
     * @param {NetworkType} networkType
     * @param {AggregateTransactionCosignature[]} cosignatures
     * @param {UInt64} maxFee - (Optional) Max fee defined by the sender
     * @return {AggregateTransaction}
     */
    static createBonded(deadline, innerTransactions, networkType, cosignatures = [], maxFee) {
        return new AggregateBondedV1TransactionBuilder()
            .networkType(networkType)
            .deadline(deadline)
            .maxFee(maxFee)
            .innerTransactions(innerTransactions)
            .cosignatures(cosignatures)
            .build();
    }
    /**
     * Create an aggregate bonded transaction object
     * @param {Deadline} deadline
     * @param {InnerTransaction[]} innerTransactions
     * @param {NetworkType} networkType
     * @param {AggregateTransactionCosignature[]} cosignatures
     * @param {UInt64} maxFee - (Optional) Max fee defined by the sender
     * @return {AggregateTransaction}
     */
    static createBondedV1(deadline, innerTransactions, networkType, cosignatures = [], maxFee) {
        return new AggregateBondedV1TransactionBuilder()
            .networkType(networkType)
            .deadline(deadline)
            .maxFee(maxFee)
            .innerTransactions(innerTransactions)
            .cosignatures(cosignatures)
            .build();
    }
    /**
     * @override Transaction.toJSON()
     * @description Serialize a transaction object - add own fields to the result of Transaction.toJSON()
     * @return {Object}
     * @memberof AggregateTransaction
     */
    toJSON() {
        const parent = super.toJSON();
        return {
            ...parent,
            transaction: {
                ...parent.transaction,
                transactions: this.innerTransactions.map((innerTransaction) => {
                    return innerTransaction.toJSON();
                }),
                cosignatures: this.cosignatures.map((cosignature) => {
                    return cosignature.toDTO();
                }),
            }
        };
    }
    /**
     * @internal
     * @returns {AggregateTransaction}
     */
    buildTransaction() {
        return new AggregateTransaction_1.Builder()
            .addSize(this.size)
            .addDeadline(this.deadline.toDTO())
            .addType(this.type)
            .addMaxFee(this.maxFee.toDTO())
            .addVersion(this.versionToDTO())
            .addTransactions(this.innerTransactions.map((transaction) => {
            return transaction.aggregateTransaction();
        })).build();
    }
    /**
     * Sign transaction with cosignatories creating a new SignedTransaction
     * @param initiatorAccount - Initiator account
     * @param cosignatories - The array of accounts that will cosign the transaction
     * @param generationHash - Network generation hash hex
     * @param {DerivationScheme} dScheme The derivation scheme
     * @returns {SignedTransaction}
     */
    newSignTransactionWithCosignatories(initiatorAccount, cosignatories, generationHash) {
        this.version.signatureDScheme = PublicAccount_1.PublicAccount.getDerivationSchemeFromAccVersion(initiatorAccount.version);
        const aggregateTransaction = this.buildTransaction();
        const signedTransactionRaw = aggregateTransaction
            .signTransactionWithCosigners(initiatorAccount, cosignatories, generationHash);
        return new SignedTransaction_1.SignedTransaction(signedTransactionRaw.payload, signedTransactionRaw.hash, initiatorAccount.publicKey, this.type, this.version.networkType);
    }
    /**
     * Sign transaction with cosignatories creating a new SignedTransaction
     * @param initiatorAccount - Initiator account
     * @param cosignatories - The array of accounts that will cosign the transaction
     * @param generationHash - Network generation hash hex
     * @param {DerivationScheme} dScheme The derivation scheme
     * @returns {SignedTransaction}
     */
    signTransactionWithCosignatories(initiatorAccount, cosignatories, generationHash) {
        return this.signTransactionWithCosignatoriesV1(initiatorAccount, cosignatories, generationHash);
    }
    /**
     * Sign transaction with cosignatories creating a new SignedTransaction
     * @param initiatorAccount - Initiator account
     * @param cosignatories - The array of accounts that will cosign the transaction
     * @param generationHash - Network generation hash hex
     * @returns {SignedTransaction}
     */
    signTransactionWithCosignatoriesV1(initiatorAccount, cosignatories, generationHash) {
        this.version.signatureDScheme = crypto_1.DerivationScheme.Unset;
        const aggregateTransaction = this.buildTransaction();
        const signedTransactionRaw = aggregateTransaction
            .signTransactionWithCosignersV1(initiatorAccount, cosignatories, generationHash);
        return new SignedTransaction_1.SignedTransaction(signedTransactionRaw.payload, signedTransactionRaw.hash, initiatorAccount.publicKey, this.type, this.version.networkType);
    }
    /**
     * Sign transaction with cosignatories collected from cosigned transactions and creating a new SignedTransaction
     * For off chain Aggregated Complete Transaction co-signing.
     * @param initiatorAccount - Initiator account
     * @param {CosignatureSignedTransaction[]} cosignatureSignedTransactions - Array of cosigned transaction
     * @param generationHash - Network generation hash hex
     * @param {DerivationScheme} dScheme The derivation scheme
     * @return {SignedTransaction}
     */
    newSignTransactionGivenSignatures(initiatorAccount, cosignatureSignedTransactions, generationHash) {
        this.version.signatureDScheme = PublicAccount_1.PublicAccount.getDerivationSchemeFromAccVersion(initiatorAccount.version);
        const aggregateTransaction = this.buildTransaction();
        const signedTransactionRaw = aggregateTransaction.signTransactionGivenSignatures(initiatorAccount, cosignatureSignedTransactions, generationHash);
        return new SignedTransaction_1.SignedTransaction(signedTransactionRaw.payload, signedTransactionRaw.hash, initiatorAccount.publicKey, this.type, this.version.networkType);
    }
    /**
     * Sign transaction with cosignatories collected from cosigned transactions and creating a new SignedTransaction
     * For off chain Aggregated Complete Transaction co-signing.
     * @param initiatorAccount - Initiator account
     * @param {CosignatureSignedTransaction[]} cosignatureSignedTransactions - Array of cosigned transaction
     * @param generationHash - Network generation hash hex
     * @param {DerivationScheme} dScheme The derivation scheme
     * @return {SignedTransaction}
     */
    signTransactionGivenSignatures(initiatorAccount, cosignatureSignedTransactions, generationHash) {
        return this.signTransactionGivenSignaturesV1(initiatorAccount, cosignatureSignedTransactions, generationHash);
    }
    /**
     * Sign transaction with cosignatories collected from cosigned transactions and creating a new SignedTransaction
     * For off chain Aggregated Complete Transaction co-signing.
     * @param initiatorAccount - Initiator account
     * @param {CosignatureSignedTransaction[]} cosignatureSignedTransactions - Array of cosigned transaction
     * @param generationHash - Network generation hash hex
     * @param {DerivationScheme} dScheme The derivation scheme
     * @return {SignedTransaction}
     */
    signTransactionGivenSignaturesV1(initiatorAccount, cosignatureSignedTransactions, generationHash) {
        this.version.signatureDScheme = crypto_1.DerivationScheme.Unset;
        const aggregateTransaction = this.buildTransaction();
        const signedTransactionRaw = aggregateTransaction.signTransactionGivenSignaturesV1(initiatorAccount, cosignatureSignedTransactions, generationHash);
        return new SignedTransaction_1.SignedTransaction(signedTransactionRaw.payload, signedTransactionRaw.hash, initiatorAccount.publicKey, this.type, this.version.networkType);
    }
    /**
     * Check if account has signed transaction
     * @param publicAccount - Signer public account
     * @returns {boolean}
     */
    signedByAccount(publicAccount) {
        return this.cosignatures.find((cosignature) => cosignature.signer.equals(publicAccount)) !== undefined
            || (this.signer !== undefined && this.signer.equals(publicAccount));
    }
    /**
     * @override Transaction.size()
     * @description get the byte size of a AggregateTransaction
     * @returns {number}
     * @memberof AggregateTransaction
     */
    get size() {
        return AggregateTransaction.calculateSize(this.innerTransactions || [], this.cosignatures.length);
    }
    static calculateSize(innerTransactions, numOfCosignatures) {
        // removes 80 bytes from each inner tx
        const innerTransactionsSumSize = innerTransactions.reduce((previous, current) => previous + current.size - 80, 0);
        const byteSize = Transaction_1.Transaction.getHeaderSize();
        // set static byte size fields
        const byteTransactionsSize = 4;
        const cosignerBytes = 96 * numOfCosignatures;
        return byteSize + byteTransactionsSize + innerTransactionsSumSize + cosignerBytes;
    }
}
exports.AggregateTransaction = AggregateTransaction;
class AggregateCompleteTransactionBuilder extends Transaction_1.TransactionBuilder {
    _cosignatures = [];
    _innerTransactions = [];
    cosignatures(cosignatures) {
        this._cosignatures = cosignatures;
        return this;
    }
    innerTransactions(innerTransactions) {
        this._innerTransactions = innerTransactions;
        return this;
    }
    build() {
        return new AggregateTransaction(this._networkType, TransactionType_1.TransactionType.AGGREGATE_COMPLETE_V2, this._version || TransactionTypeVersion_1.TransactionTypeVersion.AGGREGATE_COMPLETE_V2, this._deadline ? this._deadline : this._createNewDeadlineFn(), this._maxFee ? this._maxFee : (0, FeeCalculationStrategy_1.calculateFee)(AggregateTransaction.calculateSize(this._innerTransactions, this._cosignatures.length), this._feeCalculationStrategy), this._innerTransactions, this._cosignatures, this._signature, this._signer, this._transactionInfo);
    }
}
exports.AggregateCompleteTransactionBuilder = AggregateCompleteTransactionBuilder;
class AggregateCompleteV1TransactionBuilder extends Transaction_1.TransactionBuilder {
    _cosignatures = [];
    _innerTransactions = [];
    cosignatures(cosignatures) {
        this._cosignatures = cosignatures;
        return this;
    }
    innerTransactions(innerTransactions) {
        this._innerTransactions = innerTransactions;
        return this;
    }
    build() {
        return new AggregateTransaction(this._networkType, TransactionType_1.TransactionType.AGGREGATE_COMPLETE_V1, this._version || TransactionTypeVersion_1.TransactionTypeVersion.AGGREGATE_COMPLETE_V1, this._deadline ? this._deadline : this._createNewDeadlineFn(), this._maxFee ? this._maxFee : (0, FeeCalculationStrategy_1.calculateFee)(AggregateTransaction.calculateSize(this._innerTransactions, this._cosignatures.length), this._feeCalculationStrategy), this._innerTransactions, this._cosignatures, this._signature, this._signer, this._transactionInfo);
    }
}
exports.AggregateCompleteV1TransactionBuilder = AggregateCompleteV1TransactionBuilder;
class AggregateBondedTransactionBuilder extends Transaction_1.TransactionBuilder {
    _cosignatures = [];
    _innerTransactions = [];
    cosignatures(cosignatures) {
        this._cosignatures = cosignatures;
        return this;
    }
    innerTransactions(innerTransactions) {
        this._innerTransactions = innerTransactions;
        return this;
    }
    build() {
        return new AggregateTransaction(this._networkType, TransactionType_1.TransactionType.AGGREGATE_BONDED_V2, this._version || TransactionTypeVersion_1.TransactionTypeVersion.AGGREGATE_BONDED_V2, this._deadline ? this._deadline : this._createNewDeadlineFn(), this._maxFee ? this._maxFee : (0, FeeCalculationStrategy_1.calculateFee)(AggregateTransaction.calculateSize(this._innerTransactions, this._cosignatures.length), this._feeCalculationStrategy), this._innerTransactions, this._cosignatures, this._signature, this._signer, this._transactionInfo);
    }
}
exports.AggregateBondedTransactionBuilder = AggregateBondedTransactionBuilder;
class AggregateBondedV1TransactionBuilder extends Transaction_1.TransactionBuilder {
    _cosignatures = [];
    _innerTransactions = [];
    cosignatures(cosignatures) {
        this._cosignatures = cosignatures;
        return this;
    }
    innerTransactions(innerTransactions) {
        this._innerTransactions = innerTransactions;
        return this;
    }
    build() {
        return new AggregateTransaction(this._networkType, TransactionType_1.TransactionType.AGGREGATE_BONDED_V1, this._version || TransactionTypeVersion_1.TransactionTypeVersion.AGGREGATE_BONDED_V1, this._deadline ? this._deadline : this._createNewDeadlineFn(), this._maxFee ? this._maxFee : (0, FeeCalculationStrategy_1.calculateFee)(AggregateTransaction.calculateSize(this._innerTransactions, this._cosignatures.length), this._feeCalculationStrategy), this._innerTransactions, this._cosignatures, this._signature, this._signer, this._transactionInfo);
    }
}
exports.AggregateBondedV1TransactionBuilder = AggregateBondedV1TransactionBuilder;
