// tsjs-xpx-chain-sdk/dist/src/model/transaction/BasicTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BasicTransaction = void 0;
const Transaction_1 = require("./Transaction");
/**
 * A known transaction class that hold the basic transaction, without any extra details.
 */
class BasicTransaction extends Transaction_1.Transaction {
    /**
     * @constructor
     * @param type
     * @param networkType
     * @param version
     * @param deadline
     * @param maxFee
     * @param signature
     * @param signer
     * @param transactionInfo
     */
    constructor(
    /**
     * The transaction type.
     */
    type, 
    /**
     * The network type.
     */
    networkType, 
    /**
     * The transaction version number.
     */
    version, 
    /**
     * The deadline to include the transaction.
     */
    deadline, 
    /**
     * A sender of a transaction must specify during the transaction definition a max_fee,
     * meaning the maximum fee the account allows to spend for this transaction.
     */
    maxFee, 
    /**
     * The account of the transaction creator.
     */
    signer, 
    /**
     * The transaction signature (missing if part of an aggregate transaction).
     */
    signature, 
    /**
     * Transactions meta data object contains additional information about the transaction.
     */
    transactionInfo) {
        super(type, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
    }
    /**
     * @override Transaction.size()
     * @description get the byte size of a BasicTransaction, will return error
     * @returns {number}
     * @memberof BasicTransaction
     */
    get size() {
        return Transaction_1.Transaction.getHeaderSize();
    }
    /**
     * @override Transaction.buildTransaction()
     * @description buildTransaction of BasicTransaction, will return error
     * @returns {never}
     * @memberof BasicTransaction
     */
    buildTransaction() {
        throw new Error("SDK not yet implemented transaction");
    }
    /**
     * Convert an aggregate transaction to an inner transaction including transaction signer, will return error
     * @override Transaction.toAggregate()
     * @internal
     * @param signer - Transaction signer.
     * @returns {never}
     */
    toAggregate(signer) {
        throw new Error("SDK not yet implemented transaction");
    }
    /**
     * @override Transaction.toJSON()
     * @description Create JSON object
     * @returns {Object}
     * @memberof BasicTransaction
     */
    toJSON() {
        const commonTransactionObject = {
            type: this.type,
            networkType: this.version.networkType,
            version: this.versionToDTO(),
            maxFee: this.maxFee.toDTO(),
            deadline: this.deadline.toDTO(),
            signature: this.signature ? this.signature : '',
            signer: this.signer.publicKey
        };
        return { transaction: commonTransactionObject };
    }
}
exports.BasicTransaction = BasicTransaction;
