// tsjs-xpx-chain-sdk/dist/src/model/transaction/Deadline.js
"use strict";
/*
 * Copyright 2023 ProximaX
 * Copyright 2018 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultCreateNewDeadline = exports.Deadline = void 0;
const core_1 = require("@js-joda/core");
const UInt64_1 = require("../UInt64");
/**
 * The deadline of the transaction. The deadline is given as the number of seconds elapsed since the creation of the nemesis block.
 * If a transaction does not get included in a block before the deadline is reached, it is deleted.
 */
class Deadline {
    /**
     * @type {number}
     */
    static timestampNemesisBlock = 1459468800;
    /**
     * Deadline value
     */
    value;
    /**
     * Adjusted value
     */
    adjustedValue;
    /**
     * Create deadline model
     * @param deadline
     * @param chronoUnit
     * @returns {Deadline}
     */
    static create(deadline = 2, chronoUnit = core_1.ChronoUnit.HOURS) {
        const networkTimeStamp = (new Date()).getTime();
        const timeStampDateTime = core_1.LocalDateTime.ofInstant(core_1.Instant.ofEpochMilli(networkTimeStamp), core_1.ZoneId.SYSTEM);
        const deadlineDateTime = timeStampDateTime.plus(deadline, chronoUnit);
        if (deadline <= 0) {
            throw new Error('deadline should be greater than 0');
        }
        else if (timeStampDateTime.plus(24, core_1.ChronoUnit.HOURS).compareTo(deadlineDateTime) !== 1) {
            throw new Error('deadline should be less than 24 hours');
        }
        const adjustedValue = UInt64_1.UInt64.fromUint((deadlineDateTime.atZone(core_1.ZoneId.SYSTEM).toInstant().toEpochMilli() - Deadline.timestampNemesisBlock * 1000));
        return new Deadline(deadlineDateTime, adjustedValue);
    }
    /**
     * Create unrestricted deadline model
     * @param deadline
     * @param chronoUnit
     * @returns {Deadline}
     */
    static createForBonded(deadline = 2, chronoUnit = core_1.ChronoUnit.HOURS, maximum) {
        const networkTimeStamp = (new Date()).getTime();
        const timeStampDateTime = core_1.LocalDateTime.ofInstant(core_1.Instant.ofEpochMilli(networkTimeStamp), core_1.ZoneId.SYSTEM);
        const deadlineDateTime = timeStampDateTime.plus(deadline, chronoUnit);
        if (deadline <= 0) {
            throw new Error('deadline should be greater than 0');
        }
        const adjustedValue = UInt64_1.UInt64.fromUint((deadlineDateTime.atZone(core_1.ZoneId.SYSTEM).toInstant().toEpochMilli() - Deadline.timestampNemesisBlock * 1000));
        if (maximum && adjustedValue.toBigInt() > maximum.toBigInt()) {
            return Deadline.createFromUint64(maximum);
        }
        return new Deadline(deadlineDateTime, adjustedValue);
    }
    /**
     * @internal
     * @param value
     * @returns {Deadline}
     */
    static createFromDTO(value) {
        const adjustedValue = new UInt64_1.UInt64(value);
        const dateSeconds = adjustedValue.compact();
        const deadline = core_1.LocalDateTime.ofInstant(core_1.Instant.ofEpochMilli(Math.round(dateSeconds + Deadline.timestampNemesisBlock * 1000)), core_1.ZoneId.SYSTEM);
        return new Deadline(deadline, adjustedValue);
    }
    /**
     * Create unrestricted deadline model from UInt64
     * @param UInt64 value
     * @returns {Deadline}
     */
    static createFromUint64(value) {
        const adjustedValue = value;
        const dateSeconds = adjustedValue.compact();
        const deadline = core_1.LocalDateTime.ofInstant(core_1.Instant.ofEpochMilli(Math.round(dateSeconds + Deadline.timestampNemesisBlock * 1000)), core_1.ZoneId.SYSTEM);
        return new Deadline(deadline, adjustedValue);
    }
    /**
     * @param deadline
     */
    constructor(deadline, adjustedValue) {
        this.value = deadline;
        this.adjustedValue = adjustedValue;
    }
    /**
     * @internal
     * @param value
     * @returns {Deadline}
     */
    static createEmpty() {
        return new Deadline(null, UInt64_1.UInt64.fromUint(0));
    }
    /**
     * @internal
     */
    toDTO() {
        return this.adjustedValue.toDTO();
    }
}
exports.Deadline = Deadline;
const DefaultCreateNewDeadline = () => Deadline.create();
exports.DefaultCreateNewDeadline = DefaultCreateNewDeadline;
