// tsjs-xpx-chain-sdk/dist/src/model/transaction/FeeCalculationStrategy.js
"use strict";
// Copyright 2019 ProximaX Limited. All rights reserved.
// Use of this source code is governed by the Apache 2.0
// license that can be found in the LICENSE file
Object.defineProperty(exports, "__esModule", { value: true });
exports.calculateFee = exports.DefaultFeeCalculationStrategy = exports.FeeCalculationStrategy = void 0;
const UInt64_1 = require("../UInt64");
var FeeCalculationStrategy;
(function (FeeCalculationStrategy) {
    FeeCalculationStrategy[FeeCalculationStrategy["ZeroFeeCalculationStrategy"] = 0] = "ZeroFeeCalculationStrategy";
    FeeCalculationStrategy[FeeCalculationStrategy["LowFeeCalculationStrategy"] = 15000] = "LowFeeCalculationStrategy";
    FeeCalculationStrategy[FeeCalculationStrategy["MiddleFeeCalculationStrategy"] = 150000] = "MiddleFeeCalculationStrategy";
    FeeCalculationStrategy[FeeCalculationStrategy["HighFeeCalculationStrategy"] = 1500000] = "HighFeeCalculationStrategy";
})(FeeCalculationStrategy || (exports.FeeCalculationStrategy = FeeCalculationStrategy = {}));
// export const DefaultMaxFee = 75000000;
exports.DefaultFeeCalculationStrategy = FeeCalculationStrategy.MiddleFeeCalculationStrategy;
const calculateFee = (transactionByteSize, feeCalculationStratgy = exports.DefaultFeeCalculationStrategy) => {
    // return UInt64.fromUint(Math.min(DefaultMaxFee, transactionByteSize * feeCalculationStratgy));
    return UInt64_1.UInt64.fromUint(transactionByteSize * feeCalculationStratgy);
};
exports.calculateFee = calculateFee;
