// tsjs-xpx-chain-sdk/dist/src/model/transaction/HarvesterTransaction.js
"use strict";
/*
 * Copyright 2022 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoveHarvesterTransactionBuilder = exports.AddHarvesterTransactionBuilder = exports.HarvesterTransactionBuilder = exports.HarvesterTransaction = void 0;
const HarvesterTransaction_1 = require("../../infrastructure/builders/HarvesterTransaction");
const Transaction_1 = require("./Transaction");
const Transaction_2 = require("./Transaction");
const TransactionType_1 = require("./TransactionType");
const TransactionTypeVersion_1 = require("./TransactionTypeVersion");
const FeeCalculationStrategy_1 = require("./FeeCalculationStrategy");
/**
 * Harvester transactions contain data about add or remove harvester.
 */
class HarvesterTransaction extends Transaction_2.Transaction {
    harvesterKey;
    /**
     * Create a add harvester transaction object
     * @param deadline - The deadline to include the transaction.
     * @param harvesterKey - The public account of the harvester
     * @param networkType - The network type.
     * @param maxFee - (Optional) Max fee defined by the sender
     * @returns {TransferTransaction}
     */
    static createAdd(deadline, harvesterKey, networkType, maxFee) {
        return new AddHarvesterTransactionBuilder()
            .networkType(networkType)
            .deadline(deadline)
            .maxFee(maxFee)
            .harvesterKey(harvesterKey)
            .build();
    }
    /**
     * Create a remove harvester transaction object
     * @param deadline - The deadline to include the transaction.
     * @param harvesterKey - The public account of the harvester
     * @param networkType - The network type.
     * @param maxFee - (Optional) Max fee defined by the sender
     * @returns {TransferTransaction}
     */
    static createRemove(deadline, harvesterKey, networkType, maxFee) {
        return new RemoveHarvesterTransactionBuilder()
            .networkType(networkType)
            .deadline(deadline)
            .maxFee(maxFee)
            .harvesterKey(harvesterKey)
            .build();
    }
    /**
     * @param networkType
     * @param type
     * @param version
     * @param deadline
     * @param maxFee
     * @param harvesterKey
     * @param signature
     * @param signer
     * @param transactionInfo
     */
    constructor(networkType, type, version, deadline, maxFee, 
    /**
     * The harvester key
     */
    harvesterKey, signature, signer, transactionInfo) {
        super(type, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
        this.harvesterKey = harvesterKey;
    }
    /**
     * @override Transaction.size()
     * @description get the byte size of a HarvesterTransaction
     * @returns {number}
     * @memberof HarvesterTransaction
     */
    get size() {
        return HarvesterTransaction.calculateSize();
    }
    static calculateSize() {
        const byteSize = Transaction_2.Transaction.getHeaderSize();
        const harvesterKey = 32;
        return byteSize + harvesterKey;
    }
    /**
     * @override Transaction.toJSON()
     * @description Serialize a transaction object - add own fields to the result of Transaction.toJSON()
     * @return {Object}
     * @memberof HarvesterTransaction
     */
    toJSON() {
        const parent = super.toJSON();
        return {
            ...parent,
            transaction: {
                ...parent.transaction,
                harvesterKey: this.harvesterKey.toDTO()
            }
        };
    }
    /**
     * @internal
     * @returns {VerifiableTransaction}
     */
    buildTransaction() {
        return new HarvesterTransaction_1.Builder()
            .addType(this.type)
            .addSize(this.size)
            .addDeadline(this.deadline.toDTO())
            .addMaxFee(this.maxFee.toDTO())
            .addVersion(this.versionToDTO())
            .addHarvesterKey(this.harvesterKey.publicKey)
            .build();
    }
}
exports.HarvesterTransaction = HarvesterTransaction;
class HarvesterTransactionBuilder extends Transaction_1.TransactionBuilder {
    _harvesterKey;
    _transactionType;
    _transactionTypeVersion;
    constructor(transactionType, transactionTypeVersion) {
        super();
        this._transactionType = transactionType;
        this._transactionTypeVersion = transactionTypeVersion;
    }
    harvesterKey(harvesterKey) {
        this._harvesterKey = harvesterKey;
        return this;
    }
    build() {
        return new HarvesterTransaction(this._networkType, this._transactionType, this._transactionTypeVersion, this._deadline ? this._deadline : this._createNewDeadlineFn(), this._maxFee ? this._maxFee : (0, FeeCalculationStrategy_1.calculateFee)(HarvesterTransaction.calculateSize(), this._feeCalculationStrategy), this._harvesterKey, this._signature, this._signer, this._transactionInfo);
    }
}
exports.HarvesterTransactionBuilder = HarvesterTransactionBuilder;
class AddHarvesterTransactionBuilder extends HarvesterTransactionBuilder {
    constructor() {
        super(TransactionType_1.TransactionType.ADD_HARVESTER, TransactionTypeVersion_1.TransactionTypeVersion.ADD_HARVESTER);
    }
}
exports.AddHarvesterTransactionBuilder = AddHarvesterTransactionBuilder;
class RemoveHarvesterTransactionBuilder extends HarvesterTransactionBuilder {
    constructor() {
        super(TransactionType_1.TransactionType.REMOVE_HARVESTER, TransactionTypeVersion_1.TransactionTypeVersion.REMOVE_HARVESTER);
    }
}
exports.RemoveHarvesterTransactionBuilder = RemoveHarvesterTransactionBuilder;
