// tsjs-xpx-chain-sdk/dist/src/model/transaction/HashLockTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 * Copyright 2018 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.HashLockTransactionBuilder = exports.HashLockTransaction = void 0;
const HashLockTransaction_1 = require("../../infrastructure/builders/HashLockTransaction");
const TransactionHash_1 = require("./TransactionHash");
const Transaction_1 = require("./Transaction");
const TransactionType_1 = require("./TransactionType");
const TransactionTypeVersion_1 = require("./TransactionTypeVersion");
const FeeCalculationStrategy_1 = require("./FeeCalculationStrategy");
/**
 * Hash lock transaction is used before sending an Aggregate bonded transaction, as a deposit to announce the transaction.
 * When aggregate bonded transaction is confirmed funds are returned to HashLockTransaction signer.
 *
 * @since 1.0
 */
class HashLockTransaction extends Transaction_1.Transaction {
    mosaic;
    duration;
    /**
     * Aggregate bonded hash.
     */
    hash;
    /**
     * Create a Lock funds transaction object
     * @param deadline - The deadline to include the transaction.
     * @param mosaic - The locked mosaic.
     * @param duration - The funds lock duration.
     * @param signedTransaction - The signed transaction for which funds are locked.
     * @param networkType - The network type.
     * @param maxFee - (Optional) Max fee defined by the sender
     * @returns {HashLockTransaction}
     */
    static create(deadline, mosaic, duration, transactionHash, networkType, maxFee) {
        let txnHashToLock = transactionHash instanceof TransactionHash_1.TransactionHash ?
            transactionHash : new TransactionHash_1.TransactionHash(transactionHash.hash, transactionHash.type);
        return new HashLockTransactionBuilder()
            .networkType(networkType)
            .deadline(deadline)
            .maxFee(maxFee)
            .mosaic(mosaic)
            .duration(duration)
            .transactionHash(txnHashToLock)
            .build();
    }
    /**
     * @param networkType
     * @param version
     * @param deadline
     * @param maxFee
     * @param mosaic
     * @param duration
     * @param signedTransaction
     * @param signature
     * @param signer
     * @param transactionInfo
     */
    constructor(networkType, version, deadline, maxFee, 
    /**
     * The locked mosaic.
     */
    mosaic, 
    /**
     * The funds lock duration.
     */
    duration, transactionHash, signature, signer, transactionInfo) {
        super(TransactionType_1.TransactionType.HASH_LOCK, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
        this.mosaic = mosaic;
        this.duration = duration;
        this.hash = transactionHash.hash;
        if (transactionHash.type !== TransactionType_1.TransactionType.AGGREGATE_BONDED_V1 && transactionHash.type !== TransactionType_1.TransactionType.AGGREGATE_BONDED_V2) {
            throw new Error('Signed transaction must be Aggregate Bonded Transaction');
        }
    }
    /**
     * @override Transaction.size()
     * @description get the byte size of a HashLockTransaction
     * @returns {number}
     * @memberof HashLockTransaction
     */
    get size() {
        return HashLockTransaction.calculateSize();
    }
    static calculateSize() {
        const byteSize = Transaction_1.Transaction.getHeaderSize();
        // set static byte size fields
        const byteMosaicId = 8;
        const byteAmount = 8;
        const byteDuration = 8;
        const byteHash = 32;
        return byteSize + byteMosaicId + byteAmount + byteDuration + byteHash;
    }
    /**
     * @override Transaction.toJSON()
     * @description Serialize a transaction object - add own fields to the result of Transaction.toJSON()
     * @return {Object}
     * @memberof HashLockTransaction
     */
    toJSON() {
        const parent = super.toJSON();
        return {
            ...parent,
            transaction: {
                ...parent.transaction,
                mosaicId: this.mosaic.id.id.toDTO(),
                amount: this.mosaic.amount.toDTO(),
                duration: this.duration.toDTO(),
                hash: this.hash,
            }
        };
    }
    /**
     * @internal
     * @return {VerifiableTransaction}
     */
    buildTransaction() {
        return new HashLockTransaction_1.Builder()
            .addSize(this.size)
            .addDeadline(this.deadline.toDTO())
            .addType(this.type)
            .addMaxFee(this.maxFee.toDTO())
            .addVersion(this.versionToDTO())
            .addMosaicId(this.mosaic.id.id.toDTO())
            .addMosaicAmount(this.mosaic.amount.toDTO())
            .addDuration(this.duration.toDTO())
            .addHash(this.hash)
            .build();
    }
}
exports.HashLockTransaction = HashLockTransaction;
class HashLockTransactionBuilder extends Transaction_1.TransactionBuilder {
    _mosaic;
    _duration;
    _transactionHash;
    mosaic(mosaic) {
        this._mosaic = mosaic;
        return this;
    }
    duration(duration) {
        this._duration = duration;
        return this;
    }
    transactionHash(transactionHash) {
        this._transactionHash = transactionHash instanceof TransactionHash_1.TransactionHash ?
            transactionHash : new TransactionHash_1.TransactionHash(transactionHash.hash, transactionHash.type);
        return this;
    }
    build() {
        return new HashLockTransaction(this._networkType, this._version || TransactionTypeVersion_1.TransactionTypeVersion.HASH_LOCK, this._deadline ? this._deadline : this._createNewDeadlineFn(), this._maxFee ? this._maxFee : (0, FeeCalculationStrategy_1.calculateFee)(HashLockTransaction.calculateSize(), this._feeCalculationStrategy), this._mosaic, this._duration, this._transactionHash, this._signature, this._signer, this._transactionInfo);
    }
}
exports.HashLockTransactionBuilder = HashLockTransactionBuilder;
