// tsjs-xpx-chain-sdk/dist/src/model/transaction/Message.js
"use strict";
/*
 * Copyright 2023 ProximaX
 * Copyright 2018 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Message = void 0;
const format_1 = require("../../core/format");
/**
 * An abstract message class that serves as the base class of all message types.
 */
class Message {
    type;
    payload;
    message;
    /**
     * @internal
     * @param hex
     * @returns {string}
     */
    static decodeHex(hex) {
        let message = "";
        try {
            let uint8Array = format_1.Convert.hexToUint8(hex);
            message = new TextDecoder().decode(uint8Array);
        }
        catch (e) {
        }
        return message;
    }
    /**
     * @internal
     * @param message
     * @returns {string}
     */
    static encodeToHex(message) {
        let payload = "";
        try {
            let uint8Array = new TextEncoder().encode(message);
            payload = format_1.Convert.uint8ArrayToHex(uint8Array);
        }
        catch (e) {
        }
        return payload;
    }
    /**
     * @internal
     * @param type
     * @param payload - Hexadecimal message payload
     * @param message
     */
    constructor(/**
                 * Message type
                 */ type, 
    /**
     * Message payload
     */
    payload, 
    /**
     * Message payload
     */
    message = "") {
        this.type = type;
        this.payload = payload;
        this.message = message;
    }
    /**
     * Returns the byte size of the message
     */
    size() {
        return this.payload.length ? this.payload.length / 2 : 0;
    }
    /**
     * Create DTO object
     */
    toDTO() {
        return {
            type: this.type,
            payload: this.payload,
            message: this.message,
        };
    }
}
exports.Message = Message;
