// tsjs-xpx-chain-sdk/dist/src/model/transaction/MosaicRemoveLevyTransaction.js
"use strict";
// Copyright 2021 ProximaX Limited. All rights reserved.
// Use of this source code is governed by the Apache 2.0
// license that can be found in the LICENSE file
Object.defineProperty(exports, "__esModule", { value: true });
exports.MosaicRemoveLevyTransactionBuilder = exports.MosaicRemoveLevyTransaction = void 0;
const Transaction_1 = require("./Transaction");
const TransactionTypeVersion_1 = require("./TransactionTypeVersion");
const MosaicRemoveLevyTransaction_1 = require("../../infrastructure/builders/MosaicRemoveLevyTransaction");
const TransactionType_1 = require("./TransactionType");
const FeeCalculationStrategy_1 = require("./FeeCalculationStrategy");
/**
 * Mosaic Modify levy transaction contains information about levy of mosaic being modified.
 */
class MosaicRemoveLevyTransaction extends Transaction_1.Transaction {
    mosaicId;
    /**
     * Create a mosaic modify levy transaction object
     * @returns {MosaicRemoveLevyTransaction}
     */
    static create(deadline, mosaicId, networkType, maxFee) {
        return new MosaicRemoveLevyTransactionBuilder()
            .networkType(networkType)
            .deadline(deadline)
            .maxFee(maxFee)
            .mosaicId(mosaicId)
            .build();
    }
    /**
     * @param transactionType
     * @param networkType
     * @param deadline
     * @param maxFee
     * @param mosaicId
     * @param mosaicLevy
     * @param signature
     * @param signer
     * @param transactionInfo
     */
    constructor(networkType, version, deadline, maxFee, mosaicId, signature, signer, transactionInfo) {
        super(TransactionType_1.TransactionType.REMOVE_MOSAIC_LEVY, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
        this.mosaicId = mosaicId;
    }
    /**
     * @override Transaction.size()
     * @description get the byte size of a transaction
     * @returns {number}
     * @memberof Transaction
     */
    get size() {
        return MosaicRemoveLevyTransaction.calculateSize();
    }
    static calculateSize() {
        const byteSize = Transaction_1.Transaction.getHeaderSize()
            + 8; // mosaicId
        return byteSize;
    }
    /**
     * @override Transaction.toJSON()
     * @description Serialize a transaction object - add own fields to the result of Transaction.toJSON()
     * @return {Object}
     * @memberof MosaicRemoveLevyTransaction
     */
    toJSON() {
        const parent = super.toJSON();
        return {
            ...parent,
            transaction: {
                ...parent.transaction,
                mosaicId: this.mosaicId.toHex()
            }
        };
    }
    /**
     * @internal
     * @returns {VerifiableTransaction}
     */
    buildTransaction() {
        return new MosaicRemoveLevyTransaction_1.Builder()
            .addSize(this.size)
            .addType(this.type)
            .addVersion(this.versionToDTO())
            .addDeadline(this.deadline.toDTO())
            .addMaxFee(this.maxFee.toDTO())
            .addMosaicId(this.mosaicId.id.toDTO())
            .build();
    }
}
exports.MosaicRemoveLevyTransaction = MosaicRemoveLevyTransaction;
class MosaicRemoveLevyTransactionBuilder extends Transaction_1.TransactionBuilder {
    _transactionType;
    _mosaicId;
    constructor() {
        super();
        this._transactionType = TransactionType_1.TransactionType.REMOVE_MOSAIC_LEVY;
    }
    mosaicId(mosaicId) {
        this._mosaicId = mosaicId;
        return this;
    }
    build() {
        return new MosaicRemoveLevyTransaction(this._networkType, this._version || TransactionTypeVersion_1.TransactionTypeVersion.MOSAIC_REMOVE_LEVY, this._deadline ? this._deadline : this._createNewDeadlineFn(), this._maxFee ? this._maxFee : (0, FeeCalculationStrategy_1.calculateFee)(MosaicRemoveLevyTransaction.calculateSize(), this._feeCalculationStrategy), this._mosaicId, this._signature, this._signer, this._transactionInfo);
    }
}
exports.MosaicRemoveLevyTransactionBuilder = MosaicRemoveLevyTransactionBuilder;
