// tsjs-xpx-chain-sdk/dist/src/model/transaction/MosaicModifyLevyTransaction.js
"use strict";
// Copyright 2021 ProximaX Limited. All rights reserved.
// Use of this source code is governed by the Apache 2.0
// license that can be found in the LICENSE file
Object.defineProperty(exports, "__esModule", { value: true });
exports.MosaicModifyLevyTransactionBuilder = exports.MosaicModifyLevyTransaction = void 0;
const Transaction_1 = require("./Transaction");
const TransactionTypeVersion_1 = require("./TransactionTypeVersion");
const MosaicModifyLevyTransaction_1 = require("../../infrastructure/builders/MosaicModifyLevyTransaction");
const TransactionType_1 = require("./TransactionType");
const FeeCalculationStrategy_1 = require("./FeeCalculationStrategy");
const MosaicLevyType_1 = require("../mosaic/MosaicLevyType");
/**
 * Mosaic Modify levy transaction contains information about levy of mosaic being modified.
 */
class MosaicModifyLevyTransaction extends Transaction_1.Transaction {
    mosaicId;
    mosaicLevy;
    /**
     * Create a mosaic modify levy transaction object
     * @returns {MosaicModifyLevyTransaction}
     */
    static create(deadline, mosaicId, mosaicLevy, networkType, maxFee) {
        return new MosaicModifyLevyTransactionBuilder()
            .networkType(networkType)
            .deadline(deadline)
            .maxFee(maxFee)
            .mosaicLevy(mosaicLevy)
            .mosaicId(mosaicId)
            .build();
    }
    /**
     * @param transactionType
     * @param networkType
     * @param deadline
     * @param maxFee
     * @param mosaicId
     * @param mosaicLevy
     * @param signature
     * @param signer
     * @param transactionInfo
     */
    constructor(networkType, version, deadline, maxFee, mosaicId, mosaicLevy, signature, signer, transactionInfo) {
        if (mosaicLevy.type === MosaicLevyType_1.MosaicLevyType.LevyNone) {
            throw new Error('Levy type not allowed');
        }
        ;
        super(TransactionType_1.TransactionType.MODIFY_MOSAIC_LEVY, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
        this.mosaicId = mosaicId;
        this.mosaicLevy = mosaicLevy;
    }
    /**
     * @override Transaction.size()
     * @description get the byte size of a transaction
     * @returns {number}
     * @memberof Transaction
     */
    get size() {
        return MosaicModifyLevyTransaction.calculateSize();
    }
    static calculateSize() {
        const byteSize = Transaction_1.Transaction.getHeaderSize()
            + 8 // mosaicId
            + 1 // levy type
            + 25 // levy recipient
            + 8 // levy mosaicId
            + 8; // levy fee
        return byteSize;
    }
    /**
     * @override Transaction.toJSON()
     * @description Serialize a transaction object - add own fields to the result of Transaction.toJSON()
     * @return {Object}
     * @memberof MosaicModifyLevyTransaction
     */
    toJSON() {
        const parent = super.toJSON();
        return {
            ...parent,
            transaction: {
                ...parent.transaction,
                mosaicId: this.mosaicId.toDTO(),
                mosaicLevy: {
                    type: this.mosaicLevy.type,
                    mosaicId: this.mosaicLevy.mosaicId.toDTO(),
                    recipient: this.mosaicLevy.recipient.toDTO(),
                    fee: this.mosaicLevy.fee.toDTO(),
                }
            }
        };
    }
    /**
     * @internal
     * @returns {VerifiableTransaction}
     */
    buildTransaction() {
        return new MosaicModifyLevyTransaction_1.Builder()
            .addSize(this.size)
            .addType(this.type)
            .addVersion(this.versionToDTO())
            .addDeadline(this.deadline.toDTO())
            .addMaxFee(this.maxFee.toDTO())
            .addMosaicId(this.mosaicId.id.toDTO())
            .addLevyType(this.mosaicLevy.type)
            .addLevyRecipeint(this.mosaicLevy.recipient.plain())
            .addLevyMosaicId(this.mosaicLevy.mosaicId.id.toDTO())
            .addLevyFee(this.mosaicLevy.fee.toDTO())
            .build();
    }
}
exports.MosaicModifyLevyTransaction = MosaicModifyLevyTransaction;
class MosaicModifyLevyTransactionBuilder extends Transaction_1.TransactionBuilder {
    _transactionType;
    _mosaicId;
    _mosaicLevy;
    constructor() {
        super();
        this._transactionType = TransactionType_1.TransactionType.MODIFY_MOSAIC_LEVY;
    }
    mosaicId(mosaicId) {
        this._mosaicId = mosaicId;
        return this;
    }
    mosaicLevy(mosaicLevy) {
        this._mosaicLevy = mosaicLevy;
        return this;
    }
    build() {
        return new MosaicModifyLevyTransaction(this._networkType, this._version || TransactionTypeVersion_1.TransactionTypeVersion.MOSAIC_MODIFY_LEVY, this._deadline ? this._deadline : this._createNewDeadlineFn(), this._maxFee ? this._maxFee : (0, FeeCalculationStrategy_1.calculateFee)(MosaicModifyLevyTransaction.calculateSize(), this._feeCalculationStrategy), this._mosaicId, this._mosaicLevy, this._signature, this._signer, this._transactionInfo);
    }
}
exports.MosaicModifyLevyTransactionBuilder = MosaicModifyLevyTransactionBuilder;
