// tsjs-xpx-chain-sdk/dist/src/model/transaction/PlaceSdaExchangeOfferTransaction.js
"use strict";
// Copyright 2022 ProximaX Limited. All rights reserved.
// Use of this source code is governed by the Apache 2.0
// license that can be found in the LICENSE file
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlaceSdaExchangeOfferTransactionBuilder = exports.PlaceSdaExchangeOfferTransaction = void 0;
const Transaction_1 = require("./Transaction");
const TransactionTypeVersion_1 = require("./TransactionTypeVersion");
const PlaceSdaExchangeOfferTransaction_1 = require("../../infrastructure/builders/PlaceSdaExchangeOfferTransaction");
const TransactionType_1 = require("./TransactionType");
const FeeCalculationStrategy_1 = require("./FeeCalculationStrategy");
class PlaceSdaExchangeOfferTransaction extends Transaction_1.Transaction {
    offers;
    /**
     * Create PlaceSdaExchangeOfferTransaction object
     * @returns {PlaceSdaExchangeOfferTransaction}
     */
    static create(deadline, sdaExchangeOffers, networkType, maxFee) {
        return new PlaceSdaExchangeOfferTransactionBuilder()
            .networkType(networkType)
            .deadline(deadline)
            .maxFee(maxFee)
            .offers(sdaExchangeOffers)
            .build();
    }
    constructor(networkType, version, deadline, maxFee, sdaExchangeOffers, signature, signer, transactionInfo) {
        super(TransactionType_1.TransactionType.PLACE_SDA_EXCHANGE_OFFER, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
        this.offers = sdaExchangeOffers || [];
    }
    /**
     * @override Transaction.size()
     * @description get the byte size of a transaction
     * @returns {number}
     * @memberof PlaceSdaExchangeOfferTransaction
     */
    get size() {
        return PlaceSdaExchangeOfferTransaction.calculateSize(this.offers.length);
    }
    static calculateSize(offersCount) {
        const byteSize = Transaction_1.Transaction.getHeaderSize()
            + 1
            + (offersCount * (8 + 8 + 8 + 8 + 8));
        // mosaicId give -8
        // mosaicId give amount -8
        // mosaicId get - 8
        // mosaicId get amount - 8
        // duration - 8
        return byteSize;
    }
    /**
     * @override Transaction.toJSON()
     * @description Serialize a transaction object - add own fields to the result of Transaction.toJSON()
     * @return {Object}
     * @memberof PlaceSdaExchangeOfferTransaction
     */
    toJSON() {
        const parent = super.toJSON();
        return {
            ...parent,
            transaction: {
                ...parent.transaction,
                sdaExchangeOffers: this.offers.map(offer => offer.toDTO())
            }
        };
    }
    /**
     * @internal
     * @returns {VerifiableTransaction}
     */
    buildTransaction() {
        return new PlaceSdaExchangeOfferTransaction_1.Builder()
            .addSize(this.size)
            .addType(this.type)
            .addVersion(this.versionToDTO())
            .addDeadline(this.deadline.toDTO())
            .addMaxFee(this.maxFee.toDTO())
            .addOffers(this.offers)
            .build();
    }
}
exports.PlaceSdaExchangeOfferTransaction = PlaceSdaExchangeOfferTransaction;
class PlaceSdaExchangeOfferTransactionBuilder extends Transaction_1.TransactionBuilder {
    _offers;
    offers(sdaExchangeOffers) {
        this._offers = sdaExchangeOffers;
        return this;
    }
    build() {
        return new PlaceSdaExchangeOfferTransaction(this._networkType, this._version || TransactionTypeVersion_1.TransactionTypeVersion.PLACE_SDA_EXCHANGE_OFFER, this._deadline ? this._deadline : this._createNewDeadlineFn(), this._maxFee ? this._maxFee : (0, FeeCalculationStrategy_1.calculateFee)(PlaceSdaExchangeOfferTransaction.calculateSize(this._offers.length), this._feeCalculationStrategy), this._offers, this._signature, this._signer, this._transactionInfo);
    }
}
exports.PlaceSdaExchangeOfferTransactionBuilder = PlaceSdaExchangeOfferTransactionBuilder;
