// tsjs-xpx-chain-sdk/dist/src/model/transaction/RemoveSdaExchangeOfferTransaction.js
"use strict";
// Copyright 2023 ProximaX Limited. All rights reserved.
// Use of this source code is governed by the Apache 2.0
// license that can be found in the LICENSE file
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoveSdaExchangeOfferTransactionBuilder = exports.RemoveSdaExchangeOfferTransaction = void 0;
const Transaction_1 = require("./Transaction");
const TransactionTypeVersion_1 = require("./TransactionTypeVersion");
const RemoveSdaExchangeOfferTransaction_1 = require("../../infrastructure/builders/RemoveSdaExchangeOfferTransaction");
const TransactionType_1 = require("./TransactionType");
const FeeCalculationStrategy_1 = require("./FeeCalculationStrategy");
class RemoveSdaExchangeOfferTransaction extends Transaction_1.Transaction {
    offers;
    /**
     * Create RemoveSdaExchangeOfferTransaction object
     * @returns {RemoveSdaExchangeOfferTransaction}
     */
    static create(deadline, sdaExchangeOffers, networkType, maxFee) {
        return new RemoveSdaExchangeOfferTransactionBuilder()
            .networkType(networkType)
            .deadline(deadline)
            .maxFee(maxFee)
            .offers(sdaExchangeOffers)
            .build();
    }
    constructor(networkType, version, deadline, maxFee, sdaExchangeOffers, signature, signer, transactionInfo) {
        super(TransactionType_1.TransactionType.REMOVE_SDA_EXCHANGE_OFFER, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
        this.offers = sdaExchangeOffers || [];
    }
    /**
     * @override Transaction.size()
     * @description get the byte size of a transaction
     * @returns {number}
     * @memberof RemoveSdaExchangeOfferTransaction
     */
    get size() {
        return RemoveSdaExchangeOfferTransaction.calculateSize(this.offers.length);
    }
    static calculateSize(offersLength) {
        const byteSize = Transaction_1.Transaction.getHeaderSize()
            + 1 // num offers
            + offersLength * (8 + 8);
        // mosaicIdGive - 8
        // mosiacIdGet - 8
        return byteSize;
    }
    /**
     * @override Transaction.toJSON()
     * @description Serialize a transaction object - add own fields to the result of Transaction.toJSON()
     * @return {Object}
     * @memberof RemoveSdaExchangeOfferTransaction
     */
    toJSON() {
        const parent = super.toJSON();
        return {
            ...parent,
            transaction: {
                ...parent.transaction,
                sdaExchangeOffers: this.offers.map(offer => offer.toDTO())
            }
        };
    }
    /**
     * @internal
     * @returns {VerifiableTransaction}
     */
    buildTransaction() {
        return new RemoveSdaExchangeOfferTransaction_1.Builder()
            .addSize(this.size)
            .addType(this.type)
            .addVersion(this.versionToDTO())
            .addDeadline(this.deadline.toDTO())
            .addMaxFee(this.maxFee.toDTO())
            .addOffers(this.offers)
            .build();
    }
}
exports.RemoveSdaExchangeOfferTransaction = RemoveSdaExchangeOfferTransaction;
class RemoveSdaExchangeOfferTransactionBuilder extends Transaction_1.TransactionBuilder {
    _offers;
    offers(_sdaExchangeOffers) {
        this._offers = _sdaExchangeOffers;
        return this;
    }
    build() {
        return new RemoveSdaExchangeOfferTransaction(this._networkType, this._version || TransactionTypeVersion_1.TransactionTypeVersion.REMOVE_SDA_EXCHANGE_OFFER, this._deadline ? this._deadline : this._createNewDeadlineFn(), this._maxFee ? this._maxFee : (0, FeeCalculationStrategy_1.calculateFee)(RemoveSdaExchangeOfferTransaction.calculateSize(this._offers.length), this._feeCalculationStrategy), this._offers, this._signature, this._signer, this._transactionInfo);
    }
}
exports.RemoveSdaExchangeOfferTransactionBuilder = RemoveSdaExchangeOfferTransactionBuilder;
