// tsjs-xpx-chain-sdk/dist/src/model/transaction/SdaExchangeOffer.js
"use strict";
// Copyright 2019 ProximaX Limited. All rights reserved.
// Use of this source code is governed by the Apache 2.0
// license that can be found in the LICENSE file
Object.defineProperty(exports, "__esModule", { value: true });
exports.SdaExchangeOffer = void 0;
class SdaExchangeOffer {
    mosaicIdGive;
    mosaicAmountGive;
    mosaicIdGet;
    mosaicAmountGet;
    duration;
    constructor(
    /**
     * Id of the mosaic for which the offer should give
     */
    mosaicIdGive, 
    /**
     * Amount of mosaic that should be given, in smallest unit
     */
    mosaicAmountGive, 
    /**
     * Id of the mosaic for which the offer should receive
     */
    mosaicIdGet, 
    /**
     *  Amount of mosaic should be receive, in smallest unit
     */
    mosaicAmountGet, 
    /**
     * duration of block the offer valid
     */
    duration) {
        this.mosaicIdGive = mosaicIdGive;
        this.mosaicAmountGive = mosaicAmountGive;
        this.mosaicIdGet = mosaicIdGet;
        this.mosaicAmountGet = mosaicAmountGet;
        this.duration = duration;
    }
    /**
     * @internal
     */
    toDTO() {
        return {
            mosaicIdGive: this.mosaicIdGive.id.toDTO(),
            mosaicIdGiveAmount: this.mosaicAmountGive.toDTO(),
            mosaicIdGet: this.mosaicIdGet.id.toDTO(),
            mosaicIdGetAmount: this.mosaicAmountGet.toDTO(),
            duration: this.duration.toDTO()
        };
    }
}
exports.SdaExchangeOffer = SdaExchangeOffer;
