// tsjs-xpx-chain-sdk/dist/src/model/transaction/SecretProofTransaction.js
"use strict";
/*
 * Copyright 2018 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecretProofTransactionBuilder = exports.SecretProofTransaction = void 0;
const SecretProofTransaction_1 = require("../../infrastructure/builders/SecretProofTransaction");
const HashType_1 = require("./HashType");
const Transaction_1 = require("./Transaction");
const TransactionType_1 = require("./TransactionType");
const TransactionTypeVersion_1 = require("./TransactionTypeVersion");
const FeeCalculationStrategy_1 = require("./FeeCalculationStrategy");
const Base32_1 = require("../../core/format/Base32");
const Convert_1 = require("../../core/format/Convert");
class SecretProofTransaction extends Transaction_1.Transaction {
    hashType;
    secret;
    recipient;
    proof;
    /**
     * Create a secret proof transaction object.
     *
     * @param deadline - The deadline to include the transaction.
     * @param hashType - The hash algorithm secret is generated with.
     * @param secret - The seed proof hashed.
     * @param recipient - UnresolvedAddress
     * @param proof - The seed proof.
     * @param networkType - The network type.
     * @param maxFee - (Optional) Max fee defined by the sender
     *
     * @return a SecretProofTransaction instance
     */
    static create(deadline, hashType, secret, recipient, proof, networkType, maxFee) {
        return new SecretProofTransactionBuilder()
            .networkType(networkType)
            .deadline(deadline)
            .maxFee(maxFee)
            .hashType(hashType)
            .secret(secret)
            .recipient(recipient)
            .proof(proof)
            .build();
    }
    /**
     * @param networkType
     * @param version
     * @param deadline
     * @param maxFee
     * @param hashType
     * @param secret
     * @param recipient
     * @param proof
     * @param signature
     * @param signer
     * @param transactionInfo
     */
    constructor(networkType, version, deadline, maxFee, hashType, secret, recipient, proof, signature, signer, transactionInfo) {
        super(TransactionType_1.TransactionType.SECRET_PROOF, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
        this.hashType = hashType;
        this.secret = secret;
        this.recipient = recipient;
        this.proof = proof;
        if (!(0, HashType_1.HashTypeLengthValidator)(hashType, this.secret)) {
            throw new Error('HashType and Secret have incompatible length or not hexadecimal string');
        }
    }
    /**
     * @override Transaction.size()
     * @description get the byte size of a SecretProofTransaction
     * @returns {number}
     * @memberof SecretProofTransaction
     */
    get size() {
        return SecretProofTransaction.calculateSize(this.proof);
    }
    static calculateSize(proof) {
        const byteSize = Transaction_1.Transaction.getHeaderSize();
        // hash algorithm and proof size static byte size
        const byteAlgorithm = 1;
        const byteProofSize = 2;
        const byteRecipient = 25;
        const secretLength = 32;
        // get proof byte size
        const proofLength = proof.length / 2;
        return byteSize + byteAlgorithm + secretLength + byteRecipient + byteProofSize + proofLength;
    }
    /**
     * @override Transaction.toJSON()
     * @description Serialize a transaction object - add own fields to the result of Transaction.toJSON()
     * @return {Object}
     * @memberof SecretProofTransaction
     */
    toJSON() {
        const parent = super.toJSON();
        return {
            ...parent,
            transaction: {
                ...parent.transaction,
                hashAlgorithm: this.hashType,
                secret: this.secret,
                recipient: this.recipient.toDTO(),
                proof: this.proof,
            }
        };
    }
    /**
     * @override Transaction.toDTO()
     * @description Serialize a transaction object - add own fields to the result of Transaction.toJSON()
     * @return {Object}
     * @memberof SecretProofTransaction
     */
    toDTO() {
        const parent = super.toJSON();
        return {
            ...parent,
            transaction: {
                ...parent.transaction,
                hashAlgorithm: this.hashType,
                secret: this.secret,
                recipient: Convert_1.Convert.uint8ArrayToHex(Base32_1.Base32.Base32Decode(this.recipient.plain())),
                proof: this.proof,
            }
        };
    }
    /**
     * @internal
     * @returns {VerifiableTransaction}
     */
    buildTransaction() {
        return new SecretProofTransaction_1.Builder()
            .addSize(this.size)
            .addDeadline(this.deadline.toDTO())
            .addType(this.type)
            .addMaxFee(this.maxFee.toDTO())
            .addVersion(this.versionToDTO())
            .addHashAlgorithm(this.hashType)
            .addSecret(this.secret)
            .addRecipient(this.recipient.plain())
            .addProof(this.proof)
            .build();
    }
}
exports.SecretProofTransaction = SecretProofTransaction;
class SecretProofTransactionBuilder extends Transaction_1.TransactionBuilder {
    _hashType;
    _secret;
    _recipient;
    _proof;
    hashType(hashType) {
        this._hashType = hashType;
        return this;
    }
    secret(secret) {
        this._secret = secret;
        return this;
    }
    recipient(recipient) {
        this._recipient = recipient;
        return this;
    }
    proof(proof) {
        this._proof = proof;
        return this;
    }
    build() {
        return new SecretProofTransaction(this._networkType, this._version || TransactionTypeVersion_1.TransactionTypeVersion.SECRET_PROOF, this._deadline ? this._deadline : this._createNewDeadlineFn(), this._maxFee ? this._maxFee : (0, FeeCalculationStrategy_1.calculateFee)(SecretProofTransaction.calculateSize(this._proof), this._feeCalculationStrategy), this._hashType, this._secret, this._recipient, this._proof, this._signature, this._signer, this._transactionInfo);
    }
}
exports.SecretProofTransactionBuilder = SecretProofTransactionBuilder;
