// tsjs-xpx-chain-sdk/dist/src/model/transaction/SignedTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 * Copyright 2018 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignedTransaction = void 0;
const VerifiableTransaction_1 = require("../../infrastructure/builders/VerifiableTransaction");
const Convert_1 = require("../../core/format/Convert");
const Utilities_1 = require("./Utilities");
/**
 * SignedTransaction object is used to transfer the transaction data and the signature to the server
 * in order to initiate and broadcast a transaction.
 */
class SignedTransaction {
    payload;
    hash;
    signer;
    type;
    networkType;
    /**
     * @internal
     * @param payload
     * @param hash
     * @param signer
     * @param type
     * @param networkType
     */
    constructor(/**
                 * Transaction serialized data
                 */ payload, 
    /**
     * Transaction hash
     */
    hash, 
    /**
     * Transaction signer
     */
    signer, 
    /**
     * Transaction type
     */
    type, 
    /**
     * Signer network type
     */
    networkType) {
        this.payload = payload;
        this.hash = hash;
        this.signer = signer;
        this.type = type;
        this.networkType = networkType;
        if (hash.length !== 64) {
            throw new Error('hash must be 64 characters long');
        }
    }
    /**
     * @param transactionPayload
     * @param generationHash
     */
    static createFromPayload(transactionPayload, generationHash) {
        const transactionHash = VerifiableTransaction_1.VerifiableTransaction.createTransactionHash(transactionPayload, Array.from(Convert_1.Convert.hexToUint8(generationHash)));
        const sizeLength = 8;
        const signatureLength = 128;
        const publicKeyLength = 64;
        const versionLength = 8;
        const typeLength = 4;
        const signature = transactionPayload.substring(8, 136);
        const signer = transactionPayload.substring(136, 200);
        const networkType = (0, Utilities_1.extractNetwork)(transactionPayload.substring(200, 208));
        const type = parseInt(Convert_1.Convert.uint8ArrayToHex(Convert_1.Convert.hexToUint8(transactionPayload.substring(208, 212)).reverse()), 16);
        if (signature === "0".repeat(128) || signer === "0".repeat(64)) {
            throw new Error('No signature found in payload');
        }
        return new SignedTransaction(transactionPayload, transactionHash, signer, type, networkType);
    }
    /**
     * Create DTO object
     */
    toDTO() {
        return {
            payload: this.payload,
            hash: this.hash,
            signer: this.signer,
            type: this.type,
            networkType: this.networkType,
        };
    }
}
exports.SignedTransaction = SignedTransaction;
