// tsjs-xpx-chain-sdk/dist/src/model/transaction/TransactionType.js
"use strict";
/*
 * Copyright 2023 ProximaX
 * Copyright 2018 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License"),
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionType = void 0;
/**
 * Enum class containing transaction type constants.
 */
var TransactionType;
(function (TransactionType) {
    /**
     * Transfer Transaction transaction type.
     * @type {number}
     */
    TransactionType[TransactionType["TRANSFER"] = 16724] = "TRANSFER";
    /**
     * Register namespace transaction type.
     * @type {number}
     */
    TransactionType[TransactionType["REGISTER_NAMESPACE"] = 16718] = "REGISTER_NAMESPACE";
    /**
     * Address alias transaction type
     * @type {number}
     */
    TransactionType[TransactionType["ADDRESS_ALIAS"] = 16974] = "ADDRESS_ALIAS";
    /**
     * Mosaic alias transaction type
     * @type {number}
     */
    TransactionType[TransactionType["MOSAIC_ALIAS"] = 17230] = "MOSAIC_ALIAS";
    /**
     * Mosaic definition transaction type.
     * @type {number}
     */
    TransactionType[TransactionType["MOSAIC_DEFINITION"] = 16717] = "MOSAIC_DEFINITION";
    /**
     * Mosaic supply change transaction.
     * @type {number}
     */
    TransactionType[TransactionType["MOSAIC_SUPPLY_CHANGE"] = 16973] = "MOSAIC_SUPPLY_CHANGE";
    /**
     * Modify multisig account transaction type.
     * @type {number}
     */
    TransactionType[TransactionType["MODIFY_MULTISIG_ACCOUNT"] = 16725] = "MODIFY_MULTISIG_ACCOUNT";
    /**
     * Aggregate complete v1 transaction type.
     * @type {number}
     */
    TransactionType[TransactionType["AGGREGATE_COMPLETE_V1"] = 16705] = "AGGREGATE_COMPLETE_V1";
    /**
     * Aggregate bonded v1 transaction type
     * @type {number}
     */
    TransactionType[TransactionType["AGGREGATE_BONDED_V1"] = 16961] = "AGGREGATE_BONDED_V1";
    /**
     * Lock transaction type
     * @type {number}
     */
    TransactionType[TransactionType["HASH_LOCK"] = 16712] = "HASH_LOCK";
    /**
     * Secret Lock Transaction type
     * @type {number}
     */
    TransactionType[TransactionType["SECRET_LOCK"] = 16722] = "SECRET_LOCK";
    /**
     * Secret Proof transaction type
     * @type {number}
     */
    TransactionType[TransactionType["SECRET_PROOF"] = 16978] = "SECRET_PROOF";
    /**
     * Account restriction address transaction type
     * @type {number}
     */
    TransactionType[TransactionType["MODIFY_ACCOUNT_RESTRICTION_ADDRESS"] = 16720] = "MODIFY_ACCOUNT_RESTRICTION_ADDRESS";
    /**
     * Account restriction mosaic transaction type
     * @type {number}
     */
    TransactionType[TransactionType["MODIFY_ACCOUNT_RESTRICTION_MOSAIC"] = 16976] = "MODIFY_ACCOUNT_RESTRICTION_MOSAIC";
    /**
     * Account restriction operation transaction type
     * @type {number}
     */
    TransactionType[TransactionType["MODIFY_ACCOUNT_RESTRICTION_OPERATION"] = 17232] = "MODIFY_ACCOUNT_RESTRICTION_OPERATION";
    /**
     * Link account transaction type
     * @type {number}
     */
    TransactionType[TransactionType["LINK_ACCOUNT"] = 16716] = "LINK_ACCOUNT";
    /**
     * Modify account/address related metadata transaction type
     * @deprecated
     * @type {number}
     */
    TransactionType[TransactionType["MODIFY_ACCOUNT_METADATA"] = 16701] = "MODIFY_ACCOUNT_METADATA";
    /**
     * Modify mosaic related metadata transaction type
     * @deprecated
     * @type {number}
     */
    TransactionType[TransactionType["MODIFY_MOSAIC_METADATA"] = 16957] = "MODIFY_MOSAIC_METADATA";
    /**
     * Modify namespace related metadata transaction type
     * @deprecated
     * @type {number}
     */
    TransactionType[TransactionType["MODIFY_NAMESPACE_METADATA"] = 17213] = "MODIFY_NAMESPACE_METADATA";
    /**
     * Upgrade chain transaction type
     */
    TransactionType[TransactionType["CHAIN_UPGRADE"] = 16728] = "CHAIN_UPGRADE";
    /**
     * Configure chain transaction type
     */
    TransactionType[TransactionType["CHAIN_CONFIGURE"] = 16729] = "CHAIN_CONFIGURE";
    /**
     * Add exchange transaction type
     */
    TransactionType[TransactionType["ADD_EXCHANGE_OFFER"] = 16733] = "ADD_EXCHANGE_OFFER";
    /**
     * Exchange transaction type
     */
    TransactionType[TransactionType["EXCHANGE_OFFER"] = 16989] = "EXCHANGE_OFFER";
    /**
     * Remove exchange transaction type
     */
    TransactionType[TransactionType["REMOVE_EXCHANGE_OFFER"] = 17245] = "REMOVE_EXCHANGE_OFFER";
    /**
     * Modify account metadata transaction type - NEM
     */
    TransactionType[TransactionType["ACCOUNT_METADATA_V2"] = 16703] = "ACCOUNT_METADATA_V2";
    /**
    * Modify mosaic metadata transaction type - NEM
    * @type {number}
    */
    TransactionType[TransactionType["MOSAIC_METADATA_V2"] = 16959] = "MOSAIC_METADATA_V2";
    /**
    * Modify namespace metadata transaction type - NEM
    * @type {number}
    */
    TransactionType[TransactionType["NAMESPACE_METADATA_V2"] = 17215] = "NAMESPACE_METADATA_V2";
    /**
    * Modify mosaic levy transaction type
    * @type {number}
    */
    TransactionType[TransactionType["MODIFY_MOSAIC_LEVY"] = 17229] = "MODIFY_MOSAIC_LEVY";
    /**
    * Remove mosaic levy transaction type
    * @type {number}
    */
    TransactionType[TransactionType["REMOVE_MOSAIC_LEVY"] = 17485] = "REMOVE_MOSAIC_LEVY";
    /**
    * Add harvester transaction type
    */
    TransactionType[TransactionType["ADD_HARVESTER"] = 16737] = "ADD_HARVESTER";
    /**
    * Remove harvester transaction type
    */
    TransactionType[TransactionType["REMOVE_HARVESTER"] = 16993] = "REMOVE_HARVESTER";
    /**
     * Aggregate complete v2 transaction type.
     * @type {number}
     */
    TransactionType[TransactionType["AGGREGATE_COMPLETE_V2"] = 17217] = "AGGREGATE_COMPLETE_V2";
    /**
     * Aggregate bonded v2 transaction type
     * @type {number}
     */
    TransactionType[TransactionType["AGGREGATE_BONDED_V2"] = 17473] = "AGGREGATE_BONDED_V2";
    /**
     * Account v2 upgrade transaction type
     * @type {number}
     */
    TransactionType[TransactionType["Account_V2_Upgrade"] = 16984] = "Account_V2_Upgrade";
    /**
    * Place SDA exchange offer transaction type
    */
    TransactionType[TransactionType["PLACE_SDA_EXCHANGE_OFFER"] = 16746] = "PLACE_SDA_EXCHANGE_OFFER";
    /**
    * Remove SDA exchange offer transaction type
    */
    TransactionType[TransactionType["REMOVE_SDA_EXCHANGE_OFFER"] = 17002] = "REMOVE_SDA_EXCHANGE_OFFER";
    /*
    * Liquidity Provider transasction type
    */
    TransactionType[TransactionType["Create_Liquidity_Provider"] = 16745] = "Create_Liquidity_Provider";
    TransactionType[TransactionType["Manual_Rate_Change"] = 17001] = "Manual_Rate_Change";
    // ------------------------storage transactions type-------------------------------------
    TransactionType[TransactionType["Replicator_Onboarding"] = 18018] = "Replicator_Onboarding";
    TransactionType[TransactionType["Prepare_Bc_Drive"] = 16738] = "Prepare_Bc_Drive";
    TransactionType[TransactionType["Data_Modification"] = 16994] = "Data_Modification";
    TransactionType[TransactionType["Data_Modification_Approval"] = 17506] = "Data_Modification_Approval";
    TransactionType[TransactionType["Data_Modification_Single_Approval"] = 19298] = "Data_Modification_Single_Approval";
    TransactionType[TransactionType["Data_Modification_Cancel"] = 17762] = "Data_Modification_Cancel";
    TransactionType[TransactionType["Storage_Payment"] = 19042] = "Storage_Payment";
    TransactionType[TransactionType["Download_Payment"] = 18786] = "Download_Payment";
    TransactionType[TransactionType["Download"] = 17250] = "Download";
    TransactionType[TransactionType["Finish_Download"] = 18530] = "Finish_Download";
    TransactionType[TransactionType["Verification_Payment"] = 19554] = "Verification_Payment";
    TransactionType[TransactionType["End_Drive_Verification_V2"] = 20322] = "End_Drive_Verification_V2";
    TransactionType[TransactionType["Download_Approval"] = 19810] = "Download_Approval";
    TransactionType[TransactionType["Drive_Closure"] = 20066] = "Drive_Closure";
    TransactionType[TransactionType["Replicator_Offboarding"] = 18274] = "Replicator_Offboarding";
    // --------------------------------------------------------------------------------------
    // ---------------------------deprecated storage type------------------------------------
    TransactionType[TransactionType["Prepare_Drive"] = 16730] = "Prepare_Drive";
    TransactionType[TransactionType["Join_To_Drive"] = 16986] = "Join_To_Drive";
    TransactionType[TransactionType["Drive_File_System"] = 17242] = "Drive_File_System";
    TransactionType[TransactionType["Files_Deposit"] = 17498] = "Files_Deposit";
    TransactionType[TransactionType["End_Drive"] = 17754] = "End_Drive";
    TransactionType[TransactionType["Drive_Files_Reward"] = 18010] = "Drive_Files_Reward";
    TransactionType[TransactionType["Start_Drive_Verification"] = 18266] = "Start_Drive_Verification";
    TransactionType[TransactionType["End_Drive_Verification"] = 18522] = "End_Drive_Verification";
    TransactionType[TransactionType["Start_File_Download"] = 18778] = "Start_File_Download";
    TransactionType[TransactionType["End_File_Download"] = 19034] = "End_File_Download";
    //---------------------------------------------------------------------------------------
    // ----------------------------symbol merge type-----------------------------------------
    TransactionType[TransactionType["NetworkConfig_Absolute_Height"] = 16985] = "NetworkConfig_Absolute_Height";
    TransactionType[TransactionType["Node_Key_Link"] = 16972] = "Node_Key_Link";
    TransactionType[TransactionType["Vrf_Key_Link"] = 17228] = "Vrf_Key_Link";
    TransactionType[TransactionType["Lock_Fund_Transfer"] = 16738] = "Lock_Fund_Transfer";
    TransactionType[TransactionType["Lock_Fund_Cancel_Unlock"] = 16994] = "Lock_Fund_Cancel_Unlock";
    TransactionType[TransactionType["Account_Address_Restriction"] = 16739] = "Account_Address_Restriction";
    TransactionType[TransactionType["Account_Mosaic_Restriction"] = 16995] = "Account_Mosaic_Restriction";
    TransactionType[TransactionType["Account_Operation_Restriction"] = 17251] = "Account_Operation_Restriction";
    TransactionType[TransactionType["Mosaic_Global_Restriction"] = 16740] = "Mosaic_Global_Restriction";
    TransactionType[TransactionType["Mosaic_Address_Restriction"] = 16996] = "Mosaic_Address_Restriction";
    // --------------------------------------------------------------------------------------
    // ----------------------------internal service txn--------------------------------------
    TransactionType[TransactionType["Add_Dbrb_Process"] = 16748] = "Add_Dbrb_Process";
    TransactionType[TransactionType["Remove_Dbrb_Process"] = 17004] = "Remove_Dbrb_Process";
    TransactionType[TransactionType["Remove_Dbrb_Process_By_Network"] = 17260] = "Remove_Dbrb_Process_By_Network";
    TransactionType[TransactionType["Add_Or_Update_Dbrb_Process"] = 17516] = "Add_Or_Update_Dbrb_Process";
    // --------------------------------------------------------------------------------------
    // -------------SC -----------------------
    // OPERATION_IDENTIFY(16735),
    // START_OPERATION(16991),
    // END_OPERATION(17247),
    // DEPLOY(16736),
    // START_EXECUTE(16992),
    // END_EXECUTE(17248),
    // SUPER_CONTRACT_FILE_SYSTEM(17504),
    // DEACTIVATE(17760),
    // DEPLOY_CONTRACT(16750),
    // MANUAL_CALL(17006),
    // AUTOMATIC_EXECUTIONS_PAYMENT(17262),
    // SUCCESSFUL_END_BATCH_EXECUTION(17518),
    // UNSUCCESSFUL_END_BATCH_EXECUTION(17774)
    // ---------------------------------------
})(TransactionType || (exports.TransactionType = TransactionType = {}));
