// tsjs-xpx-chain-sdk/dist/src/model/transaction/Transaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 * Copyright 2018 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilder = exports.Transaction = void 0;
const crypto_1 = require("../../core/crypto");
const VerifiableTransaction_1 = require("../../infrastructure/builders/VerifiableTransaction");
const PublicAccount_1 = require("../account/PublicAccount");
const AggregateTransactionInfo_1 = require("./AggregateTransactionInfo");
const Deadline_1 = require("./Deadline");
const SignedTransaction_1 = require("./SignedTransaction");
const TransactionType_1 = require("./TransactionType");
const TransactionTypeAltName_1 = require("./TransactionTypeAltName");
const FeeCalculationStrategy_1 = require("./FeeCalculationStrategy");
const Utilities_1 = require("../../core/format/Utilities");
const TransactionVersion_1 = require("./TransactionVersion");
const Utilities_2 = require("./Utilities");
const Convert_1 = require("../../core/format/Convert");
/**
 * An abstract transaction class that serves as the base class of all NEM transactions.
 */
class Transaction {
    type;
    deadline;
    maxFee;
    signature;
    signer;
    transactionInfo;
    transactionName;
    version;
    /**
     * @description get the byte size of the common transaction header
     * @returns {number}
     * @memberof Transaction
     */
    static getHeaderSize() {
        const byteSize = 4 // size
            + 64 // signature
            + 32 // signer
            + 4 // version
            + 2 // type
            + 8 // maxFee
            + 8; // deadline
        return byteSize;
    }
    /**
     * @constructor
     * @param type
     * @param networkType
     * @param version
     * @param deadline
     * @param maxFee
     * @param signature
     * @param signer
     * @param transactionInfo
     */
    constructor(/**
                 * The transaction type.
                 */ type, 
    /**
     * The network type.
     */
    networkType, 
    /**
     * The transaction version number.
     */
    version, 
    /**
     * The deadline to include the transaction.
     */
    deadline, 
    /**
     * A sender of a transaction must specify during the transaction definition a max_fee,
     * meaning the maximum fee the account allows to spend for this transaction.
     */
    maxFee, 
    /**
     * The transaction signature (missing if part of an aggregate transaction).
     */
    signature, 
    /**
     * The account of the transaction creator.
     */
    signer, 
    /**
     * Transactions meta data object contains additional information about the transaction.
     */
    transactionInfo) {
        this.type = type;
        this.deadline = deadline;
        this.maxFee = maxFee;
        this.signature = signature;
        this.signer = signer;
        this.transactionInfo = transactionInfo;
        this.version = TransactionVersion_1.TransactionVersion.createInit(networkType, version);
        this.transactionName = Transaction.extractTransactionName(type);
    }
    /**
     * Serialize and sign transaction creating a new SignedTransaction
     * @param account - The account to sign the transaction
     * @param generationHash - Network generation hash hex
     * @param {DerivationScheme} dScheme The derivation scheme
     * @returns {SignedTransaction}
     */
    newSignWith(account, generationHash) {
        this.version.signatureDScheme = PublicAccount_1.PublicAccount.getDerivationSchemeFromAccVersion(account.version);
        const transaction = this.buildTransaction();
        const signedTransactionRaw = transaction.signTransaction(account, generationHash, this.version.signatureDScheme);
        return new SignedTransaction_1.SignedTransaction(signedTransactionRaw.payload, signedTransactionRaw.hash, account.publicKey, this.type, this.version.networkType);
    }
    /**
     * Serialize and sign transaction creating a new SignedTransaction
     * @param account - The account to sign the transaction
     * @param generationHash - Network generation hash hex
     * @returns {SignedTransaction}
     */
    signWith(account, generationHash) {
        this.version.signatureDScheme = crypto_1.DerivationScheme.Unset;
        const transaction = this.buildTransaction();
        const signedTransactionRaw = transaction.signTransaction(account, generationHash, crypto_1.DerivationScheme.Ed25519Sha3);
        return new SignedTransaction_1.SignedTransaction(signedTransactionRaw.payload, signedTransactionRaw.hash, account.publicKey, this.type, this.version.networkType);
    }
    static signPayload(txnPayload, account, generationHash) {
        const networkType = (0, Utilities_2.extractNetwork)(txnPayload.substring(200, 208));
        const type = parseInt(Convert_1.Convert.uint8ArrayToHex(Convert_1.Convert.hexToUint8(txnPayload.substring(208, 212)).reverse()), 16);
        const signedTransactionRaw = VerifiableTransaction_1.VerifiableTransaction.signTransactionPayload(txnPayload, account, generationHash, account.version);
        return new SignedTransaction_1.SignedTransaction(signedTransactionRaw.payload, signedTransactionRaw.hash, account.publicKey, type, networkType);
    }
    /**
     * @internal
     * @returns {Array<number>}
     */
    aggregateTransaction() {
        return this.buildTransaction().toAggregateTransaction(this.signer.publicKey);
    }
    /**
     * Convert an aggregate transaction to an inner transaction including transaction signer.
     * @param signer - Transaction signer.
     * @returns InnerTransaction
     */
    toNewAggregate(signer) {
        if (!signer.version) {
            throw new Error("Signer missing version, please specify to aggregate transaction");
        }
        if ([TransactionType_1.TransactionType.AGGREGATE_BONDED_V1, TransactionType_1.TransactionType.AGGREGATE_BONDED_V2,
            TransactionType_1.TransactionType.AGGREGATE_COMPLETE_V1, TransactionType_1.TransactionType.AGGREGATE_COMPLETE_V2].includes(this.type)) {
            throw new Error('Inner transaction cannot be an aggregated transaction.');
        }
        this.version.signatureDScheme = PublicAccount_1.PublicAccount.getDerivationSchemeFromAccVersion(signer.version);
        return Object.assign({ __proto__: Object.getPrototypeOf(this) }, this, { signer });
    }
    /**
     * Convert an aggregate transaction to an inner transaction including transaction signer.
     * @param signer - Transaction signer.
     * @returns InnerTransaction
     */
    toAggregate(signer) {
        return this.toAggregateV1(signer);
    }
    /**
     * Convert an aggregate transaction to an inner transaction including transaction signer.
     * @param signer - Transaction signer.
     * @returns InnerTransaction
     */
    toAggregateV1(signer) {
        if ([TransactionType_1.TransactionType.AGGREGATE_BONDED_V1, TransactionType_1.TransactionType.AGGREGATE_BONDED_V2,
            TransactionType_1.TransactionType.AGGREGATE_COMPLETE_V1, TransactionType_1.TransactionType.AGGREGATE_COMPLETE_V2].includes(this.type)) {
            throw new Error('Inner transaction cannot be an aggregated transaction.');
        }
        this.version.signatureDScheme = crypto_1.DerivationScheme.Unset;
        return Object.assign({ __proto__: Object.getPrototypeOf(this) }, this, { signer });
    }
    /**
     * Transaction pending to be included in a block
     * @returns {boolean}
     */
    isUnconfirmed() {
        return this.transactionInfo != null && this.transactionInfo.height.compact() === 0
            && this.transactionInfo.hash === this.transactionInfo.merkleComponentHash;
    }
    /**
     * Transaction included in a block
     * @returns {boolean}
     */
    isConfirmed() {
        return this.transactionInfo != null && this.transactionInfo.height.compact() > 0;
    }
    /**
     * Returns if a transaction has missing signatures.
     * @returns {boolean}
     */
    hasMissingSignatures() {
        return this.transactionInfo != null && this.transactionInfo.height.compact() === 0 &&
            this.transactionInfo.hash !== this.transactionInfo.merkleComponentHash;
    }
    /**
     * Transaction is not known by the network
     * @return {boolean}
     */
    isUnannounced() {
        return this.transactionInfo == null;
    }
    isEmbedded() {
        return this.transactionInfo instanceof AggregateTransactionInfo_1.AggregateTransactionInfo;
    }
    /**
     * @internal
     */
    versionToDTO() {
        return this.version.convertToDTO();
    }
    /**
     * @internal
     */
    versionToHex() {
        return '0x' + (this.versionToDTO() >>> 0).toString(16); // ">>> 0" hack makes it effectively an Uint32
    }
    /**
     * @description reapply a given value to the transaction in an immutable way
     * @param {Deadline} deadline
     * @returns {Transaction}
     * @memberof Transaction
     */
    reapplyGiven(deadline = Deadline_1.Deadline.create()) {
        if (this.isUnannounced()) {
            return Object.assign({ __proto__: Object.getPrototypeOf(this) }, this, { deadline });
        }
        throw new Error('an Announced transaction can\'t be modified');
    }
    /**
     * @description Serialize a transaction object
     * @returns {string}
     * @memberof Transaction
     */
    serialize() {
        const transaction = this.buildTransaction();
        return transaction.serializeUnsignedTransaction();
    }
    /**
     * @description Create JSON object
     * @returns {Object}
     * @memberof Transaction
     */
    toJSON() {
        const commonTransactionObject = {
            type: this.type,
            networkType: this.version.networkType,
            version: this.versionToDTO(),
            maxFee: this.maxFee.toDTO(),
            deadline: this.deadline.toDTO(),
            signature: this.signature ? this.signature : '',
        };
        if (this.signer) {
            Object.assign(commonTransactionObject, { signer: this.signer.publicKey });
        }
        return { transaction: commonTransactionObject };
    }
    static extractTransactionName = (type) => {
        let transactionTypeValues = Object.values(TransactionType_1.TransactionType);
        let transactionTypeKeys = Object.keys(TransactionType_1.TransactionType);
        let transactionAltNameKeys = Object.keys(TransactionTypeAltName_1.TransactionTypeAltName);
        let transactionAltNameValues = Object.values(TransactionTypeAltName_1.TransactionTypeAltName);
        if (transactionTypeValues.includes(type)) {
            let typeName = transactionTypeKeys[transactionTypeValues.indexOf(type)];
            if (transactionAltNameKeys.includes(typeName)) {
                return transactionAltNameValues[transactionAltNameKeys.indexOf(typeName)];
            }
            else {
                return (0, Utilities_1.ucwords)(typeName);
            }
        }
        else {
            return "Unknown";
        }
    };
}
exports.Transaction = Transaction;
class TransactionBuilder {
    _networkType;
    _version;
    _deadline;
    _generationHash;
    _feeCalculationStrategy = FeeCalculationStrategy_1.DefaultFeeCalculationStrategy;
    _maxFee;
    _signature;
    _signer;
    _transactionInfo;
    _createNewDeadlineFn = Deadline_1.DefaultCreateNewDeadline;
    networkType(networkType) {
        this._networkType = networkType;
        return this;
    }
    version(version) {
        this._version = version;
        return this;
    }
    deadline(deadline) {
        this._deadline = deadline;
        return this;
    }
    generationHash(generationHash) {
        this._generationHash = generationHash;
        return this;
    }
    feeCalculationStrategy(feeCalculationStrategy) {
        this._feeCalculationStrategy = feeCalculationStrategy;
        return this;
    }
    createNewDeadlineFn(createNewDeadlineFn) {
        this._createNewDeadlineFn = createNewDeadlineFn;
        return this;
    }
    maxFee(maxFee) {
        this._maxFee = maxFee;
        return this;
    }
    signature(signature) {
        this._signature = signature;
        return this;
    }
    signer(signer) {
        this._signer = signer;
        return this;
    }
    transactionInfo(transactionInfo) {
        this._transactionInfo = transactionInfo;
        return this;
    }
}
exports.TransactionBuilder = TransactionBuilder;
