// tsjs-xpx-chain-sdk/dist/src/model/transaction/TransactionTypeVersion.js
"use strict";
/*
 * Copyright 2023 ProximaX
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License"),
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionTypeVersion = void 0;
/**
 * Enum class containing transaction version constants.
 *
 * Transaction format versions are defined in catapult-server in
 * each transaction's plugin source code.
 *
 * You can check out the implemented version for each transaction at the Sirius blockchain network config
 */
var TransactionTypeVersion;
(function (TransactionTypeVersion) {
    /**
     * Transfer Transaction transaction version.
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["TRANSFER"] = 3] = "TRANSFER";
    /**
     * Register namespace transaction version.
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["REGISTER_NAMESPACE"] = 2] = "REGISTER_NAMESPACE";
    /**
     * Mosaic definition transaction version.
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["MOSAIC_DEFINITION"] = 3] = "MOSAIC_DEFINITION";
    /**
     * Mosaic supply change transaction.
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["MOSAIC_SUPPLY_CHANGE"] = 2] = "MOSAIC_SUPPLY_CHANGE";
    /**
     * Modify multisig account transaction version.
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["MODIFY_MULTISIG_ACCOUNT"] = 3] = "MODIFY_MULTISIG_ACCOUNT";
    /**
     * Aggregate complete v1 transaction version.
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["AGGREGATE_COMPLETE_V1"] = 3] = "AGGREGATE_COMPLETE_V1";
    /**
     * Aggregate bonded v1 transaction version
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["AGGREGATE_BONDED_V1"] = 3] = "AGGREGATE_BONDED_V1";
    /**
     * Lock transaction version
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["HASH_LOCK"] = 1] = "HASH_LOCK";
    /**
     * Secret Lock transaction version
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["SECRET_LOCK"] = 1] = "SECRET_LOCK";
    /**
     * Secret Proof transaction version
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["SECRET_PROOF"] = 1] = "SECRET_PROOF";
    /**
     * Address Alias transaction version
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["ADDRESS_ALIAS"] = 1] = "ADDRESS_ALIAS";
    /**
     * Mosaic Alias transaction version
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["MOSAIC_ALIAS"] = 1] = "MOSAIC_ALIAS";
    /**
     * Account Restriction address transaction version
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["MODIFY_ACCOUNT_RESTRICTION_ADDRESS"] = 1] = "MODIFY_ACCOUNT_RESTRICTION_ADDRESS";
    /**
     * Account Restriction mosaic transaction version
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["MODIFY_ACCOUNT_RESTRICTION_MOSAIC"] = 1] = "MODIFY_ACCOUNT_RESTRICTION_MOSAIC";
    /**
     * Account Restriction operation transaction version
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["MODIFY_ACCOUNT_RESTRICTION_ENTITY_TYPE"] = 1] = "MODIFY_ACCOUNT_RESTRICTION_ENTITY_TYPE";
    /**
     * Link account transaction version
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["LINK_ACCOUNT"] = 2] = "LINK_ACCOUNT";
    /**
     * Modify metadata transactions version
     * @deprecated
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["MODIFY_METADATA"] = 1] = "MODIFY_METADATA";
    /**
     * Modify account metadata nem transactions version
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["ACCOUNT_METADATA_V2"] = 1] = "ACCOUNT_METADATA_V2";
    /**
    * Modify mosaic metadata nem transactions version
    * @type {number}
    */
    TransactionTypeVersion[TransactionTypeVersion["MOSAIC_METADATA_V2"] = 1] = "MOSAIC_METADATA_V2";
    /**
    * Modify namespace metadata nem transactions version
    * @type {number}
    */
    TransactionTypeVersion[TransactionTypeVersion["NAMESPACE_METADATA_V2"] = 1] = "NAMESPACE_METADATA_V2";
    /**
     * Modify mosaic modify levy transactions version
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["MOSAIC_MODIFY_LEVY"] = 1] = "MOSAIC_MODIFY_LEVY";
    /**
     * Modify remove mosaic levy transactions version
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["MOSAIC_REMOVE_LEVY"] = 1] = "MOSAIC_REMOVE_LEVY";
    /**
     * Chain configuration transaction version
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["CHAIN_CONFIG"] = 1] = "CHAIN_CONFIG";
    /**
     * Chain upgrade transaction version
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["CHAIN_UPGRADE"] = 1] = "CHAIN_UPGRADE";
    /**
     * Add exchange transaction version
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["ADD_EXCHANGE_OFFER"] = 4] = "ADD_EXCHANGE_OFFER";
    /**
     * Exchange transaction version
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["EXCHANGE_OFFER"] = 2] = "EXCHANGE_OFFER";
    /**
     * Remove exchange transaction version
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["REMOVE_EXCHANGE_OFFER"] = 2] = "REMOVE_EXCHANGE_OFFER";
    /**
    * Add harvester transaction version
    * @type {number}
    */
    TransactionTypeVersion[TransactionTypeVersion["ADD_HARVESTER"] = 1] = "ADD_HARVESTER";
    /**
    * Remove harvester transaction version
    * @type {number}
    */
    TransactionTypeVersion[TransactionTypeVersion["REMOVE_HARVESTER"] = 1] = "REMOVE_HARVESTER";
    /**
    * Aggregate complete v2 transaction version.
    * @type {number}
    */
    TransactionTypeVersion[TransactionTypeVersion["AGGREGATE_COMPLETE_V2"] = 1] = "AGGREGATE_COMPLETE_V2";
    /**
     * Aggregate bonded v2 transaction version
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["AGGREGATE_BONDED_V2"] = 1] = "AGGREGATE_BONDED_V2";
    /**
     * Place SDA exchange transaction version.
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["PLACE_SDA_EXCHANGE_OFFER"] = 1] = "PLACE_SDA_EXCHANGE_OFFER";
    /**
     * Remove SDA exchange offer transaction version
     * @type {number}
     */
    TransactionTypeVersion[TransactionTypeVersion["REMOVE_SDA_EXCHANGE_OFFER"] = 1] = "REMOVE_SDA_EXCHANGE_OFFER";
    TransactionTypeVersion[TransactionTypeVersion["Create_Liquidity_Provider"] = 1] = "Create_Liquidity_Provider";
    TransactionTypeVersion[TransactionTypeVersion["Manual_Rate_Change"] = 1] = "Manual_Rate_Change";
    // ------------------------storage transactions type-------------------------------------
    TransactionTypeVersion[TransactionTypeVersion["Replicator_Onboarding"] = 1] = "Replicator_Onboarding";
    TransactionTypeVersion[TransactionTypeVersion["Prepare_Bc_Drive"] = 1] = "Prepare_Bc_Drive";
    TransactionTypeVersion[TransactionTypeVersion["Data_Modification"] = 1] = "Data_Modification";
    TransactionTypeVersion[TransactionTypeVersion["Data_Modification_Approval"] = 1] = "Data_Modification_Approval";
    TransactionTypeVersion[TransactionTypeVersion["Data_Modification_Single_Approval"] = 1] = "Data_Modification_Single_Approval";
    TransactionTypeVersion[TransactionTypeVersion["Data_Modification_Cancel"] = 1] = "Data_Modification_Cancel";
    TransactionTypeVersion[TransactionTypeVersion["Storage_Payment"] = 1] = "Storage_Payment";
    TransactionTypeVersion[TransactionTypeVersion["Download_Payment"] = 1] = "Download_Payment";
    TransactionTypeVersion[TransactionTypeVersion["Download"] = 1] = "Download";
    TransactionTypeVersion[TransactionTypeVersion["Finish_Download"] = 1] = "Finish_Download";
    TransactionTypeVersion[TransactionTypeVersion["Verification_Payment"] = 1] = "Verification_Payment";
    TransactionTypeVersion[TransactionTypeVersion["End_Drive_Verification_V2"] = 1] = "End_Drive_Verification_V2";
    TransactionTypeVersion[TransactionTypeVersion["Download_Approval"] = 1] = "Download_Approval";
    TransactionTypeVersion[TransactionTypeVersion["Drive_Closure"] = 1] = "Drive_Closure";
    TransactionTypeVersion[TransactionTypeVersion["Replicator_Offboarding"] = 1] = "Replicator_Offboarding";
    // --------------------------------------------------------------------------------------
})(TransactionTypeVersion || (exports.TransactionTypeVersion = TransactionTypeVersion = {}));
