// tsjs-xpx-chain-sdk/dist/src/model/transaction/UnknownTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnknownTransaction = void 0;
const Transaction_1 = require("./Transaction");
/**
 * An unknown transaction class that hold the transaction dto with issue,
 * these transaction exist at the blockchain.
 */
class UnknownTransaction extends Transaction_1.Transaction {
    unknownPayload;
    unknownData;
    /**
     * @constructor
     * @param unknownPayload
     * @param unknownData
     * @param type
     * @param networkType
     * @param version
     * @param deadline
     * @param maxFee
     * @param signature
     * @param signer
     * @param transactionInfo
     */
    constructor(unknownPayload, unknownData, 
    /**
     * The transaction type.
     */
    type, 
    /**
     * The network type.
     */
    networkType, 
    /**
     * The transaction version number.
     */
    version, 
    /**
     * The deadline to include the transaction.
     */
    deadline, 
    /**
     * A sender of a transaction must specify during the transaction definition a max_fee,
     * meaning the maximum fee the account allows to spend for this transaction.
     */
    maxFee, 
    /**
     * The account of the transaction creator.
     */
    signer, 
    /**
     * The transaction signature (missing if part of an aggregate transaction).
     */
    signature, 
    /**
     * Transactions meta data object contains additional information about the transaction.
     */
    transactionInfo) {
        super(type, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
        this.unknownPayload = unknownPayload;
        this.unknownData = unknownData;
    }
    /**
     * @override Transaction.size()
     * @description get the byte size of a UnknownTransaction, will return error
     * @returns {never}
     * @memberof UnknownTransaction
     */
    get size() {
        throw new Error("Cannot get size of UnknownTransaction");
    }
    /**
     * @override Transaction.buildTransaction()
     * @description buildTransaction of UnknownTransaction, will return error
     * @returns {never}
     * @memberof UnknownTransaction
     */
    buildTransaction() {
        throw new Error("Cannot buildTransaction from UnknownTransaction");
    }
    /**
     * Convert an aggregate transaction to an inner transaction including transaction signer, will return error
     * @override Transaction.toAggregate()
     * @internal
     * @param signer - Transaction signer.
     * @returns never
     */
    toAggregate(signer) {
        throw new Error('Cannot create inner transaction from UnknownTransaction.');
    }
    /**
     * Convert an aggregate transaction to an inner transaction including transaction signer, will return error
     * @override Transaction.toAggregateV1()
     * @internal
     * @param signer - Transaction signer.
     * @returns never
     */
    toAggregateV1(signer) {
        throw new Error('Cannot create inner transaction from UnknownTransaction.');
    }
    /**
     * Serialize and sign transaction creating a new SignedTransaction
     * @override Transaction.signWith()
     * @internal
     * @param account - The account to sign the transaction
     * @param generationHash - Network generation hash hex
     * @param {DerivationScheme} dScheme The Sign Schema. (SHA2 / SHA3)
     * @returns {never}
     */
    signWith() {
        throw new Error('Cannot sign UnknownTransaction');
    }
    /**
     * @override Transaction.aggregateTransaction()
     * @internal
     * @returns {never}
     */
    aggregateTransaction() {
        throw new Error('Cannot aggregate UnknownTransaction');
    }
    /**
     * @override Transaction.reapplyGiven()
     * @internal
     * @description reapply a given value to the transaction in an immutable way
     * @returns {never}
     * @memberof UnknownTransaction
     */
    reapplyGiven() {
        throw new Error('Cannot update UnknownTransaction');
    }
    /**
     * @override Transaction.reapplyGiven()
     * @internal
     * @description Serialize a transaction object
     * @returns {never}
     * @memberof UnknownTransaction
     */
    serialize() {
        throw new Error('Cannot serialize UnknownTransaction');
    }
    /**
     * @override Transaction.toJSON()
     * @description Create JSON object
     * @returns {Object}
     * @memberof UnknownTransaction
     */
    toJSON() {
        const commonTransactionObject = {
            type: this.type,
            networkType: this.version.networkType,
            version: this.versionToDTO(),
            maxFee: this.maxFee.toDTO(),
            deadline: this.deadline.toDTO(),
            signature: this.signature ? this.signature : '',
            signer: this.signer.publicKey,
            unknownData: this.unknownData,
            unknownPayload: this.unknownPayload
        };
        return { transaction: commonTransactionObject };
    }
}
exports.UnknownTransaction = UnknownTransaction;
