// tsjs-xpx-chain-sdk/dist/src/model/transaction/Utilities.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.genMetadataCompositeHash = exports.parseConfig = exports.getConfigSetting = exports.getABTMaxSafeDeadlineWithSeconds = exports.getABTMaxSafeDurationWithSeconds = exports.getABTMinConfigSeconds = exports.checkCosigner = exports.convertTimeUnit = exports.convertStringTimeUnit = exports.hasBits = exports.hasBit = exports.extractNetwork = exports.TimeUnit = void 0;
const Convert_1 = require("../../core/format/Convert");
const UInt64_1 = require("../UInt64");
const Deadline_1 = require("./Deadline");
const NetworkType_1 = require("../blockchain/NetworkType");
const core_1 = require("@js-joda/core");
const sha3_1 = require("@noble/hashes/sha3");
const utils_1 = require("@noble/hashes/utils");
const MetadataType_1 = require("../metadata/MetadataType");
const Base32_1 = require("../../core/format/Base32");
var TimeUnit;
(function (TimeUnit) {
    TimeUnit[TimeUnit["DAY"] = 4] = "DAY";
    TimeUnit[TimeUnit["HOUR"] = 3] = "HOUR";
    TimeUnit[TimeUnit["MINUTE"] = 2] = "MINUTE";
    TimeUnit[TimeUnit["SECOND"] = 1] = "SECOND";
    TimeUnit[TimeUnit["MILLISECOND"] = 0] = "MILLISECOND";
})(TimeUnit || (exports.TimeUnit = TimeUnit = {}));
/**
 * @param versionHex - Transaction version in hex
 * @returns {NetworkType}
 */
const extractNetwork = (versionHex) => {
    const networkType = Convert_1.Convert.hexToUint8(versionHex)[3];
    if (networkType === NetworkType_1.NetworkType.MAIN_NET) {
        return NetworkType_1.NetworkType.MAIN_NET;
    }
    else if (networkType === NetworkType_1.NetworkType.TEST_NET) {
        return NetworkType_1.NetworkType.TEST_NET;
    }
    else if (networkType === NetworkType_1.NetworkType.MIJIN) {
        return NetworkType_1.NetworkType.MIJIN;
    }
    else if (networkType === NetworkType_1.NetworkType.MIJIN_TEST) {
        return NetworkType_1.NetworkType.MIJIN_TEST;
    }
    else if (networkType === NetworkType_1.NetworkType.PRIVATE) {
        return NetworkType_1.NetworkType.PRIVATE;
    }
    else if (networkType === NetworkType_1.NetworkType.PRIVATE_TEST) {
        return (NetworkType_1.NetworkType.PRIVATE_TEST);
    }
    throw new Error('Unimplemented network type');
};
exports.extractNetwork = extractNetwork;
const hasBit = (numberToCompare, bitPosition) => {
    const bitsNumber = bitPosition > 1 ? 1 << (bitPosition - 1) : 1;
    return (numberToCompare & bitsNumber) == bitsNumber;
};
exports.hasBit = hasBit;
const hasBits = (numberToCompare, bitsNumber) => {
    return (numberToCompare & bitsNumber) == bitsNumber;
};
exports.hasBits = hasBits;
const convertStringTimeUnit = (startUnit, targetUnit) => {
    let initUnit = TimeUnit.SECOND;
    let unitAmount = parseInt(startUnit);
    let checkUnit = startUnit.toLowerCase();
    if (checkUnit.search('day') > -1 || checkUnit.search('d') > -1) {
        initUnit = TimeUnit.DAY;
    }
    else if (checkUnit.search('hour') > -1 || checkUnit.search('h') > -1) {
        initUnit = TimeUnit.HOUR;
    }
    else if (checkUnit.search('millisecond') > -1 || checkUnit.search('ms') > -1) {
        initUnit = TimeUnit.MILLISECOND;
    }
    else if (checkUnit.search('minute') > -1 || checkUnit.search('m') > -1) {
        initUnit = TimeUnit.MINUTE;
    }
    else if (checkUnit.search('sec') > -1 || checkUnit.search('s') > -1) {
        initUnit = TimeUnit.SECOND;
    }
    return (0, exports.convertTimeUnit)(unitAmount, initUnit, targetUnit);
};
exports.convertStringTimeUnit = convertStringTimeUnit;
const convertTimeUnit = (amount, startUnit, targetUnit) => {
    let downDirection = false;
    if (targetUnit > startUnit) {
        downDirection = false;
    }
    else if (targetUnit < startUnit) {
        downDirection = true;
    }
    else if (startUnit === targetUnit) {
        return amount;
    }
    let newAmount;
    let newUnit;
    if (downDirection) {
        if (startUnit === TimeUnit.DAY) {
            newAmount = amount * 24;
        }
        else if (startUnit === TimeUnit.HOUR) {
            newAmount = amount * 60;
        }
        else if (startUnit === TimeUnit.MINUTE) {
            newAmount = amount * 60;
        }
        else if (startUnit === TimeUnit.SECOND) {
            newAmount = amount * 1000;
        }
        else {
            return amount;
        }
    }
    else {
        if (startUnit === TimeUnit.HOUR) {
            newAmount = amount / 24;
        }
        else if (startUnit === TimeUnit.MINUTE) {
            newAmount = amount / 60;
        }
        else if (startUnit === TimeUnit.SECOND) {
            newAmount = amount / 60;
        }
        else if (startUnit === TimeUnit.MILLISECOND) {
            newAmount = amount / 1000;
        }
        else {
            return amount;
        }
    }
    newUnit = downDirection ? startUnit - 1 : startUnit + 1;
    return (0, exports.convertTimeUnit)(newAmount, newUnit, targetUnit);
};
exports.convertTimeUnit = convertTimeUnit;
const checkCosigner = (cosignerPublicKey, multisigAccPublicKey, multisigAccGraph) => {
    let levelZeroData = multisigAccGraph.multisigAccounts.get(0);
    if (!levelZeroData || levelZeroData[0].account.publicKey !== multisigAccPublicKey) {
        throw new Error("Please pass in the multisig account graph of the multisig account");
    }
    let isCosigner = false;
    let signable = false;
    let aggComplete = false;
    let minimumApproval = levelZeroData[0].minApproval;
    let cosignerLevel = 0;
    let allMultisigAccInfo = [];
    let paths = new Map();
    let iterator = multisigAccGraph.multisigAccounts.entries();
    for (const [level, multisigAccInfo] of iterator) {
        allMultisigAccInfo = allMultisigAccInfo.concat(multisigAccInfo);
        if (level > 0) {
            for (let i = 0; i < multisigAccInfo.length; ++i) {
                if (multisigAccInfo[i].account.publicKey === cosignerPublicKey) {
                    cosignerLevel = level;
                    isCosigner = true;
                    if (multisigAccInfo[i].cosignatories.length === 0) {
                        signable = true;
                    }
                }
            }
        }
    }
    if (signable && minimumApproval === 1) {
        let tracingPublicKeys = [cosignerPublicKey];
        paths.set(cosignerPublicKey, [cosignerPublicKey]);
        while (tracingPublicKeys.length) {
            let filteredEntry = allMultisigAccInfo.filter(x => x.cosignatories.length > 0 && x.cosignatories.some(y => tracingPublicKeys.includes(y.publicKey)));
            if (filteredEntry.length) {
                for (const entry of filteredEntry) {
                    let newPaths = [];
                    let entryCosigner = entry.cosignatories.map(x => x.publicKey);
                    for (const pk of tracingPublicKeys) {
                        if (entryCosigner.includes(pk)) {
                            let previousPath = paths.get(pk);
                            if (previousPath) {
                                newPaths = newPaths.concat(previousPath.map(x => x.concat("/" + entry.account.publicKey)));
                            }
                        }
                    }
                    paths.set(entry.account.publicKey, newPaths);
                }
                tracingPublicKeys = filteredEntry.map(x => x.account.publicKey);
            }
            else {
                tracingPublicKeys = [];
                break;
            }
        }
        const finalPaths = paths.get(multisigAccPublicKey);
        if (finalPaths) {
            for (const finalPath of finalPaths) {
                const splittedPublicKey = finalPath.split("/");
                let minApprovalIndicator = 0;
                for (const currentPK of splittedPublicKey) {
                    const foundData = allMultisigAccInfo.find(x => x.account.publicKey === currentPK);
                    if (foundData && foundData.minApproval !== 0) {
                        if (minApprovalIndicator === 0) {
                            minApprovalIndicator = foundData.minApproval;
                        }
                        else {
                            minApprovalIndicator = minApprovalIndicator * foundData.minApproval;
                        }
                    }
                }
                if (minApprovalIndicator === 1) {
                    aggComplete = true;
                    break;
                }
            }
        }
    }
    return {
        isCosigner: isCosigner,
        cosignerSignable: signable,
        aggComplete: aggComplete
    };
};
exports.checkCosigner = checkCosigner;
/**
 * Get lower value from maxBondedTransactionLifetime and maxHashLockDuration,
 * comparison is done with second.
 * This ensure expiration of both life time and lock duration of Aggregate Bonded Transaction
 * can be set using almost the nearest time.
 * @param maxBondedTransactionLifetime - Network config of max bonded transaction life time
 * @param maxHashLockDuration - Network config of max hash lock duration
 * @returns
 */
const getABTMinConfigSeconds = (maxBondedTransactionLifetime, maxHashLockDuration) => {
    let abtLifeTimeInSeconds = Math.min(maxBondedTransactionLifetime, maxHashLockDuration);
    return abtLifeTimeInSeconds;
};
exports.getABTMinConfigSeconds = getABTMinConfigSeconds;
/**
 * Get a maximum block duration for Aggregate Bonded Transaction for Sirius Chain
 * @param blockGenerationTargetTime - Network config of targeted block generation time
 * @param maxBondedTransactionLifetime - Network config of max bonded transaction life time
 * @param maxHashLockDuration - Network config of max hash lock duration
 * @returns
 */
const getABTMaxSafeDurationWithSeconds = (blockGenerationTargetTime, maxBondedTransactionLifetime, maxHashLockDuration) => {
    let abtLifeTimeInSeconds = (0, exports.getABTMinConfigSeconds)(maxBondedTransactionLifetime, maxHashLockDuration);
    let safeMaxBlockDuration = Math.floor(abtLifeTimeInSeconds / blockGenerationTargetTime) - 1;
    return safeMaxBlockDuration;
};
exports.getABTMaxSafeDurationWithSeconds = getABTMaxSafeDurationWithSeconds;
/**
 * Get a maximum Deadline for Aggregate Bonded Transaction for Sirius Chain
 * @param maxBondedTransactionLifetime - Network config of max bonded transaction life time
 * @param maxHashLockDuration - Network config of max hash lock duration
 * @param currentNodeTimetamp - When added, it is able to create a Deadline that is Daylight saving time safe.
 * @returns
 */
const getABTMaxSafeDeadlineWithSeconds = (maxBondedTransactionLifetime, maxHashLockDuration, currentNodeTimetamp) => {
    let abtLifeTimeInSeconds = (0, exports.getABTMinConfigSeconds)(maxBondedTransactionLifetime, maxHashLockDuration);
    let safeMaxDeadline;
    if (currentNodeTimetamp) {
        let deadlineUint64 = UInt64_1.UInt64.fromUint(currentNodeTimetamp.compact() + ((abtLifeTimeInSeconds - 5) * 1000));
        safeMaxDeadline = Deadline_1.Deadline.createFromUint64(deadlineUint64);
    }
    else {
        safeMaxDeadline = Deadline_1.Deadline.createForBonded(abtLifeTimeInSeconds - 5, core_1.ChronoUnit.SECONDS);
    }
    return safeMaxDeadline;
};
exports.getABTMaxSafeDeadlineWithSeconds = getABTMaxSafeDeadlineWithSeconds;
const getConfigSetting = (parsedConfig, keys, targetTimeUnit) => {
    let currentSection = parsedConfig;
    let value = null;
    for (let i = 0; i < keys.length; i++) {
        if (currentSection[keys[i]]) {
            value = currentSection[keys[i]];
            currentSection = value;
        }
        else {
            return null;
        }
    }
    if (targetTimeUnit) {
        value = (0, exports.convertTimeUnit)(value, TimeUnit.MILLISECOND, targetTimeUnit);
    }
    return value;
};
exports.getConfigSetting = getConfigSetting;
const parseConfig = (input) => {
    const result = {};
    let currentSection = result;
    input.split(/\n+/).forEach(line => {
        line = line.trim();
        if (!line || line.startsWith("#"))
            return; // Skip empty lines and comments
        const sectionMatch = line.match(/^\[(.+)]$/);
        if (sectionMatch) {
            currentSection = result;
            const keys = sectionMatch[1].split(':'); // Handle nested sections
            keys.forEach((key, index) => {
                if (!(key in currentSection)) {
                    currentSection[key] = {};
                }
                currentSection = currentSection[key]; // Move deeper into nested object
            });
            return;
        }
        const keyValueMatch = line.match(/^(.+?)\s*=\s*(.+)$/);
        if (keyValueMatch) {
            let key = keyValueMatch[1].trim();
            let value = keyValueMatch[2].trim();
            // Convert values intelligently
            if (/^\d+$/.test(value)) {
                value = parseInt(value, 10);
            }
            else if (/^\d+\.\d+$/.test(value)) {
                value = parseFloat(value);
            }
            else if (/^(true|false)$/i.test(value)) {
                value = value.toLowerCase() === "true";
            }
            else if (/^(\d+)(s|ms|m|h|d)$/.test(value)) {
                // Handle time units like 10s, 5m, 2h, 1d
                const [, num, unit] = value.match(/^(\d+)(s|ms|m|h|d)$/) || [];
                const timeMultipliers = { ms: 1, s: 1000, m: 60000, h: 3600000, d: 86400000 };
                value = parseInt(num, 10) * (timeMultipliers[unit] || 1);
            }
            else if (/^\d{1,3}(?:'\d{3})*$/.test(value)) {
                // Handle large numbers with apostrophes (e.g., "9'000'000'000")
                value = parseInt(value.replace(/'/g, ""), 10);
            }
            currentSection[key] = value;
        }
    });
    return result;
};
exports.parseConfig = parseConfig;
const genMetadataCompositeHash = (creatorPublicAcc, scopedMetadataKey, metadataType, targetId) => {
    const compositeHash = sha3_1.sha3_256
        .create()
        .update((Base32_1.Base32.Base32Decode(creatorPublicAcc.address.plain())))
        .update((0, utils_1.hexToBytes)(creatorPublicAcc.publicKey))
        .update((0, utils_1.hexToBytes)(scopedMetadataKey.toHex()).reverse())
        .update(metadataType === MetadataType_1.MetadataType.ACCOUNT ? new Uint8Array(8).fill(0) : (0, utils_1.hexToBytes)(targetId.toHex()).reverse())
        .update(new Uint8Array([metadataType]))
        .digest();
    return (0, utils_1.bytesToHex)(compositeHash);
};
exports.genMetadataCompositeHash = genMetadataCompositeHash;
