// tsjs-xpx-chain-sdk/dist/src/infrastructure/buffers/liquidityProvider/CreateLiquidityProviderBuffer.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateLiquidityProviderBuffer = void 0;
const flatbuffers = require("flatbuffers");
class CreateLiquidityProviderBuffer {
    bb = null;
    bb_pos = 0;
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsCreateLiquidityProviderBuffer(bb, obj) {
        return (obj || new CreateLiquidityProviderBuffer()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsCreateLiquidityProviderBuffer(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new CreateLiquidityProviderBuffer()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    size() {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    signature(index) {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? this.bb.readUint8(this.bb.__vector(this.bb_pos + offset) + index) : 0;
    }
    signatureLength() {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    signatureArray() {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? new Uint8Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    signer(index) {
        const offset = this.bb.__offset(this.bb_pos, 8);
        return offset ? this.bb.readUint8(this.bb.__vector(this.bb_pos + offset) + index) : 0;
    }
    signerLength() {
        const offset = this.bb.__offset(this.bb_pos, 8);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    signerArray() {
        const offset = this.bb.__offset(this.bb_pos, 8);
        return offset ? new Uint8Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    version() {
        const offset = this.bb.__offset(this.bb_pos, 10);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    type() {
        const offset = this.bb.__offset(this.bb_pos, 12);
        return offset ? this.bb.readUint16(this.bb_pos + offset) : 0;
    }
    maxFee(index) {
        const offset = this.bb.__offset(this.bb_pos, 14);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    maxFeeLength() {
        const offset = this.bb.__offset(this.bb_pos, 14);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    maxFeeArray() {
        const offset = this.bb.__offset(this.bb_pos, 14);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    deadline(index) {
        const offset = this.bb.__offset(this.bb_pos, 16);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    deadlineLength() {
        const offset = this.bb.__offset(this.bb_pos, 16);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    deadlineArray() {
        const offset = this.bb.__offset(this.bb_pos, 16);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    providerMosaicId(index) {
        const offset = this.bb.__offset(this.bb_pos, 18);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    providerMosaicIdLength() {
        const offset = this.bb.__offset(this.bb_pos, 18);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    providerMosaicIdArray() {
        const offset = this.bb.__offset(this.bb_pos, 18);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    currencyDeposit(index) {
        const offset = this.bb.__offset(this.bb_pos, 20);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    currencyDepositLength() {
        const offset = this.bb.__offset(this.bb_pos, 20);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    currencyDepositArray() {
        const offset = this.bb.__offset(this.bb_pos, 20);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    initialMosaicsMinting(index) {
        const offset = this.bb.__offset(this.bb_pos, 22);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    initialMosaicsMintingLength() {
        const offset = this.bb.__offset(this.bb_pos, 22);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    initialMosaicsMintingArray() {
        const offset = this.bb.__offset(this.bb_pos, 22);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    slashingPeriod() {
        const offset = this.bb.__offset(this.bb_pos, 24);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    windowSize() {
        const offset = this.bb.__offset(this.bb_pos, 26);
        return offset ? this.bb.readUint16(this.bb_pos + offset) : 0;
    }
    slashingAccount(index) {
        const offset = this.bb.__offset(this.bb_pos, 28);
        return offset ? this.bb.readUint8(this.bb.__vector(this.bb_pos + offset) + index) : 0;
    }
    slashingAccountLength() {
        const offset = this.bb.__offset(this.bb_pos, 28);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    slashingAccountArray() {
        const offset = this.bb.__offset(this.bb_pos, 28);
        return offset ? new Uint8Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    alpha() {
        const offset = this.bb.__offset(this.bb_pos, 30);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    beta() {
        const offset = this.bb.__offset(this.bb_pos, 32);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    static startCreateLiquidityProviderBuffer(builder) {
        builder.startObject(15);
    }
    static addSize(builder, size) {
        builder.addFieldInt32(0, size, 0);
    }
    static addSignature(builder, signatureOffset) {
        builder.addFieldOffset(1, signatureOffset, 0);
    }
    static createSignatureVector(builder, data) {
        builder.startVector(1, data.length, 1);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt8(data[i]);
        }
        return builder.endVector();
    }
    static startSignatureVector(builder, numElems) {
        builder.startVector(1, numElems, 1);
    }
    static addSigner(builder, signerOffset) {
        builder.addFieldOffset(2, signerOffset, 0);
    }
    static createSignerVector(builder, data) {
        builder.startVector(1, data.length, 1);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt8(data[i]);
        }
        return builder.endVector();
    }
    static startSignerVector(builder, numElems) {
        builder.startVector(1, numElems, 1);
    }
    static addVersion(builder, version) {
        builder.addFieldInt32(3, version, 0);
    }
    static addType(builder, type) {
        builder.addFieldInt16(4, type, 0);
    }
    static addMaxFee(builder, maxFeeOffset) {
        builder.addFieldOffset(5, maxFeeOffset, 0);
    }
    static createMaxFeeVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startMaxFeeVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addDeadline(builder, deadlineOffset) {
        builder.addFieldOffset(6, deadlineOffset, 0);
    }
    static createDeadlineVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startDeadlineVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addProviderMosaicId(builder, providerMosaicIdOffset) {
        builder.addFieldOffset(7, providerMosaicIdOffset, 0);
    }
    static createProviderMosaicIdVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startProviderMosaicIdVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addCurrencyDeposit(builder, currencyDepositOffset) {
        builder.addFieldOffset(8, currencyDepositOffset, 0);
    }
    static createCurrencyDepositVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startCurrencyDepositVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addInitialMosaicsMinting(builder, initialMosaicsMintingOffset) {
        builder.addFieldOffset(9, initialMosaicsMintingOffset, 0);
    }
    static createInitialMosaicsMintingVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startInitialMosaicsMintingVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addSlashingPeriod(builder, slashingPeriod) {
        builder.addFieldInt32(10, slashingPeriod, 0);
    }
    static addWindowSize(builder, windowSize) {
        builder.addFieldInt16(11, windowSize, 0);
    }
    static addSlashingAccount(builder, slashingAccountOffset) {
        builder.addFieldOffset(12, slashingAccountOffset, 0);
    }
    static createSlashingAccountVector(builder, data) {
        builder.startVector(1, data.length, 1);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt8(data[i]);
        }
        return builder.endVector();
    }
    static startSlashingAccountVector(builder, numElems) {
        builder.startVector(1, numElems, 1);
    }
    static addAlpha(builder, alpha) {
        builder.addFieldInt32(13, alpha, 0);
    }
    static addBeta(builder, beta) {
        builder.addFieldInt32(14, beta, 0);
    }
    static endCreateLiquidityProviderBuffer(builder) {
        const offset = builder.endObject();
        return offset;
    }
    static createCreateLiquidityProviderBuffer(builder, size, signatureOffset, signerOffset, version, type, maxFeeOffset, deadlineOffset, providerMosaicIdOffset, currencyDepositOffset, initialMosaicsMintingOffset, slashingPeriod, windowSize, slashingAccountOffset, alpha, beta) {
        CreateLiquidityProviderBuffer.startCreateLiquidityProviderBuffer(builder);
        CreateLiquidityProviderBuffer.addSize(builder, size);
        CreateLiquidityProviderBuffer.addSignature(builder, signatureOffset);
        CreateLiquidityProviderBuffer.addSigner(builder, signerOffset);
        CreateLiquidityProviderBuffer.addVersion(builder, version);
        CreateLiquidityProviderBuffer.addType(builder, type);
        CreateLiquidityProviderBuffer.addMaxFee(builder, maxFeeOffset);
        CreateLiquidityProviderBuffer.addDeadline(builder, deadlineOffset);
        CreateLiquidityProviderBuffer.addProviderMosaicId(builder, providerMosaicIdOffset);
        CreateLiquidityProviderBuffer.addCurrencyDeposit(builder, currencyDepositOffset);
        CreateLiquidityProviderBuffer.addInitialMosaicsMinting(builder, initialMosaicsMintingOffset);
        CreateLiquidityProviderBuffer.addSlashingPeriod(builder, slashingPeriod);
        CreateLiquidityProviderBuffer.addWindowSize(builder, windowSize);
        CreateLiquidityProviderBuffer.addSlashingAccount(builder, slashingAccountOffset);
        CreateLiquidityProviderBuffer.addAlpha(builder, alpha);
        CreateLiquidityProviderBuffer.addBeta(builder, beta);
        return CreateLiquidityProviderBuffer.endCreateLiquidityProviderBuffer(builder);
    }
}
exports.CreateLiquidityProviderBuffer = CreateLiquidityProviderBuffer;
