// tsjs-xpx-chain-sdk/dist/src/infrastructure/buffers/liquidityProvider/ManualRateChangeBuffer.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
Object.defineProperty(exports, "__esModule", { value: true });
exports.ManualRateChangeBuffer = void 0;
const flatbuffers = require("flatbuffers");
class ManualRateChangeBuffer {
    bb = null;
    bb_pos = 0;
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsManualRateChangeBuffer(bb, obj) {
        return (obj || new ManualRateChangeBuffer()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsManualRateChangeBuffer(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new ManualRateChangeBuffer()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    size() {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    signature(index) {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? this.bb.readUint8(this.bb.__vector(this.bb_pos + offset) + index) : 0;
    }
    signatureLength() {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    signatureArray() {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? new Uint8Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    signer(index) {
        const offset = this.bb.__offset(this.bb_pos, 8);
        return offset ? this.bb.readUint8(this.bb.__vector(this.bb_pos + offset) + index) : 0;
    }
    signerLength() {
        const offset = this.bb.__offset(this.bb_pos, 8);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    signerArray() {
        const offset = this.bb.__offset(this.bb_pos, 8);
        return offset ? new Uint8Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    version() {
        const offset = this.bb.__offset(this.bb_pos, 10);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    type() {
        const offset = this.bb.__offset(this.bb_pos, 12);
        return offset ? this.bb.readUint16(this.bb_pos + offset) : 0;
    }
    maxFee(index) {
        const offset = this.bb.__offset(this.bb_pos, 14);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    maxFeeLength() {
        const offset = this.bb.__offset(this.bb_pos, 14);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    maxFeeArray() {
        const offset = this.bb.__offset(this.bb_pos, 14);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    deadline(index) {
        const offset = this.bb.__offset(this.bb_pos, 16);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    deadlineLength() {
        const offset = this.bb.__offset(this.bb_pos, 16);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    deadlineArray() {
        const offset = this.bb.__offset(this.bb_pos, 16);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    providerMosaicId(index) {
        const offset = this.bb.__offset(this.bb_pos, 18);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    providerMosaicIdLength() {
        const offset = this.bb.__offset(this.bb_pos, 18);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    providerMosaicIdArray() {
        const offset = this.bb.__offset(this.bb_pos, 18);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    currencyBalanceIncrease() {
        const offset = this.bb.__offset(this.bb_pos, 20);
        return offset ? !!this.bb.readInt8(this.bb_pos + offset) : false;
    }
    currencyBalanceChange(index) {
        const offset = this.bb.__offset(this.bb_pos, 22);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    currencyBalanceChangeLength() {
        const offset = this.bb.__offset(this.bb_pos, 22);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    currencyBalanceChangeArray() {
        const offset = this.bb.__offset(this.bb_pos, 22);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    mosaicBalanceIncrease() {
        const offset = this.bb.__offset(this.bb_pos, 24);
        return offset ? !!this.bb.readInt8(this.bb_pos + offset) : false;
    }
    mosaicBalanceChange(index) {
        const offset = this.bb.__offset(this.bb_pos, 26);
        return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
    }
    mosaicBalanceChangeLength() {
        const offset = this.bb.__offset(this.bb_pos, 26);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    mosaicBalanceChangeArray() {
        const offset = this.bb.__offset(this.bb_pos, 26);
        return offset ? new Uint32Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    static startManualRateChangeBuffer(builder) {
        builder.startObject(12);
    }
    static addSize(builder, size) {
        builder.addFieldInt32(0, size, 0);
    }
    static addSignature(builder, signatureOffset) {
        builder.addFieldOffset(1, signatureOffset, 0);
    }
    static createSignatureVector(builder, data) {
        builder.startVector(1, data.length, 1);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt8(data[i]);
        }
        return builder.endVector();
    }
    static startSignatureVector(builder, numElems) {
        builder.startVector(1, numElems, 1);
    }
    static addSigner(builder, signerOffset) {
        builder.addFieldOffset(2, signerOffset, 0);
    }
    static createSignerVector(builder, data) {
        builder.startVector(1, data.length, 1);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt8(data[i]);
        }
        return builder.endVector();
    }
    static startSignerVector(builder, numElems) {
        builder.startVector(1, numElems, 1);
    }
    static addVersion(builder, version) {
        builder.addFieldInt32(3, version, 0);
    }
    static addType(builder, type) {
        builder.addFieldInt16(4, type, 0);
    }
    static addMaxFee(builder, maxFeeOffset) {
        builder.addFieldOffset(5, maxFeeOffset, 0);
    }
    static createMaxFeeVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startMaxFeeVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addDeadline(builder, deadlineOffset) {
        builder.addFieldOffset(6, deadlineOffset, 0);
    }
    static createDeadlineVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startDeadlineVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addProviderMosaicId(builder, providerMosaicIdOffset) {
        builder.addFieldOffset(7, providerMosaicIdOffset, 0);
    }
    static createProviderMosaicIdVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startProviderMosaicIdVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addCurrencyBalanceIncrease(builder, currencyBalanceIncrease) {
        builder.addFieldInt8(8, +currencyBalanceIncrease, +false);
    }
    static addCurrencyBalanceChange(builder, currencyBalanceChangeOffset) {
        builder.addFieldOffset(9, currencyBalanceChangeOffset, 0);
    }
    static createCurrencyBalanceChangeVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startCurrencyBalanceChangeVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addMosaicBalanceIncrease(builder, mosaicBalanceIncrease) {
        builder.addFieldInt8(10, +mosaicBalanceIncrease, +false);
    }
    static addMosaicBalanceChange(builder, mosaicBalanceChangeOffset) {
        builder.addFieldOffset(11, mosaicBalanceChangeOffset, 0);
    }
    static createMosaicBalanceChangeVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt32(data[i]);
        }
        return builder.endVector();
    }
    static startMosaicBalanceChangeVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static endManualRateChangeBuffer(builder) {
        const offset = builder.endObject();
        return offset;
    }
    static finishManualRateChangeBufferBuffer(builder, offset) {
        builder.finish(offset);
    }
    static finishSizePrefixedManualRateChangeBufferBuffer(builder, offset) {
        builder.finish(offset, undefined, true);
    }
    static createManualRateChangeBuffer(builder, size, signatureOffset, signerOffset, version, type, maxFeeOffset, deadlineOffset, providerMosaicIdOffset, currencyBalanceIncrease, currencyBalanceChangeOffset, mosaicBalanceIncrease, mosaicBalanceChangeOffset) {
        ManualRateChangeBuffer.startManualRateChangeBuffer(builder);
        ManualRateChangeBuffer.addSize(builder, size);
        ManualRateChangeBuffer.addSignature(builder, signatureOffset);
        ManualRateChangeBuffer.addSigner(builder, signerOffset);
        ManualRateChangeBuffer.addVersion(builder, version);
        ManualRateChangeBuffer.addType(builder, type);
        ManualRateChangeBuffer.addMaxFee(builder, maxFeeOffset);
        ManualRateChangeBuffer.addDeadline(builder, deadlineOffset);
        ManualRateChangeBuffer.addProviderMosaicId(builder, providerMosaicIdOffset);
        ManualRateChangeBuffer.addCurrencyBalanceIncrease(builder, currencyBalanceIncrease);
        ManualRateChangeBuffer.addCurrencyBalanceChange(builder, currencyBalanceChangeOffset);
        ManualRateChangeBuffer.addMosaicBalanceIncrease(builder, mosaicBalanceIncrease);
        ManualRateChangeBuffer.addMosaicBalanceChange(builder, mosaicBalanceChangeOffset);
        return ManualRateChangeBuffer.endManualRateChangeBuffer(builder);
    }
}
exports.ManualRateChangeBuffer = ManualRateChangeBuffer;
