// tsjs-xpx-chain-sdk/dist/src/model/transaction/storage/DataModificationSingleApprovalTransaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataModificationSingleApprovalTransaction = void 0;
const Transaction_1 = require("../Transaction");
const TransactionType_1 = require("../TransactionType");
class DataModificationSingleApprovalTransaction extends Transaction_1.Transaction {
    driveKey;
    dataModificationId;
    publicKeysCount;
    publicKeys;
    opinions;
    /**
     * @param networkType
     * @param version
     * @param deadline
     * @param maxFee
     * @param driveKey - Public key of the drive
     * @param dataModificationId hash
     * @param publicKeysCount
     * @param publicKeys
     * @param opinions uint64[]
     * @param signature
     * @param signer
     * @param transactionInfo
     */
    constructor(networkType, version, deadline, maxFee, driveKey, dataModificationId, // hash
    publicKeysCount, // uint8
    publicKeys, opinions, signature, signer, transactionInfo) {
        super(TransactionType_1.TransactionType.Data_Modification_Single_Approval, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
        this.driveKey = driveKey;
        this.dataModificationId = dataModificationId;
        this.publicKeysCount = publicKeysCount;
        this.publicKeys = publicKeys;
        this.opinions = opinions;
    }
    /**
     * @override Transaction.size()
     * @description get the byte size of a DataModificationSingleApprovalTransaction
     * @returns {number}
     * @memberof DataModificationSingleApprovalTransaction
     */
    get size() {
        return DataModificationSingleApprovalTransaction.calculateSize(this.publicKeys.length, this.opinions.length);
    }
    static calculateSize(publicKeysCount, opinionsCount) {
        const baseByteSize = Transaction_1.Transaction.getHeaderSize();
        // set static byte size fields
        const driveKeySize = 32;
        const dataModificationId = 32;
        const publicKeysCountSize = 1;
        const publicKeysSize = 32 * publicKeysCount;
        const opinionsSize = 8 * opinionsCount;
        return baseByteSize + driveKeySize + dataModificationId + publicKeysCountSize + publicKeysSize + opinionsSize;
    }
    /**
     * @override Transaction.toJSON()
     * @description Serialize a transaction object - add own fields to the result of Transaction.toJSON()
     * @return {Object}
     * @memberof DataModificationSingleApprovalTransaction
     */
    toJSON() {
        const parent = super.toJSON();
        return {
            ...parent,
            transaction: {
                ...parent.transaction,
                driveKey: this.driveKey.toDTO(),
                dataModificationId: this.dataModificationId,
                publicKeysCount: this.publicKeysCount,
                publicKeys: this.publicKeys.map((pa) => pa.publicKey),
                opinions: this.opinions.map((data) => data.toBigInt())
            }
        };
    }
    /**
     * @internal
     * @returns {never}
     */
    buildTransaction() {
        throw new Error("Not yet implemented");
    }
}
exports.DataModificationSingleApprovalTransaction = DataModificationSingleApprovalTransaction;
