// tsjs-xpx-chain-sdk/dist/src/model/transaction/storage/DataModificationApprovalTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataModificationApprovalTransaction = void 0;
const Transaction_1 = require("../Transaction");
const TransactionType_1 = require("../TransactionType");
class DataModificationApprovalTransaction extends Transaction_1.Transaction {
    driveKey;
    dataModificationId;
    modificationStatus;
    fileStructureCdi;
    fileStructureSizeBytes;
    metaFilesSizeBytes;
    usedDriveSizeBytes;
    judgingKeysCount;
    overlappingKeysCount;
    judgedKeysCount;
    opinionElementCount;
    publicKeys;
    signatures;
    presentOpinions;
    opinions;
    constructor(networkType, version, deadline, maxFee, driveKey, dataModificationId, // hash
    modificationStatus, // uint8
    fileStructureCdi, // hash
    fileStructureSizeBytes, metaFilesSizeBytes, usedDriveSizeBytes, judgingKeysCount, //uint8
    overlappingKeysCount, //uint8
    judgedKeysCount, //uint8
    opinionElementCount, //uint16
    publicKeys, signatures, presentOpinions, opinions, signature, signer, transactionInfo) {
        super(TransactionType_1.TransactionType.Data_Modification_Approval, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
        this.driveKey = driveKey;
        this.dataModificationId = dataModificationId;
        this.modificationStatus = modificationStatus;
        this.fileStructureCdi = fileStructureCdi;
        this.fileStructureSizeBytes = fileStructureSizeBytes;
        this.metaFilesSizeBytes = metaFilesSizeBytes;
        this.usedDriveSizeBytes = usedDriveSizeBytes;
        this.judgingKeysCount = judgingKeysCount;
        this.overlappingKeysCount = overlappingKeysCount;
        this.judgedKeysCount = judgedKeysCount;
        this.opinionElementCount = opinionElementCount;
        this.publicKeys = publicKeys;
        this.signatures = signatures;
        this.presentOpinions = presentOpinions;
        this.opinions = opinions;
    }
    /**
     * @override Transaction.size()
     * @description get the byte size of a DataModificationApprovalTransaction
     * @returns {number}
     * @memberof DataModificationApprovalTransaction
     */
    get size() {
        return DataModificationApprovalTransaction.calculateSize(this.publicKeys.length, this.signatures.length, this.presentOpinions.length, this.opinions.length);
    }
    static calculateSize(publicKeysCount, signaturesCount, presentOpinionsCount, opinionsCount) {
        const baseByteSize = Transaction_1.Transaction.getHeaderSize();
        // set static byte size fields
        const driveKeySize = 32;
        const dataModificationId = 32;
        const modificationStatus = 1;
        const fileStructureCdi = 32;
        const fileStructureSizeBytes = 8;
        const metaFilesSizeBytes = 8;
        const usedDriveSizeBytes = 8;
        const judgingKeysCount = 1;
        const overlappingKeysCount = 1;
        const judgedKeysCount = 1;
        const opinionElementCount = 2;
        const publicKeysSize = publicKeysCount * 32;
        const signaturesSize = signaturesCount * 64;
        const opinionsSize = opinionsCount * 8;
        return baseByteSize + driveKeySize + dataModificationId + modificationStatus + fileStructureCdi + fileStructureSizeBytes +
            metaFilesSizeBytes + usedDriveSizeBytes + judgingKeysCount + overlappingKeysCount + judgedKeysCount +
            opinionElementCount + publicKeysSize + signaturesSize + presentOpinionsCount + opinionsSize;
    }
    /**
     * @override Transaction.toJSON()
     * @description Serialize a transaction object - add own fields to the result of Transaction.toJSON()
     * @return {Object}
     * @memberof DataModificationApprovalTransaction
     */
    toJSON() {
        const parent = super.toJSON();
        return {
            ...parent,
            transaction: {
                ...parent.transaction,
                driveKey: this.driveKey.publicKey,
                dataModificationId: this.dataModificationId,
                fileStructureCdi: this.fileStructureCdi,
                fileStructureSizeBytes: this.fileStructureSizeBytes.toBigInt(),
                metaFilesSizeBytes: this.metaFilesSizeBytes.toBigInt(),
                usedDriveSizeBytes: this.usedDriveSizeBytes.toBigInt(),
                judgingKeysCount: this.judgingKeysCount,
                overlappingKeysCount: this.overlappingKeysCount,
                judgedKeysCount: this.judgedKeysCount,
                opinionElementCount: this.opinionElementCount,
                publicKeys: this.publicKeys.map(t => t.publicKey),
                signatures: this.signatures.map(t => t),
                presentOpinions: this.presentOpinions,
                opinions: this.opinions.map(t => t.toBigInt()),
            }
        };
    }
    /**
     * @internal
     * @returns {never}
     */
    buildTransaction() {
        throw new Error("Not yet implemented");
    }
}
exports.DataModificationApprovalTransaction = DataModificationApprovalTransaction;
