// tsjs-xpx-chain-sdk/dist/src/model/transaction/storage/DownloadApprovalTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DownloadApprovalTransaction = void 0;
const Transaction_1 = require("../Transaction");
const TransactionType_1 = require("../TransactionType");
class DownloadApprovalTransaction extends Transaction_1.Transaction {
    downloadChannelId;
    approvalTrigger;
    judgingKeysCount;
    overlappingKeysCount;
    judgedKeysCount;
    opinionElementCount;
    publicKeys;
    signatures;
    presentOpinions;
    opinions;
    /**
     * @param networkType
     * @param version
     * @param deadline
     * @param maxFee
     * @param downloadChannelId hash
     * @param approvalTrigger hash
     * @param judgingKeysCount uint8
     * @param overlappingKeysCount uint8
     * @param judgedKeysCount uint8
     * @param opinionElementCount uint16
     * @param publicKeys PublicAccount[]
     * @param signatures string[]
     * @param presentOpinions Uint8Array,
     * @param opinions UInt64[],
     * @param signature
     * @param signer
     * @param transactionInfo
     */
    constructor(networkType, version, deadline, maxFee, downloadChannelId, // hash
    approvalTrigger, // hash
    judgingKeysCount, // uint8
    overlappingKeysCount, // uint8
    judgedKeysCount, // uint8
    opinionElementCount, // uint16
    publicKeys, signatures, presentOpinions, opinions, signature, signer, transactionInfo) {
        super(TransactionType_1.TransactionType.Download_Approval, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
        this.downloadChannelId = downloadChannelId;
        this.approvalTrigger = approvalTrigger;
        this.judgingKeysCount = judgingKeysCount;
        this.overlappingKeysCount = overlappingKeysCount;
        this.judgedKeysCount = judgedKeysCount;
        this.opinionElementCount = opinionElementCount;
        this.publicKeys = publicKeys;
        this.signatures = signatures;
        this.presentOpinions = presentOpinions;
        this.opinions = opinions;
    }
    /**
     * @override Transaction.size()
     * @description get the byte size of a DownloadApprovalTransaction
     * @returns {number}
     * @memberof DownloadApprovalTransaction
     */
    get size() {
        return DownloadApprovalTransaction.calculateSize(this.publicKeys.length, this.signatures.length, this.presentOpinions.length, this.opinions.length);
    }
    static calculateSize(publicKeysCount, signaturesCount, presentOpinionsCount, opinionsCount) {
        const baseByteSize = Transaction_1.Transaction.getHeaderSize();
        // set static byte size fields
        const downloadChannelId = 32; // hash
        const approvalTrigger = 32; // hash
        const judgingKeysCount = 1; // uint8
        const overlappingKeysCount = 1; // uint8
        const judgedKeysCount = 1; // uint8
        const opinionElementCount = 2; // uint16
        const publicKeysSize = publicKeysCount * 32;
        const signaturesSize = signaturesCount * 64;
        const opinionsSize = opinionsCount * 8;
        return baseByteSize + downloadChannelId + approvalTrigger + judgingKeysCount + overlappingKeysCount + judgedKeysCount +
            opinionElementCount + publicKeysSize + signaturesSize + presentOpinionsCount + opinionsSize;
    }
    /**
     * @override Transaction.toJSON()
     * @description Serialize a transaction object - add own fields to the result of Transaction.toJSON()
     * @return {Object}
     * @memberof NewDownloadTransaction
     */
    toJSON() {
        const parent = super.toJSON();
        return {
            ...parent,
            transaction: {
                ...parent.transaction,
                downloadChannelId: this.downloadChannelId,
                approvalTrigger: this.approvalTrigger,
                judgingKeysCount: this.judgedKeysCount,
                overlappingKeysCount: this.overlappingKeysCount,
                judgedKeysCount: this.judgedKeysCount,
                opinionElementCount: this.opinionElementCount,
                publicKeys: this.publicKeys.map(pa => pa.publicKey),
                signatures: this.signatures,
                presentOpinions: this.presentOpinions,
                opinions: this.opinions.map(t => t.toBigInt()),
            }
        };
    }
    /**
     * @internal
     * @returns {VerifiableTransaction}
     */
    buildTransaction() {
        throw new Error("Not yet implemented");
    }
}
exports.DownloadApprovalTransaction = DownloadApprovalTransaction;
