// tsjs-xpx-chain-sdk/dist/src/model/transaction/storage/NewDownloadPaymentTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewDownloadPaymentTransactionBuilder = exports.NewDownloadPaymentTransaction = void 0;
const NewDownloadPaymentTransaction_1 = require("../../../infrastructure/builders/storage/NewDownloadPaymentTransaction");
const Transaction_1 = require("../Transaction");
const TransactionType_1 = require("../TransactionType");
const TransactionTypeVersion_1 = require("../TransactionTypeVersion");
const FeeCalculationStrategy_1 = require("../FeeCalculationStrategy");
const Convert_1 = require("../../../core/format/Convert");
class NewDownloadPaymentTransaction extends Transaction_1.Transaction {
    downloadChannelId;
    downloadSize;
    feedbackFeeAmount;
    /**
     * Create a new replicator onboarding transaction object
     * @param deadline - The deadline to include the transaction.
     * @param downloadChannelId - The download channel id hash
     * @param downloadSize - The download size
     * @param feedbackFeeAmount - feedback fee amount
     * @param networkType - The network type.
     * @param maxFee - (Optional) Max fee defined by the sender
     * @returns {NewDownloadPaymentTransaction}
     */
    static create(deadline, downloadChannelId, downloadSize, feedbackFeeAmount, networkType, maxFee) {
        return new NewDownloadPaymentTransactionBuilder()
            .networkType(networkType)
            .deadline(deadline)
            .downloadChannelId(downloadChannelId)
            .downloadSize(downloadSize)
            .feedbackFeeAmount(feedbackFeeAmount)
            .maxFee(maxFee)
            .build();
    }
    /**
     * @param networkType
     * @param version
     * @param deadline
     * @param maxFee
     * @param downloadChannelId - The identifier of the download channel
     * @param downloadSize - Download size in Mb to add to the prepaid size of the download channel
     * @param feedbackFeeAmount - Amount of XPXs to transfer to the download channel
     * @param signature
     * @param signer
     * @param transactionInfo
     */
    constructor(networkType, version, deadline, maxFee, downloadChannelId, downloadSize, feedbackFeeAmount, signature, signer, transactionInfo) {
        super(TransactionType_1.TransactionType.Download_Payment, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
        this.downloadChannelId = downloadChannelId;
        this.downloadSize = downloadSize;
        this.feedbackFeeAmount = feedbackFeeAmount;
        if (!Convert_1.Convert.isHexString(downloadChannelId) && downloadChannelId.length !== 64) {
            throw new Error("downloadChannelId should be 32 bytes hash string");
        }
        if (downloadSize.toBigInt() <= BigInt(0)) {
            throw new Error("downloadSize should be positive value");
        }
        if (feedbackFeeAmount.toBigInt() <= BigInt(0)) {
            throw new Error("feedbackFeeAmount should be positive value");
        }
    }
    /**
     * @override Transaction.size()
     * @description get the byte size of a NewDownloadPaymentTransaction
     * @returns {number}
     * @memberof NewDownloadPaymentTransaction
     */
    get size() {
        return NewDownloadPaymentTransaction.calculateSize();
    }
    static calculateSize() {
        const baseByteSize = Transaction_1.Transaction.getHeaderSize();
        // set static byte size fields
        const downloadChannelIdSize = 32;
        const downloadSizeSize = 8;
        const feedbackFeeAmountSize = 8;
        return baseByteSize + downloadChannelIdSize + downloadSizeSize + feedbackFeeAmountSize;
    }
    /**
     * @override Transaction.toJSON()
     * @description Serialize a transaction object - add own fields to the result of Transaction.toJSON()
     * @return {Object}
     * @memberof NewDownloadPaymentTransaction
     */
    toJSON() {
        const parent = super.toJSON();
        return {
            ...parent,
            transaction: {
                ...parent.transaction,
                downloadChannelId: this.downloadChannelId,
                downloadSize: this.downloadSize.toDTO(),
                feedbackFeeAmount: this.feedbackFeeAmount.toDTO()
            }
        };
    }
    /**
     * @internal
     * @returns {VerifiableTransaction}
     */
    buildTransaction() {
        return new NewDownloadPaymentTransaction_1.Builder()
            .addSize(this.size)
            .addDeadline(this.deadline.toDTO())
            .addMaxFee(this.maxFee.toDTO())
            .addVersion(this.versionToDTO())
            .addDownloadChannelId(this.downloadChannelId)
            .addDownloadSize(this.downloadSize.toDTO())
            .addFeedbackFeeAmount(this.feedbackFeeAmount.toDTO())
            .build();
    }
}
exports.NewDownloadPaymentTransaction = NewDownloadPaymentTransaction;
class NewDownloadPaymentTransactionBuilder extends Transaction_1.TransactionBuilder {
    _downloadChannelId;
    _downloadSize;
    _feedbackFeeAmount;
    downloadChannelId(downloadChannelId) {
        this._downloadChannelId = downloadChannelId;
        return this;
    }
    downloadSize(downloadSize) {
        this._downloadSize = downloadSize;
        return this;
    }
    feedbackFeeAmount(feedbackFeeAmount) {
        this._feedbackFeeAmount = feedbackFeeAmount;
        return this;
    }
    build() {
        return new NewDownloadPaymentTransaction(this._networkType, this._version || TransactionTypeVersion_1.TransactionTypeVersion.Download_Payment, this._deadline ? this._deadline : this._createNewDeadlineFn(), this._maxFee ? this._maxFee : (0, FeeCalculationStrategy_1.calculateFee)(NewDownloadPaymentTransaction.calculateSize(), this._feeCalculationStrategy), this._downloadChannelId, this._downloadSize, this._feedbackFeeAmount, this._signature, this._signer, this._transactionInfo);
    }
}
exports.NewDownloadPaymentTransactionBuilder = NewDownloadPaymentTransactionBuilder;
