// tsjs-xpx-chain-sdk/dist/src/model/transaction/storage/NewDataModificationTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewDataModificationTransactionBuilder = exports.NewDataModificationTransaction = void 0;
const NewDataModificationTransaction_1 = require("../../../infrastructure/builders/storage/NewDataModificationTransaction");
const Transaction_1 = require("../Transaction");
const TransactionType_1 = require("../TransactionType");
const TransactionTypeVersion_1 = require("../TransactionTypeVersion");
const FeeCalculationStrategy_1 = require("../FeeCalculationStrategy");
const Convert_1 = require("../../../core/format/Convert");
class NewDataModificationTransaction extends Transaction_1.Transaction {
    driveKey;
    downloadDataCdi;
    uploadSize;
    feedbackFeeAmount;
    /**
     * Create a new replicator onboarding transaction object
     * @param deadline - The deadline to include the transaction.
     * @param driveKey - The drive key
     * @param downloadDataCdi - The download data CDI hash
     * @param uploadSize - The upload size
     * @param feedbackFeeAmount - The feedback fee amount.
     * @param networkType - The network type.
     * @param maxFee - (Optional) Max fee defined by the sender
     * @returns {NewDataModificationTransaction}
     */
    static create(deadline, driveKey, downloadDataCdi, uploadSize, feedbackFeeAmount, networkType, maxFee) {
        return new NewDataModificationTransactionBuilder()
            .networkType(networkType)
            .deadline(deadline)
            .driveKey(driveKey)
            .downloadDataCdi(downloadDataCdi)
            .uploadSize(uploadSize)
            .feedbackFeeAmount(feedbackFeeAmount)
            .maxFee(maxFee)
            .build();
    }
    /**
     * @param networkType
     * @param version
     * @param deadline
     * @param maxFee
     * @param driveKey - Public key of the drive
     * @param downloadDataCdi - Download data CDI of modification hash
     * @param uploadSize - Size of upload in MB
     * @param feedbackFeeAmount - Amount of XPXs to transfer to the drive
     * @param signature
     * @param signer
     * @param transactionInfo
     */
    constructor(networkType, version, deadline, maxFee, driveKey, downloadDataCdi, uploadSize, feedbackFeeAmount, signature, signer, transactionInfo) {
        super(TransactionType_1.TransactionType.Data_Modification, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
        this.driveKey = driveKey;
        this.downloadDataCdi = downloadDataCdi;
        this.uploadSize = uploadSize;
        this.feedbackFeeAmount = feedbackFeeAmount;
        if (!Convert_1.Convert.isHexString(downloadDataCdi) && downloadDataCdi.length !== 64) {
            throw new Error("downloadDataCdi should be 32 bytes hash string");
        }
        if (uploadSize.toBigInt() <= BigInt(0)) {
            throw new Error("uploadSize should be positive value");
        }
        if (feedbackFeeAmount.toBigInt() <= BigInt(0)) {
            throw new Error("feedbackFeeAmount should be positive value");
        }
    }
    /**
     * @override Transaction.size()
     * @description get the byte size of a NewDataModificationTransaction
     * @returns {number}
     * @memberof NewDataModificationTransaction
     */
    get size() {
        return NewDataModificationTransaction.calculateSize();
    }
    static calculateSize() {
        const baseByteSize = Transaction_1.Transaction.getHeaderSize();
        // set static byte size fields
        const driveKeySize = 32;
        const downloadDataCdiSize = 32;
        const uploadSizeSize = 8;
        const feedbackFeeAmountSize = 8;
        return baseByteSize + driveKeySize + downloadDataCdiSize + uploadSizeSize + feedbackFeeAmountSize;
    }
    /**
     * @override Transaction.toJSON()
     * @description Serialize a transaction object - add own fields to the result of Transaction.toJSON()
     * @return {Object}
     * @memberof NewDataModificationTransaction
     */
    toJSON() {
        const parent = super.toJSON();
        return {
            ...parent,
            transaction: {
                ...parent.transaction,
                driveKey: this.driveKey.toDTO(),
                downloadDataCdi: this.downloadDataCdi,
                uploadSize: this.uploadSize.toDTO(),
                feedbackFeeAmount: this.feedbackFeeAmount.toDTO()
            }
        };
    }
    /**
     * @internal
     * @returns {VerifiableTransaction}
     */
    buildTransaction() {
        return new NewDataModificationTransaction_1.Builder()
            .addSize(this.size)
            .addDeadline(this.deadline.toDTO())
            .addMaxFee(this.maxFee.toDTO())
            .addVersion(this.versionToDTO())
            .addDriveKey(this.driveKey.publicKey)
            .addDownloadDataCdi(this.downloadDataCdi)
            .addUploadSize(this.uploadSize.toDTO())
            .addFeedbackFeeAmount(this.feedbackFeeAmount.toDTO())
            .build();
    }
}
exports.NewDataModificationTransaction = NewDataModificationTransaction;
class NewDataModificationTransactionBuilder extends Transaction_1.TransactionBuilder {
    _driveKey;
    _downloadDataCdi;
    _uploadSize;
    _feedbackFeeAmount;
    driveKey(driveKey) {
        this._driveKey = driveKey;
        return this;
    }
    downloadDataCdi(downloadDataCdi) {
        this._downloadDataCdi = downloadDataCdi;
        return this;
    }
    uploadSize(uploadSize) {
        this._uploadSize = uploadSize;
        return this;
    }
    feedbackFeeAmount(feedbackFeeAmount) {
        this._feedbackFeeAmount = feedbackFeeAmount;
        return this;
    }
    build() {
        return new NewDataModificationTransaction(this._networkType, this._version || TransactionTypeVersion_1.TransactionTypeVersion.Data_Modification, this._deadline ? this._deadline : this._createNewDeadlineFn(), this._maxFee ? this._maxFee : (0, FeeCalculationStrategy_1.calculateFee)(NewDataModificationTransaction.calculateSize(), this._feeCalculationStrategy), this._driveKey, this._downloadDataCdi, this._uploadSize, this._feedbackFeeAmount, this._signature, this._signer, this._transactionInfo);
    }
}
exports.NewDataModificationTransactionBuilder = NewDataModificationTransactionBuilder;
