// tsjs-xpx-chain-sdk/dist/src/model/transaction/storage/NewDataModificationCancelTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewDataModificationCancelTransactionBuilder = exports.NewDataModificationCancelTransaction = void 0;
const NewDataModificationCancelTransaction_1 = require("../../../infrastructure/builders/storage/NewDataModificationCancelTransaction");
const Transaction_1 = require("../Transaction");
const TransactionType_1 = require("../TransactionType");
const TransactionTypeVersion_1 = require("../TransactionTypeVersion");
const FeeCalculationStrategy_1 = require("../FeeCalculationStrategy");
const Convert_1 = require("../../../core/format/Convert");
class NewDataModificationCancelTransaction extends Transaction_1.Transaction {
    driveKey;
    dataModificationId;
    /**
     * Create a new replicator onboarding transaction object
     * @param deadline - The deadline to include the transaction.
     * @param driveKey - The drive key
     * @param downloadDataCdi - The download data CDI hash
     * @param networkType - The network type.
     * @param maxFee - (Optional) Max fee defined by the sender
     * @returns {NewDataModificationCancelTransaction}
     */
    static create(deadline, driveKey, downloadDataCdi, networkType, maxFee) {
        return new NewDataModificationCancelTransactionBuilder()
            .networkType(networkType)
            .deadline(deadline)
            .driveKey(driveKey)
            .downloadDataCdi(downloadDataCdi)
            .maxFee(maxFee)
            .build();
    }
    /**
     * @param networkType
     * @param version
     * @param deadline
     * @param maxFee
     * @param driveKey - Public key of the drive
     * @param downloadDataCdi - Identifier of the transaction that initiated the modification
     * @param signature
     * @param signer
     * @param transactionInfo
     */
    constructor(networkType, version, deadline, maxFee, driveKey, dataModificationId, signature, signer, transactionInfo) {
        super(TransactionType_1.TransactionType.Data_Modification_Cancel, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
        this.driveKey = driveKey;
        this.dataModificationId = dataModificationId;
        if (!Convert_1.Convert.isHexString(dataModificationId) || dataModificationId.length !== 64) {
            throw new Error("downloadDataCdi should be 32 bytes hexadecimal string");
        }
    }
    /**
     * @override Transaction.size()
     * @description get the byte size of a NewDataModificationCancelTransaction
     * @returns {number}
     * @memberof NewDataModificationCancelTransaction
     */
    get size() {
        return NewDataModificationCancelTransaction.calculateSize();
    }
    static calculateSize() {
        const baseByteSize = Transaction_1.Transaction.getHeaderSize();
        // set static byte size fields
        const driveKeySize = 32;
        const downloadDataCdiSize = 32;
        return baseByteSize + driveKeySize + downloadDataCdiSize;
    }
    /**
     * @override Transaction.toJSON()
     * @description Serialize a transaction object - add own fields to the result of Transaction.toJSON()
     * @return {Object}
     * @memberof NewDataModificationCancelTransaction
     */
    toJSON() {
        const parent = super.toJSON();
        return {
            ...parent,
            transaction: {
                ...parent.transaction,
                driveKey: this.driveKey.toDTO(),
                dataModificationId: this.dataModificationId
            }
        };
    }
    /**
     * @internal
     * @returns {VerifiableTransaction}
     */
    buildTransaction() {
        return new NewDataModificationCancelTransaction_1.Builder()
            .addSize(this.size)
            .addDeadline(this.deadline.toDTO())
            .addMaxFee(this.maxFee.toDTO())
            .addVersion(this.versionToDTO())
            .addDriveKey(this.driveKey.publicKey)
            .addDataModificationId(this.dataModificationId)
            .build();
    }
}
exports.NewDataModificationCancelTransaction = NewDataModificationCancelTransaction;
class NewDataModificationCancelTransactionBuilder extends Transaction_1.TransactionBuilder {
    _driveKey;
    _dataModificationId;
    driveKey(driveKey) {
        this._driveKey = driveKey;
        return this;
    }
    downloadDataCdi(downloadDataCdi) {
        this._dataModificationId = downloadDataCdi;
        return this;
    }
    build() {
        return new NewDataModificationCancelTransaction(this._networkType, this._version || TransactionTypeVersion_1.TransactionTypeVersion.Data_Modification_Cancel, this._deadline ? this._deadline : this._createNewDeadlineFn(), this._maxFee ? this._maxFee : (0, FeeCalculationStrategy_1.calculateFee)(NewDataModificationCancelTransaction.calculateSize(), this._feeCalculationStrategy), this._driveKey, this._dataModificationId, this._signature, this._signer, this._transactionInfo);
    }
}
exports.NewDataModificationCancelTransactionBuilder = NewDataModificationCancelTransactionBuilder;
