// tsjs-xpx-chain-sdk/dist/src/model/transaction/storage/NewDownloadTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewDownloadTransactionBuilder = exports.NewDownloadTransaction = void 0;
const NewDownloadTransaction_1 = require("../../../infrastructure/builders/storage/NewDownloadTransaction");
const Transaction_1 = require("../Transaction");
const TransactionType_1 = require("../TransactionType");
const TransactionTypeVersion_1 = require("../TransactionTypeVersion");
const FeeCalculationStrategy_1 = require("../FeeCalculationStrategy");
class NewDownloadTransaction extends Transaction_1.Transaction {
    driveKey;
    downloadSize;
    feedbackFeeAmount;
    listOfPublicKeys;
    /**
     * Create a new replicator onboarding transaction object
     * @param deadline - The deadline to include the transaction.
     * @param driveKey - Public key of the drive
     * @param downloadSize - Prepaid Download Size in Mb
     * @param feedbackFeeAmount - XPXs to lock for future payment for
     * @param listOfPublicKeys - List of public keys
     * @param networkType - The network type.
     * @param maxFee - (Optional) Max fee defined by the sender
     * @returns {NewDownloadTransaction}
     */
    static create(deadline, driveKey, downloadSize, feedbackFeeAmount, listOfPublicKeys, networkType, maxFee) {
        return new NewDownloadTransactionBuilder()
            .networkType(networkType)
            .deadline(deadline)
            .driveKey(driveKey)
            .downloadSize(downloadSize)
            .feedbackFeeAmount(feedbackFeeAmount)
            .listOfPublicKeys(listOfPublicKeys)
            .maxFee(maxFee)
            .build();
    }
    /**
     * @param networkType
     * @param version
     * @param deadline
     * @param maxFee
     * @param driveKey - Public key of the drive
     * @param downloadSize - Prepaid Download Size in Mb
     * @param feedbackFeeAmount - XPXs to lock for future payment for
     * @param listOfPublicKeys - List of public keys
     * @param signature
     * @param signer
     * @param transactionInfo
     */
    constructor(networkType, version, deadline, maxFee, driveKey, downloadSize, feedbackFeeAmount, listOfPublicKeys, signature, signer, transactionInfo) {
        super(TransactionType_1.TransactionType.Download, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
        this.driveKey = driveKey;
        this.downloadSize = downloadSize;
        this.feedbackFeeAmount = feedbackFeeAmount;
        this.listOfPublicKeys = listOfPublicKeys;
        if (downloadSize.toBigInt() <= BigInt(0)) {
            throw new Error("downloadSize should be positive value");
        }
        if (feedbackFeeAmount.toBigInt() < BigInt(0)) {
            throw new Error("feedbackFeeAmount should be positive value");
        }
    }
    /**
     * @override Transaction.size()
     * @description get the byte size of a NewDownloadTransaction
     * @returns {number}
     * @memberof NewDownloadTransaction
     */
    get size() {
        return NewDownloadTransaction.calculateSize(this.listOfPublicKeys.length);
    }
    static calculateSize(totalPublicKeys) {
        const baseByteSize = Transaction_1.Transaction.getHeaderSize();
        // set static byte size fields
        const driveKeySize = 32;
        const downloadSizeSize = 8;
        const feedbackFeeAmountSize = 8;
        const publicKeysSize = 2;
        const totalPublicKeysSize = totalPublicKeys * 32;
        return baseByteSize + driveKeySize + downloadSizeSize + feedbackFeeAmountSize + publicKeysSize + totalPublicKeysSize;
    }
    /**
     * @override Transaction.toJSON()
     * @description Serialize a transaction object - add own fields to the result of Transaction.toJSON()
     * @return {Object}
     * @memberof NewDownloadTransaction
     */
    toJSON() {
        const parent = super.toJSON();
        return {
            ...parent,
            transaction: {
                ...parent.transaction,
                driveKey: this.driveKey.publicKey,
                downloadSize: this.downloadSize.toDTO(),
                feedbackFeeAmount: this.feedbackFeeAmount.toDTO(),
                listOfPublicKeys: this.listOfPublicKeys.map((publicKey) => {
                    return publicKey.toDTO();
                })
            }
        };
    }
    /**
     * @internal
     * @returns {VerifiableTransaction}
     */
    buildTransaction() {
        return new NewDownloadTransaction_1.Builder()
            .addSize(this.size)
            .addDeadline(this.deadline.toDTO())
            .addMaxFee(this.maxFee.toDTO())
            .addVersion(this.versionToDTO())
            .addDriveKey(this.driveKey.publicKey)
            .addDownloadSize(this.downloadSize.toDTO())
            .addFeedbackFeeAmount(this.feedbackFeeAmount.toDTO())
            .addListOfPublicKeys(this.listOfPublicKeys.map((data) => data.publicKey))
            .build();
    }
}
exports.NewDownloadTransaction = NewDownloadTransaction;
class NewDownloadTransactionBuilder extends Transaction_1.TransactionBuilder {
    _driveKey;
    _downloadSize;
    _feedbackFeeAmount;
    _listOfPublicKeys;
    driveKey(driveKey) {
        this._driveKey = driveKey;
        return this;
    }
    downloadSize(downloadSize) {
        this._downloadSize = downloadSize;
        return this;
    }
    feedbackFeeAmount(feedbackFeeAmount) {
        this._feedbackFeeAmount = feedbackFeeAmount;
        return this;
    }
    listOfPublicKeys(listOfPublicKeys) {
        this._listOfPublicKeys = listOfPublicKeys;
        return this;
    }
    build() {
        return new NewDownloadTransaction(this._networkType, this._version || TransactionTypeVersion_1.TransactionTypeVersion.Download, this._deadline ? this._deadline : this._createNewDeadlineFn(), this._maxFee ? this._maxFee : (0, FeeCalculationStrategy_1.calculateFee)(NewDownloadTransaction.calculateSize(this._listOfPublicKeys.length), this._feeCalculationStrategy), this._driveKey, this._downloadSize, this._feedbackFeeAmount, this._listOfPublicKeys, this._signature, this._signer, this._transactionInfo);
    }
}
exports.NewDownloadTransactionBuilder = NewDownloadTransactionBuilder;
