// tsjs-xpx-chain-sdk/dist/src/model/transaction/storage/NewEndDriveVerificationTransactionV2.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewEndDriveVerificationV2TransactionBuilder = exports.NewEndDriveVerificationV2Transaction = void 0;
const NewEndDriveVerificationV2Transaction_1 = require("../../../infrastructure/builders/storage/NewEndDriveVerificationV2Transaction");
const Transaction_1 = require("../Transaction");
const TransactionType_1 = require("../TransactionType");
const TransactionTypeVersion_1 = require("../TransactionTypeVersion");
const FeeCalculationStrategy_1 = require("../FeeCalculationStrategy");
const Convert_1 = require("../../../core/format/Convert");
class NewEndDriveVerificationV2Transaction extends Transaction_1.Transaction {
    driveKey;
    verificationTrigger;
    shardId;
    keys;
    signatures;
    opinions;
    /**
     * Create a new replicator onboarding transaction object
     * @param deadline - The deadline to include the transaction.
     * @param driveKey - Public key of the drive
     * @param verificationTrigger - The hash of block that initiated the Verification
     * @param shardId - Shard identifier, type unsigned short integer
     * @param keys - Replicators' public keys
     * @param signatures - Signatures of replicators' opinions
     * @param opinions - Two-dimensional bit array of opinions (1 is success, 0 is failure)
     * @param networkType - The network type.
     * @param maxFee - (Optional) Max fee defined by the sender
     * @returns {NewEndDriveVerificationV2Transaction}
     */
    static create(deadline, driveKey, verificationTrigger, shardId, keys, signatures, opinions, networkType, maxFee) {
        return new NewEndDriveVerificationV2TransactionBuilder()
            .networkType(networkType)
            .deadline(deadline)
            .driveKey(driveKey)
            .verificationTrigger(verificationTrigger)
            .shardId(shardId)
            .keys(keys)
            .signatures(signatures)
            .opinions(opinions)
            .maxFee(maxFee)
            .build();
    }
    /**
     * @param networkType
     * @param version
     * @param deadline
     * @param maxFee
     * @param driveKey - The drive key
     * @param verificationTrigger - hash (32 bytes)
     * @param shardId - shard id of type short unsigned integer
     * @param keys - punlic keys
     * @param signatures - Signatures array (64 bytes)
     * @param opinions - uint8
     * @param signature
     * @param signer
     * @param transactionInfo
     */
    constructor(networkType, version, deadline, maxFee, driveKey, verificationTrigger, shardId, keys, signatures, opinions, signature, signer, transactionInfo) {
        super(TransactionType_1.TransactionType.End_Drive_Verification_V2, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
        this.driveKey = driveKey;
        this.verificationTrigger = verificationTrigger;
        this.shardId = shardId;
        this.keys = keys;
        this.signatures = signatures;
        this.opinions = opinions;
        if (!Convert_1.Convert.isHexString(verificationTrigger) && verificationTrigger.length !== 64) {
            throw new Error("verificationTrigger should be 32 bytes hash string");
        }
        if (shardId > 0xFFFF || shardId < 0) {
            throw new Error("shardId out of range, should be unsigned short integer");
        }
        if (opinions > 0xFF || opinions < 0) {
            throw new Error("opinions out of range, should be unsigned byte integer");
        }
    }
    /**
     * @override Transaction.size()
     * @description get the byte size of a NewEndDriveVerificationV2Transaction
     * @returns {number}
     * @memberof NewEndDriveVerificationV2Transaction
     */
    get size() {
        return NewEndDriveVerificationV2Transaction.calculateSize(this.keys.length, this.signatures.length);
    }
    static calculateSize(keysCount, judgingKeyCount) {
        const baseByteSize = Transaction_1.Transaction.getHeaderSize();
        // set static byte size fields
        const driveKeySize = 32;
        const verificationTriggerSize = 32;
        const shardId = 2;
        const keysCountSize = 1;
        const judgingKeyCountSize = 1;
        // nested size
        const keySize = 32;
        const signatureSize = 64;
        return baseByteSize + driveKeySize + verificationTriggerSize + shardId
            + keysCountSize + judgingKeyCountSize
            + (keysCount * keySize) + (judgingKeyCount * signatureSize)
            + ((keysCount * judgingKeyCount) + 7) / 8;
    }
    /**
     * @override Transaction.toJSON()
     * @description Serialize a transaction object - add own fields to the result of Transaction.toJSON()
     * @return {Object}
     * @memberof NewEndDriveVerificationV2Transaction
     */
    toJSON() {
        const parent = super.toJSON();
        return {
            ...parent,
            transaction: {
                ...parent.transaction,
                driveKey: this.driveKey.publicKey,
                verificationTrigger: this.verificationTrigger,
                shardId: this.shardId,
                keys: this.keys.map((data) => data.toDTO()),
                signatures: this.signatures.map((data) => data),
                opinions: this.opinions
            }
        };
    }
    /**
     * @internal
     * @returns {VerifiableTransaction}
     */
    buildTransaction() {
        return new NewEndDriveVerificationV2Transaction_1.Builder()
            .addSize(this.size)
            .addDeadline(this.deadline.toDTO())
            .addMaxFee(this.maxFee.toDTO())
            .addVersion(this.versionToDTO())
            .addDriveKey(this.driveKey.publicKey)
            .addVerificationTrigger(this.verificationTrigger)
            .addShardId(this.shardId)
            .addKeys(this.keys.map((key) => key.publicKey))
            .addSignatures(this.signatures.map((signature) => signature))
            .addOpinions(this.opinions)
            .build();
    }
}
exports.NewEndDriveVerificationV2Transaction = NewEndDriveVerificationV2Transaction;
class NewEndDriveVerificationV2TransactionBuilder extends Transaction_1.TransactionBuilder {
    _driveKey;
    _verificationTrigger;
    _shardId;
    _keys;
    _signatures;
    _opinions;
    driveKey(driveKey) {
        this._driveKey = driveKey;
        return this;
    }
    verificationTrigger(verificationTrigger) {
        this._verificationTrigger = verificationTrigger;
        return this;
    }
    shardId(shardId) {
        this._shardId = shardId;
        return this;
    }
    keys(keys) {
        this._keys = keys;
        return this;
    }
    signatures(signatures) {
        this._signatures = signatures;
        return this;
    }
    opinions(opinions) {
        this._opinions = opinions;
        return this;
    }
    build() {
        return new NewEndDriveVerificationV2Transaction(this._networkType, this._version || TransactionTypeVersion_1.TransactionTypeVersion.End_Drive_Verification_V2, this._deadline ? this._deadline : this._createNewDeadlineFn(), this._maxFee ? this._maxFee : (0, FeeCalculationStrategy_1.calculateFee)(NewEndDriveVerificationV2Transaction.calculateSize(this._keys.length, this._signatures.length), this._feeCalculationStrategy), this._driveKey, this._verificationTrigger, this._shardId, this._keys, this._signatures, this._opinions, this._signature, this._signer, this._transactionInfo);
    }
}
exports.NewEndDriveVerificationV2TransactionBuilder = NewEndDriveVerificationV2TransactionBuilder;
