// tsjs-xpx-chain-sdk/dist/src/model/transaction/storage/NewPrepareBcDriveTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewPrepareBcDriveTransactionBuilder = exports.NewPrepareBcDriveTransaction = void 0;
const NewPrepareBcDriveTransaction_1 = require("../../../infrastructure/builders/storage/NewPrepareBcDriveTransaction");
const Transaction_1 = require("../Transaction");
const TransactionType_1 = require("../TransactionType");
const TransactionTypeVersion_1 = require("../TransactionTypeVersion");
const FeeCalculationStrategy_1 = require("../FeeCalculationStrategy");
class NewPrepareBcDriveTransaction extends Transaction_1.Transaction {
    driveSize;
    verificationFeeAmount;
    replicatorCount;
    /**
     * Create a new replicator onboarding transaction object
     * @param deadline - The deadline to include the transaction.
     * @param driveSize - Size of drive
     * @param verificationFeeAmount - Amount of XPXs to transfer to the drive
     * @param replicatorCount - Number of replicators
     * @param networkType - The network type.
     * @param maxFee - (Optional) Max fee defined by the sender
     * @returns {NewPrepareBcDriveTransaction}
     */
    static create(deadline, driveSize, verificationFeeAmount, replicatorCount, networkType, maxFee) {
        return new NewPrepareBcDriveTransactionBuilder()
            .networkType(networkType)
            .deadline(deadline)
            .driveSize(driveSize)
            .verificationFeeAmount(verificationFeeAmount)
            .replicatorCount(replicatorCount)
            .maxFee(maxFee)
            .build();
    }
    /**
     * @param networkType
     * @param version
     * @param deadline
     * @param maxFee
     * @param driveSize
     * @param verificationFeeAmount
     * @param replicatorCount
     * @param signature
     * @param signer
     * @param transactionInfo
     */
    constructor(networkType, version, deadline, maxFee, driveSize, verificationFeeAmount, replicatorCount, signature, signer, transactionInfo) {
        super(TransactionType_1.TransactionType.Prepare_Bc_Drive, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
        this.driveSize = driveSize;
        this.verificationFeeAmount = verificationFeeAmount;
        this.replicatorCount = replicatorCount;
        if (driveSize.toBigInt() <= BigInt(0)) {
            throw new Error("driveSize should be positive value");
        }
        if (verificationFeeAmount.toBigInt() <= BigInt(0)) {
            throw new Error("verificationFeeAmount should be positive value");
        }
        if (replicatorCount <= 0) {
            throw new Error("replicatorCount should be positive value");
        }
        else if (replicatorCount > 0xFFFF) {
            throw new Error("replicatorCount out of range, should be uint16 value");
        }
    }
    /**
     * @override Transaction.size()
     * @description get the byte size of a NewPrepareBcDriveTransaction
     * @returns {number}
     * @memberof NewPrepareBcDriveTransaction
     */
    get size() {
        return NewPrepareBcDriveTransaction.calculateSize();
    }
    static calculateSize() {
        const baseByteSize = Transaction_1.Transaction.getHeaderSize();
        // set static byte size fields
        const driveSizeSize = 8;
        const verificationFeeAmountSize = 8;
        const replicatorCountSize = 2;
        return baseByteSize + driveSizeSize + verificationFeeAmountSize + replicatorCountSize;
    }
    /**
     * @override Transaction.toJSON()
     * @description Serialize a transaction object - add own fields to the result of Transaction.toJSON()
     * @return {Object}
     * @memberof NewPrepareBcDriveTransaction
     */
    toJSON() {
        const parent = super.toJSON();
        return {
            ...parent,
            transaction: {
                ...parent.transaction,
                driveSize: this.driveSize.toDTO(),
                verificationFeeAmount: this.verificationFeeAmount.toDTO(),
                replicatorCount: this.replicatorCount,
            }
        };
    }
    /**
     * @internal
     * @returns {VerifiableTransaction}
     */
    buildTransaction() {
        return new NewPrepareBcDriveTransaction_1.Builder()
            .addSize(this.size)
            .addDeadline(this.deadline.toDTO())
            .addMaxFee(this.maxFee.toDTO())
            .addVersion(this.versionToDTO())
            .addDriveSize(this.driveSize.toDTO())
            .addVerificationFeeAmount(this.verificationFeeAmount.toDTO())
            .addReplicatorCount(this.replicatorCount)
            .build();
    }
}
exports.NewPrepareBcDriveTransaction = NewPrepareBcDriveTransaction;
class NewPrepareBcDriveTransactionBuilder extends Transaction_1.TransactionBuilder {
    _driveSize;
    _verificationFeeAmount;
    _replicatorCount;
    driveSize(driveSize) {
        this._driveSize = driveSize;
        return this;
    }
    verificationFeeAmount(verificationFeeAmount) {
        this._verificationFeeAmount = verificationFeeAmount;
        return this;
    }
    replicatorCount(replicatorCount) {
        this._replicatorCount = replicatorCount;
        return this;
    }
    build() {
        return new NewPrepareBcDriveTransaction(this._networkType, this._version || TransactionTypeVersion_1.TransactionTypeVersion.Prepare_Bc_Drive, this._deadline ? this._deadline : this._createNewDeadlineFn(), this._maxFee ? this._maxFee : (0, FeeCalculationStrategy_1.calculateFee)(NewPrepareBcDriveTransaction.calculateSize(), this._feeCalculationStrategy), this._driveSize, this._verificationFeeAmount, this._replicatorCount, this._signature, this._signer, this._transactionInfo);
    }
}
exports.NewPrepareBcDriveTransactionBuilder = NewPrepareBcDriveTransactionBuilder;
