// tsjs-xpx-chain-sdk/dist/src/model/transaction/storage/NewReplicatorOnboardingTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewReplicatorOnboardingTransactionBuilder = exports.NewReplicatorOnboardingTransaction = void 0;
const NewReplicatorOnboardingTransaction_1 = require("../../../infrastructure/builders/storage/NewReplicatorOnboardingTransaction");
const Transaction_1 = require("../Transaction");
const TransactionType_1 = require("../TransactionType");
const TransactionTypeVersion_1 = require("../TransactionTypeVersion");
const FeeCalculationStrategy_1 = require("../FeeCalculationStrategy");
class NewReplicatorOnboardingTransaction extends Transaction_1.Transaction {
    capacity;
    /**
     * Create a new replicator onboarding transaction object
     * @param deadline - The deadline to include the transaction.
     * @param capacity - The Storage Size that the Replicator provides
     * @param networkType - The network type.
     * @param maxFee - (Optional) Max fee defined by the sender
     * @returns {NewReplicatorOnboardingTransaction}
     */
    static create(deadline, capacity, networkType, maxFee) {
        return new NewReplicatorOnboardingTransactionBuilder()
            .networkType(networkType)
            .deadline(deadline)
            .capacity(capacity)
            .maxFee(maxFee)
            .build();
    }
    /**
     * @param networkType
     * @param version
     * @param deadline
     * @param maxFee
     * @param capacity
     * @param signature
     * @param signer
     * @param transactionInfo
     */
    constructor(networkType, version, deadline, maxFee, capacity, signature, signer, transactionInfo) {
        super(TransactionType_1.TransactionType.Replicator_Onboarding, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
        this.capacity = capacity;
        if (capacity.toBigInt() <= BigInt(0)) {
            throw new Error("capacity should be positive value");
        }
    }
    /**
     * @override Transaction.size()
     * @description get the byte size of a NewReplicatorOnboardingTransaction
     * @returns {number}
     * @memberof NewReplicatorOnboardingTransaction
     */
    get size() {
        return NewReplicatorOnboardingTransaction.calculateSize();
    }
    static calculateSize() {
        const baseByteSize = Transaction_1.Transaction.getHeaderSize();
        // set static byte size fields
        const capacity = 8;
        return baseByteSize + capacity;
    }
    /**
     * @override Transaction.toJSON()
     * @description Serialize a transaction object - add own fields to the result of Transaction.toJSON()
     * @return {Object}
     * @memberof NewReplicatorOnboardingTransaction
     */
    toJSON() {
        const parent = super.toJSON();
        return {
            ...parent,
            transaction: {
                ...parent.transaction,
                capacity: this.capacity.toDTO(),
            }
        };
    }
    /**
     * @internal
     * @returns {VerifiableTransaction}
     */
    buildTransaction() {
        return new NewReplicatorOnboardingTransaction_1.Builder()
            .addSize(this.size)
            .addDeadline(this.deadline.toDTO())
            .addMaxFee(this.maxFee.toDTO())
            .addVersion(this.versionToDTO())
            .addCapacity(this.capacity.toDTO())
            .build();
    }
}
exports.NewReplicatorOnboardingTransaction = NewReplicatorOnboardingTransaction;
class NewReplicatorOnboardingTransactionBuilder extends Transaction_1.TransactionBuilder {
    _capacity;
    capacity(capacity) {
        this._capacity = capacity;
        return this;
    }
    build() {
        return new NewReplicatorOnboardingTransaction(this._networkType, this._version || TransactionTypeVersion_1.TransactionTypeVersion.Replicator_Onboarding, this._deadline ? this._deadline : this._createNewDeadlineFn(), this._maxFee ? this._maxFee : (0, FeeCalculationStrategy_1.calculateFee)(NewReplicatorOnboardingTransaction.calculateSize(), this._feeCalculationStrategy), this._capacity, this._signature, this._signer, this._transactionInfo);
    }
}
exports.NewReplicatorOnboardingTransactionBuilder = NewReplicatorOnboardingTransactionBuilder;
