// tsjs-xpx-chain-sdk/dist/src/model/transaction/storage/NewStoragePaymentTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewStoragePaymentTransactionBuilder = exports.NewStoragePaymentTransaction = void 0;
const NewStoragePaymentTransaction_1 = require("../../../infrastructure/builders/storage/NewStoragePaymentTransaction");
const Transaction_1 = require("../Transaction");
const TransactionType_1 = require("../TransactionType");
const TransactionTypeVersion_1 = require("../TransactionTypeVersion");
const FeeCalculationStrategy_1 = require("../FeeCalculationStrategy");
class NewStoragePaymentTransaction extends Transaction_1.Transaction {
    driveKey;
    storageUnits;
    /**
     * Create a new replicator onboarding transaction object
     * @param deadline - The deadline to include the transaction.
     * @param driveKey - Public key of the drive
     * @param storageUnits - Amount of storage units to transfer to the drive
     * @param networkType - The network type.
     * @param maxFee - (Optional) Max fee defined by the sender
     * @returns {NewStoragePaymentTransaction}
     */
    static create(deadline, driveKey, storageUnits, networkType, maxFee) {
        return new NewStoragePaymentTransactionBuilder()
            .networkType(networkType)
            .deadline(deadline)
            .driveKey(driveKey)
            .storageUnits(storageUnits)
            .maxFee(maxFee)
            .build();
    }
    /**
     * @param networkType
     * @param version
     * @param deadline
     * @param maxFee
     * @param driveKey - Public key of the drive
     * @param storageUnits - Amount of storage units to transfer to the drive
     * @param signature
     * @param signer
     * @param transactionInfo
     */
    constructor(networkType, version, deadline, maxFee, driveKey, storageUnits, signature, signer, transactionInfo) {
        super(TransactionType_1.TransactionType.Storage_Payment, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
        this.driveKey = driveKey;
        this.storageUnits = storageUnits;
        if (storageUnits.toBigInt() <= BigInt(0)) {
            throw new Error("storageUnits should be positive value");
        }
    }
    /**
     * @override Transaction.size()
     * @description get the byte size of a NewStoragePaymentTransaction
     * @returns {number}
     * @memberof NewStoragePaymentTransaction
     */
    get size() {
        return NewStoragePaymentTransaction.calculateSize();
    }
    static calculateSize() {
        const baseByteSize = Transaction_1.Transaction.getHeaderSize();
        // set static byte size fields
        const driveKeySize = 32;
        const storageUnitsSize = 8;
        return baseByteSize + driveKeySize + storageUnitsSize;
    }
    /**
     * @override Transaction.toJSON()
     * @description Serialize a transaction object - add own fields to the result of Transaction.toJSON()
     * @return {Object}
     * @memberof NewStoragePaymentTransaction
     */
    toJSON() {
        const parent = super.toJSON();
        return {
            ...parent,
            transaction: {
                ...parent.transaction,
                driveKey: this.driveKey.toDTO(),
                storageUnits: this.storageUnits.toDTO()
            }
        };
    }
    /**
     * @internal
     * @returns {VerifiableTransaction}
     */
    buildTransaction() {
        return new NewStoragePaymentTransaction_1.Builder()
            .addSize(this.size)
            .addDeadline(this.deadline.toDTO())
            .addMaxFee(this.maxFee.toDTO())
            .addVersion(this.versionToDTO())
            .addDriveKey(this.driveKey.publicKey)
            .addStorageUnits(this.storageUnits.toDTO())
            .build();
    }
}
exports.NewStoragePaymentTransaction = NewStoragePaymentTransaction;
class NewStoragePaymentTransactionBuilder extends Transaction_1.TransactionBuilder {
    _driveKey;
    _storageUnits;
    driveKey(driveKey) {
        this._driveKey = driveKey;
        return this;
    }
    storageUnits(storageUnits) {
        this._storageUnits = storageUnits;
        return this;
    }
    build() {
        return new NewStoragePaymentTransaction(this._networkType, this._version || TransactionTypeVersion_1.TransactionTypeVersion.Storage_Payment, this._deadline ? this._deadline : this._createNewDeadlineFn(), this._maxFee ? this._maxFee : (0, FeeCalculationStrategy_1.calculateFee)(NewStoragePaymentTransaction.calculateSize(), this._feeCalculationStrategy), this._driveKey, this._storageUnits, this._signature, this._signer, this._transactionInfo);
    }
}
exports.NewStoragePaymentTransactionBuilder = NewStoragePaymentTransactionBuilder;
