// tsjs-xpx-chain-sdk/dist/src/model/transaction/storage/NewVerificationPaymentTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewVerificationPaymentTransactionBuilder = exports.NewVerificationPaymentTransaction = void 0;
const NewVerificationPaymentTransaction_1 = require("../../../infrastructure/builders/storage/NewVerificationPaymentTransaction");
const Transaction_1 = require("../Transaction");
const TransactionType_1 = require("../TransactionType");
const TransactionTypeVersion_1 = require("../TransactionTypeVersion");
const FeeCalculationStrategy_1 = require("../FeeCalculationStrategy");
class NewVerificationPaymentTransaction extends Transaction_1.Transaction {
    driveKey;
    verificationFeeAmount;
    /**
     * Create a new replicator onboarding transaction object
     * @param deadline - The deadline to include the transaction.
     * @param driveKey - Public key of the drive
     * @param verificationFeeAmount - Amount of XPXs to transfer to the drive
     * @param networkType - The network type.
     * @param maxFee - (Optional) Max fee defined by the sender
     * @returns {NewVerificationPaymentTransaction}
     */
    static create(deadline, driveKey, verificationFeeAmount, networkType, maxFee) {
        return new NewVerificationPaymentTransactionBuilder()
            .networkType(networkType)
            .deadline(deadline)
            .driveKey(driveKey)
            .verificationFeeAmount(verificationFeeAmount)
            .maxFee(maxFee)
            .build();
    }
    /**
     * @param networkType
     * @param version
     * @param deadline
     * @param maxFee
     * @param driveKey - The drive key
     * @param verificationFeeAmount - The verification fee amount
     * @param signature
     * @param signer
     * @param transactionInfo
     */
    constructor(networkType, version, deadline, maxFee, driveKey, verificationFeeAmount, signature, signer, transactionInfo) {
        super(TransactionType_1.TransactionType.Verification_Payment, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
        this.driveKey = driveKey;
        this.verificationFeeAmount = verificationFeeAmount;
        if (verificationFeeAmount.toBigInt() <= BigInt(0)) {
            throw new Error("verificationFeeAmount should be positive value");
        }
    }
    /**
     * @override Transaction.size()
     * @description get the byte size of a NewVerificationPaymentTransaction
     * @returns {number}
     * @memberof NewVerificationPaymentTransaction
     */
    get size() {
        return NewVerificationPaymentTransaction.calculateSize();
    }
    static calculateSize() {
        const baseByteSize = Transaction_1.Transaction.getHeaderSize();
        // set static byte size fields
        const driveKeySize = 32;
        const verificationFeeAmountSize = 8;
        return baseByteSize + driveKeySize + verificationFeeAmountSize;
    }
    /**
     * @override Transaction.toJSON()
     * @description Serialize a transaction object - add own fields to the result of Transaction.toJSON()
     * @return {Object}
     * @memberof NewVerificationPaymentTransaction
     */
    toJSON() {
        const parent = super.toJSON();
        return {
            ...parent,
            transaction: {
                ...parent.transaction,
                driveKey: this.driveKey.toDTO(),
                verificationFeeAmount: this.verificationFeeAmount.toDTO()
            }
        };
    }
    /**
     * @internal
     * @returns {VerifiableTransaction}
     */
    buildTransaction() {
        return new NewVerificationPaymentTransaction_1.Builder()
            .addSize(this.size)
            .addDeadline(this.deadline.toDTO())
            .addMaxFee(this.maxFee.toDTO())
            .addVersion(this.versionToDTO())
            .addDriveKey(this.driveKey.publicKey)
            .addVerificationFeeAmount(this.verificationFeeAmount.toDTO())
            .build();
    }
}
exports.NewVerificationPaymentTransaction = NewVerificationPaymentTransaction;
class NewVerificationPaymentTransactionBuilder extends Transaction_1.TransactionBuilder {
    _driveKey;
    _verificationFeeAmountSize;
    driveKey(driveKey) {
        this._driveKey = driveKey;
        return this;
    }
    verificationFeeAmount(verificationFeeAmount) {
        this._verificationFeeAmountSize = verificationFeeAmount;
        return this;
    }
    build() {
        return new NewVerificationPaymentTransaction(this._networkType, this._version || TransactionTypeVersion_1.TransactionTypeVersion.Verification_Payment, this._deadline ? this._deadline : this._createNewDeadlineFn(), this._maxFee ? this._maxFee : (0, FeeCalculationStrategy_1.calculateFee)(NewVerificationPaymentTransaction.calculateSize(), this._feeCalculationStrategy), this._driveKey, this._verificationFeeAmountSize, this._signature, this._signer, this._transactionInfo);
    }
}
exports.NewVerificationPaymentTransactionBuilder = NewVerificationPaymentTransactionBuilder;
