// tsjs-xpx-chain-sdk/dist/src/model/transaction/storage/ReplicatorOffboardingTransaction.js
"use strict";
/*
 * Copyright 2023 ProximaX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReplicatorOffboardingTransactionBuilder = exports.ReplicatorOffboardingTransaction = void 0;
const ReplicatorOffboardingTransaction_1 = require("../../../infrastructure/builders/storage/ReplicatorOffboardingTransaction");
const Transaction_1 = require("../Transaction");
const TransactionType_1 = require("../TransactionType");
const TransactionTypeVersion_1 = require("../TransactionTypeVersion");
const FeeCalculationStrategy_1 = require("../FeeCalculationStrategy");
class ReplicatorOffboardingTransaction extends Transaction_1.Transaction {
    driveKey;
    /**
     * Create a new replicator onboarding transaction object
     * @param deadline - The deadline to include the transaction.
     * @param driveKey - Public key of the drive
     * @param networkType - The network type.
     * @param maxFee - (Optional) Max fee defined by the sender
     * @returns {ReplicatorOffboardingTransaction}
     */
    static create(deadline, driveKey, networkType, maxFee) {
        return new ReplicatorOffboardingTransactionBuilder()
            .networkType(networkType)
            .deadline(deadline)
            .driveKey(driveKey)
            .maxFee(maxFee)
            .build();
    }
    /**
     * @param networkType
     * @param version
     * @param deadline
     * @param maxFee
     * @param driveKey
     * @param signature
     * @param signer
     * @param transactionInfo
     */
    constructor(networkType, version, deadline, maxFee, driveKey, signature, signer, transactionInfo) {
        super(TransactionType_1.TransactionType.Replicator_Offboarding, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
        this.driveKey = driveKey;
    }
    /**
     * @override Transaction.size()
     * @description get the byte size of a ReplicatorOffboardingTransaction
     * @returns {number}
     * @memberof ReplicatorOffboardingTransaction
     */
    get size() {
        return ReplicatorOffboardingTransaction.calculateSize();
    }
    static calculateSize() {
        const baseByteSize = Transaction_1.Transaction.getHeaderSize();
        // set static byte size fields
        const driveKeySize = 32;
        return baseByteSize + driveKeySize;
    }
    /**
     * @override Transaction.toJSON()
     * @description Serialize a transaction object - add own fields to the result of Transaction.toJSON()
     * @return {Object}
     * @memberof ReplicatorOffboardingTransaction
     */
    toJSON() {
        const parent = super.toJSON();
        return {
            ...parent,
            transaction: {
                ...parent.transaction,
                driveKey: this.driveKey.toDTO(),
            }
        };
    }
    /**
     * @internal
     * @returns {VerifiableTransaction}
     */
    buildTransaction() {
        return new ReplicatorOffboardingTransaction_1.Builder()
            .addSize(this.size)
            .addDeadline(this.deadline.toDTO())
            .addMaxFee(this.maxFee.toDTO())
            .addVersion(this.versionToDTO())
            .addDriveKey(this.driveKey.publicKey)
            .build();
    }
}
exports.ReplicatorOffboardingTransaction = ReplicatorOffboardingTransaction;
class ReplicatorOffboardingTransactionBuilder extends Transaction_1.TransactionBuilder {
    _driveKey;
    driveKey(driveKey) {
        this._driveKey = driveKey;
        return this;
    }
    build() {
        return new ReplicatorOffboardingTransaction(this._networkType, this._version || TransactionTypeVersion_1.TransactionTypeVersion.Replicator_Offboarding, this._deadline ? this._deadline : this._createNewDeadlineFn(), this._maxFee ? this._maxFee : (0, FeeCalculationStrategy_1.calculateFee)(ReplicatorOffboardingTransaction.calculateSize(), this._feeCalculationStrategy), this._driveKey, this._signature, this._signer, this._transactionInfo);
    }
}
exports.ReplicatorOffboardingTransactionBuilder = ReplicatorOffboardingTransactionBuilder;
