// tsjs-xpx-chain-sdk/dist/src/model/storage/DownloadChannel.js
"use strict";
// Copyright 2023 ProximaX Limited. All rights reserved.
// Use of this source code is governed by the Apache 2.0
// license that can be found in the LICENSE file
Object.defineProperty(exports, "__esModule", { value: true });
exports.DownloadChannel = void 0;
const UInt64_1 = require("../UInt64");
class DownloadChannel {
    id;
    consumer;
    drive;
    downloadSize;
    downloadApprovalCountLeft;
    listOfPublicKeys;
    shardReplicators;
    cumulativePayments;
    metaId;
    constructor(id, consumer, drive, downloadSize, downloadApprovalCountLeft, listOfPublicKeys, shardReplicators, // need verify
    cumulativePayments, metaId) {
        this.id = id;
        this.consumer = consumer;
        this.drive = drive;
        this.downloadSize = downloadSize;
        this.downloadApprovalCountLeft = downloadApprovalCountLeft;
        this.listOfPublicKeys = listOfPublicKeys;
        this.shardReplicators = shardReplicators;
        this.cumulativePayments = cumulativePayments;
        this.metaId = metaId;
    }
    static createFromDownloadChannelInfoDTO(dto, metaId) {
        return DownloadChannel.createFromDTO(dto.id, dto.consumer, dto.drive, dto.downloadSize, dto.downloadApprovalCountLeft, dto.listOfPublicKeys, dto.shardReplicators, dto.cumulativePayments, metaId);
    }
    static createFromDTO(id, consumer, drive, downloadSize, downloadApprovalCountLeft, listOfPublicKeys, shardReplicators, // need verify
    cumulativePayments, metaId) {
        return new DownloadChannel(id, consumer, drive, new UInt64_1.UInt64(downloadSize), downloadApprovalCountLeft, listOfPublicKeys, shardReplicators, {
            replicator: cumulativePayments.replicator,
            payment: new UInt64_1.UInt64(cumulativePayments.payment)
        }, metaId);
    }
}
exports.DownloadChannel = DownloadChannel;
