// tsjs-xpx-chain-sdk/dist/src/model/storage/DriveInfo.js
"use strict";
// Copyright 2023 ProximaX Limited. All rights reserved.
// Use of this source code is governed by the Apache 2.0
// license that can be found in the LICENSE file
Object.defineProperty(exports, "__esModule", { value: true });
exports.DriveInfo = void 0;
const Address_1 = require("../account/Address");
const UInt64_1 = require("../UInt64");
class DriveInfo {
    multisig;
    multisigAddress;
    owner;
    rootHash;
    lastModificationId;
    size;
    usedSizeBytes;
    metaFilesSizeBytes;
    replicatorCount;
    ownerManagement;
    activeDataModifications;
    completedDataModifications;
    confirmedUsedSizes;
    replicators;
    offboardingReplicators;
    downloadShards;
    dataModificationShards;
    metaId;
    constructor(multisig, multisigAddress, owner, rootHash, lastModificationId, size, usedSizeBytes, metaFilesSizeBytes, replicatorCount, ownerManagement, activeDataModifications, // need verify
    completedDataModifications, confirmedUsedSizes, replicators, offboardingReplicators, // need verify
    downloadShards, // need verify
    dataModificationShards, metaId) {
        this.multisig = multisig;
        this.multisigAddress = multisigAddress;
        this.owner = owner;
        this.rootHash = rootHash;
        this.lastModificationId = lastModificationId;
        this.size = size;
        this.usedSizeBytes = usedSizeBytes;
        this.metaFilesSizeBytes = metaFilesSizeBytes;
        this.replicatorCount = replicatorCount;
        this.ownerManagement = ownerManagement;
        this.activeDataModifications = activeDataModifications;
        this.completedDataModifications = completedDataModifications;
        this.confirmedUsedSizes = confirmedUsedSizes;
        this.replicators = replicators;
        this.offboardingReplicators = offboardingReplicators;
        this.downloadShards = downloadShards;
        this.dataModificationShards = dataModificationShards;
        this.metaId = metaId;
    }
    static createFromDriveDTO(driveDTO, metaId) {
        return DriveInfo.createFromDTO(driveDTO.multisig, driveDTO.multisigAddress, driveDTO.owner, driveDTO.rootHash, driveDTO.lastModificationId, driveDTO.size, driveDTO.usedSizeBytes, driveDTO.metaFilesSizeBytes, driveDTO.replicatorCount, driveDTO.ownerManagement, driveDTO.activeDataModifications, driveDTO.completedDataModifications, driveDTO.confirmedUsedSizes, driveDTO.replicators, driveDTO.offboardingReplicators, driveDTO.downloadShards, driveDTO.dataModificationShards, metaId);
    }
    static createFromDTO(multisig, multisigAddress, owner, rootHash, lastModificationId, size, usedSizeBytes, metaFilesSizeBytes, replicatorCount, ownerManagement, activeDataModifications, // need verify
    completedDataModifications, confirmedUsedSizes, replicators, offboardingReplicators, // need verify
    downloadShards, // need verify
    dataModificationShards, metaId) {
        let activeDataMod = activeDataModifications.map((data) => {
            return {
                id: data.id,
                owner: data.owner,
                downloadDataCdi: data.downloadDataCdi,
                expectedUploadSize: new UInt64_1.UInt64(data.expectedUploadSize),
                actualUploadSize: new UInt64_1.UInt64(data.actualUploadSize),
                folderName: data.folderName,
                readyForApproval: data.readyForApproval,
                isStream: data.isStream,
                state: data.state,
                success: data.success
            };
        });
        let completedDataMod = completedDataModifications.map((data) => {
            return {
                id: data.id,
                owner: data.owner,
                downloadDataCdi: data.downloadDataCdi,
                expectedUploadSize: new UInt64_1.UInt64(data.expectedUploadSize),
                actualUploadSize: new UInt64_1.UInt64(data.actualUploadSize),
                folderName: data.folderName,
                readyForApproval: data.readyForApproval,
                isStream: data.isStream,
                state: data.state,
                success: data.success
            };
        });
        let confirmedUsedSizeData = confirmedUsedSizes.map((data) => {
            return {
                replicator: data.replicator,
                size: new UInt64_1.UInt64(data.size)
            };
        });
        let dataModificationShardData = dataModificationShards.map((data) => {
            let actualSRs = data.actualShardReplicators.map((shardData) => {
                return {
                    key: shardData.key,
                    uploadSize: new UInt64_1.UInt64(shardData.uploadSize)
                };
            });
            let formerSRs = data.formerShardReplicators.map((shardData) => {
                return {
                    key: shardData.key,
                    uploadSize: new UInt64_1.UInt64(shardData.uploadSize)
                };
            });
            return {
                replicator: data.replicator,
                actualShardReplicators: actualSRs,
                formerShardReplicators: formerSRs,
                ownerUpload: new UInt64_1.UInt64(data.ownerUpload)
            };
        });
        return new DriveInfo(multisig, Address_1.Address.createFromEncoded(multisigAddress), owner, rootHash, lastModificationId, new UInt64_1.UInt64(size), new UInt64_1.UInt64(usedSizeBytes), new UInt64_1.UInt64(metaFilesSizeBytes), replicatorCount, ownerManagement, activeDataMod, completedDataMod, confirmedUsedSizeData, replicators, offboardingReplicators, downloadShards, dataModificationShardData, metaId);
    }
}
exports.DriveInfo = DriveInfo;
