// tsjs-xpx-chain-sdk/dist/src/model/storage/Replicator.js
"use strict";
// Copyright 2023 ProximaX Limited. All rights reserved.
// Use of this source code is governed by the Apache 2.0
// license that can be found in the LICENSE file
Object.defineProperty(exports, "__esModule", { value: true });
exports.Replicator = void 0;
const UInt64_1 = require("../UInt64");
class Replicator {
    key;
    version;
    drives;
    downloadChannels;
    metaId;
    constructor(key, version, drives, downloadChannels, // need verify
    metaId) {
        this.key = key;
        this.version = version;
        this.drives = drives;
        this.downloadChannels = downloadChannels;
        this.metaId = metaId;
    }
    static createFromReplicatorDTO(dto, metaId) {
        return Replicator.createFromDTO(dto.key, dto.version, dto.drives, dto.downloadChannels, metaId);
    }
    static createFromDTO(key, version, drives, downloadChannels, // need verify
    metaId) {
        let driveWorks = drives.map((data) => {
            return {
                drive: data.drive,
                lastApprovedDataModificationId: data.lastApprovedDataModificationId,
                initialDownloadWork: new UInt64_1.UInt64(data.initialDownloadWork),
                lastCompletedCumulativeDownloadWork: new UInt64_1.UInt64(data.lastCompletedCumulativeDownloadWork)
            };
        });
        return new Replicator(key, version, driveWorks, downloadChannels, metaId);
    }
}
exports.Replicator = Replicator;
