// tsjs-xpx-chain-sdk/dist/src/model/exchangeSda/AccountSdaExchanges.js
"use strict";
// Copyright 2020 ProximaX Limited. All rights reserved.
// Use of this source code is governed by the Apache 2.0
// license that can be found in the LICENSE file
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountSdaExchanges = void 0;
const PublicAccount_1 = require("../account/PublicAccount");
const SdaOfferInfo_1 = require("./SdaOfferInfo");
class AccountSdaExchanges {
    owner;
    version;
    sdaOfferBalances;
    expiredSdaOfferBalances;
    constructor(owner, version, sdaOfferBalances, expiredSdaOfferBalances) {
        this.owner = owner;
        this.version = version;
        this.sdaOfferBalances = sdaOfferBalances;
        this.expiredSdaOfferBalances = expiredSdaOfferBalances;
    }
    static createFromDTO(accountSdaExchangeDTO, networkType) {
        return accountSdaExchangeDTO ? new AccountSdaExchanges(PublicAccount_1.PublicAccount.createFromPublicKey(accountSdaExchangeDTO.owner, networkType), accountSdaExchangeDTO.version, accountSdaExchangeDTO.sdaOfferBalances.map(sdaOfferBalance => SdaOfferInfo_1.SdaOfferInfo.createFromDTO(sdaOfferBalance)), accountSdaExchangeDTO.expiredSdaOfferBalances.map(expiredSdaOfferBalance => SdaOfferInfo_1.SdaOfferInfo.createFromDTO(expiredSdaOfferBalance))) : undefined;
    }
}
exports.AccountSdaExchanges = AccountSdaExchanges;
