// tsjs-xpx-chain-sdk/dist/src/model/exchangeSda/SdaOfferInfoWithOwner.js
"use strict";
// Copyright 2020 ProximaX Limited. All rights reserved.
// Use of this source code is governed by the Apache 2.0
// license that can be found in the LICENSE file
Object.defineProperty(exports, "__esModule", { value: true });
exports.SdaOfferInfoWithOwner = void 0;
const MosaicId_1 = require("../mosaic/MosaicId");
const PublicAccount_1 = require("../account/PublicAccount");
const UInt64_1 = require("../UInt64");
class SdaOfferInfoWithOwner {
    mosaicIdGive;
    mosaicIdGet;
    currentMosaicGiveAmount;
    currentMosaicGetAmount;
    initialMosaicGiveAmount;
    initialMosaicGetAmount;
    deadline;
    owner;
    constructor(mosaicIdGive, mosaicIdGet, currentMosaicGiveAmount, currentMosaicGetAmount, initialMosaicGiveAmount, initialMosaicGetAmount, 
    /**
    *  block height deadline
    */
    deadline, owner) {
        this.mosaicIdGive = mosaicIdGive;
        this.mosaicIdGet = mosaicIdGet;
        this.currentMosaicGiveAmount = currentMosaicGiveAmount;
        this.currentMosaicGetAmount = currentMosaicGetAmount;
        this.initialMosaicGiveAmount = initialMosaicGiveAmount;
        this.initialMosaicGetAmount = initialMosaicGetAmount;
        this.deadline = deadline;
        this.owner = owner;
    }
    static createFromDTO(offerInfoDTO, networkType) {
        return new SdaOfferInfoWithOwner(new MosaicId_1.MosaicId(offerInfoDTO.mosaicIdGive), new MosaicId_1.MosaicId(offerInfoDTO.mosaicIdGet), new UInt64_1.UInt64(offerInfoDTO.currentMosaicGiveAmount), new UInt64_1.UInt64(offerInfoDTO.currentMosaicGetAmount), new UInt64_1.UInt64(offerInfoDTO.initialMosaicGiveAmount), new UInt64_1.UInt64(offerInfoDTO.initialMosaicGetAmount), new UInt64_1.UInt64(offerInfoDTO.deadline), PublicAccount_1.PublicAccount.createFromPublicKey(offerInfoDTO.owner, networkType));
    }
}
exports.SdaOfferInfoWithOwner = SdaOfferInfoWithOwner;
